{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# UVEAL MELANOMA TUMOR CELL ANALYSIS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### Ashley Laughney ... 03-29-2018\n",
    "###### Ethan Earlie ........... 07-20-2020"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## INSTALL DEPENDENCIES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_dpi = 300"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "run_in_background_wait_default=True\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "import time\n",
    "import decimal\n",
    "import json\n",
    "import os\n",
    "from os import path\n",
    "import random\n",
    "import shlex\n",
    "import shutil\n",
    "import glob\n",
    "from subprocess import call, check_output, Popen, PIPE\n",
    "import gc\n",
    "import numpy as np\n",
    "from numpy.linalg import norm\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import networkx as nx\n",
    "import pickle\n",
    "import joblib\n",
    "import re\n",
    "import csv\n",
    "import h5py\n",
    "from copy import deepcopy\n",
    "import itertools\n",
    "from itertools import combinations\n",
    "from collections import defaultdict, OrderedDict\n",
    "from imp import reload\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "from colors import rgb, hex\n",
    "import scipy.cluster.hierarchy as hc\n",
    "#import wikipedia\n",
    "import pylab\n",
    "from scipy.stats import wasserstein_distance\n",
    "\n",
    "import matplotlib\n",
    "print(matplotlib.get_backend())\n",
    "#%matplotlib inline\n",
    "#print(matplotlib.get_backend())\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "print(matplotlib.get_backend())\n",
    "import matplotlib.gridspec as gridspec\n",
    "print(matplotlib.get_backend())\n",
    "import matplotlib.colors as mcolors\n",
    "\n",
    "from functools import partial\n",
    "from statsmodels.sandbox.stats.multicomp import multipletests\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "\n",
    "from sklearn import manifold\n",
    "from sklearn.manifold.t_sne import _joint_probabilities, _joint_probabilities_nn\n",
    "from sklearn.metrics import adjusted_rand_score\n",
    "from sklearn.metrics.pairwise import pairwise_distances\n",
    "from sklearn.neighbors import NearestNeighbors\n",
    "from sklearn import cluster\n",
    "from sklearn.svm import SVC, LinearSVC\n",
    "\n",
    "from scipy import stats\n",
    "from scipy.cluster.hierarchy import dendrogram\n",
    "from scipy.sparse import coo_matrix, csr_matrix, find, csgraph\n",
    "from scipy.sparse.linalg import eigs\n",
    "from scipy.stats import t, variation\n",
    "from scipy.stats.mstats import kruskalwallis, rankdata\n",
    "from scipy.spatial import distance\n",
    "from scipy.spatial.distance import pdist\n",
    "from scipy.cluster.hierarchy import dendrogram, fcluster, leaves_list, set_link_color_palette\n",
    "from kneed import DataGenerator, KneeLocator\n",
    "\n",
    "#import seqc.stats.analysis as tools\n",
    "\n",
    "import fastcluster\n",
    "import bhtsne\n",
    "import phenograph\n",
    "import palantir\n",
    "\n",
    "#import magic\n",
    "#import wishbone\n",
    "import seqc\n",
    "import seqc.plot\n",
    "import seqc.stats\n",
    "import seqc.stats.resampled_nonparametric as RNP\n",
    "\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "print(matplotlib.get_backend())\n",
    "\n",
    "from pylab import *\n",
    "from scipy.optimize import curve_fit\n",
    "from geode import *\n",
    "#import gseapy as gp\n",
    "from seqc.stats import gsea as GSEA\n",
    "\n",
    "from scipy.stats import zscore\n",
    "import matplotlib.ticker as plticker\n",
    "import cmocean\n",
    "from sklearn.linear_model import LinearRegression\n",
    "\n",
    "\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from scipy.cluster.hierarchy import linkage, dendrogram\n",
    "import statsmodels.formula.api as sms\n",
    "from sklearn.cluster import SpectralClustering\n",
    "from sklearn.mixture import GaussianMixture, BayesianGaussianMixture\n",
    "\n",
    "from IPython.display import HTML, display\n",
    "from IPython.display import IFrame\n",
    "#import requests, json\n",
    "clustergrammer_url = 'http://amp.pharm.mssm.edu/clustergrammer/matrix_upload/'\n",
    "\n",
    "from statistics import mode\n",
    "from pprint import pprint\n",
    "\n",
    "#from src.plotCnvArr import plotcnv\n",
    "#from src.inferCNV import infer_cnv\n",
    "\n",
    "def gauss(x,mu,sigma,A):\n",
    "    return A*exp(-(x-mu)**2/2/sigma**2)\n",
    "\n",
    "def bimodal(x,mu1,sigma1,A1,mu2,sigma2,A2):\n",
    "    return gauss(x,mu1,sigma1,A1)+gauss(x,mu2,sigma2,A2)\n",
    "\n",
    "def rgb2hex(r,g,b):\n",
    "    hex = \"#{:02x}{:02x}{:02x}\".format(r,g,b)\n",
    "    return hex\n",
    "\n",
    "def hex2rgb(hexcode):\n",
    "    rgb = tuple(map(ord,hexcode[1:].decode('hex')))\n",
    "    return rgb\n",
    "\n",
    "# Custom analysis functions, embedded here\n",
    "#import seqc.stats.analysis as tools\n",
    "\n",
    "def display_link(url):\n",
    "    raw_html = '<a href=\"%s\" target=\"_blank\">%s</a>' % (url, url)\n",
    "    return display(HTML(raw_html))\n",
    "\n",
    "import matplotlib.patches as patches\n",
    "print(matplotlib.get_backend())\n",
    "import matplotlib.colors as colors\n",
    "print(matplotlib.get_backend())\n",
    "import math\n",
    "import scipy.cluster.hierarchy as sch\n",
    "\n",
    "import re\n",
    "from kneed import DataGenerator, KneeLocator\n",
    "from natsort import natsorted, ns\n",
    "\n",
    "_nsre = re.compile('([0-9]+)')\n",
    "def natural_sort_key(s):\n",
    "    return [int(text) if text.isdigit() else text.lower()\n",
    "            for text in re.split(_nsre, s)] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# INSTALL FORCE DIRECTED LAYOUT DEPENDENCIES\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.neighbors import NearestNeighbors\n",
    "from scipy.sparse import csr_matrix, find\n",
    "from fa2 import ForceAtlas2\n",
    "\n",
    "def determine_kernel(pca_projections, knn=30, n_jobs=-1):\n",
    "    \"\"\"Run Diffusion maps using the adaptive anisotropic kernel\n",
    "\n",
    "    :param pca_projections: PCA projections of the data\n",
    "    :param n_components: Number of diffusion components\n",
    "    :return: Diffusion components, corresponding eigen values and the diffusion operator\n",
    "    \"\"\"\n",
    "\n",
    "    # Determine the kernel\n",
    "    print('Determing nearest neighbor graph...')\n",
    "    nbrs = NearestNeighbors(n_neighbors=int(knn), metric='euclidean',\n",
    "                            n_jobs=n_jobs).fit(pca_projections.values)\n",
    "    kNN = nbrs.kneighbors_graph(pca_projections.values, mode='distance')\n",
    "\n",
    "    # Adaptive k\n",
    "    adaptive_k = int(np.floor(knn / 3))\n",
    "    nbrs = NearestNeighbors(n_neighbors=int(adaptive_k),\n",
    "                            metric='euclidean', n_jobs=n_jobs).fit(pca_projections.values)\n",
    "    adaptive_std = nbrs.kneighbors_graph(\n",
    "        pca_projections.values, mode='distance').max(axis=1)\n",
    "    adaptive_std = np.ravel(adaptive_std.todense())\n",
    "\n",
    "    # Kernel\n",
    "    N = pca_projections.shape[0]\n",
    "    x, y, dists = find(kNN)\n",
    "\n",
    "    # Diffusion components\n",
    "    W = csr_matrix((np.exp(-dists), (x, y)), shape=[N, N])\n",
    "    kernel = W + W.T\n",
    "\n",
    "    return kernel\n",
    "\n",
    "\n",
    "def compute_force_atlas(adjacency, index_cells, verbose=True, iterations=1500):\n",
    "\n",
    "    init_coords = np.random.random((adjacency.shape[0], 2))\n",
    "\n",
    "    forceatlas2 = ForceAtlas2(\n",
    "        # Behavior alternatives\n",
    "        outboundAttractionDistribution=False,  # Dissuade hubs\n",
    "        linLogMode=False,  # NOT IMPLEMENTED\n",
    "        adjustSizes=False,  # Prevent overlap (NOT IMPLEMENTED)\n",
    "        edgeWeightInfluence=1.0,\n",
    "        # Performance\n",
    "        jitterTolerance=1.0,  # Tolerance\n",
    "        barnesHutOptimize=True,\n",
    "        barnesHutTheta=1.2,\n",
    "        multiThreaded=False,  # NOT IMPLEMENTED\n",
    "        # Tuning\n",
    "        scalingRatio=2.0,\n",
    "        strongGravityMode=False,\n",
    "        gravity=1.0,\n",
    "        # Log\n",
    "        verbose=verbose)\n",
    "\n",
    "    positions = forceatlas2.forceatlas2(\n",
    "        adjacency, pos=init_coords, iterations=iterations)\n",
    "    positions = np.array(positions)\n",
    "\n",
    "    # Convert to dataframe\n",
    "    positions = pd.DataFrame(positions,\n",
    "                             index=index_cells, columns=['x', 'y'])\n",
    "    return positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FUNCTIONS FOR PARSING GMT FILES\n",
    "import sys, logging\n",
    "def gsea_gmt_parser(gmt, min_size = 3, max_size = 1000, gene_list=None):\n",
    "    \"\"\"Parse gene_sets.gmt(gene set database) file or download from enrichr server.  \n",
    "    \n",
    "    :param gmt: the gene_sets.gmt file of GSEA input or an enrichr libary name.\n",
    "                checkout full enrichr library name here: http://amp.pharm.mssm.edu/Enrichr/#stats\n",
    "                \n",
    "    :param min_size: Minimum allowed number of genes from gene set also the data set. Default: 3. \n",
    "    :param max_size: Maximum allowed number of genes from gene set also the data set. Default: 5000.\n",
    "    :param gene_list: Used for filtering gene set. Only used this argument for :func:`call` method.\n",
    "    :return: Return a new filtered gene set database dictionary. \n",
    "\n",
    "    **DO NOT** filter gene sets, when use :func:`replot`. Because ``GSEA`` Desktop have already\n",
    "    do this for you.\n",
    "            \n",
    "    \"\"\"\n",
    "\n",
    "    if gmt.lower().endswith(\".gmt\"):\n",
    "        with open(gmt) as genesets:    \n",
    "             genesets_dict = { line.strip(\"\\n\").split(\"\\t\")[0]: line.strip(\"\\n\").split(\"\\t\")[2:] \n",
    "                              for line in genesets.readlines()}    \n",
    "    else:\n",
    "        logging.info(\"Downloading and generating Enrichr library gene sets...\") \n",
    "        names = get_library_name()\n",
    "        if gmt in names:\n",
    "            import requests\n",
    "            ENRICHR_URL = 'http://amp.pharm.mssm.edu/Enrichr/geneSetLibrary'\n",
    "            query_string = '?mode=text&libraryName=%s'\n",
    "            response = requests.get( ENRICHR_URL + query_string % gmt)\n",
    "        else: \n",
    "            raise Exception(\"gene_set files(.gmt) not found\")\n",
    "        if not response.ok:\n",
    "            raise Exception('Error fetching enrichment results, check internet connection first.')\n",
    "                     \n",
    "        genesets_dict = { line.split(\"\\t\")[0]: \n",
    "                          [gene.strip(\"\\n\").split(\",\")[0] for gene in line.split(\"\\t\")[2:-1]] \n",
    "                          for line in response.iter_lines(chunk_size=1024, decode_unicode='utf-8')}    \n",
    " \n",
    "    \n",
    "\n",
    "    #filtering dict\n",
    "    if sys.version_info[0] == 3 :\n",
    "        genesets_filter =  {k: v for k, v in genesets_dict.items() if len(v) >= min_size and len(v) <= max_size}\n",
    "    elif sys.version_info[0] == 2:\n",
    "        genesets_filter =  {k: v for k, v in genesets_dict.iteritems() if len(v) >= min_size and len(v) <= max_size}\n",
    "    else:\n",
    "        sys.stderr.write(\"System failure. Please Provide correct input files\")\n",
    "        sys.exit(1)    \n",
    "    if gene_list is not None:\n",
    "        subsets = sorted(genesets_filter.keys())             \n",
    "        for subset in subsets:            \n",
    "            tag_indicator = in1d(gene_list, genesets_filter.get(subset), assume_unique=True)\n",
    "            tag_len = sum(tag_indicator)      \n",
    "            if tag_len <= min_size or tag_len >= max_size:                    \n",
    "                del genesets_filter[subset]\n",
    "            else:\n",
    "                continue\n",
    "                \n",
    "    filsets_num = len(genesets_dict) - len(genesets_filter)\n",
    "    logging.info(\"{a} gene_sets have been filtered out when max_size={b} and min_size={c}\". \\\n",
    "                 format(a=filsets_num,b=max_size,c=min_size))\n",
    "    \n",
    "    if filsets_num == len(genesets_dict):\n",
    "        sys.stderr.write(\"No gene sets passed throught filtering condition!!!, try new paramters again!\\n\" +\\\n",
    "                         \"Note: Gene names for gseapy is case sensitive.\" )\n",
    "        sys.exit(1)\n",
    "    else:\n",
    "        return genesets_filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FUNCTIONS FOR PARSING DENDROGRAMS\n",
    "# Next we need to grab the labels of the leaves given the indexes. \n",
    "# But before we do that, since it's difficult to keep track of what color e.g. '#c13d3f' is, \n",
    "# we make make an IPython notebook compatible HTML representation of the dictionary holding the information. \n",
    "# Objects of this class will behave just like dictionaries, except for representing them as a HTML table.\n",
    "class Clusters(dict):\n",
    "    def _repr_html_(self):\n",
    "        html = '<table style=\"border: 0;\">'\n",
    "        for c in self:\n",
    "            hx = rgb2hex(colorConverter.to_rgb(c))\n",
    "            html += '<tr style=\"border: 0;\">' \\\n",
    "            '<td style=\"background-color: {0}; ' \\\n",
    "                       'border: 0;\">' \\\n",
    "            '<code style=\"background-color: {0};\">'.format(hx)\n",
    "            html += c + '</code></td>'\n",
    "            html += '<td style=\"border: 0\"><code>' \n",
    "            html += repr(self[c]) + '</code>'\n",
    "            html += '</td></tr>'\n",
    "        \n",
    "        html += '</table>'\n",
    "        \n",
    "        return html\n",
    "    \n",
    "def get_cluster_classes(den, label='ivl'):\n",
    "    cluster_idxs = defaultdict(list)\n",
    "    for c, pi in zip(den['color_list'], den['icoord']):\n",
    "        for leg in pi[1:3]:\n",
    "            i = (leg - 5.0) / 10.0\n",
    "            if abs(i - int(i)) < 1e-5:\n",
    "                cluster_idxs[c].append(int(i))\n",
    "    \n",
    "    cluster_classes = Clusters()\n",
    "    for c, l in cluster_idxs.items():\n",
    "        i_l = [den[label][i] for i in l]\n",
    "        cluster_classes[c] = i_l\n",
    "    \n",
    "    return cluster_classes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FUNCTION FOR RUNNING R SCRIPTS\n",
    "# run bash command in background if stdoutfile does not exist. If it does, output contents of stdoutfile.\n",
    "# If force=true, overwrite and re-run command.\n",
    "def run_in_background(command, stdoutfile, stderrfile=\"\", force=False, wait=run_in_background_wait_default):\n",
    "    if force or not path.exists(stdoutfile):\n",
    "        command = 'bash -c \\'' + command + '\\' > ' + stdoutfile\n",
    "        if stderrfile:\n",
    "            command = command + ' 2>' + stderrfile\n",
    "        else:\n",
    "            command = command + ' 2>&1'\n",
    "        if not wait:\n",
    "            command = command + ' &'\n",
    "        print('calling ' + command + '\\n')\n",
    "        os.system(command)\n",
    "    else:\n",
    "        print(\"Output from stdout file \" + stdoutfile)\n",
    "        !cat {stdoutfile}\n",
    "        if stderrfile and path.exists(stderrfile):\n",
    "            print(\"Output from stderr file \" + stderrfile)\n",
    "            !cat {stderrfile}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# MAKE CUSTOM COLR MAP\n",
    "def make_cmap(colors, position=None, bit=False):\n",
    "    '''\n",
    "    make_cmap takes a list of tuples which contain RGB values. The RGB\n",
    "    values may either be in 8-bit [0 to 255] (in which bit must be set to\n",
    "    True when called) or arithmetic [0 to 1] (default). make_cmap returns\n",
    "    a cmap with equally spaced colors.\n",
    "    Arrange your tuples so that the first color is the lowest value for the\n",
    "    colorbar and the last is the highest.\n",
    "    position contains values from 0 to 1 to dictate the location of each color.\n",
    "    '''\n",
    "    import matplotlib as mpl\n",
    "    import numpy as np\n",
    "    bit_rgb = np.linspace(0,1,256)\n",
    "    if position == None:\n",
    "        position = np.linspace(0,1,len(colors))\n",
    "    else:\n",
    "        if len(position) != len(colors):\n",
    "            sys.exit(\"position length must be the same as colors\")\n",
    "        elif position[0] != 0 or position[-1] != 1:\n",
    "            sys.exit(\"position must start with 0 and end with 1\")\n",
    "    if bit:\n",
    "        for i in range(len(colors)):\n",
    "            colors[i] = (bit_rgb[colors[i][0]],\n",
    "                         bit_rgb[colors[i][1]],\n",
    "                         bit_rgb[colors[i][2]])\n",
    "    cdict = {'red':[], 'green':[], 'blue':[]}\n",
    "    for pos, color in zip(position, colors):\n",
    "        cdict['red'].append((pos, color[0], color[0]))\n",
    "        cdict['green'].append((pos, color[1], color[1]))\n",
    "        cdict['blue'].append((pos, color[2], color[2]))\n",
    "\n",
    "    cmap = mpl.colors.LinearSegmentedColormap('my_colormap',cdict,256)\n",
    "    return cmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# H5 Wrapper\n",
    "class H5:\n",
    "\n",
    "    def __init__(self, archive_name: str):\n",
    "        \"\"\"Wrapper for the pandas HDFStore class which ensures that all interactions with\n",
    "        the archive result in a closed, flushed archive.\n",
    "\n",
    "        In order to ensure data usability, all data must be submitted in DataFrame format.\n",
    "        This decision was made to encourage users to pair metadata with sequencing data,\n",
    "        and reduce the incidence of unexpected data permutation.\n",
    "\n",
    "        :param archive_name: name of the h5 archive to open. If the archive does not exist\n",
    "          it will be created using a blosc5 filter\n",
    "\n",
    "        :method ls: list contents of the archive\n",
    "        :method save: save an object to the h5 archive\n",
    "        :method load: load an object from the archive\n",
    "        :method remove: remove a DataFrame from the archive\n",
    "        :method is_open: returns True if the h5 archive is open, else False\n",
    "        \"\"\"\n",
    "        if os.path.isfile(archive_name):\n",
    "            self._archive = pd.HDFStore(archive_name, mode='a')\n",
    "            self._archive.close()\n",
    "        else:\n",
    "            self._archive = pd.HDFStore(\n",
    "                archive_name, mode='a', complib='blosc', complevel=5)\n",
    "            self._archive.close()\n",
    "\n",
    "    def __repr__(self):\n",
    "        self._archive.open()\n",
    "        try:\n",
    "            return repr(self._archive)\n",
    "        finally:\n",
    "            self._archive.close()\n",
    "\n",
    "    def save(self, data: pd.DataFrame, location: str) -> None:\n",
    "        \"\"\"Save DataFrame data to the h5 archive in location.\n",
    "\n",
    "        :param data: DataFrame object to store\n",
    "        :param location: filepath to save the object in the h5 hierarchy\n",
    "        \"\"\"\n",
    "        if not isinstance(data, pd.DataFrame):\n",
    "            if isinstance(data, np.ndarray):\n",
    "                res = input('np.ndarray class detected. Save as pd.DataFrame with '\n",
    "                            'ascending integer indices? [y/n] ')\n",
    "                if res in ['y', 'yes', 'Y', 'YES', 'True', 'true', '1']:\n",
    "                    data = pd.DataFrame(data)\n",
    "                else:\n",
    "                    print('User elected not to save DataFrame, archive is unmodified.')\n",
    "                    return\n",
    "            else:\n",
    "                raise TypeError('only pd.DataFrame objects can be saved using this '\n",
    "                                'class. To save np.ndarray objects please see the tables '\n",
    "                                'package.')\n",
    "        self._archive.open()\n",
    "        try:\n",
    "            self._archive[location] = data\n",
    "        finally:\n",
    "            self._archive.close()\n",
    "\n",
    "    def load(self, location: str) -> None:\n",
    "        \"\"\"Load and return the dataframe found at location in the archive.\n",
    "\n",
    "        :param location: str, location of object to retrieve from h5\n",
    "        :return: pd.DataFrame, object found at location\n",
    "        \"\"\"\n",
    "        self._archive.open()\n",
    "        try:\n",
    "            return self._archive[location]\n",
    "        finally:\n",
    "            self._archive.close()\n",
    "\n",
    "    def ls(self) -> None:\n",
    "        \"\"\"list archive contents\"\"\"\n",
    "        try:\n",
    "            self._archive.open()\n",
    "            print(self._archive.keys())\n",
    "            print(self._archive)\n",
    "        finally:\n",
    "            self._archive.close()\n",
    "\n",
    "    def remove(self, location: str) -> None:\n",
    "        \"\"\"remove the DataFrame at location from the archive\n",
    "\n",
    "        Note: removing a dataframe at a branch node will remove all leaves sharing this\n",
    "        prefix. e.g. in an archive containing:\n",
    "\n",
    "        /data\n",
    "        /data/filtered\n",
    "        /data/metadata\n",
    "        /new_data/data\n",
    "\n",
    "        removing /data would remove the first three DataFrame objects from the archive.\n",
    "\n",
    "        :param location: location of DataFrame to remove\n",
    "        :return: None\n",
    "        \"\"\"\n",
    "\n",
    "        self._archive.open()\n",
    "        try:\n",
    "            if location not in self._archive.keys():\n",
    "                raise ValueError(\n",
    "                    '{} not contained in archive, nothing to remove.'.format(location))\n",
    "            else:\n",
    "                removed = [k for k in self._archive.keys()\n",
    "                           if k.startswith(location + '/')]\n",
    "                if len(removed) != 0:\n",
    "                    res = input(\n",
    "                        'Removing branch node {}, which is a prefix for {!a} will remove '\n",
    "                        'all listed DataFrames. Continue with removal? [y/n] '.format(\n",
    "                            location, removed))\n",
    "                    if res not in ['y', 'yes', 'Y', 'YES', 'True', 'true', '1']:\n",
    "                        print('returned without deletion.')\n",
    "                        return\n",
    "                self._archive.remove(location)\n",
    "        finally:\n",
    "            self._archive.close()\n",
    "\n",
    "    @property\n",
    "    def is_open(self) -> bool:\n",
    "        return self._archive.is_open"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "x = np.arange(5)\n",
    "plt.scatter(x,x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# DISPLALY ALL CURRENT DEPENEDENCY\n",
    "!pip freeze"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#s=pd.HDFStore(\"/ifs/e63data/massaguelab/ashley/data/sc_RNAseq/h5/counts/Project_MB_07384/UVMEL_MERGED_noAdditionalLibSizeFilter.h5\")\n",
    "#s.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#s.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tmp = '/ifs/e63data/massaguelab/ashley/data/sc_RNAseq/h5/counts/Project_MB_07384/UVMEL_MERGED_noAdditionalLibSizeFilter.h5'\n",
    "#!ls -l '/ifs/e63data/massaguelab/ashley/data/sc_RNAseq/h5/counts/Project_MB_07384/UVMEL_MERGED_noAdditionalLibSizeFilter.h5'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#h5_data = seqc.H5('/ifs/e63data/massaguelab/ashley/data/sc_RNAseq/h5/counts/Project_MB_07384/UVMEL_MERGED_noAdditionalLibSizeFilter.h5')\n",
    "#h5_data.ls()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#help(seqc.H5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#matplotlib.get_backend()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## LOAD SAMPLE DATA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SAMPLE INFORMATION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_PATH = os.path.expanduser('/workdir/uvmel_project/data/')\n",
    "print(DATA_PATH)\n",
    "FN  = 'UVMEL_MERGED.h5'#_noAdditionalLibSizeFilter\n",
    "ORGANISM = 'human'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SET PATH TO SAVE DIRECTORIES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TAG ANALYSIS OUTPUT FOLDERS WITH CURRENT DATE\n",
    "tag = '_ARCHETYPES'\n",
    "now = time.strftime(\"%x\") \n",
    "now = str.replace(now,'/','_')\n",
    "\n",
    "# SPECIFY OUTPUT STEMS FOR FIGURES/PATHWAY ANALYSIS\n",
    "GSEA_output_stem = DATA_PATH.replace('/data/','/GSEA/') + FN.replace('.h5','{}_'.format(tag)) + now +'/'\n",
    "print(GSEA_output_stem)\n",
    "FIG_output_stem = DATA_PATH.replace('/data/','/figures/') + FN.replace('.h5','{}_'.format(tag)) + now +'/'\n",
    "print(FIG_output_stem)\n",
    "CNV_output_stem = DATA_PATH.replace('/data/','/CNV/') + FN.replace('.h5','{}_'.format(tag)) + now +'/'\n",
    "print(CNV_output_stem)\n",
    "CSV_output_stem = DATA_PATH.replace(\"/data/\",'/CSV/') + FN.replace('.h5','{}_'.format(tag)) + now +'/'\n",
    "print(CSV_output_stem)\n",
    "\n",
    "# CREATE GSEA DIRECTORY IF IT DOES NOT EXIST\n",
    "d = os.path.dirname(GSEA_output_stem)\n",
    "if not os.path.exists(d):\n",
    "        os.makedirs(d)\n",
    "\n",
    "# CREATE FIGURE DIRECTORY IF IT DOES NOT EXIST     \n",
    "d = os.path.dirname(FIG_output_stem)\n",
    "if not os.path.exists(d):\n",
    "        os.makedirs(d)\n",
    "        \n",
    "# CREATE CNV DIRECTORY IF IT DOES NOT EXIST     \n",
    "d = os.path.dirname(CNV_output_stem)\n",
    "if not os.path.exists(d):\n",
    "        os.makedirs(d)\n",
    "        \n",
    "# CREATE CNV DIRECTORY IF IT DOES NOT EXIST     \n",
    "d = os.path.dirname(CSV_output_stem)\n",
    "if not os.path.exists(d):\n",
    "        os.makedirs(d)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### LOAD PROCESSED DATAFRAME"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "h5_data = H5(DATA_PATH+FN)\n",
    "h5_data.ls()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SPECIFY COLORMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- DEFINE ALL COLORMAPS UPFRONT FOR UNIFORMITY ---\n",
    "\n",
    "# CATEGORICAL: SAMPLE\n",
    "FLATUI_SAMPLE = [ \n",
    "                 'FF0000', #'UM01A'\n",
    "                 'CC0066', #'UM01B'\n",
    "                 'CC00CC', #'UM02'\n",
    "                 '0000CD', #'UM03'\n",
    "                 'FF8000', #'UM04'\n",
    "                 '660066', #'UM05'\n",
    "                 '7F00FF', #'UM06'\n",
    "]\n",
    "\n",
    "FLATUI_PATIENT = [ \n",
    "                 'DC143C', #'UM01'\n",
    "                 'C71585', #'UM02'\n",
    "                 '0000CD', #'UM03'\n",
    "                 'FF1493', #'UM04'\n",
    "                 'FF4500', #'UM05'\n",
    "                 '330066', #'UM06'\n",
    "]\n",
    "\n",
    "FLATUI_CASTLE = [\n",
    "                 '0000FF', # 1\n",
    "                 'FF0000', # 2\n",
    "]\n",
    "\n",
    "FLATUI_STAGE = [\n",
    "                 'C0C0C0', # 1\n",
    "                 '000000', # 2\n",
    "]\n",
    "\n",
    "FLATUI_TYPE = [\n",
    "                 '000099', # Epithelioid\n",
    "                 '990099', # Mixed\n",
    "                 '990000', # Spindle\n",
    "]\n",
    "\n",
    "FLATUI_VASCULO = [\n",
    "                 'DC143C', # Absent\n",
    "                 'B0C4DE', # Present\n",
    "]\n",
    "\n",
    "FLATUI_CLASS = [\n",
    "                 '0000FF', # 0\n",
    "                 'FF0000', # 1   \n",
    "                 'FF8C00', # 2\n",
    "                 'FFD700', # 3  \n",
    "                 '808000', # 4\n",
    "                 '9ACD32', # 5   \n",
    "                 '006400', # 6\n",
    "                 '20B2AA', # 7  \n",
    "                 '2F4F4F', # 8\n",
    "                 '00BFFF', # 9   \n",
    "                 '1E90FF', # 10\n",
    "                 '191970', # 11 \n",
    "                 '4B0082', # 12\n",
    "                 '8B008B', # 13  \n",
    "                 'FF00FF', # 14\n",
    "                 'FF1493', # 15 \n",
    "                 '8B4513', # 16\n",
    "                 'D2691E', # 17 \n",
    "                 '708090', # 18\n",
    "                 'B0C4DE', # 19  \n",
    "                 '696969', # 20\n",
    "                 'FFB6C1', # 21\n",
    "                 'FAEBD7', # 22\n",
    "                 'F5DEB3', # 23\n",
    "                 'FFFACD', # 24\n",
    "                 'D2B48C', # 25\n",
    "                 'FFA07A', # 26                \n",
    "]\n",
    "\n",
    "FLATUI_META_CELL_TYPE = [\n",
    "                 'B0C4DE', # IMMUNE\n",
    "                 'FFA07A', # PHOTORECEPTOR  \n",
    "                 '000000', # TUMOR \n",
    "]\n",
    "\n",
    "# CONVERT HEX TO RGB (FLATUI_SAMPLE)\n",
    "colors = np.zeros((len(FLATUI_META_CELL_TYPE),3))\n",
    "for ind,hexcolor in enumerate(FLATUI_META_CELL_TYPE):\n",
    "    colors[ind,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "CM_META_CELL_TYPE = LinearSegmentedColormap.from_list('FLATUI_META_CELL_TYPE', colors, N=len(colors))\n",
    "\n",
    "# CONVERT HEX TO RGB (FLATUI_SAMPLE)\n",
    "colors = np.zeros((len(FLATUI_SAMPLE),3))\n",
    "for ind,hexcolor in enumerate(FLATUI_SAMPLE):\n",
    "    colors[ind,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "CM_SAMPLES = LinearSegmentedColormap.from_list('FLATUI_SAMPLE', colors, N=len(colors))\n",
    "\n",
    "# CONVERT HEX TO RGB (FLATUI_CASTLE)\n",
    "colors = np.zeros((len(FLATUI_CASTLE),3))\n",
    "for ind,hexcolor in enumerate(FLATUI_CASTLE):\n",
    "    colors[ind,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "CM_CASTLE = LinearSegmentedColormap.from_list('FLATUI_CASTLE', colors, N=len(colors))\n",
    "\n",
    "# CONVERT HEX TO RGB (FLATUI_PATIENT)\n",
    "colors = np.zeros((len(FLATUI_PATIENT),3))\n",
    "for ind,hexcolor in enumerate(FLATUI_PATIENT):\n",
    "    colors[ind,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "CM_PATIENT = LinearSegmentedColormap.from_list('FLATUI_PATIENT', colors, N=len(colors))\n",
    "\n",
    "# CONVERT HEX TO RGB (FLATUI_STAGE)\n",
    "colors = np.zeros((len(FLATUI_STAGE),3))\n",
    "for ind,hexcolor in enumerate(FLATUI_STAGE):\n",
    "    colors[ind,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "CM_STAGE = LinearSegmentedColormap.from_list('FLATUI_STAGE', colors, N=len(colors))\n",
    "\n",
    "# CONVERT HEX TO RGB (FLATUI_SMOKING)\n",
    "colors = np.zeros((len(FLATUI_TYPE),3))\n",
    "for ind,hexcolor in enumerate(FLATUI_TYPE):\n",
    "    colors[ind,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "CM_TYPE = LinearSegmentedColormap.from_list('FLATUI_TYPE', colors, N=len(colors))\n",
    "\n",
    "# CONVERT HEX TO RGB (FLATUI_TREATMENT)\n",
    "colors = np.zeros((len(FLATUI_VASCULO),3))\n",
    "for ind,hexcolor in enumerate(FLATUI_VASCULO):\n",
    "    colors[ind,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "CM_VASCULO = LinearSegmentedColormap.from_list('FLATUI_VASCULO', colors, N=len(colors))\n",
    "\n",
    "# CONVERT HEX TO RGB (FLATUI_CLASS)\n",
    "colors = np.zeros((len(FLATUI_CLASS),3))\n",
    "for ind,hexcolor in enumerate(FLATUI_CLASS):\n",
    "    colors[ind,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "CM_CLASS = LinearSegmentedColormap.from_list('FLATUI_CLASS', colors, N=len(colors))\n",
    "\n",
    "# CONTINUOUS: DIVERGING\n",
    "CM_DIVERGING = plt.cm.RdBu_r\n",
    "\n",
    "# CONTINUOUS: SEQUENTIAL\n",
    "CM_SEQUENTIAL = cmocean.cm.thermal "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CANCER CELLS ONLY"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### SPECIFY SUBSET TYPE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_type = 'TUMOR'\n",
    "\n",
    "exec('DF_{} = h5_data.load(\\'/DF_{}\\')'.format(subset_type,subset_type))\n",
    "exec('NDF_{} = h5_data.load(\\'/NDF_{}\\')'.format(subset_type,subset_type))\n",
    "exec('INDF_{} = h5_data.load(\\'/INDF_{}\\')'.format(subset_type,subset_type))\n",
    "exec('METADATA_{} = h5_data.load(\\'/METADATA_{}\\')'.format(subset_type,subset_type))\n",
    "exec('DIMENSIONS_{} = h5_data.load(\\'/DIMENSIONS_{}\\')'.format(subset_type,subset_type))\n",
    "exec('GENE_RANK_MAST_{} = h5_data.load(\\'/GENE_RANK_MAST_{}\\')'.format(subset_type,subset_type))\n",
    "exec('ARCHETYPE_DISTANCES_{} = h5_data.load(\\'/ARCHETYPE_DISTANCES_{}\\')'.format(subset_type,subset_type))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### REPORT LIBRARY SIZE DISTRIBUTION OF SUBSET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PRINT LIBRARY SIZE AND CELL COUNT STATISTICS\n",
    "exec('median_libsize= DF_{}.sum(axis=1).median()'.format(subset_type))\n",
    "exec('num_cells = len(DF_{}.sum(axis=1))'.format(subset_type))\n",
    "\n",
    "print('All Samples, Count: {}, Median LibSize: {}'.format(num_cells, median_libsize))\n",
    "\n",
    "exec('QUERY = DF_{}'.format(subset_type))\n",
    "for index in np.unique(QUERY.index.get_level_values('Legend')):\n",
    "    tmp = QUERY.loc[QUERY.index.map(lambda x: x[1] == index)]\n",
    "    median_libsize= tmp.sum(axis=1).median()\n",
    "    num_cells = len(tmp.sum(axis=1))\n",
    "    print('Sample ID: {}, Count: {}, Median LibSize: {}'.format(index, num_cells, median_libsize))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exec('QUERY = DF_{}'.format(subset_type))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "\n",
    "plt.figure(figsize = (6,6))\n",
    "gs1 = gridspec.GridSpec(2,2)\n",
    "gs1.update(wspace=0.5, hspace=0.7) # set the spacing between axes. \n",
    "\n",
    "# (1) PLOT LOG LIBRARY SIZE\n",
    "ax = plt.subplot(gs1[0])\n",
    "bins = np.linspace(np.log10(QUERY.values.sum(axis=1).min()), np.log10(QUERY.values.sum(axis=1).max())*0.95, 100)\n",
    "for ind, label in enumerate(np.unique(QUERY.index.get_level_values('Legend'))):\n",
    "    x = np.log10(QUERY.loc[QUERY.index.map(lambda x: x[1] == label)].values.sum(axis=1)) # log cell size\n",
    "    plt.hist(x, bins, alpha=0.5, label=label,color = np.divide(tuple(hex(FLATUI_SAMPLE[ind]).rgb),255)) \n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Frequency')\n",
    "plt.xlabel('\\nLog. Cell Size')\n",
    "sns.despine()\n",
    "\n",
    "# (2) PLOT LOG NUM UNIQUE GENES\n",
    "ax = plt.subplot(gs1[1])\n",
    "bins = np.linspace(np.log10((np.sum(QUERY > 0,axis=1)).min()), \n",
    "                   np.log10((np.sum(QUERY > 0,axis=1)).max())*0.95, 100)\n",
    "for ind, label in enumerate(np.unique(QUERY.index.get_level_values('Legend'))):\n",
    "    x = np.log10(np.sum(QUERY.loc[QUERY.index.map(lambda x: x[1] == label)] > 0,axis=1))\n",
    "    plt.hist(x, bins, alpha=0.5, label=label,color = np.divide(tuple(hex(FLATUI_SAMPLE[ind]).rgb),255)) \n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Frequency')\n",
    "plt.xlabel('\\nLog. Number unique genes')\n",
    "sns.despine()\n",
    "\n",
    "# (3) PLOT LOG NUMBER OF CELLS CONTRIBUTING TO EACH GENE\n",
    "ax = plt.subplot(gs1[2])\n",
    "data = np.log10(np.sum(QUERY.values > 0,axis=0))\n",
    "bins = np.linspace(0, data.max()*0.95, 20)\n",
    "for ind, label in enumerate(np.unique(QUERY.index.get_level_values('Legend'))):\n",
    "    x = np.log10(np.sum(QUERY.loc[QUERY.index.map(lambda x: x[1] == label)] > 0,axis=0))\n",
    "    x[(np.isinf(x)) | (np.isnan(x))] = 0\n",
    "    plt.hist(x, bins, alpha=0.5, label=label,color = np.divide(tuple(hex(FLATUI_SAMPLE[ind]).rgb),255)) \n",
    "    \n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Frequency')\n",
    "plt.xlabel('\\nLog. Cells contributing to each gene')\n",
    "sns.despine()\n",
    "\n",
    "# (4) PLOT MITOCHONDRIAL FRACTION PER CELL \n",
    "ax = plt.subplot(gs1[3]) \n",
    "vals = META['mt_fraction']\n",
    "vals = np.log10(META['mt_fraction'].astype(float))\n",
    "vals[(np.isinf(vals)) | (np.isnan(vals))] = 0\n",
    "\n",
    "bins = np.linspace(vals.min(),vals.max()*0.95, 20)\n",
    "for ind, label in enumerate(np.unique(QUERY.index.get_level_values('Legend'))):\n",
    "    x = vals.loc[vals.index.map(lambda x: x[1] == label)]\n",
    "    x[(np.isinf(x)) | (np.isnan(x))] = 0\n",
    "    plt.hist(x, bins, alpha=0.5, label=label,color = np.divide(tuple(hex(FLATUI_SAMPLE[ind]).rgb),255)) \n",
    "    \n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Frequency')\n",
    "plt.xlabel('\\nLog. Mitochondrial Fraction')\n",
    "sns.despine()\n",
    "\n",
    "# Add (abbreviated) legend bottom left\n",
    "L = plt.legend(loc='upper right',prop={'size':6},bbox_to_anchor=(1.8, 0.95),fancybox=True) \n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '_library_distribution_subset_{}_log10'.format(subset_type)\n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label + '.png'\n",
    "plt.savefig(fn, dpi=fig_dpi)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### FILTER LOW ABUNDANCE GENES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exec('QUERY = DF_{}'.format(subset_type))\n",
    "\n",
    "# FIT BINOMIAL DISTRIBUTION AND FILTER BASED ON MEAN/STD OF SECOND\n",
    "plt.figure(figsize = (10,3))\n",
    "gs1 = gridspec.GridSpec(1, 2)\n",
    "gs1.update(wspace=0.7, hspace=0.7) # set the spacing between axes. \n",
    "\n",
    "# (4) PLOT LOG NUMBER OF CELLS CONTRIBUTING TO EACH GENE\n",
    "num_cells_per_gene = np.log(np.sum(QUERY.values > 0,axis=0))\n",
    "num_cells_per_gene[(np.isinf(num_cells_per_gene)) | (np.isnan(num_cells_per_gene))] = 0\n",
    "rmv_genes1 = np.where(num_cells_per_gene<=0)[0] # GENES MUST BE DETECTED IN AT LEAST 10 CELLS\n",
    "\n",
    "ax = plt.subplot(gs1[0])\n",
    "bins = np.linspace(num_cells_per_gene.min(), num_cells_per_gene.max()*0.95, 20)\n",
    "plt.hist(num_cells_per_gene, bins, alpha=0.5, label='keep')\n",
    "\n",
    "if rmv_genes1.any():\n",
    "    plt.hist(num_cells_per_gene[rmv_genes1], bins, alpha=1, label='remove')\n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Frequency')\n",
    "plt.xlabel('Gene Filter 1: Remove Genes Singletons \\n(Log. # Expressing Cells)')\n",
    "sns.despine()\n",
    "\n",
    "# (5) PLOT LOG COUNTS PER GENE - REMOVE LOW ABUNDANCE GENES\n",
    "log_counts_per_gene = np.log(np.log(np.sum(QUERY.values,axis=0)))\n",
    "log_counts_per_gene[(np.isinf(log_counts_per_gene)) | (np.isnan(log_counts_per_gene))] = 0\n",
    "data = log_counts_per_gene\n",
    "\n",
    "ax = plt.subplot(gs1[1])\n",
    "bins = np.linspace(data.min(), data.max()*0.95, 100)\n",
    "y,x,_=hist(data,bins,alpha=.3,label='data')\n",
    "\n",
    "x=(x[1:]+x[:-1])/2 \n",
    "expected=(0,.2,3500,1.5,.2,500)\n",
    "params,cov=curve_fit(bimodal,x,y,expected)\n",
    "sigma=sqrt(diag(cov))\n",
    "plot(x,bimodal(x,*params),color='red',lw=3,label='model')\n",
    "\n",
    "mu1 = params[0]\n",
    "std1 = params[1]\n",
    "mu2 = params[3]\n",
    "std2 = params[4]\n",
    "rmv_genes_neg = np.where(data<mu2-4*std2)[0]\n",
    "rmv_genes2 = np.sort(list(set(list(rmv_genes_neg))))\n",
    "\n",
    "if rmv_genes2.any():\n",
    "    plt.hist(data[rmv_genes2], bins, alpha=1, label='remove')\n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Frequency')\n",
    "plt.xlabel('Gene Filter 2: \\nLog-Log Counts/Gene')\n",
    "sns.despine()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '_filter_genes'\n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label + '.png'\n",
    "plt.savefig(fn, dpi=fig_dpi)\n",
    "print(fn)\n",
    "\n",
    "# EVALUATE NUMBER OF CELLS/GENES REMOVED\n",
    "print('Count Gene Filter 1: {}'.format(len(rmv_genes1)))\n",
    "print('Count Gene Filter 2: {}'.format(len(rmv_genes2)))\n",
    "\n",
    "# REMOVE SELECTED OUTLIER GENES(HIGHLIGHTED IN GREEN)\n",
    "CUT_DF = deepcopy(QUERY)\n",
    "rmv_genes = np.sort(list(set(list(rmv_genes1) + list(rmv_genes2))))\n",
    "if rmv_genes.any():\n",
    "    CUT_DF = CUT_DF.drop(CUT_DF.columns[rmv_genes],axis=1)\n",
    "    print(CUT_DF.shape)\n",
    "    \n",
    "# Remove empty genes if they exist\n",
    "drop_genes = np.where(CUT_DF.sum(axis=0)==0)[0]\n",
    "CUT_DF = CUT_DF.drop(CUT_DF.columns[drop_genes],axis=1)\n",
    "print(CUT_DF.shape)\n",
    "\n",
    "# GENES EXCLUDING LOW ABUNDANCE GENES\n",
    "goi = list(CUT_DF.columns)\n",
    "print(len(goi))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### LOAD KEY GENESETS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LOAD GENE LIST FROM EXCEL\n",
    "path_to_genesets = DATA_PATH + 'uveal_melanoma_MB_v3.csv'\n",
    "genesets = pd.read_csv(path_to_genesets,header='infer')\n",
    "genesets = genesets.apply(lambda x: x.astype(str).str.upper())\n",
    "print(shape(genesets)[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### EVALUATE GENE EXPRESSION PER META CLASS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_type = 'box'\n",
    "title = 'BAP1'\n",
    "meta = 'Assignment'\n",
    "datatype = 'INDF_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "\n",
    "fig = plt.figure(figsize = (1,5))\n",
    "ax = plt.gca()\n",
    "\n",
    "scale_type = 'count'\n",
    "palette = dict(zip(['Castle 1','Castle 2'],['#0000FF','#FF0000'])) \n",
    "\n",
    "\n",
    "genes = [title]\n",
    "genes = np.unique([x for x in genes if str(x) != 'NAN'])\n",
    "detected_genes = list(set(genes).intersection(set(QUERY.columns)))\n",
    "vals = QUERY[detected_genes]\n",
    "SCORE = np.nansum(vals,axis=1)\n",
    "\n",
    "# Format data structure for violin plot\n",
    "violin_data = []\n",
    "for ind,v in enumerate(SCORE):\n",
    "    violin_data.append({'gene': title, 'Z-normalized Expression': v,\n",
    "                        meta:META[meta].values[ind]}) \n",
    "violin_data = pd.DataFrame(violin_data)  \n",
    "\n",
    "# BOXPLOT GENE EXPRESSION\n",
    "if plot_type == 'box':\n",
    "    g = sns.boxplot(x=\"gene\", y=\"Z-normalized Expression\", hue=meta,data=violin_data, palette=palette,notch = True, \n",
    "                    hue_order = ['Castle 1','Castle 2'],\n",
    "                    fliersize = 4, showmeans=False,linewidth = 1, ax = ax) #order = labels, \n",
    "    g.set_ylabel(\"{}\".format(datatype),fontsize=10)\n",
    "    g.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "    g.tick_params(labelsize=10)\n",
    "    sns.despine()\n",
    "    ax.set(ylim=(0, SCORE.max()*0.95))\n",
    "    g.legend(loc='upper right',prop={'size':6},bbox_to_anchor=(2.0, 0.95),fancybox=True) \n",
    "\n",
    "\n",
    "elif plot_type == 'violin':\n",
    "    # VIOLIN GENE EXPRESSION\n",
    "    g = sns.violinplot(x=\"gene\", y=\"Z-normalized Expression\", hue=meta,data=violin_data, palette=palette,notch = True,\n",
    "                    hue_order = ['Castle 1','Castle 2'],fliersize = 4, showmeans=True,linewidth = 1) #order = labels, \n",
    "\n",
    "    g.set_ylabel(\"{}\".format(datatype),fontsize=10)\n",
    "    g.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "    ax.set(ylim=(0, SCORE.max()*0.95))\n",
    "    g.tick_params(labelsize=10)\n",
    "    sns.despine()\n",
    "\n",
    "# COMPARE DISTRIBUTIONS\n",
    "CLASS1 = violin_data.loc[violin_data[meta].isin(['Castle 1'])]['Z-normalized Expression'].values\n",
    "CLASS2 = violin_data.loc[violin_data[meta].isin(['Castle 2'])]['Z-normalized Expression'].values\n",
    "print(title + ' CLASS1 vs. CLASS2')\n",
    "print(stats.mannwhitneyu(CLASS1,CLASS2))\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '_{}_{}PLOT_{}_{}_{}'.format(subset_type, plot_type,datatype,meta,title)\n",
    "fn = FIG_output_stem + '/BOXPLOTS/' + FN.replace(\".h5\", \"\") + figure_label\n",
    "    \n",
    "# CREATE GSEA DIRECTORY IF IT DOES NOT EXIST\n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "        \n",
    "plt.savefig(fn + '.png', bbox_inches='tight',dpi=fig_dpi)\n",
    "plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=fig_dpi)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TUMOR CELL CLUSTERS AND INFERRED CASTLE CLASS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### VISUALIZE PHENOGRAPH CLUSTERS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# PLOT CATEGORICAL CLUSTER ASSIGNMENTS\n",
    "# Note: Any rows assigned -1 were identified as outliers and should not be considered as a member of any community.\n",
    "plt.figure(figsize = (8,3))\n",
    "nrow = 1\n",
    "ncol = 2\n",
    "dot_size = 1\n",
    "gs1 = gridspec.GridSpec(nrow, ncol)\n",
    "gs1.update(wspace=0.1, hspace=0.1) # set the spacing between axes. \n",
    "\n",
    "dimtype = 'ForceDirected' \n",
    "meta = 'Phenograph_Class'\n",
    "exec('DIM = DIMENSIONS_{}'.format(subset_type))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "exec('QUERY = INDF_{}'.format(subset_type))\n",
    "x = DIM['{}0'.format(dimtype)]\n",
    "y = DIM['{}1'.format(dimtype)]\n",
    "axis_min = -ceil(ceil(max(np.abs(x).max(),np.abs(y).max()))/10)*10\n",
    "axis_max = ceil(ceil(max(np.abs(x).max(),np.abs(y).max()))/10)*10\n",
    "\n",
    "# Cluster Assignment\n",
    "ax = plt.subplot(gs1[0])\n",
    "seqc.plot.scatter.categorical(x, y, c=META[meta].values, \n",
    "                              cmap= CM_CLASS,\n",
    "                              legend_kwargs={'ncol': 1}, s=dot_size, ax=ax,randomize=True);\n",
    "plt.rc('xtick', labelsize=6)\n",
    "plt.rc('ytick', labelsize=6)\n",
    "sns.despine()\n",
    "plt.xlabel('tSNE-1', fontname='Helvetica', size=10, weight='normal')\n",
    "plt.ylabel('tSNE-2', fontname='Helvetica', size=10, weight='normal')\n",
    "plt.xlim(axis_min,axis_max)\n",
    "plt.ylim(axis_min,axis_max)\n",
    "plt.title('{}'.format(meta), fontname='Helvetica', size=12, weight='normal')\n",
    "plt.axis('off')\n",
    "ax.legend_.remove()\n",
    "\n",
    "# Castle Class\n",
    "ax = plt.subplot(gs1[1])\n",
    "seqc.plot.scatter.categorical(x, y, c=META['Assignment'], cmap=CM_CASTLE,legend_kwargs={'ncol': 1}, s=dot_size, ax=ax);\n",
    "sns.despine()\n",
    "plt.xlabel('tSNE-1', fontname='Helvetica', size=10, weight='normal')\n",
    "plt.ylabel('tSNE-2', fontname='Helvetica', size=10, weight='normal')\n",
    "plt.title('Sample ID', fontname='Helvetica', size=12, weight='normal')\n",
    "plt.xlim(axis_min,axis_max)\n",
    "plt.ylim(axis_min,axis_max)\n",
    "plt.axis('off')\n",
    "ax.legend_.remove()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '_CLUSTERS_INDF_{}_{}_{}'.format(subset_type,meta,dimtype) \n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label \n",
    "plt.savefig(fn + '.png', bbox_inches='tight',dpi=fig_dpi)\n",
    "plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=fig_dpi)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MEAN EXPRESSION OF KEY GENES PER CLUSTER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta = 'Phenograph_Class'\n",
    "cluster_mean = QUERY.groupby(level = [meta],axis=0).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "genesets.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "master_title = ['CIN_signature_Bakhoum','CIN_responsive_NC_NFKB_UP_Bakhoum',\n",
    "                'CIN_responsive_NC_NFKB_DN_Bakhoum', 'NC_NFKB_REGULATORS_UP_Bakhoum',\n",
    "                'NC_NFKB_REGULATORS_DN_Bakhoum']\n",
    "class_genes = genesets.loc[:,master_title]\n",
    "class_genes = [gene for gene in class_genes.values.flatten() if gene not in ['NAN']]\n",
    "matster_title = 'NC_NFKB'\n",
    "print(len(class_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "master_title = 'MANUAL'\n",
    "class_genes = ['JARID2','TING1','TMEM173','BAP1','LPCAT1','HLA-A','HLA-B','HLA-C','HLA-E',\n",
    "              'PMEL','MLANA','TYR','TYRP1','DCT','RING1']\n",
    "print(len(class_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta = 'Phenograph_Class'\n",
    "FLATUI_PLOT = FLATUI_CLASS\n",
    "\n",
    "datatype = 'INDF_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "QUERY = QUERY.groupby(level = [meta],axis=0).median()\n",
    "\n",
    "# GENERATE LUT FOR PHENOGRAPH CLSUTERS BEFORE DOWNSAMPLING MATRIX\n",
    "# SPECIFY GENESET AND TITLE\n",
    "ind = QUERY.index\n",
    "# CONVERT HEX TO RGB (FLATUI_CLASS)\n",
    "colors = np.zeros((len(FLATUI_PLOT),3))\n",
    "for ii,hexcolor in enumerate(FLATUI_PLOT):\n",
    "    colors[ii,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "# Palatte for Class METADATA\n",
    "cix = (np.linspace(0,shape(colors)[0],len(np.unique(ind)))).astype(int)\n",
    "if cix[len(cix)-1]==shape(colors)[0]:\n",
    "    cix[len(cix)-1]=shape(colors)[0]-1\n",
    "lut = dict(zip(np.unique(ind), colors[cix,:]))\n",
    "row_colors = pd.Series(ind).map(lut)\n",
    "\n",
    "# ONLY DISPLAY METASTASIS ASSOCIATED CLUSTERS\n",
    "subset_key = meta\n",
    "ix = [ind for ind, name in enumerate(QUERY.index.names) if name==subset_key][0]\n",
    "\n",
    "genes = class_genes\n",
    "genes = [gene for gene in genes if gene in goi]\n",
    "genes = np.unique(genes)\n",
    "\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "heatmap_data = pd.DataFrame(data = zscore(QUERY[genes].values,axis=0),columns = genes)\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns\n",
    "\n",
    "# LINKAGE \n",
    "method = 'average' # average, single centroid/euclidean\n",
    "metric = 'euclidean' # cosine\n",
    "linkage = hc.linkage(heatmap_data, method=method, metric = metric)\n",
    "row_linkage = deepcopy(linkage)\n",
    "linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "col_linkage = deepcopy(linkage)\n",
    "\n",
    "# REORDER HEATMAP ACCORDING TO LINKAGE (OPTIONAL, STILL SLOW)\n",
    "r1 = hc.leaves_list(row_linkage)\n",
    "cl = hc.leaves_list(col_linkage)\n",
    "mat = heatmap_data.iloc[r1,cl]\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(10,5))\n",
    "\n",
    "# ADD ROW COLOR INDEX (CELL OF ORIGIN)\n",
    "ax1 = fig.add_axes([0.18,0.1,0.03,0.6]) # [x0,y0,width,height]\n",
    "x = 0\n",
    "y = 0\n",
    "for c in row_colors[r1]:\n",
    "    pos = (x, y / len(r1))\n",
    "    ax1.add_patch(patches.Rectangle(pos, 1, 1 / len(r1), color=c))\n",
    "    if y >= len(r1)-1:\n",
    "        x += 1\n",
    "        y = 0\n",
    "    else:\n",
    "        y += 1\n",
    "plt.axis('off')\n",
    "\n",
    "# ADD MATRIX WITH GENE NAMES\n",
    "axmatrix = fig.add_axes([0.2,0.1,0.7,0.6])\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=plt.cm.RdBu_r,vmin=-3,vmax=3)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "xtick = plt.xticks(range(len(labels)), labels, rotation = 90, fontsize = 12,fontname='Arial')\n",
    "axmatrix.grid(False)\n",
    "\n",
    "# ADD DENDROGRAM\n",
    "sch.set_link_color_palette(['#808080', '#808080', '#808080', '#808080','#808080','#808080','#808080'])\n",
    "ax2 = fig.add_axes([0.105,0.1,0.08,0.6]) # [x0,y0,width,height]\n",
    "Z1 = sch.dendrogram(row_linkage, orientation='left',above_threshold_color='#808080')\n",
    "ax2.set_xticks([])\n",
    "ax2.set_yticks([])\n",
    "plt.axis('off')\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'heatmaps/user_defined/{}_{}_HEATMAP_{}_{}_colored_{}'.format(datatype,master_title,method,metric,meta).replace('-','_').replace(' ','_')\n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MEAN EXPRESSION OF KEY GENES PER CELL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "QUERY.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# VIOLIN PLOTS\n",
    "subset_type = 'TUMOR'\n",
    "datatype = 'INDF_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "\n",
    "np.unique(QUERY.index.get_level_values('Archetype'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CLASS RELATED GENES\n",
    "class_genes = genesets.loc[:,['Castle 2', 'Castle 1']]\n",
    "class_genes = [gene for gene in class_genes.values.flatten() if gene not in ['NAN']]\n",
    "print(len(class_genes))\n",
    "\n",
    "# COLUMN INDEX AND COLORS\n",
    "genes = [gene for gene in class_genes if gene in QUERY.columns]#gene_array['Gene'].values\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "heatmap_data = pd.DataFrame(data = zscore(QUERY[genes].values,axis=0),columns = genes, index = QUERY.index)\n",
    "print(heatmap_data.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ROW INDEX\n",
    "meta = 'Archetype'\n",
    "\n",
    "FLATUI_PLOT = [\n",
    "                 'E6E6FA', # not assgined\n",
    "                 'FF66FF', # pink 0 \n",
    "                 '6600CC', # purple 1\n",
    "                 '9ACD32', # Green 2\n",
    "                 '00FFFF', # skye blue 3\n",
    "                 'FF8800', # orange 4\n",
    "                 '0000FF', # dark blue 5\n",
    "                 'FFD700', # yellow 6\n",
    "                 '20B2AA', # teal 7 \n",
    "                 '008000', # forest green 8\n",
    "                 '8B008B', # 9\n",
    "    ]\n",
    "\n",
    "ind = META.index.get_level_values(meta)\n",
    "\n",
    "# CONVERT HEX TO RGB (FLATUI_CLASS)\n",
    "colors = np.zeros((len(FLATUI_PLOT),3))\n",
    "for ii,hexcolor in enumerate(FLATUI_PLOT):\n",
    "    colors[ii,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "# Palatte for Class METADATA\n",
    "cix = (np.linspace(0,shape(colors)[0],len(np.unique(ind)))).astype(int)\n",
    "if cix[len(cix)-1]==shape(colors)[0]:\n",
    "    cix[len(cix)-1]=shape(colors)[0]-1\n",
    "lut = dict(zip(np.unique(ind), colors[cix,:]))\n",
    "row_colors = pd.Series(ind).map(lut)\n",
    "\n",
    "# COLUMN INDEX AND COLORS\n",
    "genes = [gene for gene in class_genes if gene in QUERY.columns]#gene_array['Gene'].values\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "heatmap_data = pd.DataFrame(data = zscore(QUERY[genes].values,axis=0),columns = genes, index = QUERY.index)\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmap_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LINKAGE \n",
    "method = 'average' # average, single\n",
    "metric = 'euclidean' # cosine\n",
    "linkage = hc.linkage(heatmap_data, method=method, metric = metric)\n",
    "row_linkage = deepcopy(linkage)\n",
    "linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "col_linkage = deepcopy(linkage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "QUERY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# REORDER HEATMAP ACCORDING TO LINKAGE (OPTIONAL, STILL SLOW)\n",
    "r1 = hc.leaves_list(row_linkage)\n",
    "c1 = hc.leaves_list(col_linkage)\n",
    "mat = heatmap_data.iloc[r1,c1]\n",
    "\n",
    "row_colors = pd.Series(mat.index.get_level_values(meta),index =mat.index.get_level_values(meta)).map(lut)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(4,10))\n",
    "plt.rcParams[\"axes.grid\"] = False\n",
    "\n",
    "# ADD ROW COLOR INDEX 1 (PHENOGRAPH CLASS)\n",
    "ax1 = fig.add_axes([0,0.1,0.05,0.6]) # [x0,y0,width,height]\n",
    "x = 0\n",
    "y = 0\n",
    "for c in row_colors:\n",
    "    pos = (x, y / len(row_colors))\n",
    "    ax1.add_patch(patches.Rectangle(pos, 1, 1 / len(row_colors), color=c))\n",
    "    if y >= len(row_colors)-1:\n",
    "        x += 1\n",
    "        y = 0\n",
    "    else:\n",
    "        y += 1\n",
    "plt.axis('off')\n",
    "\n",
    "# ADD MATRIX WITH GENE NAMES\n",
    "axmatrix = fig.add_axes([0.05,0.1,0.9,0.6])\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=plt.cm.RdBu_r,vmin=-1,vmax=1)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "xtick = plt.xticks(range(len(labels)), labels, rotation = 90, fontsize = 14)\n",
    "\n",
    "# ADD COLORBAR\n",
    "axcolor = fig.add_axes([1.1,0.1,0.1,0.1])\n",
    "cbar = plt.colorbar(im, cax=axcolor)\n",
    "#cbar.ax.get_yaxis().set_ticks([])\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'CASTLE_GENES'.format(meta,subset_type,method,metric)\n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ARCHETYPAL ANALYSIS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_type"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### IDENTIFY OPTIMAL NUMBER OF ARCHETYPES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Diffusion Eigenvalues\n",
    "component_prefix = 'IPC'\n",
    "exec('QUERY = DIMENSIONS_{}'.format(subset_type))\n",
    "idx = QUERY.columns.to_series().str.contains('^{}((?!TSNE).)*$'.format(component_prefix)) # select from DIMS_NDF\n",
    "exec('IPC = DIMENSIONS_{}.loc[:, idx.values]'.format(subset_type))\n",
    "IPC = IPC.iloc[:,:22]\n",
    "IPC.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip freeze"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# IDENTIFY OPTIMAL NUMBER OF ARCHETYPES BASED ON KNEE POINT OF THE MODEL EXPALINED VARIANCE\n",
    "from py_pcha.PCHA import PCHA\n",
    "n = 20\n",
    "SSE = [0] * n\n",
    "varexpl =[0]*n\n",
    "for noc in np.arange(n):\n",
    "    #SSE[noc] = noc\n",
    "    XC, S, C, SSE[noc], varexpl[noc] = PCHA(IPC.T.values, noc=noc+1, delta=0.05)\n",
    "    print('noc: {}, SSE: {}, EXPVAR: {}'.format(noc, SSE[noc], varexpl[noc]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "varexpl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# IDENTIFY POINT OF MAXIMUM CURVATURE IN CUMULATIVE EXPLAINED VARIANCE\n",
    "y = varexpl\n",
    "x = np.arange(len(y))+1\n",
    "kneedle = KneeLocator(x, y)\n",
    "kneedle.plot_knee_normalized()\n",
    "knee = ncomponents = kneedle.knee\n",
    "plt.title('knee: {}'.format(knee))\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '_{}_KNEEPOINT_ARCHETYPE_EXPLAINED_VARIANCE'.format(subset_type)\n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label \n",
    "plt.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### COMPUTE ARCHETYPES USING OPTIMAL NUMBER OF ARCHETYPES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute Archeytpes on Imputed PCS\n",
    "optimal_noc = 8\n",
    "from py_pcha.PCHA import PCHA\n",
    "XC, S, C, SSE, varexpl = PCHA(IPC.T.values, noc=optimal_noc, delta=0.05)\n",
    "archetypes = pd.DataFrame(data = np.squeeze(np.asarray(XC)).T, columns = IPC.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "archetypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### COMPUTE PAIRWISE DISTANCE BETWEEN ARCHETYPES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "distance = np.zeros([archetypes.shape[0],archetypes.shape[0]])\n",
    "for current_archetype_idx1 in np.arange(archetypes.shape[0]):\n",
    "    current_archetype1 = archetypes.loc[current_archetype_idx1,:]\n",
    "\n",
    "    for current_archetype_idx2 in np.arange(archetypes.shape[0]):\n",
    "        current_archetype2 = archetypes.loc[current_archetype_idx2,:]\n",
    "\n",
    "        distance[current_archetype_idx1,current_archetype_idx2] = \\\n",
    "        np.sqrt((current_archetype1-current_archetype2)**2).sum()\n",
    "\n",
    "distance = pd.DataFrame(data = distance, columns = np.arange(archetypes.shape[0]),\n",
    "                        index = np.arange(archetypes.shape[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clustered heatmap of euclidean distance between archetypes\n",
    "yticks = distance.index\n",
    "xticks = distance.columns\n",
    "\n",
    "# LINKAGE BASED ON NES OF SIGNIFICANT DIFFERENTIALLY EXPRESSED PATHWAYS\n",
    "method = 'average'\n",
    "metric = 'cosine'\n",
    "linkage = hc.linkage(distance, method=method, metric = metric)\n",
    "row_linkage = deepcopy(linkage)\n",
    "linkage = hc.linkage(distance.T, method=method, metric = metric)\n",
    "col_linkage = deepcopy(linkage)\n",
    "\n",
    "# REORDER HEATMAP ACCORDING TO LINKAGE (OPTIONAL, STILL SLOW)\n",
    "r1 = hc.leaves_list(row_linkage)\n",
    "cl = hc.leaves_list(col_linkage)\n",
    "mat = distance.iloc[r1,cl]\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(5,5))\n",
    "plt.rcParams[\"axes.grid\"] = False\n",
    "\n",
    "# ADD MATRIX WITH GENE NAMES\n",
    "axmatrix = fig.add_axes([0.1,0.1,0.6,0.6])\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=plt.cm.RdBu_r)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "xtick = plt.xticks(range(len(labels)), labels, rotation = 90, fontsize = 14)\n",
    "\n",
    "labels = list(mat.index)\n",
    "axmatrix.yaxis.set_ticks_position('left')\n",
    "ytick = plt.yticks(range(len(labels)), labels, rotation = 0, fontsize = 14)\n",
    "\n",
    "#axmatrix.set_yticklabels([]);\n",
    "\n",
    "# ADD COLORBAR\n",
    "axcolor = fig.add_axes([0.75,0.1,0.01,0.1])\n",
    "cbar = plt.colorbar(im, cax=axcolor)\n",
    "#cbar.ax.get_yaxis().set_ticks([])\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'euclidean_distance_btw_archetypes'\n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### FIND CELL NEAREST TO EACH ARCHETYPE BASED ON EUCLIDEAN DISTANCE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute distance between each cell and archetype\n",
    "archetype_distances = pd.DataFrame()\n",
    "for current_archetype_idx in np.arange(archetypes.shape[0]):\n",
    "    current_archetype = archetypes.loc[current_archetype_idx,:]\n",
    "    distances = np.sqrt(((current_archetype-IPC)**2).sum(axis = 1)).to_frame(name = 'Distance_Archetype_{}'.format(current_archetype_idx))\n",
    "    archetype_distances = pd.concat([archetype_distances, distances],axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### VISUALIZE CELL NEAREST TO EACH ARCHETYPE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.arange(len(archetype_distances.columns))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "list(genesets.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visuaulize distance to each archetype\n",
    "# PLOT CATEGORICAL CLUSTER ASSIGNMENTS\n",
    "# Note: Any rows assigned -1 were identified as outliers and should not be considered as a member of any community.\n",
    "plt.figure(figsize = (10,5))\n",
    "nrow = 1\n",
    "ncol = 2\n",
    "dot_size = 1\n",
    "gs1 = gridspec.GridSpec(nrow, ncol)\n",
    "gs1.update(wspace=0.01, hspace=0.1) # set the spacing between axes. \n",
    "\n",
    "cm = plt.cm.bwr \n",
    "dimtype = 'ForceDirected' \n",
    "meta = 'Phenograph_Class'\n",
    "exec('DIM = DIMENSIONS_{}'.format(subset_type))\n",
    "exec('QUERY= INDF_{}'.format(subset_type))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "x = DIM['{}0'.format(dimtype)]\n",
    "y = DIM['{}1'.format(dimtype)]\n",
    "axis_min = -ceil(ceil(max(np.abs(x).max(),np.abs(y).max()))/10)*10\n",
    "axis_max = ceil(ceil(max(np.abs(x).max(),np.abs(y).max()))/10)*10\n",
    "\n",
    "# MONOSOMY 3 SIGNATURE\n",
    "ax = plt.subplot(gs1[0])\n",
    "gene_sig = 'Monosomy 3 Up'\n",
    "signature_genes = genesets[gene_sig].values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "phase = zscore(np.nanmedian(QUERY[detected_genes],axis=1))\n",
    "# Plot scatter map with phase score\n",
    "mask = np.isnan(phase)\n",
    "    \n",
    "# Randomize/Sort Color Values before Plotting\n",
    "color_values = np.array(phase)\n",
    "i = np.random.permutation(len(color_values))\n",
    "xx = np.array(x)[i]\n",
    "yy = np.array(y)[i]\n",
    "color_values = color_values[i]\n",
    "\n",
    "plt.scatter(xx, yy, c = color_values, cmap = CM_DIVERGING, s=3, alpha =0.4)\n",
    "plt.clim(-1.5,1.5)\n",
    "plt.xlim(axis_min,axis_max)\n",
    "plt.ylim(axis_min,axis_max)\n",
    "plt.axis('off')\n",
    "plt.title('Monosomy 3', fontname='Helvetica', size=12, weight='normal')\n",
    "\n",
    "# BAP1 SIGNATURE\n",
    "ax = plt.subplot(gs1[1])\n",
    "gene_sig = 'Castle 2'\n",
    "signature_genes = genesets[gene_sig].values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "phase = zscore(np.nanmedian(QUERY[detected_genes],axis=1))\n",
    "# Plot scatter map with phase score\n",
    "mask = np.isnan(phase)\n",
    "    \n",
    "# Randomize/Sort Color Values before Plotting\n",
    "color_values = np.array(phase)\n",
    "i = np.random.permutation(len(color_values))\n",
    "xx = np.array(x)[i]\n",
    "yy = np.array(y)[i]\n",
    "color_values = color_values[i]\n",
    "\n",
    "plt.scatter(xx, yy, c = color_values, cmap = CM_DIVERGING, s=3, alpha =0.4)\n",
    "plt.clim(-1.5,1.5)\n",
    "plt.xlim(axis_min,axis_max)\n",
    "plt.ylim(axis_min,axis_max)\n",
    "plt.axis('off')\n",
    "plt.title('GEP 2', fontname='Helvetica', size=12, weight='normal')\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '_{}_MONOSOMY3_CASTLE2'.format(subset_type)\n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label \n",
    "plt.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FLATUI_ARCHETYPES = [\n",
    "                 'FF66FF', # pink\n",
    "                 '6600CC', # purple\n",
    "                 '9ACD32', # Green\n",
    "                 '00FFFF', # skye blue\n",
    "                 'FF8800', # orange\n",
    "                 '0000FF', # dark blue\n",
    "                 'FFD700', # yellow\n",
    "                 '20B2AA', # teal\n",
    "                 #'008000', # forest green\n",
    "                 #'8B008B'\n",
    "    ]\n",
    "\n",
    "# Visuaulize distance to each archetype\n",
    "# PLOT CATEGORICAL CLUSTER ASSIGNMENTS\n",
    "# Note: Any rows assigned -1 were identified as outliers and should not be considered as a member of any community.\n",
    "plt.figure(figsize = (20,10))\n",
    "nrow = 1\n",
    "ncol = 2\n",
    "dot_size = 1\n",
    "gs1 = gridspec.GridSpec(nrow, ncol)\n",
    "gs1.update(wspace=0.01, hspace=0.1) # set the spacing between axes. \n",
    "\n",
    "cm = plt.cm.bwr \n",
    "dimtype = 'ForceDirected' \n",
    "meta = 'Phenograph_Class'\n",
    "exec('DIM = DIMENSIONS_{}'.format(subset_type))\n",
    "exec('QUERY= INDF_{}'.format(subset_type))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "x = DIM['{}0'.format(dimtype)]\n",
    "y = DIM['{}1'.format(dimtype)]\n",
    "axis_min = -ceil(ceil(max(np.abs(x).max(),np.abs(y).max()))/10)*10\n",
    "axis_max = ceil(ceil(max(np.abs(x).max(),np.abs(y).max()))/10)*10\n",
    "\n",
    "# MONOSOMY 3 SIGNATURE\n",
    "ax = plt.subplot(gs1[0])\n",
    "gene_sig = 'Monosomy 3 Up'\n",
    "signature_genes = genesets[gene_sig].values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "phase = zscore(np.nanmedian(QUERY[detected_genes],axis=1))\n",
    "# Plot scatter map with phase score\n",
    "mask = np.isnan(phase)\n",
    "    \n",
    "# Randomize/Sort Color Values before Plotting\n",
    "color_values = np.array(phase)\n",
    "i = np.random.permutation(len(color_values))\n",
    "xx = np.array(x)[i]\n",
    "yy = np.array(y)[i]\n",
    "color_values = color_values[i]\n",
    "\n",
    "plt.scatter(xx, yy, c = color_values, cmap = CM_DIVERGING, s=3, alpha =0.4)\n",
    "plt.clim(-1.5,1.5)\n",
    "plt.xlim(axis_min,axis_max)\n",
    "plt.ylim(axis_min,axis_max)\n",
    "plt.axis('off')\n",
    "plt.title('Monosomy 3', fontname='Helvetica', size=12, weight='normal')\n",
    "\n",
    "# Distance to archeytpe\n",
    "#ax = plt.subplot(gs1[1])\n",
    "#archetype_distances['Archetype_Cluster'] = [0]*archetype_distances.shape[0]\n",
    "#plt.scatter(x,y,c = '#E6E6FA',s = 3)\n",
    "for ii in np.arange(len(archetype_distances.columns)):\n",
    "    c = np.log10(archetype_distances['Distance_Archetype_{}'.format(ii)])\n",
    "    #cc = c < c.mean()-2*c.std()\n",
    "    c = c==c.min()\n",
    "    #numcells = cc.sum()\n",
    "    #print(numcells)    \n",
    "    #text(x, y, s, fontsize=12)\n",
    "    plt.scatter(x[c.values],y[c.values],s= 300, color = '#'+FLATUI_ARCHETYPES[ii])\n",
    "    offset = 1000\n",
    "    plt.text(x[c.values]+offset,y[c.values]+offset,s = '{}'.format(ii), fontsize = 20)#numcells/10)\n",
    "    sns.despine()\n",
    "    plt.xlabel('tSNE-1', fontname='Helvetica', size=10, weight='normal')\n",
    "    plt.ylabel('tSNE-2', fontname='Helvetica', size=10, weight='normal')\n",
    "    plt.title('Archetypes', fontname='Helvetica', size=12, weight='normal')\n",
    "    plt.xlim(axis_min,axis_max)\n",
    "    plt.ylim(axis_min,axis_max)\n",
    "    plt.axis('off')\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '_{}_MONOSOMY3_ARCHETYPES_noText_Top5'.format(subset_type)\n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label \n",
    "plt.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FLATUI_ARCHETYPES = [\n",
    "                 'E6E6FA', # not assgined\n",
    "                 'FF66FF', # pink\n",
    "                 '6600CC', # purple\n",
    "                 '9ACD32', # Green\n",
    "                 '00FFFF', # skye blue\n",
    "                 'FF8800', # orange\n",
    "                 '0000FF', # dark blue\n",
    "                 'FFD700', # yellow\n",
    "                 '20B2AA', # teal\n",
    "                 #'008000', # forest green\n",
    "                 #'8B008B'\n",
    "    ]\n",
    "\n",
    "# CONVERT HEX TO RGB (FLATUI_SAMPLE)\n",
    "colors = np.zeros((len(FLATUI_ARCHETYPES),3))\n",
    "for ind,hexcolor in enumerate(FLATUI_ARCHETYPES):\n",
    "    colors[ind,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "CM_ARCHETYPES = LinearSegmentedColormap.from_list('FLATUI_ARCHETYPES', colors, N=len(colors))\n",
    "\n",
    "num_archetypes = len(archetype_distances.columns)\n",
    "ccc = [0]*archetype_distances.shape[0]\n",
    "for ii in np.arange(num_archetypes):\n",
    "    c = np.log10(archetype_distances['Distance_Archetype_{}'.format(ii)])\n",
    "    cc = c == c.min()\n",
    "    cc = cc*(ii+1)\n",
    "    ccc=ccc+cc\n",
    "    \n",
    "ax = plt.gca()\n",
    "seqc.plot.scatter.categorical(x, y, c= ccc, randomize = True,edgecolors='none',\n",
    "                              cmap=CM_ARCHETYPES,legend_kwargs={'ncol': 1}, s=1, ax=ax);\n",
    "plt.title('Archetypes', fontname='Helvetica', size=12, weight='normal')\n",
    "plt.xlim(axis_min,axis_max)\n",
    "plt.ylim(axis_min,axis_max)\n",
    "sns.despine()\n",
    "ax.legend_.remove()\n",
    "plt.axis('off')\n",
    "#ccc[ccc>num_archetypes] = 0 # remove cells assigned to more than one archetype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.unique(ccc)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### UPDATE INDEX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "METADATA_TUMOR.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write fuzzy clusters to metadata\n",
    "METADATA_TUMOR['Archetype_Nearest_Cell'] = ccc\n",
    "\n",
    "# Update index\n",
    "new_index = pd.MultiIndex.from_tuples(list(zip(METADATA_TUMOR.index.get_level_values('Sample ID'), \n",
    "                                               METADATA_TUMOR.index.get_level_values('Legend'),\n",
    "                                               METADATA_TUMOR.index.get_level_values('Patient'),\n",
    "                                               METADATA_TUMOR.index.get_level_values('Cell ID'),\n",
    "                                               METADATA_TUMOR.index.get_level_values('Phenograph_Class'),\n",
    "                                               METADATA_TUMOR.index.get_level_values('Archetype'),\n",
    "                                               METADATA_TUMOR.index.get_level_values('Nearest Archetype'),\n",
    "                                               METADATA_TUMOR['Archetype_Nearest_Cell'])), \n",
    "                                  names=['Sample ID','Legend', 'Patient','Cell ID','Phenograph_Class','Archetype',\n",
    "                                        'Nearest Archetype','Archetype_Nearest_Cell'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_type = 'TUMOR'\n",
    "exec('DF_{} = pd.DataFrame(DF_{}.values, columns = DF_{}.columns, index = new_index)'\\\n",
    "     .format(subset_type,subset_type,subset_type))\n",
    "exec('NDF_{} = pd.DataFrame(NDF_{}.values, columns = NDF_{}.columns, index = new_index)'\\\n",
    "     .format(subset_type,subset_type,subset_type))\n",
    "exec('INDF_{} = pd.DataFrame(INDF_{}.values, columns = INDF_{}.columns, index = new_index)'\\\n",
    "     .format(subset_type,subset_type,subset_type))\n",
    "exec('METADATA_{} = pd.DataFrame(METADATA_{}.values, columns = METADATA_{}.columns, index = new_index)'\\\n",
    "     .format(subset_type,subset_type,subset_type))\n",
    "exec('DIMENSIONS_{} = pd.DataFrame(DIMENSIONS_{}.values, columns = DIMENSIONS_{}.columns, index = new_index)'\\\n",
    "     .format(subset_type,subset_type,subset_type))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### COMPUTE DIFFUSION DISTANCE BETWEEN EACH CELL AND EACH ARCHETYPE (NEAREST CELL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PLOT DCS on TSNE, SELECT MOST RELEVANT PCS\n",
    "dimtype = 'ForceDirected' \n",
    "\n",
    "exec('x = DIMENSIONS_{}[\\'{}0\\']'.format(subset_type,dimtype))\n",
    "exec('y = DIMENSIONS_{}[\\'{}1\\']'.format(subset_type,dimtype))\n",
    "xmax = np.abs(x).max()\n",
    "ymax = np.abs(y).max()\n",
    "axis_min = -ceil(ceil(max(xmax,ymax)/10)*10)\n",
    "axis_max = ceil(ceil(max(xmax,ymax)/10)*10)\n",
    "\n",
    "cm = plt.cm.bwr \n",
    "dot_size = 1\n",
    "\n",
    "# Diffusion Eigenvalues\n",
    "component_prefix = 'DC'\n",
    "exec('QUERY = DIMENSIONS_{}'.format(subset_type))\n",
    "idx = QUERY.columns.to_series().str.contains('^{}((?!TSNE).)*$'.format(component_prefix)) # select from DIMS_NDF\n",
    "exec('DIFFUSION_EIGS = DIMENSIONS_{}.loc[:, idx.values]'.format(subset_type))\n",
    "\n",
    "nrow = 2\n",
    "ncol = 3\n",
    "plt.figure(figsize = (10,5))\n",
    "gs1 = gridspec.GridSpec(nrow, ncol)\n",
    "gs1.update(wspace=0.5, hspace=0.5)  \n",
    "\n",
    "# Plot scatter map with phase score\n",
    "for ind,label in enumerate(DIFFUSION_EIGS.columns[:5]):\n",
    "    c = DIFFUSION_EIGS[label].values\n",
    "    # CURRENT AXIS\n",
    "    ax = plt.subplot(gs1[ind])\n",
    "    plt.scatter(x, y, c=c,cmap=CM_DIVERGING, s=dot_size, alpha =0.4)             # finite data\n",
    "    plt.title(label, fontname='Helvetica', size=12, weight='normal')\n",
    "    plt.axis('off')\n",
    "    plt.clim(-0.04,0.04)\n",
    "    plt.xlim(axis_min,axis_max)\n",
    "    plt.ylim(axis_min,axis_max)\n",
    "    \n",
    "# PLOT SAMPLE_ID\n",
    "ax = plt.subplot(gs1[ind+1])\n",
    "exec('seqc.plot.scatter.categorical(x, y, c=NDF_{}.index.get_level_values(\\'Legend\\'), randomize = True, \\\n",
    "     cmap=CM_SAMPLES, s=dot_size, ax=ax)'.format(subset_type))\n",
    "plt.title('Sample ID', fontname='Helvetica', size=12, weight='normal')\n",
    "plt.xlim(axis_min,axis_max)\n",
    "plt.ylim(axis_min,axis_max)\n",
    "plt.axis('off')\n",
    "ax.legend_.remove()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '_{}_DIFFUSIONEIGS_{}'.format(subset_type,dimtype)\n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label \n",
    "plt.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SELECT MOST IMPORTANT DIFFUSION EIGS FOR MULTISCALE DISTANCE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DIFFUSION_EIGS.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SELECT EIGENVALUES, EXCLUDING ZEROITH\n",
    "eigs = DIFFUSION_EIGS\n",
    "eigs = [eig[0].astype(float) for eig in eigs.values]\n",
    "eigs = np.array(eigs)\n",
    "eigengap = np.abs(eigs[:-1]-eigs[1:])\n",
    "\n",
    "# COMPUTE EIGENGAP\n",
    "fig = plt.figure(figsize = (3,3))\n",
    "ax = plt.gca()\n",
    "ax.set_facecolor('white')\n",
    "plt.plot(np.arange(len(eigengap))+1,eigengap, marker='.', linestyle='dashed',linewidth=1, markersize=10, color = 'k')\n",
    "plt.xlim(1,18)\n",
    "plt.ylim(0,0.025)\n",
    "\n",
    "ax.set_ylabel(\"\\n\\n\\n{}\".format('Delat(Eigenvalue)'),fontsize=10)\n",
    "ax.set_xlabel(\"Diffusion Component\",fontsize=10)\n",
    "ax.tick_params(labelsize=10)\n",
    "\n",
    "plt.grid(False)\n",
    "sns.despine()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '{}_DCeigengap'.format(subset_type)\n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=fig_dpi)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PAIRWISE DIFFUSION DISTANCE BETWEEN CELLS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "L = 5 # number of eigen vectors to include in multi-scale distance\n",
    "\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "\n",
    "vec = DIFFUSION_EIGS.iloc[:,:L]\n",
    "scale_factor = (eigs/(1-eigs))#**2 # constant\n",
    "scale_factor = scale_factor[:L]\n",
    "\n",
    "distance_matrix = vec*scale_factor\n",
    "pairwise_distances = squareform(pdist(distance_matrix, lambda u,v: np.sqrt(((u-v)**2).sum()))) # MS, remove sqrt?\n",
    "\n",
    "diffusion_pairwise_distances= pd.DataFrame(index = DIFFUSION_EIGS.index, data=pairwise_distances, \n",
    "                                           columns = DIFFUSION_EIGS.index.get_level_values('Cell ID'))\n",
    "\n",
    "fig = plt.figure(figsize = (3,3))\n",
    "ax = plt.gca()\n",
    "plt.imshow(-np.log(pairwise_distances+0.001))\n",
    "plt.colorbar()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ASSIGN SOFT CLUSTER TO ARCHETYPE BASED ON DIFFUSION DISTANCE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diffusion_pairwise_distances"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Get ID of cell nearest to each archetype\n",
    "master_groupby = 'Archetype_Nearest_Cell'\n",
    "ix = [ind for ind,name in enumerate(diffusion_pairw ise_distances.index.names) if name==master_groupby][0]\n",
    "cellid_archetype = np.zeros(len(np.unique(diffusion_pairwise_distances.index.get_level_values(master_groupby)))-1)\n",
    "\n",
    "print(np.unique(diffusion_pairwise_distances.index.get_level_values(master_groupby))[1:])\n",
    "\n",
    "for arch_ind in np.unique(diffusion_pairwise_distances.index.get_level_values(master_groupby))[1:]:\n",
    "    vals = diffusion_pairwise_distances.loc[diffusion_pairwise_distances.index.map(lambda x: x[ix] in [arch_ind])]\n",
    "    print(vals)\n",
    "    print(arch_ind)\n",
    "    Thank\n",
    "    cellid_archetype[int(arch_ind)-1] = vals.index.get_level_values('Cell ID')[0]\n",
    "    \n",
    "cellid_archetype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get ID of cell nearest to each archetype\n",
    "master_groupby = 'Archetype_Nearest_Cell'\n",
    "ix = [ind for ind,name in enumerate(diffusion_pairwise_distances.index.names) if name==master_groupby][0]\n",
    "cellid_archetype = np.zeros(len(np.unique(diffusion_pairwise_distances.index.get_level_values(master_groupby)))-1)\n",
    "for arch_ind in np.unique(diffusion_pairwise_distances.index.get_level_values(master_groupby))[1:]:\n",
    "    vals = diffusion_pairwise_distances.loc[diffusion_pairwise_distances.index.map(lambda x: x[ix] in [arch_ind])]\n",
    "    cellid_archetype[int(arch_ind)-1] = vals.index.get_level_values('Cell ID')[0]\n",
    "cellid_archetype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assign soft cluster arround each archtype based on multi-scale diffusion distance\n",
    "# Radius around each archetype define as 1/2 multiscale diffusion distance to nearest archetype\n",
    "master_groupby = 'Archetype_Nearest_Cell'\n",
    "ix = [ind for ind,name in enumerate(diffusion_pairwise_distances.index.names) if name==master_groupby][0]\n",
    "\n",
    "assignments = pd.DataFrame(index = diffusion_pairwise_distances.index, \n",
    "                           columns = ['Archetype_Soft_Cluster'],\n",
    "                           data = np.zeros(diffusion_pairwise_distances.shape[0]).astype(int))\n",
    "\n",
    "for arch_ind in np.unique(diffusion_pairwise_distances.index.get_level_values(master_groupby))[1:]:\n",
    "    \n",
    "    vals = diffusion_pairwise_distances.loc[diffusion_pairwise_distances.index.map(lambda x: x[ix] in [arch_ind])]\n",
    "    \n",
    "    distance_to_other_archetypes = vals[cellid_archetype.astype(int)].\\\n",
    "                                   drop(vals.index.get_level_values('Cell ID')[0],axis=1)\n",
    "    \n",
    "    arch_radius = 0.5*distance_to_other_archetypes.min().min()\n",
    "    tmp = (vals < arch_radius)*arch_ind\n",
    "    assignments = pd.DataFrame(index = diffusion_pairwise_distances.index, \n",
    "                               columns = ['Archetype_Soft_Cluster'],\n",
    "                               data = assignments['Archetype_Soft_Cluster'].values + tmp.values[0])\n",
    "\n",
    "METADATA_TUMOR['Archetype_Soft_Cluster'] = assignments['Archetype_Soft_Cluster']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FLATUI_ARCHETYPES = [\n",
    "                 'E6E6FA', # not assgined\n",
    "                 'FF66FF', # pink\n",
    "                 '6600CC', # purple\n",
    "                 '9ACD32', # Green\n",
    "                 '00FFFF', # skye blue\n",
    "                 'FF8800', # orange\n",
    "                 '0000FF', # dark blue\n",
    "                 'FFD700', # yellow\n",
    "                 '20B2AA', # teal\n",
    "                 #'008000', # forest green\n",
    "                 #'8B008B'\n",
    "    ]\n",
    "\n",
    "# CONVERT HEX TO RGB (FLATUI_SAMPLE)\n",
    "colors = np.zeros((len(FLATUI_ARCHETYPES),3))\n",
    "for ind,hexcolor in enumerate(FLATUI_ARCHETYPES):\n",
    "    colors[ind,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "CM_ARCHETYPES = LinearSegmentedColormap.from_list('FLATUI_ARCHETYPES', colors, N=len(colors))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### PLOT CATEGORICAL CLUSTER ASSIGNMENTS\n",
    "# Note: Any rows assigned -1 were identified as outliers and should not be considered as a member of any community.\n",
    "plt.figure(figsize = (8,3))\n",
    "nrow = 1\n",
    "ncol = 2\n",
    "dot_size = 1\n",
    "gs1 = gridspec.GridSpec(nrow, ncol)\n",
    "gs1.update(wspace=0.1, hspace=0.1) # set the spacing between axes. \n",
    "\n",
    "dimtype = 'ForceDirected' \n",
    "meta = 'Phenograph_Class'\n",
    "exec('DIM = DIMENSIONS_{}'.format(subset_type))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "exec('QUERY = INDF_{}'.format(subset_type))\n",
    "x = DIM['{}0'.format(dimtype)]\n",
    "y = DIM['{}1'.format(dimtype)]\n",
    "axis_min = -ceil(ceil(max(np.abs(x).max(),np.abs(y).max()))/10)*10\n",
    "axis_max = ceil(ceil(max(np.abs(x).max(),np.abs(y).max()))/10)*10\n",
    "\n",
    "# Cluster Assignment\n",
    "ax = plt.subplot(gs1[0])\n",
    "seqc.plot.scatter.categorical(x, y, c=META[meta].values, \n",
    "                              cmap= CM_CLASS,\n",
    "                              legend_kwargs={'ncol': 1}, s=dot_size, ax=ax,randomize=True);\n",
    "plt.rc('xtick', labelsize=6)\n",
    "plt.rc('ytick', labelsize=6)\n",
    "sns.despine()\n",
    "plt.xlabel('tSNE-1', fontname='Helvetica', size=10, weight='normal')\n",
    "plt.ylabel('tSNE-2', fontname='Helvetica', size=10, weight='normal')\n",
    "plt.xlim(axis_min,axis_max)\n",
    "plt.ylim(axis_min,axis_max)\n",
    "plt.title('{}'.format(meta), fontname='Helvetica', size=12, weight='normal')\n",
    "plt.axis('off')\n",
    "ax.legend_.remove()\n",
    "\n",
    "# Castle Class\n",
    "ax = plt.subplot(gs1[1])\n",
    "seqc.plot.scatter.categorical(x, y, c=META['Archetype_Soft_Cluster'], \n",
    "                              cmap=CM_ARCHETYPES,legend_kwargs={'ncol': 1}, s=dot_size, ax=ax);\n",
    "sns.despine()\n",
    "plt.xlabel('tSNE-1', fontname='Helvetica', size=10, weight='normal')\n",
    "plt.ylabel('tSNE-2', fontname='Helvetica', size=10, weight='normal')\n",
    "plt.title('Sample ID', fontname='Helvetica', size=12, weight='normal')\n",
    "plt.xlim(axis_min,axis_max)\n",
    "plt.ylim(axis_min,axis_max)\n",
    "plt.axis('off')\n",
    "#ax.legend_.remove()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '_CLUSTERS_INDF_{}_{}_{}'.format(subset_type,meta,dimtype) \n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label \n",
    "plt.savefig(fn + '.png', bbox_inches='tight',dpi=fig_dpi)\n",
    "plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=fig_dpi)\n",
    "print(fn)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### ASSIGN ALL CELLS TO NEAREST ARCHETYPE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diffusion_pairwise_distances.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "D = pd.DataFrame(index = diffusion_pairwise_distances.index)\n",
    "for ind in np.unique(diffusion_pairwise_distances.index.get_level_values('Archetype_Nearest_Cell')):\n",
    "    D[ind]=diffusion_pairwise_distances[\\\n",
    "           diffusion_pairwise_distances.index.get_level_values('Archetype_Nearest_Cell')==ind].mean(axis=0).values\n",
    "\n",
    "METADATA_TUMOR['Nearest_Archetype'] = D.idxmin(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cm_flatui_1 = FLATUI_ARCHETYPES[1:]\n",
    "\n",
    "# CONVERT HEX TO RGB (FLATUI_SAMPLE)\n",
    "colors = np.zeros((len(cm_flatui_1),3))\n",
    "for ind,hexcolor in enumerate(cm_flatui_1):\n",
    "    colors[ind,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "CM_ARCHETYPES = LinearSegmentedColormap.from_list('cm_flatui_1', colors, N=len(colors))\n",
    "\n",
    "# PLOT CATEGORICAL CLUSTER ASSIGNMENTS\n",
    "# Note: Any rows assigned -1 were identified as outliers and should not be considered as a member of any community.\n",
    "plt.figure(figsize = (8,3))\n",
    "nrow = 1\n",
    "ncol = 2\n",
    "dot_size = 1\n",
    "gs1 = gridspec.GridSpec(nrow, ncol)\n",
    "gs1.update(wspace=0.1, hspace=0.1) # set the spacing between axes. \n",
    "\n",
    "dimtype = 'ForceDirected' \n",
    "meta = 'Phenograph_Class'\n",
    "exec('DIM = DIMENSIONS_{}'.format(subset_type))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "exec('QUERY = INDF_{}'.format(subset_type))\n",
    "x = DIM['{}0'.format(dimtype)]\n",
    "y = DIM['{}1'.format(dimtype)]\n",
    "axis_min = -ceil(ceil(max(np.abs(x).max(),np.abs(y).max()))/10)*10\n",
    "axis_max = ceil(ceil(max(np.abs(x).max(),np.abs(y).max()))/10)*10\n",
    "\n",
    "# Cluster Assignment\n",
    "ax = plt.subplot(gs1[0])\n",
    "seqc.plot.scatter.categorical(x, y, c=META[meta].values, \n",
    "                              cmap= CM_CLASS,\n",
    "                              legend_kwargs={'ncol': 1}, s=dot_size, ax=ax,randomize=True);\n",
    "plt.rc('xtick', labelsize=6)\n",
    "plt.rc('ytick', labelsize=6)\n",
    "sns.despine()\n",
    "plt.xlabel('tSNE-1', fontname='Helvetica', size=10, weight='normal')\n",
    "plt.ylabel('tSNE-2', fontname='Helvetica', size=10, weight='normal')\n",
    "plt.xlim(axis_min,axis_max)\n",
    "plt.ylim(axis_min,axis_max)\n",
    "plt.title('{}'.format(meta), fontname='Helvetica', size=12, weight='normal')\n",
    "plt.axis('off')\n",
    "ax.legend_.remove()\n",
    "\n",
    "# Castle Class\n",
    "ax = plt.subplot(gs1[1])\n",
    "seqc.plot.scatter.categorical(x, y, c=META['Nearest_Archetype'], \n",
    "                              cmap=CM_ARCHETYPES,legend_kwargs={'ncol': 1}, s=dot_size, ax=ax);\n",
    "sns.despine()\n",
    "plt.xlabel('tSNE-1', fontname='Helvetica', size=10, weight='normal')\n",
    "plt.ylabel('tSNE-2', fontname='Helvetica', size=10, weight='normal')\n",
    "plt.title('Sample ID', fontname='Helvetica', size=12, weight='normal')\n",
    "plt.xlim(axis_min,axis_max)\n",
    "plt.ylim(axis_min,axis_max)\n",
    "plt.axis('off')\n",
    "#ax.legend_.remove()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '_CLUSTERS_INDF_{}_{}_{}'.format(subset_type,meta,dimtype) \n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label \n",
    "plt.savefig(fn + '.png', bbox_inches='tight',dpi=fig_dpi)\n",
    "plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=fig_dpi)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### UPDATE MULTI-INDEX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update index\n",
    "new_index = pd.MultiIndex.from_tuples(list(zip(METADATA_TUMOR.index.get_level_values('Sample ID'), \n",
    "                                               METADATA_TUMOR.index.get_level_values('Legend'),\n",
    "                                               METADATA_TUMOR.index.get_level_values('Patient'),\n",
    "                                               METADATA_TUMOR.index.get_level_values('Cell ID'),\n",
    "                                               METADATA_TUMOR.index.get_level_values('Phenograph_Class'),\n",
    "                                               METADATA_TUMOR.index.get_level_values('Archetype'),\n",
    "                                               METADATA_TUMOR.index.get_level_values('Nearest Archetype'),\n",
    "                                               METADATA_TUMOR['Archetype_Nearest_Cell'],\n",
    "                                               METADATA_TUMOR['Archetype_Soft_Cluster'],\n",
    "                                               METADATA_TUMOR['Nearest_Archetype']\n",
    "                                              )), \n",
    "                                  names=['Sample ID','Legend', 'Patient','Cell ID','Phenograph_Class','Archetype',\n",
    "                                        'Nearest Archetype','Archetype_Nearest_Cell','Archetype_Soft_Cluster',\n",
    "                                        'Nearest_Archetype'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_type = 'TUMOR'\n",
    "exec('DF_{} = pd.DataFrame(DF_{}.values, columns = DF_{}.columns, index = new_index)'.\n",
    "     format(subset_type,subset_type,subset_type))\n",
    "exec('NDF_{} = pd.DataFrame(NDF_{}.values, columns = NDF_{}.columns, index = new_index)'.\n",
    "     format(subset_type,subset_type,subset_type))\n",
    "exec('INDF_{} = pd.DataFrame(INDF_{}.values, columns = INDF_{}.columns, index = new_index)'.\n",
    "     format(subset_type,subset_type,subset_type))\n",
    "exec('METADATA_{} = pd.DataFrame(METADATA_{}.values, columns = METADATA_{}.columns, index = new_index)'.\n",
    "     format(subset_type,subset_type,subset_type))\n",
    "exec('DIMENSIONS_{} = pd.DataFrame(DIMENSIONS_{}.values, columns = DIMENSIONS_{}.columns, index = new_index)'.\n",
    "     format(subset_type,subset_type,subset_type))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### FRACTION OF EACH ARCHETYPE PER PATIENT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FLATUI_ARCHETYPES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "groupby_type = 'Nearest_Archetype' # specify grouping\n",
    "cm_flatui = FLATUI_ARCHETYPES # specify colormap for grouping\n",
    "\n",
    "datatype = 'METADATA_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "\n",
    "# UPDATE INDEX TO CONTAIN DE_CLASS\n",
    "new_index = pd.MultiIndex.from_tuples(list(zip(META.index.get_level_values('Sample ID'), \n",
    "                                               META.index.get_level_values('Patient'),\n",
    "                                               META.index.get_level_values('Cell ID'),\n",
    "                                               META[groupby_type])), \n",
    "                                  names=['Sample ID','Patient', 'Cell ID',groupby_type])\n",
    "META = pd.DataFrame(data =META.values, index = new_index, columns = META.columns)\n",
    "\n",
    "exec('tissue_cluster_sizes = \\\n",
    "     META.groupby(level=[\\'Patient\\', \\'{}\\'], axis=0).size().unstack().fillna(0)'.format(groupby_type))\n",
    "\n",
    "# PRINT NUMBER OF CELLS PER META-CELL TYPE\n",
    "print(tissue_cluster_sizes.sum(axis=0))\n",
    "print('\\n')\n",
    "print(tissue_cluster_sizes.sum(axis=0).sum())\n",
    "# PRINT NUMBER OF CELLS PER PATIENT\n",
    "print(tissue_cluster_sizes.sum(axis=1))\n",
    "\n",
    "# CONVERT HEX TO RGB (FLATUI_CLASS)\n",
    "colors = np.zeros((len(cm_flatui),3))\n",
    "for ii,hexcolor in enumerate(cm_flatui):\n",
    "    colors[ii,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "metacell_colors = [rgb2hex(int(color[0]*255), int(color[1]*255), int(color[2]*255)) for color in colors]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a pie chart\n",
    "D = tissue_cluster_sizes\n",
    "plt.figure(figsize = (3,1))\n",
    "ax = plt.gca()\n",
    "D.plot.bar(stacked=True, color=metacell_colors, ax = ax, width = 0.85)\n",
    "ax.legend_.remove()\n",
    "plt.axis('off')\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '_{}Distribution_PerCellType_{}_PathwayMerged_retainN_noaxis'.\\\n",
    "                format(meta,subset_type).replace('.','_')\n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label \n",
    "plt.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue_cluster_sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "D = tissue_cluster_sizes\n",
    "plt.figure(figsize = (3,1))\n",
    "ax = plt.gca()\n",
    "D.div(D.sum(axis=1),axis=0).plot.bar(stacked=True, color=metacell_colors, ax = ax, width = 0.85)\n",
    "ax.legend_.remove()\n",
    "plt.axis('off')\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '_{}Distribution_PerCellType_{}_PathwayMerged_fraction'.format(meta,subset_type).replace('.','_')\n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label \n",
    "plt.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "D.div(D.sum(axis=1),axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### REPRESENTATION OF EACH ARCHETYPE WITHIN EACH PATIENT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "patient_fraction = np.log10(D.div(D.sum(axis=1),axis=0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "patient_fraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exec('tissue_cluster_sizes = INDF_{}.groupby(level=[\\'Patient\\', \\'{}\\'], axis=0).size().unstack().fillna(0)'.format(subset_type,meta))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue_cluster_sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(tissue_cluster_sizes.div(tissue_cluster_sizes.sum(axis=1),axis =0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_size_array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#meta = 'Archetype_Soft_Cluster'\n",
    "#FLATUI_PLOT = FLATUI_ARCHETYPES\n",
    "\n",
    "meta = 'Nearest_Archetype'\n",
    "FLATUI_PLOT = FLATUI_NEAREST_ARCHETYPES\n",
    "\n",
    "# Phenograph clusters\n",
    "ind = META[meta].values\n",
    "colors = np.zeros((len(FLATUI_PLOT),3))\n",
    "for ii,hexcolor in enumerate(FLATUI_PLOT):\n",
    "    colors[ii,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "# Palatte for Class METADATA\n",
    "cix = (np.linspace(0,shape(colors)[0],len(np.unique(ind)))).astype(int)\n",
    "if cix[len(cix)-1]==shape(colors)[0]:\n",
    "    cix[len(cix)-1]=shape(colors)[0]-1\n",
    "lut = dict(zip(np.unique(ind), colors[cix,:]))\n",
    "\n",
    "exec('tissue_cluster_sizes = INDF_{}.groupby(level=[\\'Patient\\', \\'{}\\'], axis=0).size().unstack().fillna(0)'.format(subset_type,meta))\n",
    "tissue_cluster_fraction = (tissue_cluster_sizes.div(tissue_cluster_sizes.sum(axis=1),axis =0))\n",
    "\n",
    "exec('sample_size_array = \\\n",
    "     META.groupby(level=[\\'Patient\\', \\'{}\\'], axis=0).size().unstack().fillna(0).sum(axis=1).values'.format(groupby_type))\n",
    "\n",
    "#SCORE = tissue_cluster_fraction[cluster_rank]\n",
    "\n",
    "violin_data = pd.DataFrame()\n",
    "for class_name in tissue_cluster_fraction.columns:\n",
    "    a = pd.DataFrame({meta:[class_name]*tissue_cluster_fraction.shape[0], \n",
    "                      'Fraction':tissue_cluster_fraction[class_name].values,\n",
    "                      'Sample Size':tissue_cluster_sizes[class_name].values,\n",
    "                      'Patient': tissue_cluster_fraction[class_name].index})\n",
    "    violin_data = violin_data.append(a)\n",
    "\n",
    "cluster_rank = violin_data.groupby(meta).median().sort_values('Fraction',ascending = False).index#[0]+order\n",
    "\n",
    "fig = plt.figure(figsize = (3,5))\n",
    "ax = plt.gca()\n",
    "\n",
    "# Plot violin plot\n",
    "g = sns.boxplot(x=meta, y=\"Fraction\" ,data=violin_data,\n",
    "                linewidth = 1, ax = ax, order = cluster_rank, notch = False,fliersize =1,\n",
    "               palette = lut);\n",
    "g.set_ylabel(\"{} Fraction per Patient\".format(meta,fontsize=10))\n",
    "g.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "g.tick_params(labelsize=10)\n",
    "sns.despine()\n",
    "\n",
    "# Add in points to show each observation\n",
    "sns.swarmplot(x=meta, y=\"Fraction\" ,data=violin_data,\n",
    "              size=4, color=\"k\", linewidth=0, order = cluster_rank, marker = 'o')\n",
    "\n",
    "g.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "plt.ylim((0,1))\n",
    "\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['bottom'].set_visible(True)\n",
    "ax.spines['left'].set_visible(True)\n",
    "ax.spines['bottom'].set_linewidth(0.75)\n",
    "ax.spines['left'].set_linewidth(0.75)\n",
    "\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'fraction_{}_per_patient_ranked_size'.format(meta)\n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "plt.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue_cluster_fraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "violin_data.to_csv('/workdir/uvmel_project/data/Archetype_per_Patient_Fractions_Ranked_Boxplot_Data.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MULTI-SCALE DIFFUSION DISTANCE BETWEEN ARCHETYPES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diffusion_pairwise_distances.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "master_groupby = 'Archetype_Nearest_Cell'\n",
    "ix = [ind for ind,name in enumerate(diffusion_pairwise_distances.index.names) if name==master_groupby][0]\n",
    "distance_to_other_archetypes = pd.DataFrame()\n",
    "for arch_ind in np.unique(diffusion_pairwise_distances.index.get_level_values(master_groupby))[1:]:\n",
    "    vals = diffusion_pairwise_distances.loc[diffusion_pairwise_distances.index.map(lambda x: x[ix] in [arch_ind])]\n",
    "    distance_to_other_archetypes = distance_to_other_archetypes.append(vals[cellid_archetype.astype(int)])\n",
    "\n",
    "distance_to_other_archetypes = pd.DataFrame(index = \\\n",
    "                                            distance_to_other_archetypes.index.get_level_values(master_groupby), \n",
    "                                            data = \\\n",
    "                                            distance_to_other_archetypes.values,\n",
    "                                            columns = \\\n",
    "                                            distance_to_other_archetypes.index.get_level_values(master_groupby))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clustered heatmap of euclidean distance between archetypes\n",
    "yticks = distance_to_other_archetypes.index\n",
    "xticks = distance_to_other_archetypes.columns\n",
    "\n",
    "# LINKAGE BASED ON NES OF SIGNIFICANT DIFFERENTIALLY EXPRESSED PATHWAYS\n",
    "method = 'average'\n",
    "metric = 'cosine'\n",
    "linkage = hc.linkage(distance_to_other_archetypes, method=method, metric = metric)\n",
    "row_linkage = deepcopy(linkage)\n",
    "linkage = hc.linkage(distance_to_other_archetypes.T, method=method, metric = metric)\n",
    "col_linkage = deepcopy(linkage)\n",
    "\n",
    "# REORDER HEATMAP ACCORDING TO LINKAGE (OPTIONAL, STILL SLOW)\n",
    "r1 = hc.leaves_list(row_linkage)\n",
    "cl = hc.leaves_list(col_linkage)\n",
    "mat = distance_to_other_archetypes.iloc[r1,cl]\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(5,5))\n",
    "plt.rcParams[\"axes.grid\"] = False\n",
    "\n",
    "# ADD MATRIX WITH GENE NAMES\n",
    "axmatrix = fig.add_axes([0.1,0.1,0.6,0.6])\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=plt.cm.RdBu_r)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "xtick = plt.xticks(range(len(labels)), labels, rotation = 90, fontsize = 14)\n",
    "\n",
    "labels = list(mat.index)\n",
    "axmatrix.yaxis.set_ticks_position('left')\n",
    "ytick = plt.yticks(range(len(labels)), labels, rotation = 0, fontsize = 14)\n",
    "\n",
    "#axmatrix.set_yticklabels([]);\n",
    "\n",
    "# ADD COLORBAR\n",
    "axcolor = fig.add_axes([0.75,0.1,0.01,0.1])\n",
    "cbar = plt.colorbar(im, cax=axcolor)\n",
    "#cbar.ax.get_yaxis().set_ticks([])\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'diffusion_distance_ongraph_btw_archetypes'\n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "distance_to_other_archetypes[4].sort_values().index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "distance_to_other_archetypes[5].sort_values().index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "distance_to_other_archetypes[3].sort_values().index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RANKED BOXPLOTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RANK BY DIFFUSION DISTANCE\n",
    "order = distance_to_other_archetypes[4].sort_values().index\n",
    "order = order.append(pd.Index([0])) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genesets.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta = 'Archetype_Soft_Cluster'\n",
    "\n",
    "FLATUI_ARCHETYPES = [\n",
    "                 'E6E6FA', # not assgined\n",
    "                 'FF66FF', # pink\n",
    "                 '6600CC', # purple\n",
    "                 '9ACD32', # Green\n",
    "                 '00FFFF', # skye blue\n",
    "                 'FF8800', # orange\n",
    "                 '0000FF', # dark blue\n",
    "                 'FFD700', # yellow\n",
    "                 '20B2AA', # teal\n",
    "                 #'008000', # forest green\n",
    "                 #'8B008B'\n",
    "    ]\n",
    "\n",
    "FLATUI_PLOT = FLATUI_ARCHETYPES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_genes = ['TMEM173', 'BAP1', 'JARID2', 'RING1', 'MYC','PRAME', 'JUN','CDH1', 'VIM', 'TYR', \n",
    "              'TYRP1','MLANA', 'PMEL', 'DCT','ASCL1', 'SOX10', 'CHD7', 'FOXD3', 'PAX3', 'NUMB',\n",
    "              'SNAI1','SNAI2','NEUROG2','HLA-A','HLA-B','HLA-C', 'B2M', 'HLA-G', 'ABCG1','PPARG',\n",
    "              'DDIT3','NUPR1','RAB3B','IGFBP4','LRRC8C','TCP11L2','MAFK','NRG1','F2R','KRT19','CTGF','ZFC3H1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# INDIVIDUAL GENE EXPRESSION\n",
    "meta = 'Archetype_Soft_Cluster'\n",
    "datatype = 'INDF_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "\n",
    "# Meta-Grouping\n",
    "exec('ind = METADATA_{}[meta].values'.format(subset_type))\n",
    "colors = np.zeros((len(FLATUI_PLOT),3))\n",
    "for ii,hexcolor in enumerate(FLATUI_PLOT):\n",
    "    colors[ii,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "\n",
    "cix = (np.linspace(0,shape(colors)[0],len(np.unique(ind)))).astype(int)\n",
    "if cix[len(cix)-1]==shape(colors)[0]:\n",
    "    cix[len(cix)-1]=shape(colors)[0]-1\n",
    "palette = dict(zip(np.unique(ind), colors[cix,:]))\n",
    "\n",
    "for title in plot_genes: # PLOT INDIVIDUAL GENES\n",
    "#for title in genesets:  # PLOT MEDIAN EXPRESSION OF PATHWAY\n",
    "    \n",
    "    # PLOT INDIVIDUAL GENES\n",
    "    genes = [title]\n",
    "    \n",
    "    # PLOT MEDIAN EXPRESSION OF SIGNATURE\n",
    "    #genes = genesets[title].values\n",
    "    \n",
    "    genes = np.unique([x for x in genes if str(x) != 'NAN'])\n",
    "    detected_genes = list(set(genes).intersection(set(QUERY.columns)))\n",
    "    vals = QUERY[detected_genes]\n",
    "    SCORE = np.nanmedian(vals,axis=1)\n",
    "    \n",
    "    # Format data structure for violin plot\n",
    "    violin_data = []\n",
    "    for ind,v in enumerate(SCORE):\n",
    "        violin_data.append({'gene': title, 'Z-normalized Expression': v,meta:META[meta].values[ind]}) \n",
    "    violin_data = pd.DataFrame(violin_data)\n",
    "    \n",
    "    # DENSITY PLOT\n",
    "    fig = plt.figure(figsize = (10,3))\n",
    "    gs1 = gridspec.GridSpec(2, 2)\n",
    "    gs1.update(wspace=0.2, hspace=0) # set the spacing between axes. \n",
    "\n",
    "    # BGEP RANKED ARPLOTS\n",
    "    ax_barplot = plt.subplot(gs1[:, 0])\n",
    "    g = sns.boxplot(x=\"gene\", y=\"Z-normalized Expression\", hue=meta,data=violin_data, palette=palette,\n",
    "                    notch = True, fliersize = 2, showmeans=False,linewidth = 1, hue_order = order, \n",
    "                    showfliers=True, whis = 0.95) \n",
    "    g.set_ylabel(\"{} {}\".format(genes[0],datatype),fontsize=10)\n",
    "    g.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "    g.tick_params(labelsize=10)\n",
    "    ax_barplot.legend_.remove()\n",
    "    vs = violin_data.iloc[[ind for ind,val in enumerate(violin_data[meta]) if val in order]]['Z-normalized Expression']\n",
    "    ax_barplot.set(ylim=(vs.min()*0.9, vs.max()*0.95))\n",
    "    ax_barplot.spines['right'].set_visible(False)\n",
    "    ax_barplot.spines['top'].set_visible(False)\n",
    "    ax_barplot.spines['bottom'].set_visible(True)\n",
    "    ax_barplot.spines['left'].set_visible(True)\n",
    "    ax_barplot.spines['bottom'].set_linewidth(0.75)\n",
    "    ax_barplot.spines['left'].set_linewidth(0.75)\n",
    "\n",
    "    ax1 = plt.subplot(gs1[0, 1])\n",
    "    ax1.spines['right'].set_visible(False)\n",
    "    ax1.spines['top'].set_visible(False)\n",
    "    ax1.spines['bottom'].set_visible(True)\n",
    "    ax1.spines['left'].set_visible(True)\n",
    "    ax1.spines['bottom'].set_linewidth(0.75)\n",
    "    ax1.spines['left'].set_linewidth(0.75)\n",
    "    bool_array = [val in order for val in violin_data[meta]]\n",
    "    bins = np.linspace(violin_data['Z-normalized Expression'][bool_array].min(), \\\n",
    "                       violin_data['Z-normalized Expression'][bool_array].max()*0.90, 100)\n",
    "    for archetype_ind in order:\n",
    "        vals = violin_data[violin_data[meta]==archetype_ind]['Z-normalized Expression']\n",
    "        plt.hist(vals,bins,density = False, alpha=0.75,color = palette[archetype_ind])\n",
    "        plt.tick_params(labelsize=10)\n",
    "        plt.xticks([])\n",
    "\n",
    "    archetype_ind = 0\n",
    "    vals = violin_data[violin_data[meta]==archetype_ind]['Z-normalized Expression']\n",
    "    ax2 = plt.subplot(gs1[1, 1])\n",
    "    plt.hist(vals,bins,density = False, alpha=1,color = palette[archetype_ind])\n",
    "    ax2.set_ylim(ax2.get_ylim()[::-1])\n",
    "    ax2.spines['right'].set_visible(False)\n",
    "    ax2.spines['top'].set_visible(True)\n",
    "    ax2.spines['bottom'].set_visible(False)\n",
    "    ax2.spines['left'].set_visible(True)\n",
    "    ax2.spines['top'].set_linewidth(0.75)\n",
    "    ax2.spines['left'].set_linewidth(0.75)\n",
    "\n",
    "    plt.xticks(rotation=70)\n",
    "    plt.tick_params(labelsize=10)\n",
    "    \n",
    "    # SAVE FIGURE\n",
    "    figure_label = '_{}_{}PLOT_{}_{}_{}'.format(subset_type, plot_type,datatype,meta,title)\n",
    "    fn = FIG_output_stem + 'BOXPLOTS_RANKED_MEDIAN_GEP2_NODENSITY/{}'.format(meta) + FN.replace(\".h5\", \"\") + \\\n",
    "    figure_label\n",
    "    fn = fn.replace(' ','_')\n",
    "    \n",
    "    # CREATE GSEA DIRECTORY IF IT DOES NOT EXIST\n",
    "    d = os.path.dirname(fn)\n",
    "    if not os.path.exists(d):\n",
    "        os.makedirs(d)\n",
    "        \n",
    "    plt.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "    #plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "    print(fn)\n",
    "    \n",
    "    # Close Figure\n",
    "    plt.close(fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### NK CELL HETEROGENEITY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Phenograph Cluster 19 = NK/NKT\n",
    "exec('INDF_{} = h5_data.load(\\'/INDF_{}\\')'.format('ALL','ALL'))\n",
    "groupby_type = 'Phenograph_Class'\n",
    "exec('tissue_cluster_sizes = INDF_ALL.groupby(level=[\\'Patient\\', \\'{}\\'], axis=0).size().unstack().fillna(0)'.\\\n",
    "     format(groupby_type))\n",
    "fraction_cell_detected = tissue_cluster_sizes.div(tissue_cluster_sizes.sum(axis=1),axis=0)\n",
    "NK_CELL_FRACTION = fraction_cell_detected[19]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "NK_CELL_FRACTION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mat2 = pd.DataFrame(index = NK_CELL_FRACTION.index, columns = ['NK FRACTION'],data = (NK_CELL_FRACTION.values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(1,5))\n",
    "plt.rcParams[\"axes.grid\"] = False\n",
    "\n",
    "# ADD MATRIX WITH GENE NAMES\n",
    "axmatrix = fig.add_axes([0.1,0.1,0.6,0.6])\n",
    "im = axmatrix.matshow(mat2, aspect='auto', origin='lower', cmap=plt.cm.RdBu_r, vmin = 0, vmax = 0.2)\n",
    "\n",
    "labels = list(mat2.index)\n",
    "axmatrix.yaxis.set_ticks_position('left')\n",
    "ytick = plt.yticks(range(len(labels)), labels, rotation = 0, fontsize = 14)\n",
    "\n",
    "#axmatrix.set_yticklabels([]);\n",
    "\n",
    "# ADD COLORBAR\n",
    "axcolor = fig.add_axes([0.75,0.1,0.1,0.1])\n",
    "cbar = plt.colorbar(im, cax=axcolor)\n",
    "#cbar.ax.get_yaxis().set_ticks([])\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'NK_NKT_FRACTION'\n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fraction of each tumor cel archetype\n",
    "groupby_type = 'Archetype_Soft_Cluster'\n",
    "exec('tissue_cluster_sizes = QUERY.groupby(level=[\\'Patient\\', \\'{}\\'], axis=0).size().unstack().fillna(0)'.format(groupby_type))\n",
    "fraction_cell_detected = tissue_cluster_sizes.div(tissue_cluster_sizes.sum(axis=1),axis=0)\n",
    "fraction_cell_detected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize = (5,5))\n",
    "ax = plt.gca()\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['bottom'].set_visible(True)\n",
    "ax.spines['left'].set_visible(True)\n",
    "\n",
    "x = np.log10(NK_CELL_FRACTION+0.001)[:5]\n",
    "\n",
    "for arch_ind in order:\n",
    "    y = np.log10(fraction_cell_detected[arch_ind]+0.001)[:5]\n",
    "     \n",
    "    #(r, p) = stats.pearsonr(x, y)\n",
    "    (slope, intercept, r, p, std_err) = stats.linregress(x,y)\n",
    "    print('Archetype{} vs NK/NKT Abundance, R: {}, p: {}'.format(arch_ind,r,p))\n",
    "    \n",
    "    xi =np.linspace(x.min(),x.max(),20)\n",
    "    yi = slope*xi+intercept\n",
    "    \n",
    "    if p < 0.05:\n",
    "        plt.scatter(x,y,c = palette[arch_ind])\n",
    "        plt.plot(xi,yi,c = palette[arch_ind])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### SAVE TO DATAFRAME"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_type = 'TUMOR'\n",
    "\n",
    "# WRITE RANKED GENE LIST TO H5\n",
    "exec('h5_data.save(DF_{}, \\'DF_{}\\')'.format(subset_type,subset_type)) # update directory\n",
    "exec('h5_data.save(NDF_{}, \\'NDF_{}\\')'.format(subset_type,subset_type)) # update directory\n",
    "exec('h5_data.save(INDF_{}, \\'INDF_{}\\')'.format(subset_type,subset_type)) # update directory\n",
    "exec('h5_data.save(METADATA_{}, \\'METADATA_{}\\')'.format(subset_type,subset_type)) # update directory\n",
    "exec('h5_data.save(DIMENSIONS_{}, \\'DIMENSIONS_{}\\')'.format(subset_type,subset_type)) # update directory\n",
    "h5_data.ls() # list contents of directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "METADATA_TUMOR.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## INTRA-PATIENT HETEROGENEITY"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ARCHETYPE DISTRIBUTION PER PATIENT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "groupby_type = 'Nearest_Archetype' # specify grouping\n",
    "cm_flatui = FLATUI_ARCHETYPES[1:] # specify colormap for grouping\n",
    "\n",
    "datatype = 'METADATA_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "\n",
    "\n",
    "exec('tissue_cluster_sizes = \\\n",
    "     META.groupby(level=[\\'Patient\\', \\'{}\\'], axis=0).size().unstack().fillna(0)'.format(groupby_type))\n",
    "\n",
    "# PRINT NUMBER OF CELLS PER META-CELL TYPE\n",
    "print(tissue_cluster_sizes.sum(axis=0))\n",
    "print('\\n')\n",
    "print(tissue_cluster_sizes.sum(axis=0).sum())\n",
    "# PRINT NUMBER OF CELLS PER PATIENT\n",
    "print(tissue_cluster_sizes.sum(axis=1))\n",
    "\n",
    "# CONVERT HEX TO RGB (FLATUI_CLASS)\n",
    "colors = np.zeros((len(cm_flatui),3))\n",
    "for ii,hexcolor in enumerate(cm_flatui):\n",
    "    colors[ii,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "metacell_colors = [rgb2hex(int(color[0]*255), int(color[1]*255), int(color[2]*255)) for color in colors]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue_cluster_sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue_cluster_sizes.div(tissue_cluster_sizes.sum(axis=1),axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "D = tissue_cluster_sizes\n",
    "plt.figure(figsize = (3,1))\n",
    "ax = plt.gca()\n",
    "D.div(D.sum(axis=1),axis=0).plot.bar(stacked=True, color=metacell_colors, ax = ax, width = 0.85)\n",
    "ax.legend_.remove()\n",
    "plt.axis('off')\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '_{}Distribution_PerCellType_{}_PathwayMerged_fraction'.format(meta,subset_type).replace('.','_')\n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label \n",
    "plt.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metacell_colors"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RANKED BOXPLOT PER PATIENT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# INDIVIDUAL GENE EXPRESSION\n",
    "meta = 'Archetype_Soft_Cluster'#'Nearest_Archetype'\n",
    "FLATUI_PLOT = FLATUI_ARCHETYPES#FLATUI_ARCHETYPES[1:]\n",
    "\n",
    "#meta = 'Nearest_Archetype'\n",
    "#FLATUI_PLOT = FLATUI_ARCHETYPES[1:]\n",
    "\n",
    "patient_ind = 'UM02'\n",
    "subset_key = 'Patient'\n",
    "datatype = 'INDF_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "\n",
    "# Meta-Grouping\n",
    "ind = META[meta].values\n",
    "colors = np.zeros((len(FLATUI_PLOT),3))\n",
    "for ii,hexcolor in enumerate(FLATUI_PLOT):\n",
    "    colors[ii,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "\n",
    "cix = (np.linspace(0,shape(colors)[0],len(np.unique(ind)))).astype(int)\n",
    "if cix[len(cix)-1]==shape(colors)[0]:\n",
    "    cix[len(cix)-1]=shape(colors)[0]-1\n",
    "palette = dict(zip(np.unique(ind), colors[cix,:]))\n",
    "\n",
    "# SUBSET  PATIENT\n",
    "ix = [ind for ind, name in enumerate(QUERY.index.names) if name==subset_key][0]\n",
    "QUERY = QUERY.loc[QUERY.index.map(lambda x: x[ix] in [patient_ind])]\n",
    "META = META.loc[META.index.map(lambda x: x[ix] in [patient_ind])]\n",
    "\n",
    "ind = META[meta].values\n",
    "\n",
    "new_order = [o for o in order if o in np.unique(META[meta])]\n",
    "\n",
    "#for title in plot_genes: # LOT INDIVIDUAL GENES\n",
    "for title in genesets: # PLOT MEDIAN EXPRESSION OF PATHWAY\n",
    "    \n",
    "    # PLOT INDIVIDUAL GENES\n",
    "    #genes = [title]\n",
    "    \n",
    "    # PLOT MEDIAN EXPRESSION OF SIGNATURE\n",
    "    genes = genesets[title].values\n",
    "    \n",
    "    genes = np.unique([x for x in genes if str(x) != 'NAN'])\n",
    "    detected_genes = list(set(genes).intersection(set(QUERY.columns)))\n",
    "    vals = QUERY[detected_genes]\n",
    "    SCORE = np.nanmedian(vals,axis=1)\n",
    "    \n",
    "    # Format data structure for violin plot\n",
    "    violin_data = []\n",
    "    for ind,v in enumerate(SCORE):\n",
    "        violin_data.append({'gene': title, 'Z-normalized Expression': v,meta:META[meta].values[ind]}) \n",
    "    violin_data = pd.DataFrame(violin_data)\n",
    "    \n",
    "    # DENSITY PLOT\n",
    "    fig = plt.figure(figsize = (10,3))\n",
    "    gs1 = gridspec.GridSpec(2, 2)\n",
    "    gs1.update(wspace=0.2, hspace=0) # set the spacing between axes. \n",
    "\n",
    "    # BGEP RANKED ARPLOTS\n",
    "    ax_barplot = plt.subplot(gs1[:, 0])\n",
    "    g = sns.boxplot(x=\"gene\", y=\"Z-normalized Expression\", hue=meta, data=violin_data, palette=palette,\n",
    "                    notch = False, fliersize = 2, showmeans=False,linewidth = 1, \n",
    "                    hue_order = new_order, \n",
    "                    showfliers=True, whis = 0.95) \n",
    "    g.set_ylabel(\"{} {}\".format(genes[0],datatype),fontsize=10)\n",
    "    g.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "    g.tick_params(labelsize=10)\n",
    "    ax_barplot.legend_.remove()\n",
    "    vs = violin_data.iloc[[ind for ind,val in enumerate(violin_data[meta]) \n",
    "                           if val in order]]['Z-normalized Expression']\n",
    "    ax_barplot.set(ylim=(vs.min()*0.9, vs.max()*0.95))\n",
    "    ax_barplot.spines['right'].set_visible(False)\n",
    "    ax_barplot.spines['top'].set_visible(False)\n",
    "    ax_barplot.spines['bottom'].set_visible(True)\n",
    "    ax_barplot.spines['left'].set_visible(True)\n",
    "    ax_barplot.spines['bottom'].set_linewidth(0.75)\n",
    "    ax_barplot.spines['left'].set_linewidth(0.75)\n",
    "\n",
    "    ax1 = plt.subplot(gs1[0, 1])\n",
    "    ax1.spines['right'].set_visible(False)\n",
    "    ax1.spines['top'].set_visible(False)\n",
    "    ax1.spines['bottom'].set_visible(True)\n",
    "    ax1.spines['left'].set_visible(True)\n",
    "    ax1.spines['bottom'].set_linewidth(0.75)\n",
    "    ax1.spines['left'].set_linewidth(0.75)\n",
    "    bool_array = [val in order for val in violin_data[meta]]\n",
    "    bins = np.linspace(violin_data['Z-normalized Expression'][bool_array].min(), \n",
    "                       violin_data['Z-normalized Expression'][bool_array].max()*0.90, 100)#\n",
    "    for archetype_ind in new_order:\n",
    "        vals = violin_data[violin_data[meta]==archetype_ind]['Z-normalized Expression']\n",
    "        plt.hist(vals,bins,density = False, alpha=0.75,color = palette[archetype_ind])\n",
    "        plt.tick_params(labelsize=10)\n",
    "        plt.xticks([])\n",
    "    \n",
    "    if (violin_data[meta]==0).sum()>0:\n",
    "        archetype_ind = 0\n",
    "        vals = violin_data[violin_data[meta]==archetype_ind]['Z-normalized Expression']\n",
    "        ax2 = plt.subplot(gs1[1, 1])\n",
    "        plt.hist(vals,bins,density =  False, alpha=1,color = palette[archetype_ind])\n",
    "        ax2.set_ylim(ax1.get_ylim()[::-1])\n",
    "        ax2.spines['right'].set_visible(False)\n",
    "        ax2.spines['top'].set_visible(True)\n",
    "        ax2.spines['bottom'].set_visible(False)\n",
    "        ax2.spines['left'].set_visible(True)\n",
    "        ax2.spines['top'].set_linewidth(0.75)\n",
    "        ax2.spines['left'].set_linewidth(0.75)\n",
    "\n",
    "    plt.xticks(rotation=70)\n",
    "    plt.tick_params(labelsize=10)\n",
    "    \n",
    "    # SAVE FIGURE\n",
    "    figure_label = '_{}_{}PLOT_{}_{}_{}_Patient{}'.format(subset_type, plot_type,datatype,meta,title,patient_ind)\n",
    "    fn = FIG_output_stem + 'BOXPLOTS_RANKED_MEDIAN_GEP2_NODENSITY_PATIENT{}/{}'.format(patient_ind,meta) + \\\n",
    "         FN.replace(\".h5\", \"\") + figure_label\n",
    "    fn = fn.replace(' ','_')\n",
    "    \n",
    "    # CREATE GSEA DIRECTORY IF IT DOES NOT EXIST\n",
    "    d = os.path.dirname(fn)\n",
    "    if not os.path.exists(d):\n",
    "        os.makedirs(d)\n",
    "        \n",
    "    plt.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "    #plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "    print(fn)\n",
    "    \n",
    "    # Close Figure\n",
    "    plt.close(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(violin_data[meta]==0).sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RECREATING MIXTURE MODEL"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### MIXTURE MODEL FOR TCGA CLASSIFICATION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "TCGA_CLASS_LIST = ['TCGA1', 'TCGA2', 'TCGA3', 'TCGA4']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make the TCGA marker_file dataframe containing the signature genes for each TCGA subtype:\n",
    "\n",
    "TCGA_markerfile_df = pd.DataFrame()\n",
    "\n",
    "TCGA_markerfile_df['TCGA1'] = [1 for x in range(15)] + [0 for x in range(45)]\n",
    "TCGA_markerfile_df['TCGA2'] = [0 for x in range(15)] + [1 for x in range(15)] + [0 for x in range(30)]\n",
    "TCGA_markerfile_df['TCGA3'] = [0 for x in range(30)] + [1 for x in range(15)] + [0 for x in range(15)]\n",
    "TCGA_markerfile_df['TCGA4'] = [0 for x in range(45)] + [1 for x in range(15)]\n",
    "\n",
    "TCGA_markerfile_df.index = ['PCDH20', 'CNTN3', 'CCDC68', 'ZNF204P', 'STXBP5L', 'PDE3A', 'ZNF883', 'CHL1', 'SLC25A27', \n",
    "                            'ZNF702P', 'ZNF667', 'ZNF391', 'KYNU', 'MPPED2', 'PCDHA13', 'PYY', 'CRYBB2', 'EYA2', \n",
    "                            'ERVFRDE1', 'MYO7B', 'LOC643719', 'C20orf26', 'DLL3', 'RHCG', 'COX6A2', 'SOSTDC1', \n",
    "                            'TMEM151A', 'FXYD1', 'LOC441869', 'LOC100188947', 'SSX5', 'C1orf116', 'GSG1L', 'VTN', \n",
    "                            'COL9A3', 'AMN', 'DYSF', 'NECAB2', 'CDH4', 'ADAM11', 'CAMK1G', 'NXPH4', 'NPAS1', 'SUSD2', \n",
    "                            'BAI1', 'CXCL13', 'RIMS2', 'KCNV2', 'GAD1', 'IDO1', 'GBP1', 'UBD', 'JAKMIP1', \n",
    "                            'SH2D1A', 'CXCL9', 'COL22A1', 'GBP5', 'TM7SF4', 'LOC96610', 'TIGIT']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# WORKING WITH TUMOR SUBSET:\n",
    "subset_type = 'TUMOR'\n",
    "\n",
    "# WORKING WITH IMPUTED NORMALIZED DATA:\n",
    "datatype = 'INDF_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "\n",
    "# Get imputed normalized gene expression means for each TCGA subtype:\n",
    "\n",
    "gene1 = 'TCGA1'\n",
    "signature_genes = TCGA_markerfile_df.loc[TCGA_markerfile_df[gene1] == 1].index.values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "vals1 = (np.nanmean(QUERY[detected_genes],axis=1))\n",
    "\n",
    "gene2 = 'TCGA2'\n",
    "signature_genes = TCGA_markerfile_df.loc[TCGA_markerfile_df[gene2] == 1].index.values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "vals2 = (np.nanmean(QUERY[detected_genes],axis=1))\n",
    "\n",
    "gene3 = 'TCGA3'\n",
    "signature_genes = TCGA_markerfile_df.loc[TCGA_markerfile_df[gene3] == 1].index.values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "vals3 = (np.nanmean(QUERY[detected_genes],axis=1))\n",
    "\n",
    "gene4 = 'TCGA4'\n",
    "signature_genes = TCGA_markerfile_df.loc[TCGA_markerfile_df[gene4] == 1].index.values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "vals4 = (np.nanmean(QUERY[detected_genes],axis=1))\n",
    "\n",
    "X = pd.DataFrame(data = {gene1: vals1, gene2: vals2, gene3: vals3, gene4: vals4}, index = QUERY.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### RUN BGMM MODEL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fit BGMM using dataframe X:\n",
    "bgmm = BayesianGaussianMixture(covariance_type='spherical', n_components=4, random_state=1)\n",
    "bgmm.fit(X[['TCGA1','TCGA2','TCGA3','TCGA4']])\n",
    "\n",
    "# Predict TCGA classification for each cell:\n",
    "clusters = bgmm.predict(X[['TCGA1','TCGA2','TCGA3','TCGA4']])\n",
    "\n",
    "# Use T-SNE to visualize cell classifications:\n",
    "X_embedded = manifold.TSNE(n_components=2, random_state=34).fit_transform(X[['TCGA1','TCGA2','TCGA3','TCGA4']])\n",
    "X['tsne1'] = X_embedded[:,0]\n",
    "X['tsne2'] = X_embedded[:,1]\n",
    "\n",
    "X['Assignment'] = clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make Class Assignments\n",
    "lut ={}\n",
    "mean_array = []\n",
    "for ii in np.unique(X['Assignment']):\n",
    "    mean_array.append([X[X['Assignment']==ii][x].mean() for x in TCGA_CLASS_LIST])\n",
    "mean_array = np.array(mean_array)\n",
    "\n",
    "for i in range(len(np.unique(X['Assignment']))):\n",
    "    max_ind = np.unravel_index(np.argmax(mean_array, axis=None), mean_array.shape)\n",
    "    lut[max_ind[0]] = TCGA_CLASS_LIST[max_ind[1]]\n",
    "    mean_array[:,max_ind[1]] = 0\n",
    "    mean_array[max_ind[0],:] = 0\n",
    "\n",
    "# MAP ASSIGNMENTS\n",
    "X['Assignment'] = X['Assignment'].map(lut)\n",
    "\n",
    "sns.set(font_scale=2.5)\n",
    "sns.set_style(\"white\")\n",
    "g = sns.pairplot(X[['tsne1','tsne2','Assignment']], \n",
    "                 hue = 'Assignment',\n",
    "                 palette={\"TCGA1\": \"#0000FF\",\n",
    "                          \"TCGA2\": \"#A4D3FC\",\n",
    "                          \"TCGA3\": \"#FF7DC2\", \n",
    "                          \"TCGA4\": \"#FF0000\",},\n",
    "            diag_kind = 'kde', aspect = 1, size = 10);# SAVE FIGURE\n",
    "\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'MSK_BayesianGMM_TCGA'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "\n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "#g.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "#g.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# UPDATE DIRECTORY\n",
    "META['TCGA_Assignment'] = X.Assignment\n",
    "METADATA_TUMOR = META.copy()\n",
    "exec('h5_data.save(METADATA_{}, \\'METADATA_{}\\')'.format(subset_type,subset_type))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### INTRATUMOR TCGA HETEROGENEITY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalize mean TCGA signature gene expression for each cell:\n",
    "X_norm = X[TCGA_CLASS_LIST].copy()\n",
    "X_norm = X_norm.div(X_norm.sum(axis=1), axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### TCGA DISTRIBUTION PER PATIENT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "METADATA_TUMOR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "METADATA_TUMOR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GENE-GENE COVARIANCE WITHIN NORMAL SUBSET\n",
    "n = 500\n",
    "reps = np.arange(20)\n",
    "\n",
    "boxplot_patient_list = []\n",
    "boxplot_fract_list = []\n",
    "boxplot_class_list = []\n",
    "master_groupby = 'Patient'\n",
    "\n",
    "for rep in reps:\n",
    "    for patient in METADATA_TUMOR.index.get_level_values('Patient').unique():\n",
    "        SUBSET = pd.DataFrame()\n",
    "        for class_selection in np.unique(METADATA_TUMOR.index.get_level_values(master_groupby)):\n",
    "            CSUBSET = METADATA_TUMOR.loc[METADATA_TUMOR.index.map(lambda x: x[ix] in \n",
    "                                                                  [class_selection])].sample(n=n,replace=True)\n",
    "            SUBSET = SUBSET.append(CSUBSET)\n",
    "        \n",
    "        tmp = [SUBSET.loc[SUBSET.index.get_level_values('Patient') == patient].\\\n",
    "               TCGA_Assignment.value_counts().reindex(SUBSET.TCGA_Assignment.unique(), fill_value=0)[x] \n",
    "               for x in TCGA_CLASS_LIST]\n",
    "        tmp = list(tmp/sum(tmp))\n",
    "    \n",
    "        boxplot_patient_list = boxplot_patient_list + [patient for x in range(len(TCGA_CLASS_LIST))]\n",
    "        boxplot_fract_list = boxplot_fract_list + tmp\n",
    "        boxplot_class_list = boxplot_class_list + TCGA_CLASS_LIST\n",
    "\n",
    "boxplot_data = pd.DataFrame()\n",
    "boxplot_data['Patient'] = ['MSK_' + x for x in boxplot_patient_list]\n",
    "boxplot_data['Fraction'] = boxplot_fract_list\n",
    "boxplot_data['Class'] = boxplot_class_list\n",
    "\n",
    "boxplot_data.Fraction = [x if x > 0.0001 else 1e-4 for x in boxplot_data.Fraction]\n",
    "\n",
    "# Set plot\n",
    "fig = plt.figure(figsize = (7,4))\n",
    "ax = plt.gca()\n",
    "\n",
    "# Plot bar plot\n",
    "\n",
    "sns.barplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class', ci=95, capsize=0.15,\n",
    "                palette={\"TCGA1\": \"#0000FF\",\n",
    "                         \"TCGA2\": \"#A4D3FC\",\n",
    "                         \"TCGA3\": \"#FF7DC2\", \n",
    "                         \"TCGA4\": \"#FF0000\",},\n",
    "                linewidth=1, ax = ax, alpha=1,\n",
    "                errcolor='black',\n",
    "                errwidth=1)\n",
    "\n",
    "g = sns.stripplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class',\n",
    "                palette={\"TCGA1\": \"#0000FF\",\n",
    "                         \"TCGA2\": \"#A4D3FC\",\n",
    "                         \"TCGA3\": \"#FF7DC2\", \n",
    "                         \"TCGA4\": \"#FF0000\",},\n",
    "                linewidth = 0.5, edgecolor='white',\n",
    "                ax = ax, alpha=0.7,\n",
    "                dodge=True)\n",
    "\n",
    "fig.get_axes()[0].set_yscale('log')\n",
    "\n",
    "plt.ylim((1e-4,1.15))\n",
    "ax.set_ylabel(\"Cell Type Fraction\",fontsize=10)\n",
    "ax.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "ax.tick_params(labelsize=10)\n",
    "ax.set_xticklabels(boxplot_data['Patient'].unique(),rotation=90)\n",
    "sns.despine()\n",
    "\n",
    "ax.legend_.remove()\n",
    "plt.tight_layout()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'TCGA_Distribution_by_Patient_LOG'\n",
    "fn = '/workdir/uvmel_project/figures/Revision_2_Figures/' + figure_label \n",
    "plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n = 500\n",
    "reps = np.arange(20)\n",
    "\n",
    "master_groupby = 'Patient'\n",
    "ix = [ind for ind,name in enumerate(X.index.names) if name==master_groupby][0]\n",
    "fraction_TCGA1 = np.zeros((len(reps),len(np.unique(X.index.get_level_values(master_groupby)))))\n",
    "fraction_TCGA2 = np.zeros((len(reps),len(np.unique(X.index.get_level_values(master_groupby)))))\n",
    "fraction_TCGA3 = np.zeros((len(reps),len(np.unique(X.index.get_level_values(master_groupby)))))\n",
    "fraction_TCGA4 = np.zeros((len(reps),len(np.unique(X.index.get_level_values(master_groupby)))))\n",
    "\n",
    "for rep in reps:\n",
    "    # RANDOMLY SAMPLE SAME NUMBER OF CELLS PER PATIENT\n",
    "    SUBSET = pd.DataFrame()\n",
    "    for class_selection in np.unique(X.index.get_level_values(master_groupby)):\n",
    "        CSUBSET = X.loc[X.index.map(lambda x: x[ix] in [class_selection])].sample(n=n,replace=True)\n",
    "        SUBSET = SUBSET.append(CSUBSET)\n",
    "    \n",
    "    bgmm = BayesianGaussianMixture(covariance_type='spherical', n_components=4, random_state=1)\n",
    "    bgmm.fit(SUBSET[TCGA_CLASS_LIST])\n",
    "    clusters = bgmm.predict(SUBSET[TCGA_CLASS_LIST])\n",
    "    SUBSET['Assignment'] = clusters\n",
    "\n",
    "    # Make Class Assignments\n",
    "    lut ={}\n",
    "    mean_array = []\n",
    "    for ii in np.unique(SUBSET['Assignment']):\n",
    "        mean_array.append([SUBSET[SUBSET['Assignment']==ii][x].mean() for x in TCGA_CLASS_LIST])\n",
    "    mean_array = np.array(mean_array)\n",
    "    \n",
    "    for i in range(len(np.unique(SUBSET['Assignment']))):\n",
    "        max_ind = np.unravel_index(np.argmax(mean_array, axis=None), mean_array.shape)\n",
    "        lut[max_ind[0]] = TCGA_CLASS_LIST[max_ind[1]]\n",
    "        mean_array[:,max_ind[1]] = 0\n",
    "        mean_array[max_ind[0],:] = 0\n",
    "    \n",
    "    # MAP ASSIGNMENTS\n",
    "    SUBSET['Assignment'] = SUBSET['Assignment'].map(lut)\n",
    "\n",
    "    # Update multi-index\n",
    "    new_index = pd.MultiIndex.from_tuples(list(zip(SUBSET.index.get_level_values('Sample ID'), \n",
    "                                                   SUBSET.index.get_level_values('Legend'),\n",
    "                                                   SUBSET.index.get_level_values('Patient'),\n",
    "                                                   SUBSET.index.get_level_values('Cell ID'),\n",
    "                                                   SUBSET.index.get_level_values('Phenograph_Class'),\n",
    "                                                   SUBSET['Assignment'])), \n",
    "                                          names=['Sample ID','Legend', 'Patient','Cell ID',\n",
    "                                                 'Phenograph_Class','Assignment'])\n",
    "    SUBSET = pd.DataFrame(data = SUBSET.values, columns = SUBSET.columns, index = new_index)\n",
    "\n",
    "    # CASTLE ASSIGNMENT PER PATIENT\n",
    "    meta = 'Assignment'\n",
    "    exec('tmp = SUBSET.groupby(level=[\\'Patient\\', \\'{}\\'], axis=0).size().unstack().fillna(0)'.format(meta))\n",
    "\n",
    "    # SAVE PER DOWNSAMPLE\n",
    "    cell_type_fraction = tmp.div(tmp.sum(axis=1),axis=0)\n",
    "    fraction_TCGA1[rep,:] = cell_type_fraction['TCGA1'].values\n",
    "    fraction_TCGA2[rep,:] = cell_type_fraction['TCGA2'].values\n",
    "    fraction_TCGA3[rep,:] = cell_type_fraction['TCGA3'].values\n",
    "    fraction_TCGA4[rep,:] = cell_type_fraction['TCGA4'].values\n",
    "    \n",
    "    \n",
    "# CONVER TO PANDAS DATAFRAME\n",
    "boxplot_data_fraction_TCGA1 = pd.DataFrame(data = fraction_TCGA1, \n",
    "                                             columns = ['{}_{}'.format(master_groupby,ind) \n",
    "                                                        for ind in \n",
    "                                                        np.unique(X.index.get_level_values(master_groupby))])\n",
    "boxplot_data_fraction_TCGA1 = boxplot_data_fraction_TCGA1.stack().rename_axis(('Rep', 'Patient')).\\\n",
    "                                                                      reset_index(name='Fraction')\n",
    "\n",
    "boxplot_data_fraction_TCGA2 = pd.DataFrame(data = fraction_TCGA2, \n",
    "                                             columns = ['{}_{}'.format(master_groupby,ind) \n",
    "                                                        for ind in \n",
    "                                                        np.unique(X.index.get_level_values(master_groupby))])\n",
    "boxplot_data_fraction_TCGA2 = boxplot_data_fraction_TCGA2.stack().rename_axis(('Rep', 'Patient')).\\\n",
    "                                                                      reset_index(name='Fraction')\n",
    "\n",
    "boxplot_data_fraction_TCGA3 = pd.DataFrame(data = fraction_TCGA3, \n",
    "                                             columns = ['{}_{}'.format(master_groupby,ind) \n",
    "                                                        for ind in \n",
    "                                                        np.unique(X.index.get_level_values(master_groupby))])\n",
    "boxplot_data_fraction_TCGA3 = boxplot_data_fraction_TCGA3.stack().rename_axis(('Rep', 'Patient')).\\\n",
    "                                                                      reset_index(name='Fraction')\n",
    "\n",
    "boxplot_data_fraction_TCGA4 = pd.DataFrame(data = fraction_TCGA4, \n",
    "                                             columns = ['{}_{}'.format(master_groupby,ind) \n",
    "                                                        for ind in \n",
    "                                                        np.unique(X.index.get_level_values(master_groupby))])\n",
    "boxplot_data_fraction_TCGA4 = boxplot_data_fraction_TCGA4.stack().rename_axis(('Rep', 'Patient')).\\\n",
    "                                                                      reset_index(name='Fraction')\n",
    "\n",
    "boxplot_data_fraction_TCGA1['Class'] = ['TCGA1']*len(boxplot_data_fraction_TCGA1)\n",
    "boxplot_data_fraction_TCGA2['Class'] = ['TCGA2']*len(boxplot_data_fraction_TCGA2)\n",
    "boxplot_data_fraction_TCGA3['Class'] = ['TCGA3']*len(boxplot_data_fraction_TCGA3)\n",
    "boxplot_data_fraction_TCGA4['Class'] = ['TCGA4']*len(boxplot_data_fraction_TCGA4)\n",
    "\n",
    "boxplot_data = boxplot_data_fraction_TCGA1.append([boxplot_data_fraction_TCGA2,\n",
    "                                                   boxplot_data_fraction_TCGA3,\n",
    "                                                   boxplot_data_fraction_TCGA4])\n",
    "boxplot_data['Patient'] = [x.replace('Patient','MSK') for x in boxplot_data.Patient]\n",
    "\n",
    "#boxplot_data[boxplot_data.Fraction < 0.0001] = 0\n",
    "boxplot_data.Fraction = [x if x > 0.001 else 1e-4 for x in boxplot_data.Fraction]\n",
    "\n",
    "\n",
    "# VIOLIN PLOT\n",
    "fig = plt.figure(figsize = (10,4))\n",
    "ax = plt.gca()\n",
    "\n",
    "# Plot violin plot\n",
    "sns.barplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class', ci=95, capsize=.1,\n",
    "                palette={\"TCGA1\": \"#0000FF\",\n",
    "                         \"TCGA2\": \"#A4D3FC\",\n",
    "                         \"TCGA3\": \"#FF7DC2\", \n",
    "                         \"TCGA4\": \"#FF0000\",},\n",
    "                linewidth=1, ax = ax, \n",
    "                alpha=1)\n",
    "\n",
    "g = sns.stripplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class',\n",
    "                palette={\"TCGA1\": \"#0000FF\",\n",
    "                         \"TCGA2\": \"#A4D3FC\",\n",
    "                         \"TCGA3\": \"#FF7DC2\", \n",
    "                         \"TCGA4\": \"#FF0000\",},\n",
    "                linewidth = 0.5, edgecolor='white',\n",
    "                ax = ax, alpha=0.7,\n",
    "                dodge=True)\n",
    "\n",
    "fig.get_axes()[0].set_yscale('log')\n",
    "\n",
    "plt.ylim((1e-4,1.15))\n",
    "g.set_ylabel(\"Cell Type Fraction\",fontsize=10)\n",
    "g.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "g.tick_params(labelsize=10)\n",
    "g.set_xticklabels(ax.get_xticklabels(),rotation=90)\n",
    "sns.despine()\n",
    "\n",
    "ax.legend_.remove()\n",
    "#plt.tight_layout()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'TCGA_Distribution_by_Patient_LOG'\n",
    "fn = '/workdir/uvmel_project/figures/Revision_2_Figures/' + figure_label \n",
    "plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "boxplot_patient_list = []\n",
    "boxplot_fract_list = []\n",
    "boxplot_class_list = []\n",
    "\n",
    "for patient in METADATA_TUMOR.index.get_level_values('Patient').unique():\n",
    "    tmp = [METADATA_TUMOR.loc[METADATA_TUMOR.index.get_level_values('Patient') == patient].\\\n",
    "           TCGA_Assignment.value_counts().reindex(METADATA_TUMOR.TCGA_Assignment.unique(), fill_value=0)[x] \n",
    "           for x in TCGA_CLASS_LIST]\n",
    "    tmp = list(tmp/sum(tmp))\n",
    "    \n",
    "    boxplot_patient_list = boxplot_patient_list + [patient for x in range(len(TCGA_CLASS_LIST))]\n",
    "    boxplot_fract_list = boxplot_fract_list + tmp\n",
    "    boxplot_class_list = boxplot_class_list + TCGA_CLASS_LIST\n",
    "\n",
    "boxplot_data = pd.DataFrame()\n",
    "boxplot_data['Patient'] = boxplot_patient_list\n",
    "boxplot_data['Fraction'] = boxplot_fract_list\n",
    "boxplot_data['Class'] = boxplot_class_list\n",
    "\n",
    "# Set plot\n",
    "fig = plt.figure(figsize = (10,4))\n",
    "ax = plt.gca()\n",
    "\n",
    "# Plot bar plot\n",
    "g = sns.barplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class', ci=95, capsize=.2,\n",
    "                palette={\"TCGA1\": \"#0000FF\",\n",
    "                         \"TCGA2\": \"#A4D3FC\",\n",
    "                         \"TCGA3\": \"#FF7DC2\", \n",
    "                         \"TCGA4\": \"#FF0000\",},\n",
    "                linewidth = 1, ax = ax);\n",
    "fig.get_axes()[0].set_yscale('log')\n",
    "\n",
    "plt.ylim((0,1))\n",
    "g.set_ylabel(\"Cell Type Fraction\",fontsize=10)\n",
    "g.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "g.tick_params(labelsize=10)\n",
    "g.set_xticklabels(ax.get_xticklabels(),rotation=90)\n",
    "sns.despine()\n",
    "\n",
    "ax.legend_.remove()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'TCGA_Distribution_by_Patient_LOG'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### MIXTURE MODEL FOR GEP CLASSIFICATION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create GEP marker_file dataframe containing gene signatures for each GEP class:\n",
    "\n",
    "GEP_markerfile_df = pd.DataFrame() \n",
    "GEP_markerfile_df['GEP1'] = [1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0] \n",
    "GEP_markerfile_df['GEP2'] = [0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1] \n",
    "GEP_markerfile_df.index = ['ID2', 'FXR1', 'EIF1B', 'LTA4H', 'MTUS1', 'SATB1', 'LMCD1', 'ROBO1', \n",
    "                           'CDH1', 'ECM1', 'HTR2B', 'RAB31']\n",
    "\n",
    "pd.DataFrame(columns=['GEP1','GEP2']).to_csv(DATA_PATH + 'cellAssign/GEP_markerfile.csv', index=False) \n",
    "GEP_markerfile_df.to_csv(DATA_PATH + 'cellAssign/GEP_markerfile.csv', header=None, mode='a')\n",
    "GEP_markerfile_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_type = 'TUMOR'\n",
    "# LOAD GENE LIST FROM EXCEL\n",
    "path_to_genesets = DATA_PATH+'uveal_melanoma_MB_v3.csv'\n",
    "genesets = pd.read_csv(path_to_genesets,header='infer')\n",
    "genesets = genesets.apply(lambda x: x.astype(str).str.upper())\n",
    "print(shape(genesets)[1])\n",
    "\n",
    "# Using imputed. normalized data:\n",
    "datatype = 'INDF_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "exec('META = METADATA_{}'.format(subset_type))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get mean gene expression for signature genes for each GEP class:\n",
    "gene1 = 'Castle 1'\n",
    "signature_genes = genesets[gene1].values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "vals1 = (np.nanmean(QUERY[detected_genes],axis=1))\n",
    "\n",
    "gene2 = 'Castle 2'\n",
    "signature_genes = genesets[gene2].values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "vals2 = (np.nanmean(QUERY[detected_genes],axis=1))\n",
    "\n",
    "X = pd.DataFrame(data = {gene1: vals1, gene2: vals2}, index = QUERY.index)\n",
    "\n",
    "# Train BGMM model:\n",
    "bgmm = BayesianGaussianMixture(covariance_type='diag', n_components=2, random_state=1)\n",
    "bgmm.fit(X[['Castle 1','Castle 2']])\n",
    "\n",
    "# Use trained BGMM to make predictions:\n",
    "clusters = bgmm.predict(X[['Castle 1','Castle 2']])\n",
    "X['Assignment'] = clusters\n",
    "\n",
    "# Make Class Assignments\n",
    "lut ={}\n",
    "for ii in np.unique(X['Assignment']):\n",
    "    if X[X['Assignment']==ii]['Castle 1'].mean()>X[X['Assignment']==ii]['Castle 2'].mean():\n",
    "        lut[ii] = 'Castle 1'\n",
    "    else:\n",
    "        lut[ii] = 'Castle 2'\n",
    "    \n",
    "X['Assignment'] = X['Assignment'].map(lut)\n",
    "\n",
    "\n",
    "# Plot results:\n",
    "sns.set(font_scale=2.5)\n",
    "sns.set_style(\"white\")\n",
    "g = sns.pairplot(X[['Castle 1','Castle 2','Assignment']], \n",
    "                 hue = 'Assignment',\n",
    "                 palette={\"Castle 1\": \"#0000FF\", \"Castle 2\": \"#FF0000\", 'Mixed': '#000000'},\n",
    "                 diag_kind = 'kde', aspect = 1, size = 10,\n",
    "                 diag_kws={\"shade\": False}) # SAVE FIGURE\n",
    "\n",
    "g.set(xlabel='', ylabel='')\n",
    "\n",
    "# SAVE FIGURE\n",
    "#figure_label = 'MSK_BayesianGMM'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "\n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "#g.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "#g.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# UPDATE DIRECTORY\n",
    "META['Assignment'] = X.Assignment\n",
    "METADATA_TUMOR = META.copy()\n",
    "exec('h5_data.save(METADATA_{}, \\'METADATA_{}\\')'.format(subset_type,subset_type))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### COLOR GEP PLOT BY TCGA CLASS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X['TCGA_Class'] = METADATA_TUMOR.TCGA_Assignment\n",
    "\n",
    "sns.set(font_scale=2.5)\n",
    "sns.set_style(\"white\")\n",
    "g = sns.pairplot(X[['Castle 1','Castle 2','Assignment', 'TCGA_Class']], \n",
    "                 hue = 'TCGA_Class',\n",
    "                 palette={\"TCGA1\": \"#0000FF\",\n",
    "                          \"TCGA2\": \"#A4D3FC\",\n",
    "                          \"TCGA3\": \"#FF7DC2\", \n",
    "                          \"TCGA4\": \"#FF0000\",},\n",
    "            diag_kind = 'kde', aspect = 1, size = 10);# SAVE FIGURE\n",
    "\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'MSK_Mixture_Model_FULL_TCGA_colored'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "\n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "g.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "g.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### MOVE ON WITH INTRATUMOR HETEROGENEITY FOR GEP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Second GEP assignment scatter plot:\n",
    "\n",
    "plt.figure(figsize = (5,5))\n",
    "sns.despine()\n",
    "sns.set(rc={'axes.facecolor':'white', 'figure.facecolor':'white'})\n",
    "ax = plt.gca()\n",
    "\n",
    "ax = plt.scatter(X[X['Assignment']=='Castle 1']['Castle 2'], X[X['Assignment']=='Castle 1']['Castle 1'],\n",
    "                 s = 1, c ='blue',alpha = 0.5)\n",
    "ax = sns.kdeplot(X[X['Assignment']=='Castle 1']['Castle 2'], X[X['Assignment']=='Castle 1']['Castle 1'],\n",
    "                 cmap=\"Blues\", shade=True, shade_lowest=False,n_levels = 20, alpha = 0.5)\n",
    "\n",
    "ax = plt.scatter(X[X['Assignment']=='Castle 2']['Castle 2'], X[X['Assignment']=='Castle 2']['Castle 1'],\n",
    "                 s = 1, c ='red',alpha = 0.5)\n",
    "ax = sns.kdeplot(X[X['Assignment']=='Castle 2']['Castle 2'], X[X['Assignment']=='Castle 2']['Castle 1'],\n",
    "                 cmap=\"Reds\", shade=True, shade_lowest=False, n_levels = 20, alpha = 0.5)\n",
    "\n",
    "\n",
    "# SAVE FIGURE\n",
    "# SAVE FIGURE\n",
    "figure_label = 'MSK_Mixture_Model'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "\n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update multi-index\n",
    "new_index = pd.MultiIndex.from_tuples(list(zip(X.index.get_level_values('Sample ID'), \n",
    "                                               X.index.get_level_values('Legend'),\n",
    "                                               X.index.get_level_values('Patient'),\n",
    "                                               X.index.get_level_values('Cell ID'),\n",
    "                                               X.index.get_level_values('Phenograph_Class'),\n",
    "                                               X['Assignment'])), \n",
    "                                  names=['Sample ID','Legend', 'Patient','Cell ID','Phenograph_Class','Assignment'])\n",
    "X = pd.DataFrame(data = X.values, columns = X.columns, index = new_index)\n",
    "\n",
    "# CLASS ASSIGNMENT BY PATIENT \n",
    "meta = 'Assignment'\n",
    "exec('tmp = X.groupby(level=[\\'Patient\\', \\'{}\\'], axis=0).size().unstack().fillna(0)'.format(meta))\n",
    "colors = [\"#0000FF\",\"#FF0000\"]\n",
    "\n",
    "plt.figure(figsize = (5,3))\n",
    "ax = plt.gca()\n",
    "tmp.div(tmp.sum(axis=1),axis=0).plot.barh(stacked=True, color=colors, ax = ax, width = 0.95)\n",
    "ax.legend_.remove()\n",
    "ax.set_frame_on(False)\n",
    "plt.xticks(fontsize=8)\n",
    "plt.yticks(fontsize=8)\n",
    "plt.xlabel('Cell Fraction', fontsize=12)\n",
    "plt.ylabel('Patient', fontsize=12)\n",
    "\n",
    "# SAVE FIGURE\n",
    "#figure_label = 'GEP_Distribution_by_Patient'\n",
    "#fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label\n",
    "#plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "#plt.savefig(fn + '.pdf', dpi=400)\n",
    "#print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Down sampled assignment distribution (not used in paper anymore) \n",
    "\n",
    "# GENE-GENE COVARIANCE WITHIN NORMAL SUBSET\n",
    "n = 500\n",
    "reps = np.arange(20)\n",
    "\n",
    "master_groupby = 'Patient'\n",
    "ix = [ind for ind,name in enumerate(X.index.names) if name==master_groupby][0]\n",
    "fraction_castle1 = np.zeros((len(reps),len(np.unique(X.index.get_level_values(master_groupby)))))\n",
    "fraction_castle2 = np.zeros((len(reps),len(np.unique(X.index.get_level_values(master_groupby)))))\n",
    "\n",
    "for rep in reps:\n",
    "    # RANDOMLY SAMPLE SAME NUMBER OF CELLS PER PATIENT\n",
    "    SUBSET = pd.DataFrame()\n",
    "    for class_selection in np.unique(X.index.get_level_values(master_groupby)):\n",
    "        CSUBSET = X.loc[X.index.map(lambda x: x[ix] in [class_selection])].sample(n=n,replace=True)\n",
    "        SUBSET = SUBSET.append(CSUBSET)\n",
    "    \n",
    "    bgmm = BayesianGaussianMixture(covariance_type='diag', n_components=2, random_state=1)\n",
    "    bgmm.fit(SUBSET[['Castle 1','Castle 2']])\n",
    "    clusters = bgmm.predict(SUBSET[['Castle 1','Castle 2']])\n",
    "    SUBSET['Assignment'] = clusters\n",
    "\n",
    "    # Make Class Assignments\n",
    "    lut ={}\n",
    "    for ii in np.unique(SUBSET['Assignment']):\n",
    "        if SUBSET[SUBSET['Assignment']==ii]['Castle 1'].mean()>SUBSET[SUBSET['Assignment']==ii]['Castle 2'].mean():\n",
    "            lut[ii] = 'Castle 1'\n",
    "        else:\n",
    "            lut[ii] = 'Castle 2'\n",
    "    \n",
    "    SUBSET['Assignment'] = SUBSET['Assignment'].map(lut)\n",
    "\n",
    "    # Update multi-index\n",
    "    new_index = pd.MultiIndex.from_tuples(list(zip(SUBSET.index.get_level_values('Sample ID'), \n",
    "                                                   SUBSET.index.get_level_values('Legend'),\n",
    "                                                   SUBSET.index.get_level_values('Patient'),\n",
    "                                                   SUBSET.index.get_level_values('Cell ID'),\n",
    "                                                   SUBSET.index.get_level_values('Phenograph_Class'),\n",
    "                                                   SUBSET['Assignment'])), \n",
    "                                          names=['Sample ID','Legend', 'Patient','Cell ID',\n",
    "                                                 'Phenograph_Class','Assignment'])\n",
    "    SUBSET = pd.DataFrame(data = SUBSET.values, columns = SUBSET.columns, index = new_index)\n",
    "\n",
    "    # CASTLE ASSIGNMENT PER PATIENT\n",
    "    meta = 'Assignment'\n",
    "    exec('tmp = SUBSET.groupby(level=[\\'Patient\\', \\'{}\\'], axis=0).size().unstack().fillna(0)'.format(meta))\n",
    "\n",
    "    # SAVE PER DOWNSAMPLE\n",
    "    cell_type_fraction = tmp.div(tmp.sum(axis=1),axis=0)\n",
    "    fraction_castle1[rep,:] = cell_type_fraction['Castle 1'].values\n",
    "    fraction_castle2[rep,:] = cell_type_fraction['Castle 2'].values\n",
    "    \n",
    "# CONVER TO PANDAS DATAFRAME\n",
    "boxplot_data_fraction_castle1 = pd.DataFrame(data = fraction_castle1, \n",
    "                                             columns = ['{}_{}'.format(master_groupby,ind) \n",
    "                                                        for ind in \n",
    "                                                        np.unique(X.index.get_level_values(master_groupby))])\n",
    "boxplot_data_fraction_castle1 = boxplot_data_fraction_castle1.stack().rename_axis(('Rep', 'Patient')).\\\n",
    "                                                                      reset_index(name='Fraction')\n",
    "\n",
    "boxplot_data_fraction_castle2 = pd.DataFrame(data = fraction_castle2, \n",
    "                                             columns = ['{}_{}'.format(master_groupby,ind) \n",
    "                                                        for ind in \n",
    "                                                        np.unique(X.index.get_level_values(master_groupby))])\n",
    "boxplot_data_fraction_castle2 = boxplot_data_fraction_castle2.stack().rename_axis(('Rep', 'Patient')).\\\n",
    "                                                                      reset_index(name='Fraction')\n",
    "\n",
    "boxplot_data_fraction_castle1['Class'] = ['Castle 1']*len(boxplot_data_fraction_castle1)\n",
    "boxplot_data_fraction_castle2['Class'] = ['Castle 2']*len(boxplot_data_fraction_castle2)\n",
    "boxplot_data = boxplot_data_fraction_castle1.append(boxplot_data_fraction_castle2)\n",
    "boxplot_data['Patient'] = [x.replace('Patient','MSK') for x in boxplot_data.Patient]\n",
    "\n",
    "#boxplot_data[boxplot_data.Fraction < 0.0001] = 0\n",
    "boxplot_data.Fraction = [x if x > 0.001 else 1e-4 for x in boxplot_data.Fraction]\n",
    "\n",
    "# VIOLIN PLOT\n",
    "fig = plt.figure(figsize = (10,4))\n",
    "ax = plt.gca()\n",
    "\n",
    "# Plot violin plot\n",
    "sns.barplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class', ci=95, capsize=.2,\n",
    "                palette={\"Castle 1\": \"#0000FF\", \"Castle 2\": \"#FF0000\", 'Mixed': '#000000'},\n",
    "                linewidth=1, ax = ax, alpha=1)\n",
    "\n",
    "g = sns.stripplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class',\n",
    "                palette={\"Castle 1\": \"#0000FF\", \"Castle 2\": \"#FF0000\", 'Mixed': '#000000'},\n",
    "                linewidth = 0.5, edgecolor='white',\n",
    "                ax = ax, alpha=0.7,\n",
    "                dodge=True)\n",
    "\n",
    "fig.get_axes()[0].set_yscale('log')\n",
    "\n",
    "plt.ylim((1e-4,1.15))\n",
    "g.set_ylabel(\"Cell Type Fraction\",fontsize=10)\n",
    "g.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "g.tick_params(labelsize=10)\n",
    "g.set_xticklabels(ax.get_xticklabels(),rotation=90)\n",
    "sns.despine()\n",
    "\n",
    "ax.legend_.remove()\n",
    "#plt.tight_layout()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'GEP_Distribution_by_Patient_LOG'\n",
    "fn = '/workdir/uvmel_project/figures/Revision_2_Figures/' + figure_label \n",
    "plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "boxplot_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SUBSET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GENE-GENE COVARIANCE WITHIN NORMAL SUBSET\n",
    "n = 500\n",
    "reps = np.arange(20)\n",
    "\n",
    "boxplot_patient_list = []\n",
    "boxplot_fract_list = []\n",
    "boxplot_class_list = []\n",
    "GEP_CLASS_LIST = ['Castle 1', 'Castle 2']\n",
    "master_groupby = 'Patient'\n",
    "\n",
    "for rep in reps:\n",
    "    for patient in METADATA_TUMOR.index.get_level_values('Patient').unique():\n",
    "        SUBSET = pd.DataFrame()\n",
    "        for class_selection in np.unique(METADATA_TUMOR.index.get_level_values(master_groupby)):\n",
    "            CSUBSET = METADATA_TUMOR.loc[METADATA_TUMOR.index.map(lambda x: x[ix] in \n",
    "                                                                  [class_selection])].sample(n=n,replace=True)\n",
    "            SUBSET = SUBSET.append(CSUBSET)\n",
    "        \n",
    "        tmp = [SUBSET.loc[SUBSET.index.get_level_values('Patient') == patient].\\\n",
    "               Assignment.value_counts().reindex(SUBSET.Assignment.unique(), fill_value=0)[x] \n",
    "               for x in GEP_CLASS_LIST]\n",
    "        tmp = list(tmp/sum(tmp))\n",
    "    \n",
    "        boxplot_patient_list = boxplot_patient_list + [patient for x in range(len(GEP_CLASS_LIST))]\n",
    "        boxplot_fract_list = boxplot_fract_list + tmp\n",
    "        boxplot_class_list = boxplot_class_list + GEP_CLASS_LIST\n",
    "\n",
    "boxplot_data = pd.DataFrame()\n",
    "boxplot_data['Patient'] = ['MSK_' + x for x in boxplot_patient_list]\n",
    "boxplot_data['Fraction'] = boxplot_fract_list\n",
    "boxplot_data['Class'] = boxplot_class_list\n",
    "\n",
    "boxplot_data.Fraction = [x if x > 0.0001 else 1e-4 for x in boxplot_data.Fraction]\n",
    "\n",
    "# Set plot\n",
    "fig = plt.figure(figsize = (7,4))\n",
    "ax = plt.gca()\n",
    "\n",
    "# Plot bar plot\n",
    "\n",
    "sns.barplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class', ci=95, capsize=0.15,\n",
    "                palette={\"Castle 1\": \"#0000FF\", \"Castle 2\": \"#FF0000\", 'Mixed': '#000000'},\n",
    "                linewidth=1, ax = ax, alpha=1,\n",
    "                errcolor='black',\n",
    "                errwidth=1)\n",
    "\n",
    "g = sns.stripplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class',\n",
    "                palette={\"Castle 1\": \"#0000FF\", \"Castle 2\": \"#FF0000\", 'Mixed': '#000000'},\n",
    "                linewidth = 0.5, edgecolor='white',\n",
    "                ax = ax, alpha=0.7,\n",
    "                dodge=True)\n",
    "\n",
    "fig.get_axes()[0].set_yscale('log')\n",
    "\n",
    "plt.ylim((1e-4,1.15))\n",
    "ax.set_ylabel(\"Cell Type Fraction\",fontsize=10)\n",
    "ax.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "ax.tick_params(labelsize=10)\n",
    "ax.set_xticklabels(boxplot_data['Patient'].unique(),rotation=90)\n",
    "sns.despine()\n",
    "\n",
    "ax.legend_.remove()\n",
    "plt.tight_layout()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'GEP_Distribution_by_Patient_LOG'\n",
    "fn = '/workdir/uvmel_project/figures/Revision_2_Figures/' + figure_label \n",
    "plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "boxplot_patient_list = []\n",
    "boxplot_fract_list = []\n",
    "boxplot_class_list = []\n",
    "GEP_CLASS_LIST = ['Castle 1', 'Castle 2']\n",
    "\n",
    "for patient in METADATA_TUMOR.index.get_level_values('Patient').unique():\n",
    "    tmp = [METADATA_TUMOR.loc[METADATA_TUMOR.index.get_level_values('Patient') == patient].\\\n",
    "           Assignment.value_counts().reindex(METADATA_TUMOR.Assignment.unique(), fill_value=0)[x] \n",
    "           for x in GEP_CLASS_LIST]\n",
    "    tmp = list(tmp/sum(tmp))\n",
    "    \n",
    "    boxplot_patient_list = boxplot_patient_list + [patient for x in range(len(GEP_CLASS_LIST))]\n",
    "    boxplot_fract_list = boxplot_fract_list + tmp\n",
    "    boxplot_class_list = boxplot_class_list + GEP_CLASS_LIST\n",
    "\n",
    "boxplot_data = pd.DataFrame()\n",
    "boxplot_data['Patient'] = ['MSK_' + x for x in boxplot_patient_list]\n",
    "boxplot_data['Fraction'] = boxplot_fract_list\n",
    "boxplot_data['Class'] = boxplot_class_list\n",
    "\n",
    "# Set plot\n",
    "fig = plt.figure(figsize = (7,4))\n",
    "ax = plt.gca()\n",
    "\n",
    "# Plot bar plot\n",
    "\n",
    "# Castle 1:\n",
    "ax.bar(x = boxplot_data.loc[boxplot_data.Class == 'Castle 1']['Patient'], \n",
    "        height = boxplot_data.loc[boxplot_data.Class == 'Castle 1']['Fraction'], capsize=.2,\n",
    "        color=\"#0505C5\",\n",
    "        linewidth = 1,\n",
    "        width=0.5)\n",
    "\n",
    "# Castle 2:\n",
    "ax.bar(x = boxplot_data.loc[boxplot_data.Class == 'Castle 2']['Patient'], \n",
    "        height = boxplot_data.loc[boxplot_data.Class == 'Castle 2']['Fraction'], capsize=.2,\n",
    "        color=\"#D60808\",\n",
    "        linewidth = 1,\n",
    "        bottom=boxplot_data.loc[boxplot_data.Class == 'Castle 1']['Fraction'],\n",
    "        width=0.5)\n",
    "\n",
    "fig.get_axes()[0].set_yscale('log')\n",
    "\n",
    "plt.ylim((0,1))\n",
    "ax.set_ylabel(\"Cell Type Fraction\",fontsize=10)\n",
    "ax.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "ax.tick_params(labelsize=10)\n",
    "ax.set_xticklabels(boxplot_data['Patient'],rotation=90)\n",
    "sns.despine()\n",
    "\n",
    "#ax.legend_.remove()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'GEP_Distribution_by_Patient_LOG_NO_ds'\n",
    "fn = '/workdir/uvmel_project/figures/Revision_2_Figures/' + figure_label \n",
    "plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get GEP distributions per patient:\n",
    "\n",
    "boxplot_patient_list = []\n",
    "boxplot_fract_list = []\n",
    "boxplot_class_list = []\n",
    "GEP_CLASS_LIST = ['Castle 1', 'Castle 2']\n",
    "\n",
    "for patient in METADATA_TUMOR.index.get_level_values('Patient').unique():\n",
    "    tmp = [METADATA_TUMOR.loc[METADATA_TUMOR.index.get_level_values('Patient') == patient].\\\n",
    "           Assignment.value_counts().reindex(METADATA_TUMOR.Assignment.unique(), fill_value=0)[x] \n",
    "           for x in GEP_CLASS_LIST]\n",
    "    tmp = list(tmp/sum(tmp))\n",
    "    \n",
    "    boxplot_patient_list = boxplot_patient_list + [patient for x in range(len(GEP_CLASS_LIST))]\n",
    "    boxplot_fract_list = boxplot_fract_list + tmp\n",
    "    boxplot_class_list = boxplot_class_list + GEP_CLASS_LIST\n",
    "\n",
    "boxplot_data = pd.DataFrame()\n",
    "boxplot_data['Patient'] = ['MSK_' + x for x in boxplot_patient_list]\n",
    "boxplot_data['Fraction'] = boxplot_fract_list\n",
    "boxplot_data['Class'] = boxplot_class_list\n",
    "\n",
    "# VIOLIN PLOT\n",
    "fig = plt.figure(figsize = (8,6))\n",
    "ax = plt.gca()\n",
    "\n",
    "# Plot violin plot\n",
    "g = sns.barplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class', ci=95, capsize=.2,\n",
    "                palette={\"Castle 1\": \"#0000FF\", \"Castle 2\": \"#FF0000\", 'Mixed': '#000000'},\n",
    "                linewidth = 1, ax = ax);\n",
    "fig.get_axes()[0].set_yscale('log')\n",
    "\n",
    "plt.ylim((0,1))\n",
    "g.set_ylabel(\"scRNA-seq \\n Tumor Fraction\", fontsize=18)\n",
    "g.set_xlabel(\" \", fontsize=10, rotation = 90)\n",
    "g.tick_params(labelsize=18)\n",
    "g.set_xticklabels(ax.get_xticklabels(),rotation=90)\n",
    "#sns.set_style(\"noticks\")\n",
    "sns.despine()\n",
    "\n",
    "ax.legend_.remove()\n",
    "plt.tight_layout()\n",
    "g.axis(True)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'GEP_Distribution_by_Patient_LOG_NO_ds'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "boxplot_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### TCGA DIAMOND PLOTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CONVERT HEX TO RGB (FLATUI_CASTLE)\n",
    "ind_GEP = ['Castle 1', 'Castle 2']\n",
    "colors = np.zeros((len(FLATUI_CASTLE),3))\n",
    "for ii,hexcolor in enumerate(FLATUI_CASTLE):\n",
    "    colors[ii,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "cix = (np.linspace(0,shape(colors)[0],len(np.unique(ind_GEP)))).astype(int)\n",
    "if cix[len(cix)-1]==shape(colors)[0]:\n",
    "    cix[len(cix)-1]=shape(colors)[0]-1\n",
    "lut_CASTLE = dict(zip(np.unique(ind_GEP), colors[cix,:]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GEP_color_set = [pd.Series(METADATA_TUMOR.loc[METADATA_TUMOR.index.map(lambda x: x[2] == y)].\\\n",
    "                           Assignment.values.tolist()).map(lut_CASTLE)\n",
    "                 for y in METADATA_TUMOR.index.get_level_values('Patient').unique()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diamond_list = [plt.Rectangle((0,-1), np.sqrt(2), np.sqrt(2), 45, color='black', fill=False, linewidth=0.5) \n",
    "                for x in range(6)]\n",
    "\n",
    "fig1, axes = plt.subplots(2,3)\n",
    "\n",
    "TCGA_probs_x_MSK = [X_norm.loc[X_norm.index.map(lambda x: x[2] == y)].TCGA2.values - \n",
    "                    X_norm.loc[X_norm.index.map(lambda x: x[2] == y)].TCGA4.values\n",
    "                    for y in X_norm.index.unique(level='Patient')]\n",
    "\n",
    "TCGA_probs_y_MSK = [X_norm.loc[X_norm.index.map(lambda x: x[2] == y)].TCGA1.values - \n",
    "                    X_norm.loc[X_norm.index.map(lambda x: x[2] == y)].TCGA3.values\n",
    "                    for y in X_norm.index.unique(level='Patient')]\n",
    "\n",
    "TCGA_probs_per_patient_MSK = [[(TCGA_probs_x_MSK[x][y], TCGA_probs_y_MSK[x][y]) \n",
    "                               for y in range(len(TCGA_probs_x_MSK[x]))]\n",
    "                               for x in range(len(TCGA_probs_x_MSK))]\n",
    "\n",
    "for i, ax in enumerate(axes.flatten()):\n",
    "    ax.scatter(*zip(*TCGA_probs_per_patient_MSK[i]), s=1, color=GEP_color_set[i],\n",
    "               alpha=0.25)\n",
    "    ax.set_frame_on(False)\n",
    "    ax.axes.get_xaxis().set_visible(False)\n",
    "    ax.axes.get_yaxis().set_visible(False)\n",
    "    ax.add_patch(diamond_list[i])\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'TCGA_PROBS_DiamondPlot'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig1.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig1.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diamond_list = [plt.Rectangle((0,-1), np.sqrt(2), np.sqrt(2), 45, color='black', fill=False, linewidth=0.5)\n",
    "                for x in range(6)]\n",
    "\n",
    "fig1, axes = plt.subplots(2,3)\n",
    "\n",
    "TCGA_probs_x_MSK = [X_norm.loc[X_norm.index.map(lambda x: x[2] == y)].TCGA2.values - \n",
    "                    X_norm.loc[X_norm.index.map(lambda x: x[2] == y)].TCGA4.values\n",
    "                    for y in X_norm.index.unique(level='Patient')]\n",
    "\n",
    "TCGA_probs_y_MSK = [X_norm.loc[X_norm.index.map(lambda x: x[2] == y)].TCGA1.values - \n",
    "                    X_norm.loc[X_norm.index.map(lambda x: x[2] == y)].TCGA3.values\n",
    "                    for y in X_norm.index.unique(level='Patient')]\n",
    "\n",
    "TCGA_probs_per_patient_MSK = [[(TCGA_probs_x_MSK[x][y], TCGA_probs_y_MSK[x][y]) \n",
    "                               for y in range(len(TCGA_probs_x_MSK[x]))]\n",
    "                               for x in range(len(TCGA_probs_x_MSK))]\n",
    "\n",
    "for i, ax in enumerate(axes.flatten()):\n",
    "    ax.scatter(*zip(*TCGA_probs_per_patient_MSK[i]), s=0.2, color='black',\n",
    "               alpha=0.3)\n",
    "    ax.set_frame_on(False)\n",
    "    ax.axes.get_xaxis().set_visible(False)\n",
    "    ax.axes.get_yaxis().set_visible(False)\n",
    "    ax.add_patch(diamond_list[i])\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'TCGA_PROBS_DiamondPlot_monochrome'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig1.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig1.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CONVERT HEX TO RGB (FLATUI_CASTLE)\n",
    "FLATUI_TCGA = ['0000FF', 'A4D3FC', 'FF7DC2', 'FF0000']\n",
    "\n",
    "ind_TCGA = TCGA_CLASS_LIST\n",
    "colors = np.zeros((len(FLATUI_TCGA),3))\n",
    "for ii,hexcolor in enumerate(FLATUI_TCGA):\n",
    "    colors[ii,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "cix = (np.linspace(0,shape(colors)[0],len(np.unique(ind_TCGA)))).astype(int)\n",
    "if cix[len(cix)-1]==shape(colors)[0]:\n",
    "    cix[len(cix)-1]=shape(colors)[0]-1\n",
    "lut_TCGA = dict(zip(np.unique(ind_TCGA), colors[cix,:]))\n",
    "CM_TCGA = LinearSegmentedColormap.from_list('FLATUI_TCGA', colors, N=len(colors))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "TCGA_color_set = [pd.Series(METADATA_TUMOR.loc[METADATA_TUMOR.index.map(lambda x: x[2] == y)].\\\n",
    "                           TCGA_Assignment.values.tolist()).map(lut_TCGA)\n",
    "                 for y in METADATA_TUMOR.index.get_level_values('Patient').unique()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diamond_list = [plt.Rectangle((0,-1), np.sqrt(2), np.sqrt(2), 45, color='black', fill=False, linewidth=0.5)\n",
    "                for x in range(6)]\n",
    "\n",
    "fig1, axes = plt.subplots(2,3)\n",
    "\n",
    "for i, ax in enumerate(axes.flatten()):\n",
    "    ax.scatter(*zip(*TCGA_probs_per_patient_MSK[i]), s=1, color=TCGA_color_set[i],\n",
    "               alpha=0.25)\n",
    "    ax.set_frame_on(False)\n",
    "    ax.axes.get_xaxis().set_visible(False)\n",
    "    ax.axes.get_yaxis().set_visible(False)\n",
    "    ax.add_patch(diamond_list[i])\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'TCGA_PROBS_DiamondPlot_tcga_colored'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig1.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig1.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GEP CLINICAL PROGNOSTIFICATION OF INDIVIDUAL TUMOR CELLS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GEP SIGNATURE CLUSTERED HEATMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class_genes = GEP_markerfile_df.index.values.tolist()\n",
    "\n",
    "# Working with imputed normalized tumor cell data:\n",
    "subset_type = 'TUMOR'\n",
    "datatype = 'INDF_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "\n",
    "# COLUMN INDEX AND COLORS\n",
    "genes = [gene for gene in class_genes if gene in QUERY.columns]#gene_array['Gene'].values\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "heatmap_data = pd.DataFrame(data = zscore(QUERY[genes].values,axis=0),columns = genes, index = QUERY.index)\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns\n",
    "\n",
    "# LINKAGE \n",
    "method = 'average' # average, single\n",
    "metric = 'euclidean' # cosine\n",
    "linkage = hc.linkage(heatmap_data, method=method, metric = metric)\n",
    "row_linkage = deepcopy(linkage)\n",
    "linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "col_linkage = deepcopy(linkage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# REORDER HEATMAP ACCORDING TO LINKAGE (OPTIONAL, STILL SLOW)\n",
    "r1 = hc.leaves_list(row_linkage)\n",
    "c1 = hc.leaves_list(col_linkage)\n",
    "mat = heatmap_data.iloc[r1,c1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GEP_assignments_sorted = [METADATA_TUMOR.loc[METADATA_TUMOR.index.get_level_values('Cell ID') == x].\\\n",
    "                          Assignment.\\\n",
    "                          values[0] for x in mat.index.get_level_values('Cell ID').values]\n",
    "\n",
    "row_colors = pd.Series(GEP_assignments_sorted, index=GEP_assignments_sorted).map(lut_CASTLE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(4,10))\n",
    "plt.rcParams[\"axes.grid\"] = False\n",
    "\n",
    "# ADD ROW COLOR INDEX 1 (PHENOGRAPH CLASS)\n",
    "ax1 = fig.add_axes([0,0.1,0.05,0.6]) # [x0,y0,width,height]\n",
    "x = 0\n",
    "y = 0\n",
    "for c in row_colors:\n",
    "    pos = (x, y / len(row_colors))\n",
    "    ax1.add_patch(patches.Rectangle(pos, 1, 1 / len(row_colors), color=c))\n",
    "    if y >= len(row_colors)-1:\n",
    "        x += 1\n",
    "        y = 0\n",
    "    else:\n",
    "        y += 1\n",
    "plt.axis('off')\n",
    "\n",
    "# ADD MATRIX WITH GENE NAMES\n",
    "axmatrix = fig.add_axes([0.05,0.1,0.9,0.6])\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=plt.cm.RdBu_r,vmin=-1,vmax=1)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_frame_on(False)\n",
    "axmatrix.set_yticklabels([]);\n",
    "xtick = plt.xticks(range(len(labels)), labels, rotation = 90, fontsize = 14)\n",
    "plt.tick_params(size=0)\n",
    "\n",
    "# ADD COLORBAR\n",
    "axcolor = fig.add_axes([1.1,0.1,0.1,0.1])\n",
    "cbar = plt.colorbar(im, cax=axcolor)\n",
    "#cbar.ax.get_yaxis().set_ticks([])\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'GEP_Heatmap_BayesianGMM'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PAIRWISE CORRELATION BETWEEN MONOSOMY 3 SIGNATURE AND GEP2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using imputed normalized data for tumor cells:\n",
    "subset_type = 'TUMOR'\n",
    "datatype = 'INDF_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "\n",
    "meta = 'Assignment'\n",
    "\n",
    "gene1 = 'Castle 2'\n",
    "signature_genes = genesets[gene1].values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "G1 = (np.nansum(QUERY[detected_genes],axis=1))#*complexity\n",
    "\n",
    "gene2 = 'Monosomy 3 Up'\n",
    "signature_genes = genesets[gene2].values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "G2 = (np.nansum(QUERY[detected_genes],axis=1))#*complexity\n",
    "\n",
    "# ROW INDEX\n",
    "meta = 'Assignment'\n",
    "groupby_type = meta # 'Class', 'Meta-Source', 'Legend'\n",
    "FLATUI = ['0000FF','FF0000']#FLATUI_CLASS, 'FLATUI_SOURCE, FLATUI_SAMPLES\n",
    "\n",
    "violin_data = pd.DataFrame({\n",
    "                            meta:list(META[meta]),\n",
    "                            gene1: G1,\n",
    "                            gene2: G2,\n",
    "                           })\n",
    "\n",
    "ind = violin_data[groupby_type]\n",
    "\n",
    "# CONVERT HEX TO RGB (FLATUI_CLASS)\n",
    "colors = np.zeros((len(FLATUI),3))\n",
    "for ii,hexcolor in enumerate(FLATUI):\n",
    "    colors[ii,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "# Palatte for Class METADATA\n",
    "cix = (np.linspace(0,shape(colors)[0],len(np.unique(ind)))).astype(int)\n",
    "if cix[len(cix)-1]==shape(colors)[0]:\n",
    "    cix[len(cix)-1]=shape(colors)[0]-1\n",
    "lut = dict(zip(np.unique(ind), colors[cix,:]))\n",
    "dot_colors = pd.Series(ind).map(lut)\n",
    "metacell_colors = [rgb2hex(int(color[0]*255), int(color[1]*255), int(color[2]*255)) for color in dot_colors]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# DISTRIBUTION OF SOX2/SOX9\n",
    "x = gene1\n",
    "y = gene2\n",
    "\n",
    "plt.figure(figsize = (20,20))\n",
    "sns.despine()\n",
    "sns.set(style='ticks')\n",
    "\n",
    "# JOINTPLOT\n",
    "g = sns.jointplot(x, y, data=violin_data.fillna(0), kind=\"reg\", stat_func=None, ratio=2, color='k', size=10)\n",
    "g.ax_joint.cla() # or g.ax_joint.collections[0].set_visible(False), as per mwaskom's comment\n",
    "\n",
    "# REPLOT SCATTER WITH LINEAGE COLORED\n",
    "plt.sca(g.ax_joint)\n",
    "plt.scatter(violin_data[x], violin_data[y], c=dot_colors)\n",
    "\n",
    "# ADD REGRESSION\n",
    "xx = violin_data[x]\n",
    "yy = violin_data[y]\n",
    "f = lambda x, *p: polyval(p, x)\n",
    "p, cov = curve_fit(f, xx, yy, [1, 1])\n",
    "\n",
    "# simulated draws from the probability density function of the regression\n",
    "xi = linspace(np.min(xx), np.max(xx), 100)\n",
    "ps = np.random.multivariate_normal(p, cov, 10000)\n",
    "ysample = np.asarray([f(xi, *pi) for pi in ps])\n",
    "lower = percentile(ysample, 10, axis=0)\n",
    "upper = percentile(ysample, 90, axis=0)\n",
    "\n",
    "# regression estimate line\n",
    "y_fit = poly1d(p)(xi)\n",
    "\n",
    "# PLOT REGRESSION LINE\n",
    "plt.fill_between(xi, lower, upper, facecolor='k', alpha=0.5)\n",
    "plt.plot(xi, y_fit, 'k-')\n",
    "plt.tick_params(size=0)\n",
    "plt.yticks(ticks=[0, 5, 10, 15, 20, 25, 30, 35], \n",
    "           labels=['0', '5', '10', '15', '20', '25', '30', '35'], size=20)\n",
    "plt.xticks(ticks=[0, 5, 10, 15, 20], labels=['0', '5', '10', '15', '20'], size=20)\n",
    "#plt.xlim((0,5))\n",
    "#plt.ylim((0,5))\n",
    "\n",
    "# ADD AXIS LABELS\n",
    "g.set_axis_labels('GEP2 Signature', 'Monosomy 3 Signature', fontsize=28, fontname = 'Arial')\n",
    "\n",
    "# COMPUTE STATISTICS\n",
    "(r, p) = stats.pearsonr(violin_data[x].fillna(0), violin_data[y].fillna(0))\n",
    "print('{} vs {}, R: {}, p: {}'.format(x,y,r,p))\n",
    "\n",
    "# SAVE FIGURE\n",
    "#figure_label = 'M3_GEP2_Correlation_BGMM_UPDATED'.format(meta,subset_type,method,metric)\n",
    "#fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "#    \n",
    "#d = os.path.dirname(fn)\n",
    "#if not os.path.exists(d):\n",
    "#    os.makedirs(d)\n",
    "#    \n",
    "#plt.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "#plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stats.pearsonr(violin_data[x].fillna(0), violin_data[y].fillna(0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "violin_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stats.pearsonr(violin_data[:1088][x].fillna(0), violin_data[:1088][y].fillna(0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.nextafter(0, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.float_info"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### EVALUATE GENE EXPRESSION PER GEP CLASS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_type = 'box'\n",
    "title = 'BAP1'\n",
    "meta = 'Assignment'\n",
    "datatype = 'INDF_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "\n",
    "fig = plt.figure(figsize = (1,3))\n",
    "ax = plt.gca()\n",
    "\n",
    "scale_type = 'count'\n",
    "palette = dict(zip(['Castle 1','Castle 2'],['#0000FF','#FF0000'])) \n",
    "\n",
    "\n",
    "genes = [title]\n",
    "genes = np.unique([x for x in genes if str(x) != 'NAN'])\n",
    "detected_genes = list(set(genes).intersection(set(QUERY.columns)))\n",
    "vals = QUERY[detected_genes]\n",
    "SCORE = np.nansum(vals,axis=1)\n",
    "\n",
    "# Format data structure for violin plot\n",
    "violin_data = []\n",
    "for ind,v in enumerate(SCORE):\n",
    "    violin_data.append({'gene': title, 'Z-normalized Expression': v,\n",
    "                        meta:META[meta].values[ind]}) \n",
    "violin_data = pd.DataFrame(violin_data)  \n",
    "\n",
    "# BOXPLOT GENE EXPRESSION\n",
    "if plot_type == 'box':\n",
    "    g = sns.boxplot(x=\"gene\", y=\"Z-normalized Expression\", \n",
    "                    hue=meta,data=violin_data, palette=palette,notch = True, \n",
    "                    hue_order = ['Castle 1','Castle 2'], \n",
    "                    fliersize = 4, showmeans=False,linewidth = 1, ax = ax) #order = labels, \n",
    "    g.set_ylabel('BAP1 Imputed Expression'.format(datatype),fontsize=14)\n",
    "    g.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "    g.tick_params(labelsize=10)\n",
    "    sns.despine()\n",
    "    ax.set(ylim=(0, SCORE.max()*0.95))\n",
    "    g.legend(loc='upper right',prop={'size':6},bbox_to_anchor=(2.0, 0.95),fancybox=True) \n",
    "\n",
    "\n",
    "elif plot_type == 'violin':\n",
    "    # VIOLIN GENE EXPRESSION\n",
    "    g = sns.violinplot(x=\"gene\", y=\"Z-normalized Expression\", \n",
    "                       hue=meta,data=violin_data, palette=palette,notch = True, \n",
    "                    hue_order = ['Castle 1','Castle 2'],fliersize = 4, showmeans=True,linewidth = 1) #order = labels, \n",
    "\n",
    "    g.set_ylabel('BAP1 Imputed Expression'.format(datatype),fontsize=14)\n",
    "    g.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "    ax.set(ylim=(0, SCORE.max()*0.95))\n",
    "    g.tick_params(labelsize=10)\n",
    "    sns.despine()\n",
    "    \n",
    "plt.tick_params(size=0)\n",
    "plt.yticks(ticks=[0, 0.5, 1, 1.5, 2.0], \n",
    "           labels=['0.0', '0.5', '1.0', '1.5', '2.0'], size=11)\n",
    "\n",
    "# COMPARE DISTRIBUTIONS\n",
    "CLASS1 = violin_data.loc[violin_data[meta].isin(['Castle 1'])]['Z-normalized Expression'].values\n",
    "CLASS2 = violin_data.loc[violin_data[meta].isin(['Castle 2'])]['Z-normalized Expression'].values\n",
    "print(title + ' CLASS1 vs. CLASS2')\n",
    "print(stats.mannwhitneyu(CLASS1,CLASS2))\n",
    "\n",
    "# SAVE FIGURE\n",
    "#figure_label = 'BAP1_Expression_by_GEP_BGMM_UPDATED'.format(meta,subset_type,method,metric)\n",
    "#fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "#    \n",
    "#d = os.path.dirname(fn)\n",
    "#if not os.path.exists(d):\n",
    "#    os.makedirs(d)\n",
    "#    \n",
    "#plt.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "#plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "#print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "METADATA_TUMOR.Assignment.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "violin_data.Assignment.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PHENOTYPIC VOLUME BY PATIENT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### The commented out code produces the phenotypic volume results, which were then pickled using joblib and imported to save time"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "subset_type = 'TUMOR'\n",
    "meta ='Assignment'\n",
    "\n",
    "exec('QUERY = NDF_{}'.format(subset_type))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "new_index = pd.MultiIndex.from_tuples(list(zip(META.index.get_level_values('Sample ID'), \n",
    "                                               META.index.get_level_values('Legend'),\n",
    "                                               META.index.get_level_values('Patient'),\n",
    "                                               META.index.get_level_values('Cell ID'),\n",
    "                                               META.index.get_level_values('Phenograph_Class'),\n",
    "                                               META[meta])), \n",
    "                                  names=['Sample ID','Legend', 'Patient','Cell ID','Phenograph_Class',meta])\n",
    "QUERY = pd.DataFrame(data = QUERY.values, index = new_index, columns = QUERY.columns)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# SPECIFY SUBSET\n",
    "fig_dpi = 400\n",
    "\n",
    "# FIT BINOMIAL DISTRIBUTION AND FILTER BASED ON MEAN/STD OF SECOND\n",
    "plt.figure(figsize = (10,3))\n",
    "gs1 = gridspec.GridSpec(1, 3)\n",
    "gs1.update(wspace=0.7, hspace=0.7) # set the spacing between axes. \n",
    "\n",
    "# (4) PLOT LOG NUMBER OF CELLS CONTRIBUTING TO EACH GENE\n",
    "num_cells_per_gene = np.log(np.sum(QUERY.values > 0,axis=0))\n",
    "num_cells_per_gene[(np.isinf(num_cells_per_gene)) | (np.isnan(num_cells_per_gene))] = 0\n",
    "rmv_genes1 = np.where(num_cells_per_gene<=0)[0] # GENES MUST BE DETECTED IN AT LEAST 10 CELLS\n",
    "\n",
    "ax = plt.subplot(gs1[0])\n",
    "bins = np.linspace(num_cells_per_gene.min(), num_cells_per_gene.max()*0.95, 20)\n",
    "plt.hist(num_cells_per_gene, bins, alpha=0.5, label='keep')\n",
    "\n",
    "if rmv_genes1.any():\n",
    "    plt.hist(num_cells_per_gene[rmv_genes1], bins, alpha=1, label='remove')\n",
    "    \n",
    "#ax.set_axis_bgcolor('white')\n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Frequency')\n",
    "plt.xlabel('Gene Filter 1: Remove Genes Singletons \\n(Log. # Expressing Cells)')\n",
    "plt.grid(True)\n",
    "sns.despine()\n",
    "\n",
    "# (2) PLOT LOG COUNTS PER GENE - REMOVE LOW ABUNDANCE GENES\n",
    "log_counts_per_gene = np.log(np.log(np.sum(QUERY.values,axis=0)))\n",
    "log_counts_per_gene[(np.isinf(log_counts_per_gene)) | (np.isnan(log_counts_per_gene))] = 0\n",
    "data = log_counts_per_gene\n",
    "\n",
    "ax = plt.subplot(gs1[1])\n",
    "bins = np.linspace(data.min(), data.max()*0.95, 100)\n",
    "y,x,_=hist(data,bins,alpha=.3,label='data')\n",
    "\n",
    "x=(x[1:]+x[:-1])/2 \n",
    "expected=(0,.1,3500,1.75,.2,500)\n",
    "params,cov=curve_fit(bimodal,x,y,expected)\n",
    "sigma=sqrt(diag(cov))\n",
    "plot(x,bimodal(x,*params),color='red',lw=3,label='model')\n",
    "\n",
    "mu1 = params[0]\n",
    "std1 = params[1]\n",
    "mu2 = params[3]\n",
    "std2 = params[4]\n",
    "rmv_genes_neg = np.where(data<mu2-3*std2)[0]\n",
    "rmv_genes2 = np.sort(list(set(list(rmv_genes_neg))))\n",
    "\n",
    "if rmv_genes2.any():\n",
    "    plt.hist(data[rmv_genes2], bins, alpha=1, label='remove')\n",
    "    \n",
    "#ax.set_axis_bgcolor('white')\n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Frequency')\n",
    "plt.xlabel('Gene Filter 2: \\nLog-Log Counts/Gene')\n",
    "plt.grid(True)\n",
    "sns.despine()\n",
    "\n",
    "# (3) PLOT STDEV PER GENE\n",
    "pseudocount = 0.001\n",
    "std_per_gene = np.log10(QUERY.std(axis=0, skipna=False)+pseudocount)\n",
    "data = std_per_gene\n",
    "\n",
    "ax = plt.subplot(gs1[2])\n",
    "bins = np.linspace(data.min(), data.max()*0.95, 100)\n",
    "y,x,_=hist(data,bins,alpha=.3,label='data')\n",
    "\n",
    "x=(x[1:]+x[:-1])/2 \n",
    "expected=(0,-3,3000,-1,.2,500)\n",
    "params,cov=curve_fit(bimodal,x,y,expected)\n",
    "sigma=sqrt(diag(cov))\n",
    "plot(x,bimodal(x,*params),color='red',lw=3,label='model')\n",
    "\n",
    "mu1 = params[0]\n",
    "std1 = params[1]\n",
    "mu2 = params[3]\n",
    "std2 = params[4]\n",
    "rmv_genes_neg = np.where(data<mu2-3*abs(std2))[0]\n",
    "rmv_genes3 = np.sort(list(set(list(rmv_genes_neg))))\n",
    "\n",
    "if rmv_genes3.any():\n",
    "    plt.hist(data[rmv_genes3], bins, alpha=1, label='remove')\n",
    "    \n",
    "#ax.set_axis_bgcolor('white')\n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Frequency')\n",
    "plt.xlabel('Gene Filter 3: \\nSDEV/Gene')\n",
    "plt.grid(True)\n",
    "sns.despine()\n",
    "\n",
    "# Add (abbreviated) legend bottom left\n",
    "L = plt.legend(loc='upper right',prop={'size':12},bbox_to_anchor=(1.6, 1.05),fancybox=True) \n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '_filter_genes'\n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label + '.png'\n",
    "#plt.savefig(fn, dpi=fig_dpi)\n",
    "print(fn)\n",
    "\n",
    "# EVALUATE NUMBER OF CELLS/GENES REMOVED\n",
    "print('Count Gene Filter 1: {}'.format(len(rmv_genes1)))\n",
    "print('Count Gene Filter 2: {}'.format(len(rmv_genes2)))\n",
    "print('Count Gene Filter 3: {}'.format(len(rmv_genes3)))\n",
    "\n",
    "# REMOVE SELECTED OUTLIER GENES(HIGHLIGHTED IN GREEN)\n",
    "CUT_DF = deepcopy(QUERY)\n",
    "rmv_genes = np.sort(list(set(list(rmv_genes1) + list(rmv_genes2)+list(rmv_genes3))))\n",
    "if rmv_genes.any():\n",
    "    CUT_DF = CUT_DF.drop(CUT_DF.columns[rmv_genes],axis=1)\n",
    "    print(CUT_DF.shape)\n",
    "    \n",
    "# Remove empty genes if they exist\n",
    "drop_genes = np.where(CUT_DF.sum(axis=0)==0)[0]\n",
    "CUT_DF = CUT_DF.drop(CUT_DF.columns[drop_genes],axis=1)\n",
    "print(CUT_DF.shape)\n",
    "\n",
    "# GENES EXCLUDING LOW ABUNDANCE GENES\n",
    "variable_goi = list(CUT_DF.columns)\n",
    "print(len(variable_goi))"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "n = 150\n",
    "reps = np.arange(100)\n",
    "gene_subset = variable_goi"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# GENE-GENE COVARIANCE WITHIN PATIENT 1\n",
    "master_groupby = 'Patient'\n",
    "class_selection = ['UM01']\n",
    "ix = [ind for ind,name in enumerate(QUERY.index.names) if name==master_groupby][0]\n",
    "logvolume_castle1 = np.zeros(len(reps))\n",
    "\n",
    "for rep in reps:\n",
    "    \n",
    "    SUBSET = QUERY.loc[QUERY.index.map(lambda x: x[ix] in class_selection)][gene_subset].sample(n=n,replace=True)\n",
    "    GENE_COV = SUBSET.cov(min_periods=None)\n",
    "    eigenvalues = np.linalg.eigvals(GENE_COV)\n",
    "    logvolume = np.float(np.sum(0.5*np.log10(eigenvalues[eigenvalues>0]**2))/len(eigenvalues))\n",
    "    logvolume_castle1[rep] = logvolume\n",
    "    print(rep)\n",
    "    \n",
    "# GENE-GENE COVARIANCE WITHIN PATIENT 2\n",
    "master_groupby = 'Patient'\n",
    "class_selection = ['UM02']\n",
    "ix = [ind for ind,name in enumerate(QUERY.index.names) if name==master_groupby][0]\n",
    "logvolume_castle2 = np.zeros(len(reps))\n",
    "\n",
    "for rep in reps:\n",
    "    \n",
    "    SUBSET = QUERY.loc[QUERY.index.map(lambda x: x[ix] in class_selection)][gene_subset].sample(n=n,replace=True)\n",
    "    GENE_COV = SUBSET.cov(min_periods=None)\n",
    "    eigenvalues = np.linalg.eigvals(GENE_COV)\n",
    "    logvolume = np.float(np.sum(0.5*np.log10(eigenvalues[eigenvalues>0]**2))/len(eigenvalues))\n",
    "    logvolume_castle2[rep] = logvolume\n",
    "    print(rep)\n",
    "    \n",
    "# GENE-GENE COVARIANCE WITHIN PATIENT 3\n",
    "master_groupby = 'Patient'\n",
    "class_selection = ['UM03']\n",
    "ix = [ind for ind,name in enumerate(QUERY.index.names) if name==master_groupby][0]\n",
    "logvolume_castle3 = np.zeros(len(reps))\n",
    "\n",
    "for rep in reps:\n",
    "    \n",
    "    SUBSET = QUERY.loc[QUERY.index.map(lambda x: x[ix] in class_selection)][gene_subset].sample(n=n,replace=True)\n",
    "    GENE_COV = SUBSET.cov(min_periods=None)\n",
    "    eigenvalues = np.linalg.eigvals(GENE_COV)\n",
    "    logvolume = np.float(np.sum(0.5*np.log10(eigenvalues[eigenvalues>0]**2))/len(eigenvalues))\n",
    "    logvolume_castle3[rep] = logvolume\n",
    "    print(rep)\n",
    "    \n",
    "# GENE-GENE COVARIANCE WITHIN PATIENT 4\n",
    "master_groupby = 'Patient'\n",
    "class_selection = ['UM04']\n",
    "ix = [ind for ind,name in enumerate(QUERY.index.names) if name==master_groupby][0]\n",
    "logvolume_castle4 = np.zeros(len(reps))\n",
    "\n",
    "for rep in reps:\n",
    "    \n",
    "    SUBSET = QUERY.loc[QUERY.index.map(lambda x: x[ix] in class_selection)][gene_subset].sample(n=n,replace=True)\n",
    "    GENE_COV = SUBSET.cov(min_periods=None)\n",
    "    eigenvalues = np.linalg.eigvals(GENE_COV)\n",
    "    logvolume = np.float(np.sum(0.5*np.log10(eigenvalues[eigenvalues>0]**2))/len(eigenvalues))\n",
    "    logvolume_castle4[rep] = logvolume\n",
    "    print(rep)\n",
    "    \n",
    "# GENE-GENE COVARIANCE WITHIN PATIENT 5\n",
    "master_groupby = 'Patient'\n",
    "class_selection = ['UM05']\n",
    "ix = [ind for ind,name in enumerate(QUERY.index.names) if name==master_groupby][0]\n",
    "logvolume_castle5 = np.zeros(len(reps))\n",
    "\n",
    "for rep in reps:\n",
    "    \n",
    "    SUBSET = QUERY.loc[QUERY.index.map(lambda x: x[ix] in class_selection)][gene_subset].sample(n=n,replace=True)\n",
    "    GENE_COV = SUBSET.cov(min_periods=None)\n",
    "    eigenvalues = np.linalg.eigvals(GENE_COV)\n",
    "    logvolume = np.float(np.sum(0.5*np.log10(eigenvalues[eigenvalues>0]**2))/len(eigenvalues))\n",
    "    logvolume_castle5[rep] = logvolume\n",
    "    print(rep)\n",
    "    \n",
    "# GENE-GENE COVARIANCE WITHIN PATIENT 6\n",
    "master_groupby = 'Patient'\n",
    "class_selection = ['UM06']\n",
    "ix = [ind for ind,name in enumerate(QUERY.index.names) if name==master_groupby][0]\n",
    "logvolume_castle6 = np.zeros(len(reps))\n",
    "\n",
    "for rep in reps:\n",
    "    \n",
    "    SUBSET = QUERY.loc[QUERY.index.map(lambda x: x[ix] in class_selection)][gene_subset].sample(n=n,replace=True)\n",
    "    GENE_COV = SUBSET.cov(min_periods=None)\n",
    "    eigenvalues = np.linalg.eigvals(GENE_COV)\n",
    "    logvolume = np.float(np.sum(0.5*np.log10(eigenvalues[eigenvalues>0]**2))/len(eigenvalues))\n",
    "    logvolume_castle6[rep] = logvolume\n",
    "    print(rep)\n",
    "\n",
    "# WRITE RESULTS TO DATAFRAME\n",
    "boxplot_data_goi = pd.DataFrame()\n",
    "boxplot_data_goi['Patient'] = ['UM01']*len(logvolume_castle1)+['UM02']*len(logvolume_castle2)+\\\n",
    "                                 ['UM03']*len(logvolume_castle3)+['UM04']*len(logvolume_castle4)+\\\n",
    "                                 ['UM05']*len(logvolume_castle5)+['UM06']*len(logvolume_castle6)\n",
    "boxplot_data_goi['Log Phenotypic Volume'] = np.concatenate([logvolume_castle1,logvolume_castle2,\n",
    "                                                            logvolume_castle3,logvolume_castle4,\n",
    "                                                            logvolume_castle5,logvolume_castle6],axis = 0)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# STATISTIC\n",
    "C1 = boxplot_data_goi.loc[boxplot_data_goi['Patient'].isin(['UM01'])]['Log Phenotypic Volume'].values\n",
    "C2 = boxplot_data_goi.loc[boxplot_data_goi['Patient'].isin(['UM02'])]['Log Phenotypic Volume'].values\n",
    "C3 = boxplot_data_goi.loc[boxplot_data_goi['Patient'].isin(['UM03'])]['Log Phenotypic Volume'].values\n",
    "C4 = boxplot_data_goi.loc[boxplot_data_goi['Patient'].isin(['UM04'])]['Log Phenotypic Volume'].values\n",
    "C5 = boxplot_data_goi.loc[boxplot_data_goi['Patient'].isin(['UM05'])]['Log Phenotypic Volume'].values\n",
    "C6 = boxplot_data_goi.loc[boxplot_data_goi['Patient'].isin(['UM06'])]['Log Phenotypic Volume'].values\n",
    "\n",
    "# Pickle results incase needed later:\n",
    "joblib.dump(C1,'/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UM01')\n",
    "joblib.dump(C2,'/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UM02')\n",
    "joblib.dump(C3,'/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UM03')\n",
    "joblib.dump(C4,'/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UM04')\n",
    "joblib.dump(C5,'/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UM05')\n",
    "joblib.dump(C6,'/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UM06')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import phenotypic volume results per patient:\n",
    "\n",
    "PHEN_VOL_UM01 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UM01')\n",
    "PHEN_VOL_UM02 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UM02')\n",
    "PHEN_VOL_UM03 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UM03')\n",
    "PHEN_VOL_UM04 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UM04')\n",
    "PHEN_VOL_UM05 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UM05')\n",
    "PHEN_VOL_UM06 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UM06')\n",
    "\n",
    "PHEN_VOL_LIST = [PHEN_VOL_UM01, PHEN_VOL_UM02, PHEN_VOL_UM03, PHEN_VOL_UM04, PHEN_VOL_UM05, PHEN_VOL_UM06]\n",
    "\n",
    "phen_vol_data = pd.DataFrame()\n",
    "\n",
    "phen_vol_data['Patient'] = ['MSK-UM01']*len(PHEN_VOL_LIST[0])+['MSK-UM02']*len(PHEN_VOL_LIST[1])+\\\n",
    "                           ['MSK-UM03']*len(PHEN_VOL_LIST[2])+['MSK-UM04']*len(PHEN_VOL_LIST[3])+\\\n",
    "                           ['MSK-UM05']*len(PHEN_VOL_LIST[4])+['MSK-UM06']*len(PHEN_VOL_LIST[5])\n",
    "\n",
    "phen_vol_data['Log Phenotypic Volume'] = [x for l in PHEN_VOL_LIST for x in l]\n",
    "\n",
    "phen_vol_data['Status'] = ['Alive' if x in ['MSK-UM01', 'MSK-UM05', 'MSK-UM03', 'MSK-UM04']\n",
    "                            else 'Deceased' for x in phen_vol_data.Patient.values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phen_vol_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize using violin plot:\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(7,7))\n",
    "\n",
    "ax = sns.violinplot(x='Patient', y='Log Phenotypic Volume', data=phen_vol_data,\n",
    "                    order=phen_vol_data.groupby('Patient').mean().\\\n",
    "                          sort_values(by='Log Phenotypic Volume').index.values,\n",
    "                    hue='Status',\n",
    "                    palette={'Alive': 'white', 'Deceased': 'gray'},\n",
    "                    dodge=False,\n",
    "                    scale='count')\n",
    "\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "plt.xticks(rotation=90)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'Phenotypic_Volume_Violin_MSK_count_scale'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label\n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PHENOTYPIC VOLUME INTRAPATIENT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### The commented out code produces the phenotypic volume results, which were then pickled using joblib and imported to save time"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "n = 50\n",
    "reps = np.arange(100)\n",
    "gene_subset = variable_goi"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# GENE-GENE COVARIANCE WITHIN PATIENT 1\n",
    "master_groupby = 'Assignment'\n",
    "class_selection = ['Castle 1']\n",
    "ix = [ind for ind,name in enumerate(QUERY.index.names) if name==master_groupby][0]\n",
    "logvolume_castle1 = np.zeros(len(reps))\n",
    "\n",
    "for rep in reps:\n",
    "    \n",
    "    SUBSET = QUERY.loc[QUERY.index.map(lambda x: x[ix] in class_selection)][gene_subset].sample(n=n,replace=True)\n",
    "    GENE_COV = SUBSET.cov(min_periods=None)\n",
    "    eigenvalues = np.linalg.eigvals(GENE_COV)\n",
    "    logvolume = np.float(np.sum(0.5*np.log10(eigenvalues[eigenvalues>0]**2))/len(eigenvalues))\n",
    "    logvolume_castle1[rep] = logvolume\n",
    "    print(rep)\n",
    "    \n",
    "# GENE-GENE COVARIANCE WITHIN PATIENT 2\n",
    "master_groupby = 'Assignment'\n",
    "class_selection = ['Castle 2']\n",
    "ix = [ind for ind,name in enumerate(QUERY.index.names) if name==master_groupby][0]\n",
    "logvolume_castle2 = np.zeros(len(reps))\n",
    "\n",
    "for rep in reps:\n",
    "    \n",
    "    SUBSET = QUERY.loc[QUERY.index.map(lambda x: x[ix] in class_selection)][gene_subset].sample(n=n,replace=True)\n",
    "    GENE_COV = SUBSET.cov(min_periods=None)\n",
    "    eigenvalues = np.linalg.eigvals(GENE_COV)\n",
    "    logvolume = np.float(np.sum(0.5*np.log10(eigenvalues[eigenvalues>0]**2))/len(eigenvalues))\n",
    "    logvolume_castle2[rep] = logvolume\n",
    "    print(rep)\n",
    "    \n",
    "# WRITE RESULTS TO DATAFRAME\n",
    "boxplot_data_goi_UM03 = pd.DataFrame()\n",
    "boxplot_data_goi_UM03['Assignment'] = ['Castle 1']*len(logvolume_castle1)+['Castle 2']*len(logvolume_castle2)\n",
    "boxplot_data_goi_UM03['Log Phenotypic Volume'] = np.concatenate([logvolume_castle1,logvolume_castle2],axis = 0)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# STATISTIC\n",
    "C1_UM03 = boxplot_data_goi_UM03.loc[boxplot_data_goi_UM03['Assignment'].\\\n",
    "                                    isin(['Castle 1'])]['Log Phenotypic Volume'].values\n",
    "C2_UM03 = boxplot_data_goi_UM03.loc[boxplot_data_goi_UM03['Assignment'].\\\n",
    "                                    isin(['Castle 2'])]['Log Phenotypic Volume'].values\n",
    "\n",
    "# Pickle results incase needed later:\n",
    "joblib.dump(C1_UM03,'/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UM03_GEP1')\n",
    "joblib.dump(C2_UM03,'/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UM03_GEP2')\n",
    "\n",
    "stats.mannwhitneyu(C1_UM03, C2_UM03)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# IMport phenotypic volume results for GEP1 vs. GEP2 for UM03:\n",
    "\n",
    "C1_UM03 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UM03_GEP1')\n",
    "C2_UM03 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UM03_GEP2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PLot results:\n",
    "\n",
    "sns.set_style(\"white\")\n",
    "fig = plt.figure(figsize = (5,5))\n",
    "ax = plt.gca()\n",
    "\n",
    "sns.kdeplot(C1_UM03, shade=True, color=\"#0000FF\")\n",
    "sns.kdeplot(C2_UM03, shade=True, color=\"#FF0000\")\n",
    "\n",
    "ax.set_ylabel(\"Frequency\",fontsize=10)\n",
    "ax.set_xlabel(\"Log Phenotypic Volume\",fontsize=10)\n",
    "ax.tick_params(labelsize=10)\n",
    "#g.set_xticklabels(ax.get_xticklabels(),rotation=90)\n",
    "sns.despine()\n",
    "plt.ylim((0,4))\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'Log_Phenotypic_Volume_UM03'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label\n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PRC1 TARGET GENES HEATMAPS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GEP2 RANKED HEATMAP FOR PRC1 TARGET GENES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RANK BY GEP 2\n",
    "meta = 'Nearest_Archetype'#'Archetype_Soft_Cluster'\n",
    "subset_type = 'TUMOR'\n",
    "exec('QUERY = INDF_{}'.format(subset_type))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "rank_by = 'Castle 2'#'Class 2 Onken v2'\n",
    "genes = np.unique(genesets[rank_by].values.ravel().tolist())\n",
    "detected_genes = [gene for gene in genes if gene in list(QUERY.columns)]\n",
    "\n",
    "QUERY = QUERY.loc[:, (QUERY != 0).any(axis=0)]\n",
    "\n",
    "vals = QUERY[detected_genes]\n",
    "SCORE = np.nanmedian(vals,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "PRC1_target_genes = pd.read_csv(DATA_PATH+'PRC1_targets_Genesets_and_Chipseq.csv')\n",
    "plot_genes = PRC1_target_genes['PRC1_targets_Genesets_and_Chipseq'].dropna().values.tolist()\n",
    "plot_genes = [x for x in plot_genes if x in QUERY.columns.values.tolist()]\n",
    "\n",
    "# LOAD GENESETS\n",
    "path_to_genesets = DATA_PATH+'uveal_melanoma_MB_v3.csv'\n",
    "genesets = pd.read_csv(path_to_genesets,header='infer')\n",
    "\n",
    "subset_type = 'TUMOR'\n",
    "datatype = 'INDF_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "\n",
    "# HEATMAP OF INDIVIDUAL CELLS RANKED BY LUNGE EP\n",
    "rank_by = 'Castle 2'#\n",
    "genesets_include = [rank_by]\n",
    "genes = np.unique(genesets[genesets_include].values.ravel().tolist())\n",
    "genes = [gene for gene in genes if gene in list(set(QUERY.columns))]\n",
    "vals = QUERY[genes]\n",
    "QUERY['GEP2_RANK'] =np.nanmean(vals,axis=1)\n",
    "\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "heatmap_data = pd.DataFrame( data = zscore(QUERY.sort_values(by=['GEP2_RANK'])[plot_genes],axis=0), \n",
    "                   columns = plot_genes, index = QUERY.sort_values(by=['GEP2_RANK']).index)\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns\n",
    "\n",
    "# LINKAGE \n",
    "method = 'average' # average, single centroid/euclidean\n",
    "metric = 'cosine' # cosine\n",
    "linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "col_linkage = deepcopy(linkage)\n",
    "cl = hc.leaves_list(col_linkage)\n",
    "mat = heatmap_data.iloc[:,cl]\n",
    "\n",
    "window = 20\n",
    "mat = mat.rolling(window, win_type='triang',center = True).sum()\n",
    "half_window = int(window/2)\n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[ind] = mat.iloc[half_window]    \n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[-ind] = mat.iloc[-half_window]\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(4,10))\n",
    "\n",
    "# ADD MATRIX WITH LINEAGE NAMES\n",
    "axmatrix = fig.add_axes([0.12,0.1,0.4,0.6])\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=CM_DIVERGING,vmin=-10,vmax=10)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "axmatrix.set_xticklabels([]);\n",
    "axmatrix.grid(False)\n",
    "plt.tick_params(axis='both', size=0)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'PRC1_Targets_Heatmap_UPDATED_average_cosine'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/GEP2_Ranked_Heatmaps/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "H2A_target_genes = pd.read_csv(DATA_PATH+'H2AK119Ub_in_921_3_23.txt', header=None)\n",
    "H2A_target_genes.columns = ['Genes']\n",
    "plot_genes = H2A_target_genes['Genes'].dropna().values.tolist()\n",
    "plot_genes = [x for x in plot_genes if x in QUERY.columns.values.tolist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(plot_genes) / len(H2A_target_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "H2A_target_genes = pd.read_csv(DATA_PATH+'H2AK119Ub_in_921_3_23.txt', header=None)\n",
    "H2A_target_genes.columns = ['Genes']\n",
    "plot_genes = H2A_target_genes['Genes'].dropna().values.tolist()\n",
    "plot_genes = [x for x in plot_genes if x in QUERY.columns.values.tolist()]\n",
    "\n",
    "# LOAD GENESETS\n",
    "path_to_genesets = DATA_PATH+'uveal_melanoma_MB_v3.csv'\n",
    "genesets = pd.read_csv(path_to_genesets,header='infer')\n",
    "\n",
    "subset_type = 'TUMOR'\n",
    "datatype = 'INDF_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "\n",
    "# HEATMAP OF INDIVIDUAL CELLS RANKED BY LUNGE EP\n",
    "rank_by = 'Castle 2'#\n",
    "genesets_include = [rank_by]\n",
    "genes = np.unique(genesets[genesets_include].values.ravel().tolist())\n",
    "genes = [gene for gene in genes if gene in list(set(QUERY.columns))]\n",
    "vals = QUERY[genes]\n",
    "QUERY['GEP2_RANK'] =np.nanmean(vals,axis=1)\n",
    "\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "heatmap_data = pd.DataFrame(data = zscore(QUERY.sort_values(by=['GEP2_RANK'])[plot_genes],axis=0), \n",
    "                   columns = plot_genes, index = QUERY.sort_values(by=['GEP2_RANK']).index)\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns\n",
    "\n",
    "# LINKAGE \n",
    "method = 'average' # average, single centroid/euclidean\n",
    "metric = 'cosine' # cosine\n",
    "linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "col_linkage = deepcopy(linkage)\n",
    "cl = hc.leaves_list(col_linkage)\n",
    "mat = heatmap_data.iloc[:,cl]\n",
    "\n",
    "window = 20\n",
    "mat = mat.rolling(window, win_type='triang',center = True).sum()\n",
    "half_window = int(window/2)\n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[ind] = mat.iloc[half_window]    \n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[-ind] = mat.iloc[-half_window]\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(4,10))\n",
    "\n",
    "# ADD MATRIX WITH LINEAGE NAMES\n",
    "axmatrix = fig.add_axes([0.12,0.1,0.4,0.6])\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=CM_DIVERGING,vmin=-10,vmax=10)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "axmatrix.set_xticklabels([]);\n",
    "axmatrix.grid(False)\n",
    "plt.tick_params(axis='both', size=0)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'H2AK119Ub_Heatmap_UPDATED_average_cosine'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/Revision_2_Figures/GEP2_Ranked_Heatmaps/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "H2A_target_genes = pd.read_csv(DATA_PATH+'H2AK119Ub_in_921_broad_3_23.txt', header=None)\n",
    "H2A_target_genes.columns = ['Genes']\n",
    "plot_genes = H2A_target_genes['Genes'].dropna().values.tolist()\n",
    "plot_genes = [x for x in plot_genes if x in QUERY.columns.values.tolist()]\n",
    "\n",
    "# LOAD GENESETS\n",
    "path_to_genesets = DATA_PATH+'uveal_melanoma_MB_v3.csv'\n",
    "genesets = pd.read_csv(path_to_genesets,header='infer')\n",
    "\n",
    "subset_type = 'TUMOR'\n",
    "datatype = 'INDF_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "\n",
    "# HEATMAP OF INDIVIDUAL CELLS RANKED BY LUNGE EP\n",
    "rank_by = 'Castle 2'#\n",
    "genesets_include = [rank_by]\n",
    "genes = np.unique(genesets[genesets_include].values.ravel().tolist())\n",
    "genes = [gene for gene in genes if gene in list(set(QUERY.columns))]\n",
    "vals = QUERY[genes]\n",
    "QUERY['GEP2_RANK'] =np.nanmean(vals,axis=1)\n",
    "\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "heatmap_data = pd.DataFrame(data = zscore(QUERY.sort_values(by=['GEP2_RANK'])[plot_genes],axis=0), \n",
    "                   columns = plot_genes, index = QUERY.sort_values(by=['GEP2_RANK']).index)\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns\n",
    "\n",
    "# LINKAGE \n",
    "method = 'average' # average, single centroid/euclidean\n",
    "metric = 'cosine' # cosine\n",
    "linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "col_linkage = deepcopy(linkage)\n",
    "cl = hc.leaves_list(col_linkage)\n",
    "mat = heatmap_data.iloc[:,cl]\n",
    "\n",
    "window = 20\n",
    "mat = mat.rolling(window, win_type='triang',center = True).sum()\n",
    "half_window = int(window/2)\n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[ind] = mat.iloc[half_window]    \n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[-ind] = mat.iloc[-half_window]\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(4,10))\n",
    "\n",
    "# ADD MATRIX WITH LINEAGE NAMES\n",
    "axmatrix = fig.add_axes([0.12,0.1,0.4,0.6])\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=CM_DIVERGING,vmin=-10,vmax=10)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "axmatrix.set_xticklabels([]);\n",
    "axmatrix.grid(False)\n",
    "plt.tick_params(axis='both', size=0)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'H2AK119Ub_Broad_Heatmap_UPDATED_average_cosine'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/Revision_2_Figures/GEP2_Ranked_Heatmaps/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(plot_genes) / len(H2A_target_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### GEP2 RANKED HEATMAP FOR NEW GENESET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RANK BY GEP 2\n",
    "meta = 'Nearest_Archetype'#'Archetype_Soft_Cluster'\n",
    "subset_type = 'TUMOR'\n",
    "exec('QUERY = INDF_{}'.format(subset_type))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "rank_by = 'Castle 2'#'Class 2 Onken v2'\n",
    "genes = np.unique(genesets[rank_by].values.ravel().tolist())\n",
    "detected_genes = [gene for gene in genes if gene in list(QUERY.columns)]\n",
    "\n",
    "vals = QUERY[detected_genes]\n",
    "SCORE = np.nanmedian(vals,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "target_genes = pd.read_csv(DATA_PATH+'829_PRE_92_1_geneset.txt', \n",
    "                                sep='\\t', header=None, names=['genes'])\n",
    "plot_genes = target_genes['genes'].dropna().values.tolist()\n",
    "plot_genes = [x for x in plot_genes if x in QUERY.columns.values.tolist()]\n",
    "\n",
    "# LOAD GENESETS\n",
    "path_to_genesets = DATA_PATH+'uveal_melanoma_MB_v3.csv'\n",
    "genesets = pd.read_csv(path_to_genesets,header='infer')\n",
    "\n",
    "subset_type = 'TUMOR'\n",
    "datatype = 'INDF_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "\n",
    "# HEATMAP OF INDIVIDUAL CELLS RANKED BY LUNGE EP\n",
    "rank_by = 'Castle 2'#\n",
    "genesets_include = [rank_by]\n",
    "genes = np.unique(genesets[genesets_include].values.ravel().tolist())\n",
    "genes = [gene for gene in genes if gene in list(set(QUERY.columns))]\n",
    "vals = QUERY[genes]\n",
    "QUERY['GEP2_RANK'] =np.nanmean(vals,axis=1)\n",
    "\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "heatmap_data = pd.DataFrame( data = zscore(QUERY.sort_values(by=['GEP2_RANK'])[plot_genes],axis=0), \n",
    "                   columns = plot_genes, index = QUERY.sort_values(by=['GEP2_RANK']).index)\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns\n",
    "\n",
    "# LINKAGE \n",
    "method = 'average' # average, single centroid/euclidean\n",
    "metric = 'euclidean' # cosine\n",
    "linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "col_linkage = deepcopy(linkage)\n",
    "cl = hc.leaves_list(col_linkage)\n",
    "mat = heatmap_data.iloc[:,cl]\n",
    "\n",
    "window = 20\n",
    "mat = mat.rolling(window, win_type='triang',center = True).sum()\n",
    "half_window = int(window/2)\n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[ind] = mat.iloc[half_window]    \n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[-ind] = mat.iloc[-half_window]\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(4,10))\n",
    "\n",
    "# ADD MATRIX WITH LINEAGE NAMES\n",
    "axmatrix = fig.add_axes([0.12,0.1,0.4,0.6])\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=CM_DIVERGING,vmin=-10,vmax=10)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "axmatrix.set_xticklabels([]);\n",
    "axmatrix.grid(False)\n",
    "plt.tick_params(axis='both', size=0)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'average_euclidean_heatmap_NEW_GENESET'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/GEP2_Ranked_Heatmaps/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### HEAT MAP FOR NC NFkB GENESETS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RANK BY GEP 2\n",
    "subset_type = 'TUMOR'\n",
    "exec('QUERY = INDF_{}'.format(subset_type))\n",
    "exec('META = METADATA_{}'.format(subset_type))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "target_genes = pd.read_csv(DATA_PATH+'Genesets_NC_NFkB.txt', \n",
    "                                sep='\\t', header=0)\n",
    "plot_genes = target_genes['CIN-responsive-NC-NFkB'].dropna().values.tolist()\n",
    "plot_genes = [x for x in plot_genes if x in QUERY.columns.values.tolist()]\n",
    "\n",
    "# LOAD GENESETS\n",
    "path_to_genesets = DATA_PATH+'uveal_melanoma_MB_v3.csv'\n",
    "genesets = pd.read_csv(path_to_genesets,header='infer')\n",
    "\n",
    "subset_type = 'TUMOR'\n",
    "datatype = 'INDF_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "\n",
    "# HEATMAP OF INDIVIDUAL CELLS RANKED BY LUNGE EP\n",
    "rank_by = 'Castle 2'#\n",
    "genesets_include = [rank_by]\n",
    "genes = np.unique(genesets[genesets_include].values.ravel().tolist())\n",
    "genes = [gene for gene in genes if gene in list(set(QUERY.columns))]\n",
    "vals = QUERY[genes]\n",
    "QUERY['GEP2_RANK'] =np.nanmean(vals,axis=1)\n",
    "\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "heatmap_data = pd.DataFrame( data = zscore(QUERY.sort_values(by=['GEP2_RANK'])[plot_genes],axis=0), \n",
    "                   columns = plot_genes, index = QUERY.sort_values(by=['GEP2_RANK']).index)\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns\n",
    "\n",
    "# LINKAGE \n",
    "method = 'average' # average, single centroid/euclidean\n",
    "metric = 'euclidean' # cosine\n",
    "linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "col_linkage = deepcopy(linkage)\n",
    "cl = hc.leaves_list(col_linkage)\n",
    "mat = heatmap_data.iloc[:,cl]\n",
    "\n",
    "window = 20\n",
    "mat = mat.rolling(window, win_type='triang',center = True).sum()\n",
    "half_window = int(window/2)\n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[ind] = mat.iloc[half_window]    \n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[-ind] = mat.iloc[-half_window]\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(4,10))\n",
    "\n",
    "# ADD MATRIX WITH LINEAGE NAMES\n",
    "axmatrix = fig.add_axes([0.12,0.1,0.4,0.6])\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=CM_DIVERGING,vmin=-10,vmax=10)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "#axmatrix.set_xticklabels([]);\n",
    "xtick = plt.xticks(range(len(labels)), labels, rotation = 90, fontsize = 6,fontname='Arial')\n",
    "axmatrix.grid(False)\n",
    "plt.tick_params(axis='y', size=0)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'average_euclidean_heatmap_NC_NFkB'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/GEP2_Ranked_Heatmaps/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### HEATMAP FOR NEW GENESETS 4/15 and 5/5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RANK BY GEP 2\n",
    "subset_type = 'TUMOR'\n",
    "exec('QUERY = INDF_{}'.format(subset_type))\n",
    "exec('META = METADATA_{}'.format(subset_type))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# LOAD GENESETS\n",
    "target_genes = pd.read_csv(DATA_PATH+'Genesets_4_15.csv', \n",
    "                                sep=',', header=0)\n",
    "\n",
    "# Rank QUERY by GEP2\n",
    "rank_by = 'Castle 2'#\n",
    "genesets_include = [rank_by]\n",
    "genes = np.unique(genesets[genesets_include].values.ravel().tolist())\n",
    "genes = [gene for gene in genes if gene in list(set(QUERY.columns))]\n",
    "vals = QUERY[genes]\n",
    "QUERY['GEP2_RANK'] =np.nanmean(vals,axis=1)\n",
    "QUERY = QUERY.sort_values(by=['GEP2_RANK'])\n",
    "\n",
    "# Iterate through each geneset:\n",
    "for col in target_genes.columns:\n",
    "    print(col)\n",
    "    plot_genes = target_genes[col].dropna().values.tolist()\n",
    "    plot_genes = [x for x in plot_genes if x in QUERY.columns.values.tolist()]\n",
    "    \n",
    "    if len(plot_genes) > 0:\n",
    "    \n",
    "        # CONSTRUCT HEATMAP DATA\n",
    "        heatmap_data = pd.DataFrame( data = zscore(QUERY[plot_genes],axis=0), \n",
    "                           columns = plot_genes, index = QUERY.index)\n",
    "        heatmap_data = heatmap_data.dropna(axis=1, how='any') \n",
    "        yticks = heatmap_data.index\n",
    "        xticks = heatmap_data.columns\n",
    "        \n",
    "        # LINKAGE \n",
    "        for m in [('average', 'euclidean'), ('average', 'cosine'), ('centroid', 'euclidean')]:\n",
    "            method = m[0] # average, single centroid/euclidean\n",
    "            metric = m[1] # cosine\n",
    "            linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "            col_linkage = deepcopy(linkage)\n",
    "            cl = hc.leaves_list(col_linkage)\n",
    "            mat = heatmap_data.iloc[:,cl]\n",
    "            \n",
    "            window = 20\n",
    "            mat = mat.rolling(window, win_type='triang',center = True).sum()\n",
    "            half_window = int(window/2)\n",
    "            for ind in np.arange(half_window):\n",
    "                mat.iloc[ind] = mat.iloc[half_window]    \n",
    "            for ind in np.arange(half_window):\n",
    "                mat.iloc[-ind] = mat.iloc[-half_window]\n",
    "            \n",
    "            # VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "            fig = plt.figure(figsize=(4,10))\n",
    "            \n",
    "            # ADD MATRIX WITH LINEAGE NAMES\n",
    "            axmatrix = fig.add_axes([0.12,0.1,0.4,0.6])\n",
    "            im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=CM_DIVERGING,vmin=-10,vmax=10)\n",
    "            labels = list(mat.columns)\n",
    "            axmatrix.xaxis.set_ticks_position('bottom')\n",
    "            axmatrix.set_yticklabels([]);\n",
    "            axmatrix.set_xticklabels([]);\n",
    "            #xtick = plt.xticks(range(len(labels)), labels, rotation = 90, fontsize = 6,fontname='Arial')\n",
    "            axmatrix.grid(False)\n",
    "            plt.tick_params(axis='both', size=0)\n",
    "            \n",
    "            # SAVE FIGURE\n",
    "            figure_label = col+'_'+method+'_'+metric\n",
    "            fn = '/workdir/uvmel_project/figures/Revision_2_Figures/Genesets_4_15_Heatmaps/' + figure_label \n",
    "                \n",
    "            d = os.path.dirname(fn)\n",
    "            if not os.path.exists(d):\n",
    "                os.makedirs(d)\n",
    "                \n",
    "            fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "            #fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "            print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RANK BY GEP 2\n",
    "subset_type = 'TUMOR'\n",
    "exec('QUERY = INDF_{}'.format(subset_type))\n",
    "exec('META = METADATA_{}'.format(subset_type))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LOAD GENESETS\n",
    "target_genes = pd.read_csv(DATA_PATH+'Genesets_5_5.csv', \n",
    "                                sep=',', header=0)\n",
    "\n",
    "# Rank QUERY by GEP2\n",
    "rank_by = 'Castle 2'#\n",
    "genesets_include = [rank_by]\n",
    "genes = np.unique(genesets[genesets_include].values.ravel().tolist())\n",
    "genes = [gene for gene in genes if gene in list(set(QUERY.columns))]\n",
    "vals = QUERY[genes]\n",
    "QUERY['GEP2_RANK'] =np.nanmean(vals,axis=1)\n",
    "QUERY = QUERY.sort_values(by=['GEP2_RANK'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('MSK Data')\n",
    "for col in target_genes.columns:\n",
    "    print(col,':')\n",
    "    plot_genes = target_genes[col].dropna().values.tolist()\n",
    "    print('Number of genes in set:',len(plot_genes))\n",
    "    plot_genes = [x for x in plot_genes if x in QUERY.columns.values.tolist()]\n",
    "    print('Number of genes in set in data:',len(plot_genes))\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LOAD GENESETS\n",
    "target_genes = pd.read_csv(DATA_PATH+'Genesets_5_5.csv', \n",
    "                                sep=',', header=0)\n",
    "\n",
    "# Rank QUERY by GEP2\n",
    "rank_by = 'Castle 2'#\n",
    "genesets_include = [rank_by]\n",
    "genes = np.unique(genesets[genesets_include].values.ravel().tolist())\n",
    "genes = [gene for gene in genes if gene in list(set(QUERY.columns))]\n",
    "vals = QUERY[genes]\n",
    "QUERY['GEP2_RANK'] =np.nanmean(vals,axis=1)\n",
    "QUERY = QUERY.sort_values(by=['GEP2_RANK'])\n",
    "\n",
    "# Iterate through each geneset:\n",
    "for col in target_genes.columns:\n",
    "    print(col)\n",
    "    plot_genes = target_genes[col].dropna().values.tolist()\n",
    "    plot_genes = [x for x in plot_genes if x in QUERY.columns.values.tolist()]\n",
    "    \n",
    "    if len(plot_genes) > 0:\n",
    "    \n",
    "        # CONSTRUCT HEATMAP DATA\n",
    "        heatmap_data = pd.DataFrame( data = zscore(QUERY[plot_genes],axis=0), \n",
    "                           columns = plot_genes, index = QUERY.index)\n",
    "        heatmap_data = heatmap_data.dropna(axis=1, how='any') \n",
    "        yticks = heatmap_data.index\n",
    "        xticks = heatmap_data.columns\n",
    "        \n",
    "        # LINKAGE \n",
    "        for m in [('average', 'euclidean'), ('average', 'cosine'), ('centroid', 'euclidean')]:\n",
    "            method = m[0] # average, single centroid/euclidean\n",
    "            metric = m[1] # cosine\n",
    "            linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "            col_linkage = deepcopy(linkage)\n",
    "            cl = hc.leaves_list(col_linkage)\n",
    "            mat = heatmap_data.iloc[:,cl]\n",
    "            \n",
    "            window = 20\n",
    "            mat = mat.rolling(window, win_type='triang',center = True).sum()\n",
    "            half_window = int(window/2)\n",
    "            for ind in np.arange(half_window):\n",
    "                mat.iloc[ind] = mat.iloc[half_window]    \n",
    "            for ind in np.arange(half_window):\n",
    "                mat.iloc[-ind] = mat.iloc[-half_window]\n",
    "            \n",
    "            # VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "            fig = plt.figure(figsize=(4,10))\n",
    "            \n",
    "            # ADD MATRIX WITH LINEAGE NAMES\n",
    "            axmatrix = fig.add_axes([0.12,0.1,0.4,0.6])\n",
    "            im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=CM_DIVERGING,vmin=-10,vmax=10)\n",
    "            labels = list(mat.columns)\n",
    "            axmatrix.xaxis.set_ticks_position('bottom')\n",
    "            axmatrix.set_yticklabels([]);\n",
    "            axmatrix.set_xticklabels([]);\n",
    "            #xtick = plt.xticks(range(len(labels)), labels, rotation = 90, fontsize = 6,fontname='Arial')\n",
    "            axmatrix.grid(False)\n",
    "            plt.tick_params(axis='both', size=0)\n",
    "            \n",
    "            # SAVE FIGURE\n",
    "            figure_label = col+'_'+method+'_'+metric\n",
    "            fn = '/workdir/uvmel_project/figures/Revision_2_Figures/Genesets_5_5_Heatmaps/' + figure_label \n",
    "                \n",
    "            d = os.path.dirname(fn)\n",
    "            if not os.path.exists(d):\n",
    "                os.makedirs(d)\n",
    "                \n",
    "            fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "            #fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "            print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### GEP2 RANK CURVE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LOAD GENESETS\n",
    "plot_genes = ['IRF3', 'RELA', 'RELB', 'MB21D1', 'IKBKB', 'NFKB1', 'TMEM173', 'STAT2', 'CXCL10', 'NFKB2',\n",
    "              'CCL5', 'TBK1', 'JAK1', 'JAK2', 'IRF9', 'TYK2', 'IRF7', 'IRF1', 'STAT1'\n",
    "             ]\n",
    "\n",
    "path_to_genesets = DATA_PATH+'uveal_melanoma_MB_v3.csv'\n",
    "genesets = pd.read_csv(path_to_genesets,header='infer')\n",
    "\n",
    "subset_type = 'TUMOR'\n",
    "datatype = 'INDF_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "\n",
    "# HEATMAP OF INDIVIDUAL CELLS RANKED BY LUNGE EP\n",
    "rank_by = 'Castle 2'#\n",
    "genesets_include = [rank_by]\n",
    "genes = np.unique(genesets[genesets_include].values.ravel().tolist())\n",
    "genes = [gene for gene in genes if gene in list(set(QUERY.columns))]\n",
    "vals = QUERY[genes]\n",
    "QUERY['GEP2_RANK'] =np.nanmean(vals,axis=1)\n",
    "heatmap_data = pd.DataFrame( data = QUERY.sort_values(by=['GEP2_RANK'])[plot_genes +['GEP2_RANK']], \n",
    "                   columns = plot_genes +['GEP2_RANK'], index = QUERY.sort_values(by=['GEP2_RANK']).index)\n",
    "\n",
    "y = heatmap_data['GEP2_RANK'].values\n",
    "x = np.arange(len(y))\n",
    "\n",
    "fig = plt.figure(figsize=(10,4))\n",
    "ax = plt.gca()\n",
    "\n",
    "ax.fill_between(x, 0, y,facecolor='#E0E0E0', interpolate=True)\n",
    "plt.xlim((0,x.max()))\n",
    "plt.ylim((0,y.max()))\n",
    "sns.despine()\n",
    "\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'PRC1_Targets_UPDATED_GEP2_RANK_curve'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/GEP2_Ranked_Heatmaps/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PRC1 TARGET GENE DEPREPRESSION IN HIGH RISK UM"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### ORIGINAL PRC1 TARGET GENES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LOAD GENESETS\n",
    "path_to_genesets = DATA_PATH + 'mRNA_BAP1_DELETED_INTACT.csv'\n",
    "znormalized_data = pd.read_csv(path_to_genesets,header='infer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LOAD GENESETS\n",
    "path_to_genesets = DATA_PATH + 'mRNA_cell_lines.csv'\n",
    "znormalized_data_cells = pd.read_csv(path_to_genesets,header='infer')\n",
    "znormalized_data_cells.set_index('Gene Symbol', inplace=True)\n",
    "znormalized_data_cells  = znormalized_data_cells.T\n",
    "\n",
    "znormalized_data_cells.fillna(0, inplace = True)\n",
    "znormalized_data_cells = znormalized_data_cells.loc[(znormalized_data_cells.sum(axis=1) != 0), \n",
    "                                                    (znormalized_data_cells.sum(axis=0) != 0)]\n",
    "znormalized_data_cells.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "znormalized_data.set_index('BAP1: Putative copy-number alterations from GISTIC', inplace=True)\n",
    "znormalized_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "znormalized_data.fillna(0, inplace = True)\n",
    "znormalized_data = znormalized_data.loc[(znormalized_data.sum(axis=1) != 0), (znormalized_data.sum(axis=0) != 0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "znormalized_data_sorted = znormalized_data.sort_index(axis = 0)\n",
    "znormalized_data_sorted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "znormalized_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmap_data = znormalized_data\n",
    "\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns\n",
    "\n",
    "# LINKAGE \n",
    "method = 'average' # average, single centroid/euclidean\n",
    "metric = 'correlation' # cosine\n",
    "linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "col_linkage = deepcopy(linkage)\n",
    "cl = hc.leaves_list(col_linkage)\n",
    "linkage = hc.linkage(heatmap_data, method=method, metric = metric)\n",
    "row_linkage = deepcopy(linkage)\n",
    "rl = hc.leaves_list(row_linkage)\n",
    "mat = heatmap_data.iloc[rl,cl]\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(20,3))\n",
    "\n",
    "# ADD MATRIX WITH LINEAGE NAMES\n",
    "axmatrix = fig.add_axes([0.12,0.1,0.4,0.8])\n",
    "colors = [(1,1,1), np.divide(tuple(hex('FF0000').rgb),255)]\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=CM_DIVERGING,vmin=-2,vmax=2)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "#axmatrix.set_xticklabels([]);\n",
    "#xtick = plt.xticks(range(len(labels)), labels, rotation = 90, fontsize = 6,fontname='Arial')\n",
    "xtick = plt.xticks([])\n",
    "axmatrix.grid(False)\n",
    "axmatrix.spines['top'].set_visible(False)\n",
    "axmatrix.spines['right'].set_visible(False)\n",
    "axmatrix.spines['bottom'].set_visible(False)\n",
    "axmatrix.spines['left'].set_visible(False)\n",
    "axmatrix.tick_params(axis=u'both', which=u'both',length=0)\n",
    "\n",
    "# ROW1 COLORS\n",
    "# COLORMAP LUT\n",
    "lut = {\n",
    "                 'Shallow Deletion':'red', # pink\n",
    "                 'Diploid': 'blue'\n",
    "    }\n",
    "\n",
    "ind1 = mat.index\n",
    "row_colors1 = pd.Series(ind1).map(lut)\n",
    "\n",
    "# ADD ROW COLOR INDEX (CELL OF ORIGIN)\n",
    "ax1 = fig.add_axes([0.08,0.1,0.02,0.8]) # [x0,y0,width,height]\n",
    "x = 0\n",
    "y = 0\n",
    "for c in row_colors1:\n",
    "    pos = (x, y / len(row_colors1))\n",
    "    ax1.add_patch(patches.Rectangle(pos, 1, 1 / len(row_colors1), color=c))\n",
    "    if y >= len(row_colors1)-1:\n",
    "        x += 1\n",
    "        y = 0\n",
    "    else:\n",
    "        y += 1\n",
    "plt.axis('off')\n",
    "\n",
    "\n",
    "# ADD COLORBAR\n",
    "axcolor = fig.add_axes([0.525,0.1,0.007,0.6])\n",
    "cbar = plt.colorbar(im, cax=axcolor)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '/PRC1_Targets_Derepressed_highriskUM_ORIGINAL'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/PRC1_Target_Genes/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "QUERY = znormalized_data_cells\n",
    "heatmap_data = pd.DataFrame( data = zscore(QUERY,axis=0), columns = QUERY.columns, index = QUERY.index)\n",
    "\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns\n",
    "\n",
    "# LINKAGE \n",
    "method = 'average' # average, single centroid/euclidean\n",
    "metric = 'correlation' # cosine\n",
    "linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "col_linkage = deepcopy(linkage)\n",
    "cl = hc.leaves_list(col_linkage)\n",
    "linkage = hc.linkage(heatmap_data, method=method, metric = metric)\n",
    "row_linkage = deepcopy(linkage)\n",
    "rl = hc.leaves_list(row_linkage)\n",
    "mat = heatmap_data.iloc[rl,cl]\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(20,2))\n",
    "\n",
    "# ADD MATRIX WITH LINEAGE NAMES\n",
    "axmatrix = fig.add_axes([0.12,0.1,0.4,0.6])\n",
    "colors = [(1,1,1), np.divide(tuple(hex('FF0000').rgb),255)]\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=CM_DIVERGING,vmin=-2,vmax=2)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "#axmatrix.set_xticklabels([]);\n",
    "#xtick = plt.xticks(range(len(labels)), labels, rotation = 90, fontsize = 6,fontname='Arial')\n",
    "xtick = plt.xticks([])\n",
    "axmatrix.grid(False)\n",
    "axmatrix.spines['top'].set_visible(False)\n",
    "axmatrix.spines['right'].set_visible(False)\n",
    "axmatrix.spines['bottom'].set_visible(False)\n",
    "axmatrix.spines['left'].set_visible(False)\n",
    "axmatrix.tick_params(axis=u'both', which=u'both',length=0)\n",
    "\n",
    "# ROW1 COLORS\n",
    "# COLORMAP LUT\n",
    "lut = {\n",
    "                 'DM1_2592':'blue', # pink\n",
    "                 'DM2_2592': 'blue',\n",
    "                 'S_2538_22':'red', # pink\n",
    "                 'S_2538_23': 'red'\n",
    "    }\n",
    "\n",
    "ind1 = mat.index\n",
    "row_colors1 = pd.Series(ind1).map(lut)\n",
    "\n",
    "# ADD ROW COLOR INDEX (CELL OF ORIGIN)\n",
    "ax1 = fig.add_axes([0.08,0.1,0.035,0.6]) # [x0,y0,width,height]\n",
    "x = 0\n",
    "y = 0\n",
    "for c in row_colors1:\n",
    "    pos = (x, y / len(row_colors1))\n",
    "    ax1.add_patch(patches.Rectangle(pos, 1, 1 / len(row_colors1), color=c))\n",
    "    if y >= len(row_colors1)-1:\n",
    "        x += 1\n",
    "        y = 0\n",
    "    else:\n",
    "        y += 1\n",
    "plt.axis('off')\n",
    "\n",
    "# ADD DENDROGRAM\n",
    "ax2 = fig.add_axes([0.00,0.1,0.08,0.6]) # [x0,y0,width,height]\n",
    "Z1 = sch.dendrogram(row_linkage, orientation='left',color_threshold = 0, above_threshold_color='#808080')\n",
    "ax2.set_xticks([])\n",
    "ax2.set_yticks([])\n",
    "plt.axis('off')\n",
    "\n",
    "# ADD COLORBAR\n",
    "axcolor = fig.add_axes([0.525,0.1,0.007,0.6])\n",
    "cbar = plt.colorbar(im, cax=axcolor)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '/PRC1_Targets_unsupervised_clustering_ORIGINAL'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/PRC1_Target_Genes/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### UPDATED PRC1 TARGET GENES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LOAD GENESETS\n",
    "path_to_genesets = DATA_PATH + 'TCGA_new PRC1 targets.csv'\n",
    "znormalized_data = pd.read_csv(path_to_genesets,header='infer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LOAD GENESETS\n",
    "path_to_genesets = DATA_PATH + '921vsMP38.csv'\n",
    "znormalized_data_cells = pd.read_csv(path_to_genesets,header='infer')\n",
    "znormalized_data_cells.set_index('gene_name', inplace=True)\n",
    "znormalized_data_cells  = znormalized_data_cells.T\n",
    "\n",
    "znormalized_data_cells.fillna(0, inplace = True)\n",
    "znormalized_data_cells = znormalized_data_cells.loc[(znormalized_data_cells.sum(axis=1) != 0), \n",
    "                                                    (znormalized_data_cells.sum(axis=0) != 0)]\n",
    "znormalized_data_cells.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "znormalized_data.set_index('Chromosome 3', inplace=True)\n",
    "znormalized_data.drop(columns=['Sample name'], inplace=True)\n",
    "znormalized_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "znormalized_data.fillna(0, inplace = True)\n",
    "znormalized_data = znormalized_data.loc[(znormalized_data.sum(axis=1) != 0), (znormalized_data.sum(axis=0) != 0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "znormalized_data_sorted = znormalized_data.sort_index(axis = 0)\n",
    "znormalized_data_sorted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "znormalized_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmap_data = znormalized_data#_sorted\n",
    "\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns\n",
    "\n",
    "# LINKAGE \n",
    "method = 'average' # average, single centroid/euclidean\n",
    "metric = 'correlation' # cosine\n",
    "linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "col_linkage = deepcopy(linkage)\n",
    "cl = hc.leaves_list(col_linkage)\n",
    "linkage = hc.linkage(heatmap_data, method=method, metric = metric)\n",
    "row_linkage = deepcopy(linkage)\n",
    "rl = hc.leaves_list(row_linkage)\n",
    "mat = heatmap_data.iloc[rl,cl]\n",
    "\n",
    "genes_to_show = ['ZEB2', 'SLC1A4', 'CSPG4', 'FADS3', 'VLDLR', 'ASPH', 'LHFPL2', 'TRIO', 'CHST3', 'C2CD2', 'NT5E']\n",
    "xlabel_list = [x if x in genes_to_show else '' for x in mat.columns.values]\n",
    "xtick_list = [x for x in range(len(xlabel_list)) if xlabel_list[x] !='']\n",
    "xlabel_list = [x for x in xlabel_list if x !='']\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(20,3))\n",
    "\n",
    "# ADD MATRIX WITH LINEAGE NAMES\n",
    "axmatrix = fig.add_axes([0.12,0.1,0.4,0.7])\n",
    "colors = [(1,1,1), np.divide(tuple(hex('FF0000').rgb),255)]\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=CM_DIVERGING,vmin=-2,vmax=2)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "#axmatrix.set_xticklabels([]);\n",
    "xtick = plt.xticks(xtick_list, xlabel_list, rotation = 90, fontsize = 6,fontname='Arial')\n",
    "#xtick = plt.xticks([])\n",
    "axmatrix.grid(False)\n",
    "axmatrix.spines['top'].set_visible(False)\n",
    "axmatrix.spines['right'].set_visible(False)\n",
    "axmatrix.spines['bottom'].set_visible(False)\n",
    "axmatrix.spines['left'].set_visible(False)\n",
    "axmatrix.tick_params(axis=u'both', which=u'both',length=3)\n",
    "\n",
    "# ROW1 COLORS\n",
    "# COLORMAP LUT\n",
    "lut = {\n",
    "                 'Monosomy 3':'red', # pink\n",
    "                 'Diploid': 'blue'\n",
    "    }\n",
    "\n",
    "ind1 = mat.index\n",
    "row_colors1 = pd.Series(ind1).map(lut)\n",
    "\n",
    "# ADD ROW COLOR INDEX (CELL OF ORIGIN)\n",
    "ax1 = fig.add_axes([0.1,0.1,0.02,0.7]) # [x0,y0,width,height]\n",
    "x = 0\n",
    "y = 0\n",
    "for c in row_colors1:\n",
    "    pos = (x, y / len(row_colors1))\n",
    "    ax1.add_patch(patches.Rectangle(pos, 1, 1 / len(row_colors1), color=c))\n",
    "    if y >= len(row_colors1)-1:\n",
    "        x += 1\n",
    "        y = 0\n",
    "    else:\n",
    "        y += 1\n",
    "plt.axis('off')\n",
    "\n",
    "# ADD COLORBAR\n",
    "axcolor = fig.add_axes([0.525,0.1,0.0095,0.25])\n",
    "cbar = plt.colorbar(im, cax=axcolor)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'PRC1_Targets_Derepressed_highriskUM_UPDATED'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/PRC1_Target_Genes/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "QUERY = znormalized_data_cells\n",
    "heatmap_data = pd.DataFrame( data = zscore(QUERY,axis=0), columns = QUERY.columns, index = QUERY.index)\n",
    "\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns\n",
    "\n",
    "# LINKAGE \n",
    "method = 'average' # average, single centroid/euclidean\n",
    "metric = 'correlation' # cosine\n",
    "linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "col_linkage = deepcopy(linkage)\n",
    "cl = hc.leaves_list(col_linkage)\n",
    "#linkage = hc.linkage(heatmap_data, method=method, metric = metric)\n",
    "#row_linkage = deepcopy(linkage)\n",
    "#rl = hc.leaves_list(row_linkage)\n",
    "mat = heatmap_data.iloc[:,cl]\n",
    "\n",
    "genes_to_show = ['ZEB2', 'SLC1A4', 'CSPG4', 'FADS3', 'VLDLR', 'ASPH', 'LHFPL2', 'TRIO', 'CHST3', 'C2CD2', 'NT5E']\n",
    "xlabel_list = [x if x in genes_to_show else '' for x in mat.columns.values]\n",
    "xtick_list = [x for x in range(len(xlabel_list)) if xlabel_list[x] !='']\n",
    "xlabel_list = [x for x in xlabel_list if x !='']\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(20,2))\n",
    "\n",
    "# ADD MATRIX WITH LINEAGE NAMES\n",
    "axmatrix = fig.add_axes([0.12,0.1,0.4,0.6])\n",
    "colors = [(1,1,1), np.divide(tuple(hex('FF0000').rgb),255)]\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=CM_DIVERGING,vmin=-2,vmax=2)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "#axmatrix.set_xticklabels([]);\n",
    "xtick = plt.xticks(xtick_list, xlabel_list, rotation = 90, fontsize = 6,fontname='Arial')\n",
    "#xtick = plt.xticks([])\n",
    "axmatrix.grid(False)\n",
    "axmatrix.spines['top'].set_visible(False)\n",
    "axmatrix.spines['right'].set_visible(False)\n",
    "axmatrix.spines['bottom'].set_visible(False)\n",
    "axmatrix.spines['left'].set_visible(False)\n",
    "axmatrix.tick_params(axis=u'x', which=u'both',length=3)\n",
    "axmatrix.tick_params(axis=u'y', which=u'both',length=0)\n",
    "\n",
    "# ROW1 COLORS\n",
    "# COLORMAP LUT\n",
    "lut = {\n",
    "                 '92.1':'blue', # pink\n",
    "                 '92.1.1': 'blue',\n",
    "                 'MP38':'red', # pink\n",
    "                 'MP38.1': 'red'\n",
    "    }\n",
    "\n",
    "ind1 = mat.index\n",
    "row_colors1 = pd.Series(ind1).map(lut)\n",
    "\n",
    "# ADD ROW COLOR INDEX (CELL OF ORIGIN)\n",
    "ax1 = fig.add_axes([0.09,0.1,0.03,0.6]) # [x0,y0,width,height]\n",
    "x = 0\n",
    "y = 0\n",
    "for c in row_colors1:\n",
    "    pos = (x, y / len(row_colors1))\n",
    "    ax1.add_patch(patches.Rectangle(pos, 1, 1 / len(row_colors1), color=c))\n",
    "    if y >= len(row_colors1)-1:\n",
    "        x += 1\n",
    "        y = 0\n",
    "    else:\n",
    "        y += 1\n",
    "plt.axis('off')\n",
    "\n",
    "# ADD COLORBAR\n",
    "axcolor = fig.add_axes([0.525,0.1,0.007,0.6])\n",
    "cbar = plt.colorbar(im, cax=axcolor)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'PRC1_Targets_unsupervised_clustering_UPDATED'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/PRC1_Target_Genes/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmap_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## STING PATHWAY GENES HEATMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "plot_genes = ['MB21D1','TMEM173','IRF3','RELA','RELB','IKBKB','NFKB1',\n",
    "              'STAT2','CXCL10','NFKB2','CCL5',\n",
    "              'TBK1','JAK1','JAK2','IRF9',\n",
    "              'TYK2','IRF7','IRF1','STAT1',\n",
    "              'PPARG','DDIT3','NUPR1','RAB3B','IGFBP4','LRRC8C',\n",
    "              'TCP11L2','MAFK','NRG1','F2R','KRT19','CTGF','ZFC3H1']\n",
    "\n",
    "print(len(plot_genes))\n",
    "\n",
    "# LOAD GENESETS\n",
    "path_to_genesets = DATA_PATH+'uveal_melanoma_MB_v3.csv'\n",
    "genesets = pd.read_csv(path_to_genesets,header='infer')\n",
    "\n",
    "subset_type = 'TUMOR'\n",
    "datatype = 'INDF_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "\n",
    "# HEATMAP OF INDIVIDUAL CELLS RANKED BY LUNGE EP\n",
    "rank_by = 'Castle 2'#\n",
    "genesets_include = [rank_by]\n",
    "genes = np.unique(genesets[genesets_include].values.ravel().tolist())\n",
    "genes = [gene for gene in genes if gene in list(set(QUERY.columns))]\n",
    "vals = QUERY[genes]\n",
    "QUERY['GEP2_RANK'] =np.nanmean(vals,axis=1)\n",
    "\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "heatmap_data = pd.DataFrame( data = zscore(QUERY.sort_values(by=['GEP2_RANK'])[plot_genes],axis=0), \n",
    "                   columns = plot_genes, index = QUERY.sort_values(by=['GEP2_RANK']).index)\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns\n",
    "\n",
    "# LINKAGE \n",
    "method = 'average' # average, single centroid/euclidean\n",
    "metric = 'cosine' # cosine\n",
    "linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "col_linkage = deepcopy(linkage)\n",
    "cl = hc.leaves_list(col_linkage)\n",
    "mat = heatmap_data.iloc[:,cl]\n",
    "#mat = heatmap_data.iloc[:,:]\n",
    "\n",
    "window = 20\n",
    "mat = mat.rolling(window, win_type='triang',center = True).sum()\n",
    "half_window = int(window/2)\n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[ind] = mat.iloc[half_window]    \n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[-ind] = mat.iloc[-half_window]\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(5,10))\n",
    "\n",
    "# ADD MATRIX WITH LINEAGE NAMES\n",
    "axmatrix = fig.add_axes([0.12,0.1,0.5,0.7])\n",
    "colors = [(1,1,1), np.divide(tuple(hex('FF0000').rgb),255)]\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=CM_DIVERGING,vmin=-10,vmax=10)\n",
    "labels = list(mat.columns)\n",
    "#labels[3] = 'CGAS'\n",
    "#labels[6] = 'STING'\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "#axmatrix.set_xticklabels([]);\n",
    "xtick = plt.xticks(range(len(labels)), labels, rotation = 90, fontsize = 5,fontname='Arial')\n",
    "axmatrix.grid(False)\n",
    "plt.tick_params(size=0)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'STING_PATHWAY_Heatmap_EXTENDED_CLUSTERED_average_cosine'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/GEP2_Ranked_Heatmaps/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HARBOUR DATASET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_PATH_HARBOUR = DATA_PATH + '/uveal_melanoma_harbour2020/'\n",
    "\n",
    "ADATA_HARBOUR = sc.read(DATA_PATH_HARBOUR+'merged_pipeline_out/filter2.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ADATA_HARBOUR.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Primary and metastatic sample sizes before filtering:\n",
    "ADATA_HARBOUR.obs.sampleType.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sample sizes by patient before filtering:\n",
    "ADATA_HARBOUR.obs['sample'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### FILTER TO JUST TUMOR CELLS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_HARBOUR_CELLTYPES = pd.read_csv(DATA_PATH_HARBOUR+'scUM_NComms_Fig1B_metadata_MD.csv', \n",
    "                                   names=['Barcode', 'Patient', 'cellType'], header=0)\n",
    "DF_HARBOUR_CELLTYPES.set_index('Barcode', inplace=True)\n",
    "\n",
    "tumor_barcodes = DF_HARBOUR_CELLTYPES.loc[DF_HARBOUR_CELLTYPES.cellType.map(lambda x: 'Tumor' in x)].index.values\n",
    "\n",
    "adata_harbour_barcode_list = [ADATA_HARBOUR.obs.index[x].split('-')[0] +'-'+ \n",
    "                              ADATA_HARBOUR.obs['sample'][x].split('_')[1] \n",
    "                              for x in range(len(ADATA_HARBOUR.obs))]\n",
    "\n",
    "tumor_cell_indices = [ADATA_HARBOUR.obs.index.values[x] for x in range(len(adata_harbour_barcode_list)) \n",
    "                      if adata_harbour_barcode_list[x] in tumor_barcodes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_HARBOUR_CELLTYPES.cellType.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ADATA_HARBOUR = ADATA_HARBOUR[tumor_cell_indices,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get cellTypes list for tumor cells:\n",
    "adata_barcodes_cleaned = [ADATA_HARBOUR.obs.index[x].split('-')[0] +'-'+ \n",
    "                          ADATA_HARBOUR.obs['sample'][x].split('_')[1] \n",
    "                          for x in range(len(ADATA_HARBOUR.obs))]\n",
    "\n",
    "cellTypes_list = [DF_HARBOUR_CELLTYPES.loc[DF_HARBOUR_CELLTYPES.index == x].cellType.values.tolist()[0]\n",
    "                  for x in adata_barcodes_cleaned]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Primary and metastatic sample sizes after filtering:\n",
    "ADATA_HARBOUR.obs.sampleType.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show metastasis sample counts:\n",
    "ADATA_HARBOUR.obs.loc[ADATA_HARBOUR.obs.sampleType.map(lambda x: x == 'metastatic')]['sample'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sample sizes by patient after filtering:\n",
    "ADATA_HARBOUR.obs['sample'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### CREATE PANDAS DF TO WORK WITH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "harbour_matrix = ADATA_HARBOUR.X.toarray()\n",
    "harbour_gene_names = ADATA_HARBOUR.var.index.values.tolist()\n",
    "DF_HARBOUR = pd.DataFrame(harbour_matrix, columns=harbour_gene_names)\n",
    "    \n",
    "DF_HARBOUR['Sample ID'] = ADATA_HARBOUR.obs['sample_num'].values.tolist()\n",
    "DF_HARBOUR['Legend'] = ADATA_HARBOUR.obs['sampleType'].values.tolist()\n",
    "DF_HARBOUR['Patient'] = [x.split('_')[1] for x in ADATA_HARBOUR.obs['sample'].values.tolist()]\n",
    "DF_HARBOUR['Cell ID'] = DF_HARBOUR.index.values.tolist()\n",
    "DF_HARBOUR['cellType'] = cellTypes_list\n",
    "DF_HARBOUR.set_index(['Sample ID', 'Legend', 'Patient', 'Cell ID', 'cellType'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_HARBOUR.index.get_level_values('cellType').value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_HARBOUR.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### FILTER OUT GENES THAT AREN'T EXPRESSED IN ANY CELLS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_HARBOUR = DF_HARBOUR.loc[:, (DF_HARBOUR != 0).any(axis=0)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### GET DF OF JUST PRIMARY TUMOR CELLS:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_HARBOUR_PRIMARY = DF_HARBOUR.loc[DF_HARBOUR.index.map(lambda x: x[1] == 'primary')]\n",
    "DF_HARBOUR_METASTASIS = DF_HARBOUR.loc[DF_HARBOUR.index.map(lambda x: x[1] == 'metastatic')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### WRITE DATAFRAMES TO h5 FOR CELLASSIGN"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# This command only needs to be run once to save the h5 file:\n",
    "\n",
    "DF_HARBOUR.to_hdf(DATA_PATH_HARBOUR+'UVMEL_HARBOUR.h5', key='DF_TUMOR')\n",
    "DF_HARBOUR_PRIMARY.to_hdf(DATA_PATH_HARBOUR+'UVMEL_HARBOUR_PRIMARY.h5', key='DF_TUMOR')\n",
    "DF_HARBOUR_METASTASIS.to_hdf(DATA_PATH_HARBOUR+'UVMEL_HARBOUR_METASTASIS.h5', key='DF_TUMOR')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "h5_data_harbour = seqc.H5(DATA_PATH_HARBOUR+'UVMEL_HARBOUR.h5')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### NORMALIZE HARBOUR DATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "NDF_HARBOUR = DF_HARBOUR.copy()\n",
    "molecule_sums = NDF_HARBOUR.sum(axis=1)\n",
    "\n",
    "NDF_HARBOUR = NDF_HARBOUR.div(molecule_sums, axis=0).mul(np.median(molecule_sums), axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "NDF_HARBOUR.to_hdf(DATA_PATH_HARBOUR+'UVMEL_HARBOUR.h5', key='NDF_TUMOR')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "NDF_HARBOUR_PRIMARY = DF_HARBOUR_PRIMARY.copy()\n",
    "molecule_sums_PRIMARY = NDF_HARBOUR_PRIMARY.sum(axis=1)\n",
    "\n",
    "NDF_HARBOUR_PRIMARY = NDF_HARBOUR_PRIMARY.div(molecule_sums_PRIMARY, axis=0).mul(np.median(molecule_sums_PRIMARY), \n",
    "                                                                                 axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "NDF_HARBOUR_PRIMARY.to_hdf(DATA_PATH_HARBOUR+'UVMEL_HARBOUR.h5', key='NDF_TUMOR_PRIMARY')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### PERFORM IMPUTATION ON NDF_HARBOUR"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### FILTER LOW ABUNDANCE GENES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FIT BINOMIAL DISTRIBUTION AND FILTER BASED ON MEAN/STD OF SECOND\n",
    "plt.figure(figsize = (10,3))\n",
    "gs1 = gridspec.GridSpec(1, 2)\n",
    "gs1.update(wspace=0.7, hspace=0.7) # set the spacing between axes. \n",
    "\n",
    "# (4) PLOT LOG NUMBER OF CELLS CONTRIBUTING TO EACH GENE\n",
    "num_cells_per_gene = np.log(np.sum(DF_HARBOUR.values > 0,axis=0))\n",
    "num_cells_per_gene[(np.isinf(num_cells_per_gene)) | (np.isnan(num_cells_per_gene))] = 0\n",
    "rmv_genes1 = np.where(num_cells_per_gene<=0)[0] # GENES MUST BE DETECTED IN AT LEAST 10 CELLS\n",
    "\n",
    "ax = plt.subplot(gs1[0])\n",
    "bins = np.linspace(num_cells_per_gene.min(), num_cells_per_gene.max()*0.95, 20)\n",
    "plt.hist(num_cells_per_gene, bins, alpha=0.5, label='keep')\n",
    "\n",
    "if rmv_genes1.any():\n",
    "    plt.hist(num_cells_per_gene[rmv_genes1], bins, alpha=1, label='remove')\n",
    "    \n",
    "#ax.set_axis_facecolor('white')\n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Frequency')\n",
    "plt.xlabel('Gene Filter 1: Remove Genes Singletons \\n(Log. # Expressing Cells)')\n",
    "plt.grid(True)\n",
    "sns.despine()\n",
    "\n",
    "# (5) PLOT LOG COUNTS PER GENE - REMOVE LOW ABUNDANCE GENES\n",
    "log_counts_per_gene = np.log(np.log(np.sum(DF_HARBOUR.values,axis=0)))\n",
    "log_counts_per_gene[(np.isinf(log_counts_per_gene)) | (np.isnan(log_counts_per_gene))] = 0\n",
    "data = log_counts_per_gene\n",
    "\n",
    "ax = plt.subplot(gs1[1])\n",
    "bins = np.linspace(data.min(), data.max()*0.95, 100)\n",
    "y,x,_=hist(data,bins,alpha=.3,label='data')\n",
    "\n",
    "x=(x[1:]+x[:-1])/2 \n",
    "expected=(0,.2,3500,1.5,.2,500)\n",
    "params,cov=curve_fit(bimodal,x,y,expected)\n",
    "sigma=sqrt(diag(cov))\n",
    "plot(x,bimodal(x,*params),color='red',lw=3,label='model')\n",
    "\n",
    "mu1 = params[0]\n",
    "std1 = params[1]\n",
    "mu2 = params[3]\n",
    "std2 = params[4]\n",
    "rmv_genes_neg = np.where(data<mu2-4*std2)[0]\n",
    "rmv_genes2 = np.sort(list(set(list(rmv_genes_neg))))\n",
    "\n",
    "if rmv_genes2.any():\n",
    "    plt.hist(data[rmv_genes2], bins, alpha=1, label='remove')\n",
    "    \n",
    "#ax.set_axis_facecolor('white')\n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Frequency')\n",
    "plt.xlabel('Gene Filter 2: \\nLog-Log Counts/Gene')\n",
    "plt.grid(True)\n",
    "sns.despine()\n",
    "\n",
    "# Add (abbreviated) legend bottom left\n",
    "L = plt.legend(loc='upper right',prop={'size':12},bbox_to_anchor=(1.6, 1.05),fancybox=True) \n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '_filter_genes'\n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label + '.png'\n",
    "plt.savefig(fn, dpi=fig_dpi)\n",
    "print(fn)\n",
    "\n",
    "# EVALUATE NUMBER OF CELLS/GENES REMOVED\n",
    "print('Count Gene Filter 1: {}'.format(len(rmv_genes1)))\n",
    "print('Count Gene Filter 2: {}'.format(len(rmv_genes2)))\n",
    "\n",
    "# REMOVE SELECTED OUTLIER GENES(HIGHLIGHTED IN GREEN)\n",
    "CUT_DF = deepcopy(DF_HARBOUR)\n",
    "rmv_genes = np.sort(list(set(list(rmv_genes1) + list(rmv_genes2))))\n",
    "if rmv_genes.any():\n",
    "    CUT_DF = CUT_DF.drop(CUT_DF.columns[rmv_genes],axis=1)\n",
    "    print(CUT_DF.shape)\n",
    "    \n",
    "# Remove empty genes if they exist\n",
    "drop_genes = np.where(CUT_DF.sum(axis=0)==0)[0]\n",
    "CUT_DF = CUT_DF.drop(CUT_DF.columns[drop_genes],axis=1)\n",
    "print(CUT_DF.shape)\n",
    "\n",
    "# GENES EXCLUDING LOW ABUNDANCE GENES\n",
    "goi = list(CUT_DF.columns)\n",
    "print(len(goi))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### RANDOMIZED PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = NDF_HARBOUR[goi]\n",
    "pca = PCA(n_components=100, whiten = False, svd_solver='randomized')\n",
    "PCA_FIT = pca.fit(data)# cells/observations x genes/features\n",
    "\n",
    "explained_variance = PCA_FIT.explained_variance_ratio_ # explained variance\n",
    "cumsum_explained_variance = np.cumsum(explained_variance) # cumulative sum of explained variance\n",
    "PCA_DATA = PCA_FIT.transform(data) # transformed data\n",
    "\n",
    "# Store output of pca in dictionary for WISHBONE\n",
    "eigenvalues = pd.Series(pca.explained_variance_,index = np.arange(pca.n_components))\n",
    "loadings = pd.DataFrame(PCA_FIT.components_.T, index = data.columns, columns = np.arange(pca.n_components))\n",
    "pca_dict = {'eigenvalues':eigenvalues,'loadings':loadings}\n",
    "\n",
    "# Correlation of each PC component with library size\n",
    "\n",
    "x = molecule_sums.values.tolist()\n",
    "\n",
    "R = np.zeros(PCA_DATA.shape[1])\n",
    "for ind in np.arange(PCA_DATA.shape[1]):\n",
    "    y = PCA_DATA[:,ind]\n",
    "    R[ind] = np.corrcoef(x,y)[1,0] # correlation coefficient\n",
    "\n",
    "nrow = 2\n",
    "ncol = 2\n",
    "plt.figure(figsize=(7,7))\n",
    "gs1 = gridspec.GridSpec(nrow,ncol)\n",
    "gs1.update(wspace = 0.5, hspace = 0.5)\n",
    "\n",
    "# Variance of each gene\n",
    "gene_variance = np.var(data, axis=0)\n",
    "variance_sort_idx =gene_variance.argsort()[::-1]\n",
    "ranked_gene_variance = gene_variance[variance_sort_idx]\n",
    "\n",
    "# Plot explained variance\n",
    "ax = plt.subplot(gs1[0])\n",
    "plt.plot(explained_variance)\n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Ratio Explained Variance')\n",
    "plt.xlabel('Principle Component')\n",
    "plt.grid(True)\n",
    "sns.despine()\n",
    "\n",
    "# Plot cumulative sum of explained variance\n",
    "ax = plt.subplot(gs1[1])\n",
    "plt.plot(cumsum_explained_variance)\n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Cumsum Explained Variance')\n",
    "plt.xlabel('Principle Component')\n",
    "plt.grid(True)\n",
    "sns.despine()\n",
    "\n",
    "# Plot correlatin of library size with each PC\n",
    "ax = plt.subplot(gs1[2])\n",
    "plt.plot(np.abs(R))\n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Correlation with Library Size (abs)')\n",
    "plt.xlabel('Principle Component')\n",
    "plt.grid(True)\n",
    "sns.despine()\n",
    "\n",
    "# Plot ranked variance of each gene\n",
    "ax = plt.subplot(gs1[3])\n",
    "plt.plot(np.log(ranked_gene_variance.values))\n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Log. Variance')\n",
    "plt.xlabel('Ranked Genes')\n",
    "plt.grid(True)\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# IDENTIFY POINT OF MAXIMUM CURVATURE IN CUMULATIVE EXPLAINED VARIANCE\n",
    "y = cumsum_explained_variance\n",
    "x = np.arange(len(y))\n",
    "kneedle = KneeLocator(x, y)\n",
    "kneedle.plot_knee_normalized()\n",
    "knee = kneedle.knee\n",
    "plt.title('knee: {}'.format(knee))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cumsum_explained_variance[31]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PCA_DATA.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SELECT DATA ASSOCIATED WITH RELEVANT PCS (REMOVE PC0 if highly correlated with cell size)\n",
    "ncomponents = knee*2\n",
    "COMPONENTS = np.arange(ncomponents)\n",
    "print(COMPONENTS)\n",
    "\n",
    "# LOADING PER COMPONENT\n",
    "loadings = PCA_FIT.components_.T[:,COMPONENTS]\n",
    "\n",
    "labels = []\n",
    "for i in COMPONENTS:\n",
    "    labels.append('PC{}'.format(i))\n",
    "\n",
    "data = NDF_HARBOUR[goi]\n",
    "data -= np.min(np.ravel(data))\n",
    "data /= np.max(np.ravel(data))\n",
    "data = pd.DataFrame(np.dot(data, loadings).astype(np.float64),index=data.index, columns = labels)\n",
    "\n",
    "\n",
    "# WRITE TO H5\n",
    "DIMENSIONS_HARBOUR_TUMOR = data.copy()\n",
    "exec('h5_data_harbour.save(DIMENSIONS_HARBOUR_{}, \\'DIMENSIONS_HARBOUR_{}\\')'\n",
    "     .format(subset_type,subset_type)) # update directory\n",
    "h5_data_harbour.ls() # list contents of directory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### REGRESS LIBRARY SIZE OUT OF PCS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exec('QUERY = DIMENSIONS_HARBOUR_{}'.format(subset_type))\n",
    "idx = QUERY.columns.to_series().str.contains('^PC((?!TSNE).)*$') # select from DIMS_NDF\n",
    "PCA_DATA = QUERY.loc[:, idx.values]\n",
    "counts = DF_HARBOUR.copy()\n",
    "\n",
    "num_comps = PCA_DATA.shape[1]\n",
    "\n",
    "pca_regressed_out_lib_size = deepcopy(PCA_DATA)\n",
    "for c in range(num_comps):\n",
    "    lm = LinearRegression(normalize=False)\n",
    "    X = counts.sum(axis=1).values.reshape(counts.shape[0], 1)\n",
    "    Y = PCA_DATA.iloc[:, c]\n",
    "    lm.fit(X, Y)\n",
    "    pca_regressed_out_lib_size.iloc[:,c] = Y-lm.predict(X)\n",
    "    \n",
    "# Add SUFIX FOR REGRESSED LIB SIZE (RLS)\n",
    "pca_regressed_out_lib_size = pca_regressed_out_lib_size.add_prefix('RLS_')\n",
    "\n",
    "for label in pca_regressed_out_lib_size.columns:\n",
    "    exec('DIMENSIONS_HARBOUR_{}[label]= pca_regressed_out_lib_size[label]'.format(subset_type))\n",
    "    \n",
    "# WRITE TO H5\n",
    "exec('h5_data_harbour.save(DIMENSIONS_HARBOUR_{}, \\'DIMENSIONS_HARBOUR_{}\\')'.\n",
    "     format(subset_type,subset_type)) # update directory\n",
    "h5_data_harbour.ls() # list contents of directory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### CALCULATE DIFFUSIONS COEFFICIENTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "knn = 27 # truncate the neighbors to the k closest points\n",
    "\n",
    "component_prefix = 'RLS_PC'\n",
    "exec('QUERY = DIMENSIONS_HARBOUR_{}'.format(subset_type))\n",
    "idx = QUERY.columns.to_series().str.contains('^{}((?!TSNE).)*$'.format(component_prefix)) # select from DIMS_NDF\n",
    "PCA_DATA = QUERY.loc[:, idx.values]\n",
    "\n",
    "DIFFUSION_EIGS = palantir.utils.run_diffusion_maps(PCA_DATA.astype(np.float64), knn=knn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_HARBOUR_PRIMARY.index.get_level_values('Patient').value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_TUMOR.index.get_level_values('Patient').value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### IMPUTE MISSING VALUES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "imputed_data_matrix = palantir.utils.run_magic_imputation(NDF_HARBOUR, DIFFUSION_EIGS, n_steps=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "INDF_HARBOUR = imputed_data_matrix.copy()\n",
    "exec('h5_data_harbour.save(INDF_HARBOUR,\\'INDF_HARBOUR\\')')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### PERFORM IMPUTATION ON NDF_HARBOUR_PRIMARY"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### FILTER LOW ABUNDANCE GENES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FIT BINOMIAL DISTRIBUTION AND FILTER BASED ON MEAN/STD OF SECOND\n",
    "plt.figure(figsize = (10,3))\n",
    "gs1 = gridspec.GridSpec(1, 2)\n",
    "gs1.update(wspace=0.7, hspace=0.7) # set the spacing between axes. \n",
    "\n",
    "# (4) PLOT LOG NUMBER OF CELLS CONTRIBUTING TO EACH GENE\n",
    "num_cells_per_gene = np.log(np.sum(DF_HARBOUR_PRIMARY.values > 0,axis=0))\n",
    "num_cells_per_gene[(np.isinf(num_cells_per_gene)) | (np.isnan(num_cells_per_gene))] = 0\n",
    "rmv_genes1 = np.where(num_cells_per_gene<=0)[0] # GENES MUST BE DETECTED IN AT LEAST 10 CELLS\n",
    "\n",
    "ax = plt.subplot(gs1[0])\n",
    "bins = np.linspace(num_cells_per_gene.min(), num_cells_per_gene.max()*0.95, 20)\n",
    "plt.hist(num_cells_per_gene, bins, alpha=0.5, label='keep')\n",
    "\n",
    "if rmv_genes1.any():\n",
    "    plt.hist(num_cells_per_gene[rmv_genes1], bins, alpha=1, label='remove')\n",
    "    \n",
    "#ax.set_axis_facecolor('white')\n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Frequency')\n",
    "plt.xlabel('Gene Filter 1: Remove Genes Singletons \\n(Log. # Expressing Cells)')\n",
    "plt.grid(True)\n",
    "sns.despine()\n",
    "\n",
    "# (5) PLOT LOG COUNTS PER GENE - REMOVE LOW ABUNDANCE GENES\n",
    "log_counts_per_gene = np.log(np.log(np.sum(DF_HARBOUR_PRIMARY.values,axis=0)))\n",
    "log_counts_per_gene[(np.isinf(log_counts_per_gene)) | (np.isnan(log_counts_per_gene))] = 0\n",
    "data = log_counts_per_gene\n",
    "\n",
    "ax = plt.subplot(gs1[1])\n",
    "bins = np.linspace(data.min(), data.max()*0.95, 100)\n",
    "y,x,_=hist(data,bins,alpha=.3,label='data')\n",
    "\n",
    "x=(x[1:]+x[:-1])/2 \n",
    "expected=(0,.2,3500,1.5,.2,500)\n",
    "params,cov=curve_fit(bimodal,x,y,expected)\n",
    "sigma=sqrt(diag(cov))\n",
    "plot(x,bimodal(x,*params),color='red',lw=3,label='model')\n",
    "\n",
    "mu1 = params[0]\n",
    "std1 = params[1]\n",
    "mu2 = params[3]\n",
    "std2 = params[4]\n",
    "rmv_genes_neg = np.where(data<mu2-4*std2)[0]\n",
    "rmv_genes2 = np.sort(list(set(list(rmv_genes_neg))))\n",
    "\n",
    "if rmv_genes2.any():\n",
    "    plt.hist(data[rmv_genes2], bins, alpha=1, label='remove')\n",
    "    \n",
    "#ax.set_axis_facecolor('white')\n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Frequency')\n",
    "plt.xlabel('Gene Filter 2: \\nLog-Log Counts/Gene')\n",
    "plt.grid(True)\n",
    "sns.despine()\n",
    "\n",
    "# Add (abbreviated) legend bottom left\n",
    "L = plt.legend(loc='upper right',prop={'size':12},bbox_to_anchor=(1.6, 1.05),fancybox=True) \n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '_filter_genes'\n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label + '.png'\n",
    "plt.savefig(fn, dpi=fig_dpi)\n",
    "print(fn)\n",
    "\n",
    "# EVALUATE NUMBER OF CELLS/GENES REMOVED\n",
    "print('Count Gene Filter 1: {}'.format(len(rmv_genes1)))\n",
    "print('Count Gene Filter 2: {}'.format(len(rmv_genes2)))\n",
    "\n",
    "# REMOVE SELECTED OUTLIER GENES(HIGHLIGHTED IN GREEN)\n",
    "CUT_DF = deepcopy(DF_HARBOUR_PRIMARY)\n",
    "rmv_genes = np.sort(list(set(list(rmv_genes1) + list(rmv_genes2))))\n",
    "if rmv_genes.any():\n",
    "    CUT_DF = CUT_DF.drop(CUT_DF.columns[rmv_genes],axis=1)\n",
    "    print(CUT_DF.shape)\n",
    "    \n",
    "# Remove empty genes if they exist\n",
    "drop_genes = np.where(CUT_DF.sum(axis=0)==0)[0]\n",
    "CUT_DF = CUT_DF.drop(CUT_DF.columns[drop_genes],axis=1)\n",
    "print(CUT_DF.shape)\n",
    "\n",
    "# GENES EXCLUDING LOW ABUNDANCE GENES\n",
    "goi = list(CUT_DF.columns)\n",
    "print(len(goi))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### RANDOMIZED PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = NDF_HARBOUR_PRIMARY[goi]\n",
    "pca = PCA(n_components=100, whiten = False, svd_solver='randomized')\n",
    "PCA_FIT = pca.fit(data)# cells/observations x genes/features\n",
    "\n",
    "explained_variance = PCA_FIT.explained_variance_ratio_ # explained variance\n",
    "cumsum_explained_variance = np.cumsum(explained_variance) # cumulative sum of explained variance\n",
    "PCA_DATA = PCA_FIT.transform(data) # transformed data\n",
    "\n",
    "# Store output of pca in dictionary for WISHBONE\n",
    "eigenvalues = pd.Series(pca.explained_variance_,index = np.arange(pca.n_components))\n",
    "loadings = pd.DataFrame(PCA_FIT.components_.T, index = data.columns, columns = np.arange(pca.n_components))\n",
    "pca_dict = {'eigenvalues':eigenvalues,'loadings':loadings}\n",
    "\n",
    "# Correlation of each PC component with library size\n",
    "\n",
    "x = molecule_sums_PRIMARY.values.tolist()\n",
    "\n",
    "R = np.zeros(PCA_DATA.shape[1])\n",
    "for ind in np.arange(PCA_DATA.shape[1]):\n",
    "    y = PCA_DATA[:,ind]\n",
    "    R[ind] = np.corrcoef(x,y)[1,0] # correlation coefficient\n",
    "\n",
    "nrow = 2\n",
    "ncol = 2\n",
    "plt.figure(figsize=(7,7))\n",
    "gs1 = gridspec.GridSpec(nrow,ncol)\n",
    "gs1.update(wspace = 0.5, hspace = 0.5)\n",
    "\n",
    "# Variance of each gene\n",
    "gene_variance = np.var(data, axis=0)\n",
    "variance_sort_idx =gene_variance.argsort()[::-1]\n",
    "ranked_gene_variance = gene_variance[variance_sort_idx]\n",
    "\n",
    "# Plot explained variance\n",
    "ax = plt.subplot(gs1[0])\n",
    "plt.plot(explained_variance)\n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Ratio Explained Variance')\n",
    "plt.xlabel('Principle Component')\n",
    "plt.grid(True)\n",
    "sns.despine()\n",
    "\n",
    "# Plot cumulative sum of explained variance\n",
    "ax = plt.subplot(gs1[1])\n",
    "plt.plot(cumsum_explained_variance)\n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Cumsum Explained Variance')\n",
    "plt.xlabel('Principle Component')\n",
    "plt.grid(True)\n",
    "sns.despine()\n",
    "\n",
    "# Plot correlatin of library size with each PC\n",
    "ax = plt.subplot(gs1[2])\n",
    "plt.plot(np.abs(R))\n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Correlation with Library Size (abs)')\n",
    "plt.xlabel('Principle Component')\n",
    "plt.grid(True)\n",
    "sns.despine()\n",
    "\n",
    "# Plot ranked variance of each gene\n",
    "ax = plt.subplot(gs1[3])\n",
    "plt.plot(np.log(ranked_gene_variance.values))\n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Log. Variance')\n",
    "plt.xlabel('Ranked Genes')\n",
    "plt.grid(True)\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# IDENTIFY POINT OF MAXIMUM CURVATURE IN CUMULATIVE EXPLAINED VARIANCE\n",
    "y = cumsum_explained_variance\n",
    "x = np.arange(len(y))\n",
    "kneedle = KneeLocator(x, y)\n",
    "kneedle.plot_knee_normalized()\n",
    "knee = kneedle.knee\n",
    "plt.title('knee: {}'.format(knee))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SELECT DATA ASSOCIATED WITH RELEVANT PCS (REMOVE PC0 if highly correlated with cell size)\n",
    "ncomponents = knee*2\n",
    "COMPONENTS = np.arange(ncomponents)\n",
    "print(COMPONENTS)\n",
    "\n",
    "# LOADING PER COMPONENT\n",
    "loadings = PCA_FIT.components_.T[:,COMPONENTS]\n",
    "\n",
    "labels = []\n",
    "for i in COMPONENTS:\n",
    "    labels.append('PC{}'.format(i))\n",
    "\n",
    "data = NDF_HARBOUR_PRIMARY[goi]\n",
    "data -= np.min(np.ravel(data))\n",
    "data /= np.max(np.ravel(data))\n",
    "data = pd.DataFrame(np.dot(data, loadings).astype(np.float64),index=data.index, columns = labels)\n",
    "\n",
    "\n",
    "# WRITE TO H5\n",
    "DIMENSIONS_HARBOUR_TUMOR_PRIMARY = data.copy()\n",
    "exec('h5_data_harbour.save(DIMENSIONS_HARBOUR_TUMOR_PRIMARY, \\'DIMENSIONS_HARBOUR_TUMOR_PRIMARY\\')')# update directory\n",
    "h5_data_harbour.ls() # list contents of directory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### REGRESS LIBRARY SIZE OUT OF PCS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exec('QUERY = DIMENSIONS_HARBOUR_TUMOR_PRIMARY')\n",
    "idx = QUERY.columns.to_series().str.contains('^PC((?!TSNE).)*$') # select from DIMS_NDF\n",
    "PCA_DATA = QUERY.loc[:, idx.values]\n",
    "counts = DF_HARBOUR_PRIMARY.copy()\n",
    "\n",
    "num_comps = PCA_DATA.shape[1]\n",
    "\n",
    "pca_regressed_out_lib_size = deepcopy(PCA_DATA)\n",
    "for c in range(num_comps):\n",
    "    lm = LinearRegression(normalize=False)\n",
    "    X = counts.sum(axis=1).values.reshape(counts.shape[0], 1)\n",
    "    Y = PCA_DATA.iloc[:, c]\n",
    "    lm.fit(X, Y)\n",
    "    pca_regressed_out_lib_size.iloc[:,c] = Y-lm.predict(X)\n",
    "    \n",
    "# Add SUFIX FOR REGRESSED LIB SIZE (RLS)\n",
    "pca_regressed_out_lib_size = pca_regressed_out_lib_size.add_prefix('RLS_')\n",
    "\n",
    "for label in pca_regressed_out_lib_size.columns:\n",
    "    exec('DIMENSIONS_HARBOUR_TUMOR_PRIMARY[label]= pca_regressed_out_lib_size[label]')\n",
    "    \n",
    "# WRITE TO H5\n",
    "exec('h5_data_harbour.save(DIMENSIONS_HARBOUR_TUMOR_PRIMARY, \\'DIMENSIONS_HARBOUR_TUMOR_PRIMARY\\')')# update directory\n",
    "h5_data_harbour.ls() # list contents of directory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### CALCULATE DIFFUSION COEFFICIENTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "knn = 27 # truncate the neighbors to the k closest points\n",
    "\n",
    "component_prefix = 'RLS_PC'\n",
    "exec('QUERY = DIMENSIONS_HARBOUR_TUMOR_PRIMARY')\n",
    "idx = QUERY.columns.to_series().str.contains('^{}((?!TSNE).)*$'.format(component_prefix)) # select from DIMS_NDF\n",
    "PCA_DATA = QUERY.loc[:, idx.values]\n",
    "\n",
    "DIFFUSION_EIGS = palantir.utils.run_diffusion_maps(PCA_DATA.astype(np.float64), knn=knn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### IMPUTE MISSING VALUES "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "imputed_data_matrix = palantir.utils.run_magic_imputation(NDF_HARBOUR_PRIMARY, DIFFUSION_EIGS, n_steps=3)\n",
    "INDF_HARBOUR_PRIMARY = imputed_data_matrix.copy()\n",
    "exec('h5_data_harbour.save(INDF_HARBOUR_PRIMARY,\\'INDF_HARBOUR_PRIMARY\\')')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MIXTURE MODEL - HARBOUR DATASET"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### MIXTURE MODEL FOR TCGA CLASSIFICATION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using imputed normalized Harbour data:\n",
    "datatype = 'INDF_HARBOUR'\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "\n",
    "# Using tumor cell subset:\n",
    "subset_type = 'TUMOR'\n",
    "\n",
    "gene1 = 'TCGA1'\n",
    "signature_genes = TCGA_markerfile_df.loc[TCGA_markerfile_df[gene1] == 1].index.values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "vals1 = (np.nanmean(QUERY[detected_genes],axis=1))\n",
    "\n",
    "gene2 = 'TCGA2'\n",
    "signature_genes = TCGA_markerfile_df.loc[TCGA_markerfile_df[gene2] == 1].index.values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "vals2 = (np.nanmean(QUERY[detected_genes],axis=1))\n",
    "\n",
    "gene3 = 'TCGA3'\n",
    "signature_genes = TCGA_markerfile_df.loc[TCGA_markerfile_df[gene3] == 1].index.values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "vals3 = (np.nanmean(QUERY[detected_genes],axis=1))\n",
    "\n",
    "gene4 = 'TCGA4'\n",
    "signature_genes = TCGA_markerfile_df.loc[TCGA_markerfile_df[gene4] == 1].index.values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "vals4 = (np.nanmean(QUERY[detected_genes],axis=1))\n",
    "\n",
    "X = pd.DataFrame(data = {gene1: vals1, gene2: vals2, gene3: vals3, gene4: vals4}, index = QUERY.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find optimal random seed for knn initialization step of the BGMM:\n",
    "rs_opt = 0\n",
    "while rs in range(1,50) and rs_opt == 0:\n",
    "    X_test = X.copy()\n",
    "    bgmm = BayesianGaussianMixture(covariance_type='tied', n_components=4, random_state=rs)\n",
    "    bgmm.fit(X_test[['TCGA1','TCGA2','TCGA3','TCGA4']])\n",
    "    clusters = bgmm.predict(X_test[['TCGA1','TCGA2','TCGA3','TCGA4']])\n",
    "    \n",
    "    X_test['Assignment'] = clusters\n",
    "    \n",
    "    # Make Class Assignments\n",
    "    lut ={}\n",
    "    mean_array = []\n",
    "    for ii in np.unique(X_test['Assignment']):\n",
    "        mean_array.append([X_test[X_test['Assignment']==ii][x].mean() for x in TCGA_CLASS_LIST])\n",
    "    mean_array = np.array(mean_array)\n",
    "\n",
    "    for i in range(len(np.unique(X_test['Assignment']))):\n",
    "        max_ind = np.unravel_index(np.argmax(mean_array, axis=None), mean_array.shape)\n",
    "        lut[max_ind[0]] = TCGA_CLASS_LIST[max_ind[1]]\n",
    "        mean_array[:,max_ind[1]] = 0\n",
    "        mean_array[max_ind[0],:] = 0\n",
    "\n",
    "    # MAP ASSIGNMENTS\n",
    "    X_test['Assignment'] = X_test['Assignment'].map(lut)\n",
    "    \n",
    "    if (X_test.Assignment.value_counts().keys()[0] in ['TCGA3', 'TCGA4']):\n",
    "        rs_opt = rs\n",
    "        X_test = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Train BGMM:\n",
    "# NOTE: Occasionally the bgmm results can vary here, changing the random_state value will help (recommend using: 33)\n",
    "\n",
    "bgmm = BayesianGaussianMixture(covariance_type='tied', n_components=4, random_state=rs_opt)\n",
    "bgmm.fit(X[['TCGA1','TCGA2','TCGA3','TCGA4']])\n",
    "clusters = bgmm.predict(X[['TCGA1','TCGA2','TCGA3','TCGA4']])\n",
    "\n",
    "# Create T-SNE:\n",
    "X_embedded = manifold.TSNE(n_components=2, random_state=2).fit_transform(X[['TCGA1','TCGA2','TCGA3','TCGA4']])\n",
    "X['tsne1'] = X_embedded[:,0]\n",
    "X['tsne2'] = X_embedded[:,1]\n",
    "\n",
    "X['Assignment'] = clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make Class Assignments\n",
    "lut ={}\n",
    "mean_array = []\n",
    "for ii in np.unique(X['Assignment']):\n",
    "    mean_array.append([X[X['Assignment']==ii][x].mean() for x in TCGA_CLASS_LIST])\n",
    "mean_array = np.array(mean_array)\n",
    "\n",
    "for i in range(len(np.unique(X['Assignment']))):\n",
    "    max_ind = np.unravel_index(np.argmax(mean_array, axis=None), mean_array.shape)\n",
    "    lut[max_ind[0]] = TCGA_CLASS_LIST[max_ind[1]]\n",
    "    mean_array[:,max_ind[1]] = 0\n",
    "    mean_array[max_ind[0],:] = 0\n",
    "\n",
    "# MAP ASSIGNMENTS\n",
    "X['Assignment'] = X['Assignment'].map(lut)\n",
    "\n",
    "sns.set(font_scale=2.5)\n",
    "sns.set_style(\"white\")\n",
    "g = sns.pairplot(X[['tsne1','tsne2','Assignment']], \n",
    "                 hue = 'Assignment',\n",
    "                 palette={\"TCGA1\": \"#0000FF\",\n",
    "                          \"TCGA2\": \"#A4D3FC\",\n",
    "                          \"TCGA3\": \"#FF7DC2\", \n",
    "                          \"TCGA4\": \"#FF0000\",},\n",
    "            diag_kind = 'kde', aspect = 1, size = 10);\n",
    "\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'HARBOUR_BayesianGMM_TCGA'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "\n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "g.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "g.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# UPDATE DIRECTORY\n",
    "X['TCGA_Assignment'] = X.Assignment.values\n",
    "METADATA_HARBOUR = X[['TCGA_Assignment']].copy()\n",
    "exec('h5_data_harbour.save(METADATA_HARBOUR, \\'METADATA_HARBOUR\\')')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### INTRATUMOR TCGA HETEROGENEITY"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### TCGA DISTRIBUTION PER PATIENT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "boxplot_patient_list = []\n",
    "boxplot_fract_list = []\n",
    "boxplot_class_list = []\n",
    "\n",
    "for patient in METADATA_HARBOUR.index.get_level_values('Patient').unique():\n",
    "    tmp = [METADATA_HARBOUR.loc[METADATA_HARBOUR.index.get_level_values('Patient') == patient].\\\n",
    "           TCGA_Assignment.value_counts().reindex(METADATA_HARBOUR.TCGA_Assignment.unique(), fill_value=0)[x] \n",
    "           for x in TCGA_CLASS_LIST]\n",
    "    tmp = list(tmp/sum(tmp))\n",
    "    \n",
    "    boxplot_patient_list = boxplot_patient_list + [patient for x in range(len(TCGA_CLASS_LIST))]\n",
    "    boxplot_fract_list = boxplot_fract_list + tmp\n",
    "    boxplot_class_list = boxplot_class_list + TCGA_CLASS_LIST\n",
    "\n",
    "boxplot_data = pd.DataFrame()\n",
    "boxplot_data['Patient'] = boxplot_patient_list\n",
    "boxplot_data['Fraction'] = boxplot_fract_list\n",
    "boxplot_data['Class'] = boxplot_class_list\n",
    "\n",
    "# VIOLIN PLOT\n",
    "fig = plt.figure(figsize = (10,4))\n",
    "ax = plt.gca()\n",
    "\n",
    "# Plot violin plot\n",
    "g = sns.barplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class', ci=95, capsize=.2,\n",
    "                palette={\"TCGA1\": \"#0000FF\",\n",
    "                         \"TCGA2\": \"#A4D3FC\",\n",
    "                         \"TCGA3\": \"#FF7DC2\", \n",
    "                         \"TCGA4\": \"#FF0000\",},\n",
    "                linewidth = 1, ax = ax);\n",
    "fig.get_axes()[0].set_yscale('log')\n",
    "\n",
    "plt.ylim((0,1))\n",
    "g.set_ylabel(\"Cell Type Fraction\",fontsize=10)\n",
    "g.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "g.tick_params(labelsize=10)\n",
    "g.set_xticklabels(ax.get_xticklabels(),rotation=90)\n",
    "sns.despine()\n",
    "\n",
    "ax.legend_.remove()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'TCGA_Distribution_by_Patient_LOG_HARBOUR'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### MIXTURE MODEL FOR GEP CLASSIFICATION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_type = 'TUMOR'\n",
    "# LOAD GENE LIST FROM EXCEL\n",
    "path_to_genesets = DATA_PATH+'uveal_melanoma_MB_v3.csv'\n",
    "genesets = pd.read_csv(path_to_genesets,header='infer')\n",
    "genesets = genesets.apply(lambda x: x.astype(str).str.upper())\n",
    "print(shape(genesets)[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# VIOLIN PLOTS\n",
    "datatype = 'INDF_HARBOUR'\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "\n",
    "# VIOLIN PLOTS\n",
    "gene1 = 'Castle 1'\n",
    "signature_genes = genesets[gene1].values\n",
    "signature_genes = [x for x in signature_genes if str(x).upper() != 'NAN']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "vals1 = (np.nanmean(QUERY[detected_genes],axis=1))\n",
    "\n",
    "gene2 = 'Castle 2'\n",
    "signature_genes = genesets[gene2].values\n",
    "signature_genes = [x for x in signature_genes if str(x).upper() != 'NAN']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "vals2 = (np.nanmean(QUERY[detected_genes],axis=1))\n",
    "\n",
    "X = pd.DataFrame(data = {gene1: vals1, gene2: vals2}, index = QUERY.index)\n",
    "\n",
    "bgmm = BayesianGaussianMixture(covariance_type='diag', n_components=2, random_state=1)\n",
    "bgmm.fit(X[['Castle 1','Castle 2']])\n",
    "clusters = bgmm.predict(X[['Castle 1','Castle 2']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# VIOLIN PLOTS\n",
    "datatype = 'INDF_HARBOUR'\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "\n",
    "# VIOLIN PLOTS\n",
    "gene1 = 'Castle 1'\n",
    "signature_genes = genesets[gene1].values\n",
    "signature_genes = [x for x in signature_genes if str(x).upper() != 'NAN']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "vals1 = (np.nanmean(QUERY[detected_genes],axis=1))\n",
    "\n",
    "gene2 = 'Castle 2'\n",
    "signature_genes = genesets[gene2].values\n",
    "signature_genes = [x for x in signature_genes if str(x).upper() != 'NAN']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "vals2 = (np.nanmean(QUERY[detected_genes],axis=1))\n",
    "\n",
    "X = pd.DataFrame(data = {gene1: vals1, gene2: vals2}, index = QUERY.index)\n",
    "\n",
    "bgmm = BayesianGaussianMixture(covariance_type='diag', n_components=2)\n",
    "bgmm.fit(X[['Castle 1','Castle 2']])\n",
    "clusters = bgmm.predict(X[['Castle 1','Castle 2']])\n",
    "X['Assignment'] = clusters\n",
    "\n",
    "# Make Class Assignments\n",
    "lut ={}\n",
    "for ii in np.unique(X['Assignment']):\n",
    "    if X[X['Assignment']==ii]['Castle 1'].mean()>X[X['Assignment']==ii]['Castle 2'].mean():\n",
    "        lut[ii] = 'Castle 1'\n",
    "    else:\n",
    "        lut[ii] = 'Castle 2'\n",
    "    \n",
    "X['Assignment'] = X['Assignment'].map(lut)\n",
    "\n",
    "sns.set(font_scale=2.5)\n",
    "sns.set_style(\"white\")\n",
    "g = sns.pairplot(X[['Castle 1','Castle 2','Assignment']], \n",
    "                 hue = 'Assignment',\n",
    "                 palette={\"Castle 1\": \"#0000FF\", \"Castle 2\": \"#FF0000\", 'Mixed': '#000000'},\n",
    "                 diag_kind = 'kde', aspect = 1, size = 10); # SAVE FIGURE\n",
    "\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'HARBOUR_BayesianGMM'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "\n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "g.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "g.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# UPDATE DIRECTORY\n",
    "METADATA_HARBOUR['Assignment'] = X.Assignment\n",
    "METADATA_HARBOUR.to_hdf(DATA_PATH_HARBOUR+'UVMEL_HARBOUR.h5', key='METADATA_HARBOUR')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### COLOR GEP PLOT BY TCGA CLASS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X['TCGA_Class'] = METADATA_HARBOUR.TCGA_Assignment\n",
    "\n",
    "sns.set(font_scale=2.5)\n",
    "sns.set_style(\"white\")\n",
    "g = sns.pairplot(X[['Castle 1', 'Castle 2', 'Assignment', 'TCGA_Class']], \n",
    "                 hue = 'TCGA_Class',\n",
    "                 palette={\"TCGA1\": \"#0000FF\",\n",
    "                          \"TCGA2\": \"#A4D3FC\",\n",
    "                          \"TCGA3\": \"#FF7DC2\", \n",
    "                          \"TCGA4\": \"#FF0000\",},\n",
    "                 diag_kind = 'kde', aspect = 1, size = 10);# SAVE FIGURE\n",
    "\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'HARBOUR_FULL_TCGA_colored'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "\n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "g.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "g.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make Class Assignments\n",
    "X['cellType'] = X.index.get_level_values('cellType').values\n",
    "\n",
    "sns.set(font_scale=2.5)\n",
    "sns.set_style(\"white\")\n",
    "g = sns.pairplot(X[['Castle 1', 'Castle 2', 'Assignment', 'cellType']], \n",
    "                 hue = 'cellType',\n",
    "                 palette={\"Class 1A Primary Tumor Cells\": \"#2F75DF\",\n",
    "                          \"Class 1B PRAME+ Met Tumor Cells\": \"#21C108\",\n",
    "                          \"Class 2 PRAME- Primary Tumor Cells\": \"#B696B9\", \n",
    "                          \"Class 2 PRAME+ Primary Tumor Cells\": \"#D17070\",},\n",
    "                 diag_kind = 'kde', aspect = 1, size = 10); # SAVE FIGURE\n",
    "\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'HABROUR_FULL_given_annotations'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "\n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "g.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "g.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### CONTINUE WITH INTRATUMOR HETEROGENEITY FOR GEP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize = (5,5))\n",
    "sns.despine()\n",
    "sns.set(rc={'axes.facecolor':'white', 'figure.facecolor':'white'})\n",
    "ax = plt.gca()\n",
    "\n",
    "ax = plt.scatter(X[X['Assignment']=='Castle 1']['Castle 2'], X[X['Assignment']=='Castle 1']['Castle 1'],\n",
    "                 s = 1, c ='blue',alpha = 0.5)\n",
    "ax = sns.kdeplot(X[X['Assignment']=='Castle 1']['Castle 2'], X[X['Assignment']=='Castle 1']['Castle 1'],\n",
    "                 cmap=\"Blues\", shade=True, shade_lowest=False,n_levels = 20, alpha = 0.5)\n",
    "\n",
    "ax = plt.scatter(X[X['Assignment']=='Castle 2']['Castle 2'], X[X['Assignment']=='Castle 2']['Castle 1'],\n",
    "                 s = 1, c ='red',alpha = 0.5)\n",
    "ax = sns.kdeplot(X[X['Assignment']=='Castle 2']['Castle 2'], X[X['Assignment']=='Castle 2']['Castle 1'],\n",
    "                 cmap=\"Reds\", shade=True, shade_lowest=False, n_levels = 20, alpha = 0.5)\n",
    "\n",
    "\n",
    "# SAVE FIGURE\n",
    "# SAVE FIGURE\n",
    "figure_label = 'Mixture_Model_HARBOUR'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "\n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update multi-index\n",
    "new_index = pd.MultiIndex.from_tuples(list(zip(X.index.get_level_values('Sample ID'), \n",
    "                                               X.index.get_level_values('Legend'),\n",
    "                                               X.index.get_level_values('Patient'),\n",
    "                                               X.index.get_level_values('Cell ID'),\n",
    "                                               X['Assignment'])), \n",
    "                                  names=['Sample ID','Legend', 'Patient','Cell ID','Assignment'])\n",
    "X = pd.DataFrame(data = X.values, columns = X.columns, index = new_index)\n",
    "\n",
    "# CLASS ASSIGNMENT BY PATIENT \n",
    "meta = 'Assignment'\n",
    "exec('tmp = X.groupby(level=[\\'Patient\\', \\'{}\\'], axis=0).size().unstack().fillna(0)'.format(meta))\n",
    "colors = [\"#0000FF\",\"#FF0000\"]\n",
    "\n",
    "plt.figure(figsize = (5,3))\n",
    "ax = plt.gca()\n",
    "tmp.div(tmp.sum(axis=1),axis=0).plot.barh(stacked=True, color=colors, ax = ax, width = 0.95)\n",
    "ax.legend_.remove()\n",
    "ax.set_frame_on(False)\n",
    "plt.xticks(fontsize=8)\n",
    "plt.yticks(fontsize=8)\n",
    "plt.xlabel('Cell Fraction', fontsize=12)\n",
    "plt.ylabel('Patient', fontsize=12)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'GEP_Distribution_by_Patient_HARBOUR'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label\n",
    "plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GENE-GENE COVARIANCE WITHIN NORMAL SUBSET\n",
    "n = 150\n",
    "reps = np.arange(100)\n",
    "\n",
    "master_groupby = 'Patient'\n",
    "ix = [ind for ind,name in enumerate(X.index.names) if name==master_groupby][0]\n",
    "fraction_castle1 = np.zeros((len(reps),len(np.unique(X.index.get_level_values(master_groupby)))))\n",
    "fraction_castle2 = np.zeros((len(reps),len(np.unique(X.index.get_level_values(master_groupby)))))\n",
    "\n",
    "for rep in reps:\n",
    "    # RANDOMLY SAMPLE SAME NUMBER OF CELLS PER PATIENT\n",
    "    SUBSET = pd.DataFrame()\n",
    "    for class_selection in np.unique(X.index.get_level_values(master_groupby)):\n",
    "        CSUBSET = X.loc[X.index.map(lambda x: x[ix] in [class_selection])].sample(n=n,replace=True)\n",
    "        SUBSET = SUBSET.append(CSUBSET)\n",
    "    \n",
    "    bgmm = BayesianGaussianMixture(covariance_type='diag', n_components=2, random_state=1)\n",
    "    bgmm.fit(SUBSET[['Castle 1','Castle 2']])\n",
    "    clusters = bgmm.predict(SUBSET[['Castle 1','Castle 2']])\n",
    "    SUBSET['Assignment'] = clusters\n",
    "\n",
    "    # Make Class Assignments\n",
    "    lut ={}\n",
    "    for ii in np.unique(SUBSET['Assignment']):\n",
    "        if SUBSET[SUBSET['Assignment']==ii]['Castle 1'].mean()>SUBSET[SUBSET['Assignment']==ii]['Castle 2'].mean():\n",
    "            lut[ii] = 'Castle 1'\n",
    "        else:\n",
    "            lut[ii] = 'Castle 2'\n",
    "    \n",
    "    SUBSET['Assignment'] = SUBSET['Assignment'].map(lut)\n",
    "\n",
    "    # Update multi-index\n",
    "    new_index = pd.MultiIndex.from_tuples(list(zip(SUBSET.index.get_level_values('Sample ID'), \n",
    "                                                   SUBSET.index.get_level_values('Legend'),\n",
    "                                                   SUBSET.index.get_level_values('Patient'),\n",
    "                                                   SUBSET.index.get_level_values('Cell ID'),\n",
    "                                                   SUBSET['Assignment'])), \n",
    "                                          names=['Sample ID','Legend', 'Patient','Cell ID',\n",
    "                                                 'Assignment'])\n",
    "    SUBSET = pd.DataFrame(data = SUBSET.values, columns = SUBSET.columns, index = new_index)\n",
    "\n",
    "    # CASTLE ASSIGNMENT PER PATIENT\n",
    "    meta = 'Assignment'\n",
    "    exec('tmp = SUBSET.groupby(level=[\\'Patient\\', \\'{}\\'], axis=0).size().unstack().fillna(0)'.format(meta))\n",
    "\n",
    "    # SAVE PER DOWNSAMPLE\n",
    "    cell_type_fraction = tmp.div(tmp.sum(axis=1),axis=0)\n",
    "    fraction_castle1[rep,:] = cell_type_fraction['Castle 1'].values\n",
    "    fraction_castle2[rep,:] = cell_type_fraction['Castle 2'].values\n",
    "    \n",
    "# CONVER TO PANDAS DATAFRAME\n",
    "boxplot_data_fraction_castle1 = pd.DataFrame(data = fraction_castle1, \n",
    "                                             columns = ['{}_{}'.format(master_groupby,ind) \n",
    "                                                        for ind in \n",
    "                                                        np.unique(X.index.get_level_values(master_groupby))])\n",
    "boxplot_data_fraction_castle1 = boxplot_data_fraction_castle1.stack().rename_axis(('Rep', 'Patient')).\\\n",
    "                                                                      reset_index(name='Fraction')\n",
    "\n",
    "boxplot_data_fraction_castle2 = pd.DataFrame(data = fraction_castle2, \n",
    "                                             columns = ['{}_{}'.format(master_groupby,ind) \n",
    "                                                        for ind in \n",
    "                                                        np.unique(X.index.get_level_values(master_groupby))])\n",
    "boxplot_data_fraction_castle2 = boxplot_data_fraction_castle2.stack().rename_axis(('Rep', 'Patient')).\\\n",
    "                                                                      reset_index(name='Fraction')\n",
    "\n",
    "boxplot_data_fraction_castle1['Class'] = ['Castle 1']*len(boxplot_data_fraction_castle1)\n",
    "boxplot_data_fraction_castle2['Class'] = ['Castle 2']*len(boxplot_data_fraction_castle2)\n",
    "boxplot_data = boxplot_data_fraction_castle1.append(boxplot_data_fraction_castle2)\n",
    "boxplot_data['Patient'] = [x.replace('Patient_','') for x in boxplot_data.Patient]\n",
    "\n",
    "# VIOLIN PLOT\n",
    "fig = plt.figure(figsize = (10,4))\n",
    "ax = plt.gca()\n",
    "\n",
    "# Plot violin plot\n",
    "g = sns.barplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class', ci=95, capsize=.2,\n",
    "                palette={\"Castle 1\": \"#0000FF\", \"Castle 2\": \"#FF0000\", 'Mixed': '#000000'},\n",
    "                linewidth = 1, ax = ax);\n",
    "fig.get_axes()[0].set_yscale('log')\n",
    "\n",
    "plt.ylim((0,1))\n",
    "g.set_ylabel(\"Cell Type Fraction\",fontsize=10)\n",
    "g.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "g.tick_params(labelsize=10)\n",
    "g.set_xticklabels(ax.get_xticklabels(),rotation=90)\n",
    "sns.despine()\n",
    "\n",
    "ax.legend_.remove()\n",
    "plt.tight_layout()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'GEP_Distribution_by_Patient_LOG_HARBOUR'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "boxplot_patient_list = []\n",
    "boxplot_fract_list = []\n",
    "boxplot_class_list = []\n",
    "\n",
    "for patient in sort(METADATA_HARBOUR.index.get_level_values('Patient').unique().tolist()):\n",
    "    tmp = [METADATA_HARBOUR.loc[METADATA_HARBOUR.index.get_level_values('Patient') == patient].\\\n",
    "           Assignment.value_counts().reindex(METADATA_HARBOUR.Assignment.unique(), fill_value=0)[x] \n",
    "           for x in GEP_CLASS_LIST]\n",
    "    tmp = list(tmp/sum(tmp))\n",
    "    \n",
    "    boxplot_patient_list = boxplot_patient_list + [patient for x in range(len(GEP_CLASS_LIST))]\n",
    "    boxplot_fract_list = boxplot_fract_list + tmp\n",
    "    boxplot_class_list = boxplot_class_list + GEP_CLASS_LIST\n",
    "\n",
    "boxplot_data = pd.DataFrame()\n",
    "boxplot_data['Patient'] = boxplot_patient_list\n",
    "boxplot_data['Fraction'] = boxplot_fract_list\n",
    "boxplot_data['Class'] = boxplot_class_list\n",
    "\n",
    "# VIOLIN PLOT\n",
    "fig = plt.figure(figsize = (10,4))\n",
    "ax = plt.gca()\n",
    "\n",
    "# Plot violin plot\n",
    "g = sns.barplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class', ci=95, capsize=.2,\n",
    "                palette={\"Castle 1\": \"#0000FF\", \"Castle 2\": \"#FF0000\", 'Mixed': '#000000'},\n",
    "                linewidth = 1, ax = ax);\n",
    "fig.get_axes()[0].set_yscale('log')\n",
    "\n",
    "plt.ylim((0,1))\n",
    "g.set_ylabel(\"Cell Type Fraction\",fontsize=10)\n",
    "g.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "g.tick_params(labelsize=10)\n",
    "g.set_xticklabels(ax.get_xticklabels(),rotation=90)\n",
    "sns.despine()\n",
    "\n",
    "ax.legend_.remove()\n",
    "plt.tight_layout()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'GEP_Distribution_by_Patient_LOG_HARBOUR_NO_ds'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MIXTURE MODEL - HARBOUR DATA PRIMARY ONLY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "METADATA_TUMOR.drop(columns='Patient').groupby('Patient').Assignment.value_counts(normalize=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### MIXTURE MODEL FOR TCGA CLASSIFICATION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exec('INDF_HARBOUR_PRIMARY = h5_data_harbour.load(\\'/INDF_HARBOUR_PRIMARY\\')')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# VIOLIN PLOTS\n",
    "datatype = 'INDF_HARBOUR_PRIMARY'\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "\n",
    "# VIOLIN PLOTS\n",
    "subset_type = 'TUMOR'\n",
    "\n",
    "gene1 = 'TCGA1'\n",
    "signature_genes = TCGA_markerfile_df.loc[TCGA_markerfile_df[gene1] == 1].index.values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "vals1 = (np.nanmean(QUERY[detected_genes],axis=1))\n",
    "\n",
    "gene2 = 'TCGA2'\n",
    "signature_genes = TCGA_markerfile_df.loc[TCGA_markerfile_df[gene2] == 1].index.values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "vals2 = (np.nanmean(QUERY[detected_genes],axis=1))\n",
    "\n",
    "gene3 = 'TCGA3'\n",
    "signature_genes = TCGA_markerfile_df.loc[TCGA_markerfile_df[gene3] == 1].index.values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "vals3 = (np.nanmean(QUERY[detected_genes],axis=1))\n",
    "\n",
    "gene4 = 'TCGA4'\n",
    "signature_genes = TCGA_markerfile_df.loc[TCGA_markerfile_df[gene4] == 1].index.values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "vals4 = (np.nanmean(QUERY[detected_genes],axis=1))\n",
    "\n",
    "X = pd.DataFrame(data = {gene1: vals1, gene2: vals2, gene3: vals3, gene4: vals4}, index = QUERY.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Train BGMM Model:\n",
    "bgmm = BayesianGaussianMixture(covariance_type='spherical', n_components=4, random_state=37)\n",
    "bgmm.fit(X[['TCGA1','TCGA2','TCGA3','TCGA4']])\n",
    "clusters = bgmm.predict(X[['TCGA1','TCGA2','TCGA3','TCGA4']])\n",
    "\n",
    "# Create T-SNE:\n",
    "X_embedded = manifold.TSNE(n_components=2, random_state=2).fit_transform(X[['TCGA1','TCGA2','TCGA3','TCGA4']])\n",
    "X['tsne1'] = X_embedded[:,0]\n",
    "X['tsne2'] = X_embedded[:,1]\n",
    "\n",
    "X['Assignment'] = clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make Class Assignments\n",
    "lut ={}\n",
    "mean_array = []\n",
    "for ii in np.unique(X['Assignment']):\n",
    "    mean_array.append([X[X['Assignment']==ii][x].mean() for x in TCGA_CLASS_LIST])\n",
    "mean_array = np.array(mean_array)\n",
    "\n",
    "for i in range(len(np.unique(X['Assignment']))):\n",
    "    max_ind = np.unravel_index(np.argmax(mean_array, axis=None), mean_array.shape)\n",
    "    lut[max_ind[0]] = TCGA_CLASS_LIST[max_ind[1]]\n",
    "    mean_array[:,max_ind[1]] = 0\n",
    "    mean_array[max_ind[0],:] = 0\n",
    "\n",
    "# MAP ASSIGNMENTS\n",
    "X['Assignment'] = X['Assignment'].map(lut)\n",
    "\n",
    "sns.set(font_scale=2.5)\n",
    "sns.set_style(\"white\")\n",
    "g = sns.pairplot(X[['tsne1','tsne2','Assignment']], \n",
    "                 hue = 'Assignment',\n",
    "                 palette={\"TCGA1\": \"#0000FF\",\n",
    "                          \"TCGA2\": \"#A4D3FC\",\n",
    "                          \"TCGA3\": \"#FF7DC2\", \n",
    "                          \"TCGA4\": \"#FF0000\",},\n",
    "            diag_kind = 'kde', aspect = 1, size = 10);# SAVE FIGURE\n",
    "\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'HARBOUR_PRIMARY_BayesianGMM_TCGA'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "\n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "g.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "g.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_test = X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# UPDATE DIRECTORY\n",
    "X['TCGA_Assignment'] = X.Assignment\n",
    "#METADATA_HARBOUR_PRIMARY = X[['TCGA_Assignment']].copy()\n",
    "#exec('h5_data_harbour.save(METADATA_HARBOUR_PRIMARY, \\'METADATA_HARBOUR_PRIMARY\\')')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_norm = X[TCGA_CLASS_LIST].copy()\n",
    "X_norm = X_norm.div(X_norm.sum(axis=1), axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### MIXTURE MODEL FOR GEP CLASSIFICATION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_type = 'TUMOR'\n",
    "# LOAD GENE LIST FROM EXCEL\n",
    "path_to_genesets = DATA_PATH+'uveal_melanoma_MB_v3.csv'\n",
    "genesets = pd.read_csv(path_to_genesets,header='infer')\n",
    "genesets = genesets.apply(lambda x: x.astype(str).str.upper())\n",
    "print(shape(genesets)[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# VIOLIN PLOTS\n",
    "datatype = 'INDF_HARBOUR_PRIMARY'\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "\n",
    "# VIOLIN PLOTS\n",
    "gene1 = 'Castle 1'\n",
    "signature_genes = genesets[gene1].values\n",
    "signature_genes = [x for x in signature_genes if str(x).upper() != 'NAN']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "vals1 = (np.nanmean(QUERY[detected_genes],axis=1))\n",
    "\n",
    "gene2 = 'Castle 2'\n",
    "signature_genes = genesets[gene2].values\n",
    "signature_genes = [x for x in signature_genes if str(x).upper() != 'NAN']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "vals2 = (np.nanmean(QUERY[detected_genes],axis=1))\n",
    "\n",
    "X = pd.DataFrame(data = {gene1: vals1, gene2: vals2}, index = QUERY.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# VIOLIN PLOTS\n",
    "datatype = 'INDF_HARBOUR_PRIMARY'\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "\n",
    "# VIOLIN PLOTS\n",
    "gene1 = 'Castle 1'\n",
    "signature_genes = genesets[gene1].values\n",
    "signature_genes = [x for x in signature_genes if str(x).upper() != 'NAN']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "vals1 = (np.nanmean(QUERY[detected_genes],axis=1))\n",
    "\n",
    "gene2 = 'Castle 2'\n",
    "signature_genes = genesets[gene2].values\n",
    "signature_genes = [x for x in signature_genes if str(x).upper() != 'NAN']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "vals2 = (np.nanmean(QUERY[detected_genes],axis=1))\n",
    "\n",
    "X = pd.DataFrame(data = {gene1: vals1, gene2: vals2}, index = QUERY.index)\n",
    "\n",
    "bgmm = BayesianGaussianMixture(covariance_type='diag', n_components=2, random_state=1)\n",
    "bgmm.fit(X[['Castle 1','Castle 2']])\n",
    "clusters = bgmm.predict(X[['Castle 1','Castle 2']])\n",
    "X['Assignment'] = clusters\n",
    "\n",
    "# Make Class Assignments\n",
    "lut ={}\n",
    "for ii in np.unique(X['Assignment']):\n",
    "    if X[X['Assignment']==ii]['Castle 1'].mean()>X[X['Assignment']==ii]['Castle 2'].mean():\n",
    "        lut[ii] = 'Castle 1'\n",
    "    else:\n",
    "        lut[ii] = 'Castle 2'\n",
    "    \n",
    "X['Assignment'] = X['Assignment'].map(lut)\n",
    "\n",
    "sns.set(font_scale=2.5)\n",
    "sns.set_style(\"white\")\n",
    "g = sns.pairplot(X[['Castle 1','Castle 2','Assignment']], \n",
    "                 hue = 'Assignment',\n",
    "                 palette={\"Castle 1\": \"#0000FF\", \"Castle 2\": \"#FF0000\", 'Mixed': '#000000'},\n",
    "                 diag_kind = 'kde', aspect = 1, size = 10); # SAVE FIGURE\n",
    "\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'HARBOUR_PRIMARY_BayesianGMM'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "\n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "#g.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "#g.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# UPDATE DIRECTORY\n",
    "METADATA_HARBOUR_PRIMARY['Assignment'] = X.Assignment\n",
    "#exec('h5_data_harbour.save(METADATA_HARBOUR_PRIMARY, \\'METADATA_HARBOUR_PRIMARY\\')')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### COLOR GEP PLOT BY TCGA CLASS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X['TCGA_Class'] = METADATA_HARBOUR_PRIMARY.TCGA_Assignment\n",
    "\n",
    "sns.set(font_scale=2.5)\n",
    "sns.set_style(\"white\")\n",
    "g = sns.pairplot(X[['Castle 1','Castle 2','Assignment', 'TCGA_Class']], \n",
    "                 hue = 'TCGA_Class',\n",
    "                 palette={\"TCGA1\": \"#0000FF\",\n",
    "                          \"TCGA2\": \"#A4D3FC\",\n",
    "                          \"TCGA3\": \"#FF7DC2\", \n",
    "                          \"TCGA4\": \"#FF0000\",},\n",
    "            diag_kind = 'kde', aspect = 1, size = 10); # SAVE FIGURE\n",
    "\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'HARBOUR_PRIMARY_FULL_TCGA_colored'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "\n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "g.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "g.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X['TCGA_Class'] = METADATA_HARBOUR_PRIMARY.TCGA_Assignment\n",
    "\n",
    "sns.set(font_scale=2.5)\n",
    "sns.set_style(\"white\")\n",
    "g = sns.pairplot(X[['Castle 1','Castle 2','TCGA_Class']], \n",
    "                 hue = 'TCGA_Class',\n",
    "                 palette={\"TCGA1\": \"#0000FF\",\n",
    "                          \"TCGA2\": \"#A4D3FC\",\n",
    "                          \"TCGA3\": \"#FF7DC2\", \n",
    "                          \"TCGA4\": \"#FF0000\",},\n",
    "            diag_kind = 'kde', aspect = 1, size = 10); # SAVE FIGURE\n",
    "\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'HARBOUR_PRIMARY_FULL_TCGA_colored'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "\n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "#g.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "#g.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n = 2000\n",
    "reps = np.arange(10)\n",
    "\n",
    "master_groupby = 'Patient'\n",
    "ix = [ind for ind,name in enumerate(X.index.names) if name==master_groupby][0]\n",
    "fraction_TCGA1 = np.zeros((len(reps),len(np.unique(X.index.get_level_values(master_groupby)))))\n",
    "fraction_TCGA2 = np.zeros((len(reps),len(np.unique(X.index.get_level_values(master_groupby)))))\n",
    "fraction_TCGA3 = np.zeros((len(reps),len(np.unique(X.index.get_level_values(master_groupby)))))\n",
    "fraction_TCGA4 = np.zeros((len(reps),len(np.unique(X.index.get_level_values(master_groupby)))))\n",
    "\n",
    "for rep in reps:\n",
    "    # RANDOMLY SAMPLE SAME NUMBER OF CELLS PER PATIENT\n",
    "    SUBSET = pd.DataFrame()\n",
    "    for class_selection in np.unique(X.index.get_level_values(master_groupby)):\n",
    "        CSUBSET = X.loc[X.index.map(lambda x: x[ix] in [class_selection])].sample(n=n,replace=True)\n",
    "        SUBSET = SUBSET.append(CSUBSET)\n",
    "    \n",
    "    bgmm = BayesianGaussianMixture(covariance_type='spherical', n_components=4, random_state=37)\n",
    "    bgmm.fit(SUBSET[TCGA_CLASS_LIST])\n",
    "    clusters = bgmm.predict(SUBSET[TCGA_CLASS_LIST])\n",
    "    SUBSET['Assignment'] = clusters\n",
    "\n",
    "    # Make Class Assignments\n",
    "    lut ={}\n",
    "    mean_array = []\n",
    "    for ii in np.unique(SUBSET['Assignment']):\n",
    "        mean_array.append([SUBSET[SUBSET['Assignment']==ii][x].mean() for x in TCGA_CLASS_LIST])\n",
    "    mean_array = np.array(mean_array)\n",
    "    \n",
    "    for i in range(len(np.unique(SUBSET['Assignment']))):\n",
    "        max_ind = np.unravel_index(np.argmax(mean_array, axis=None), mean_array.shape)\n",
    "        lut[max_ind[0]] = TCGA_CLASS_LIST[max_ind[1]]\n",
    "        mean_array[:,max_ind[1]] = 0\n",
    "        mean_array[max_ind[0],:] = 0\n",
    "    \n",
    "    # MAP ASSIGNMENTS\n",
    "    SUBSET['Assignment'] = SUBSET['Assignment'].map(lut)\n",
    "\n",
    "    # Update multi-index\n",
    "    new_index = pd.MultiIndex.from_tuples(list(zip(SUBSET.index.get_level_values('Sample ID'), \n",
    "                                                   SUBSET.index.get_level_values('Legend'),\n",
    "                                                   SUBSET.index.get_level_values('Patient'),\n",
    "                                                   SUBSET.index.get_level_values('Cell ID'),\n",
    "                                                   SUBSET['Assignment'])), \n",
    "                                          names=['Sample ID','Legend', 'Patient','Cell ID',\n",
    "                                                 'Assignment'])\n",
    "    SUBSET = pd.DataFrame(data = SUBSET.values, columns = SUBSET.columns, index = new_index)\n",
    "\n",
    "    # CASTLE ASSIGNMENT PER PATIENT\n",
    "    meta = 'Assignment'\n",
    "    exec('tmp = SUBSET.groupby(level=[\\'Patient\\', \\'{}\\'], axis=0).size().unstack().fillna(0)'.format(meta))\n",
    "\n",
    "    # SAVE PER DOWNSAMPLE\n",
    "    cell_type_fraction = tmp.div(tmp.sum(axis=1),axis=0)\n",
    "    fraction_TCGA1[rep,:] = cell_type_fraction['TCGA1'].values\n",
    "    fraction_TCGA2[rep,:] = cell_type_fraction['TCGA2'].values\n",
    "    fraction_TCGA3[rep,:] = cell_type_fraction['TCGA3'].values\n",
    "    fraction_TCGA4[rep,:] = cell_type_fraction['TCGA4'].values\n",
    "    \n",
    "    \n",
    "# CONVER TO PANDAS DATAFRAME\n",
    "boxplot_data_fraction_TCGA1 = pd.DataFrame(data = fraction_TCGA1, \n",
    "                                             columns = ['{}_{}'.format(master_groupby,ind) \n",
    "                                                        for ind in \n",
    "                                                        np.unique(X.index.get_level_values(master_groupby))])\n",
    "boxplot_data_fraction_TCGA1 = boxplot_data_fraction_TCGA1.stack().rename_axis(('Rep', 'Patient')).\\\n",
    "                                                                      reset_index(name='Fraction')\n",
    "\n",
    "boxplot_data_fraction_TCGA2 = pd.DataFrame(data = fraction_TCGA2, \n",
    "                                             columns = ['{}_{}'.format(master_groupby,ind) \n",
    "                                                        for ind in \n",
    "                                                        np.unique(X.index.get_level_values(master_groupby))])\n",
    "boxplot_data_fraction_TCGA2 = boxplot_data_fraction_TCGA2.stack().rename_axis(('Rep', 'Patient')).\\\n",
    "                                                                      reset_index(name='Fraction')\n",
    "\n",
    "boxplot_data_fraction_TCGA3 = pd.DataFrame(data = fraction_TCGA3, \n",
    "                                             columns = ['{}_{}'.format(master_groupby,ind) \n",
    "                                                        for ind in \n",
    "                                                        np.unique(X.index.get_level_values(master_groupby))])\n",
    "boxplot_data_fraction_TCGA3 = boxplot_data_fraction_TCGA3.stack().rename_axis(('Rep', 'Patient')).\\\n",
    "                                                                      reset_index(name='Fraction')\n",
    "\n",
    "boxplot_data_fraction_TCGA4 = pd.DataFrame(data = fraction_TCGA4, \n",
    "                                             columns = ['{}_{}'.format(master_groupby,ind) \n",
    "                                                        for ind in \n",
    "                                                        np.unique(X.index.get_level_values(master_groupby))])\n",
    "boxplot_data_fraction_TCGA4 = boxplot_data_fraction_TCGA4.stack().rename_axis(('Rep', 'Patient')).\\\n",
    "                                                                      reset_index(name='Fraction')\n",
    "\n",
    "boxplot_data_fraction_TCGA1['Class'] = ['TCGA1']*len(boxplot_data_fraction_TCGA1)\n",
    "boxplot_data_fraction_TCGA2['Class'] = ['TCGA2']*len(boxplot_data_fraction_TCGA2)\n",
    "boxplot_data_fraction_TCGA3['Class'] = ['TCGA3']*len(boxplot_data_fraction_TCGA3)\n",
    "boxplot_data_fraction_TCGA4['Class'] = ['TCGA4']*len(boxplot_data_fraction_TCGA4)\n",
    "\n",
    "boxplot_data = boxplot_data_fraction_TCGA1.append([boxplot_data_fraction_TCGA2,\n",
    "                                                   boxplot_data_fraction_TCGA3,\n",
    "                                                   boxplot_data_fraction_TCGA4])\n",
    "boxplot_data['Patient'] = [x.replace('Patient','MSK') for x in boxplot_data.Patient]\n",
    "\n",
    "#boxplot_data[boxplot_data.Fraction < 0.0001] = 0\n",
    "boxplot_data.Fraction = [x if x > 0.001 else 1e-4 for x in boxplot_data.Fraction]\n",
    "\n",
    "\n",
    "# VIOLIN PLOT\n",
    "fig = plt.figure(figsize = (10,4))\n",
    "ax = plt.gca()\n",
    "\n",
    "# Plot violin plot\n",
    "sns.barplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class', ci=95, capsize=.1,\n",
    "                palette={\"TCGA1\": \"#0000FF\",\n",
    "                         \"TCGA2\": \"#A4D3FC\",\n",
    "                         \"TCGA3\": \"#FF7DC2\", \n",
    "                         \"TCGA4\": \"#FF0000\",},\n",
    "                linewidth=1, ax = ax, \n",
    "                alpha=1)\n",
    "\n",
    "g = sns.stripplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class',\n",
    "                palette={\"TCGA1\": \"#0000FF\",\n",
    "                         \"TCGA2\": \"#A4D3FC\",\n",
    "                         \"TCGA3\": \"#FF7DC2\", \n",
    "                         \"TCGA4\": \"#FF0000\",},\n",
    "                linewidth = 0.5, edgecolor='white',\n",
    "                ax = ax, alpha=0.7,\n",
    "                dodge=True)\n",
    "\n",
    "fig.get_axes()[0].set_yscale('log')\n",
    "\n",
    "plt.ylim((1e-4,1.15))\n",
    "g.set_ylabel(\"Cell Type Fraction\",fontsize=10)\n",
    "g.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "g.tick_params(labelsize=10)\n",
    "g.set_xticklabels(ax.get_xticklabels(),rotation=90)\n",
    "sns.despine()\n",
    "\n",
    "ax.legend_.remove()\n",
    "#plt.tight_layout()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'TCGA_Distribution_by_Patient_LOG_HARBOUR_PRIMARY'\n",
    "fn = '/workdir/uvmel_project/figures/Revision_2_Figures/' + figure_label \n",
    "plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GENE-GENE COVARIANCE WITHIN NORMAL SUBSET\n",
    "n = 1000\n",
    "reps = np.arange(10)\n",
    "\n",
    "boxplot_patient_list = []\n",
    "boxplot_fract_list = []\n",
    "boxplot_class_list = []\n",
    "master_groupby = 'Patient'\n",
    "\n",
    "for rep in reps:\n",
    "    for patient in np.unique(METADATA_HARBOUR_PRIMARY.index.get_level_values('Patient')):\n",
    "        SUBSET = pd.DataFrame()\n",
    "        for class_selection in np.unique(METADATA_HARBOUR_PRIMARY.index.get_level_values(master_groupby)):\n",
    "            CSUBSET = METADATA_HARBOUR_PRIMARY.loc[METADATA_HARBOUR_PRIMARY.index.map(lambda x: x[ix] in \n",
    "                                                                  [class_selection])].sample(n=n,replace=True)\n",
    "            SUBSET = SUBSET.append(CSUBSET)\n",
    "        \n",
    "        tmp = [SUBSET.loc[SUBSET.index.get_level_values('Patient') == patient].\\\n",
    "               Assignment.value_counts().reindex(SUBSET.TCGA_Assignment.unique(), fill_value=0)[x] \n",
    "               for x in TCGA_CLASS_LIST]\n",
    "        tmp = list(tmp/sum(tmp))\n",
    "    \n",
    "        boxplot_patient_list = boxplot_patient_list + [patient for x in range(len(TCGA_CLASS_LIST))]\n",
    "        boxplot_fract_list = boxplot_fract_list + tmp\n",
    "        boxplot_class_list = boxplot_class_list + TCGA_CLASS_LIST\n",
    "\n",
    "boxplot_data = pd.DataFrame()\n",
    "boxplot_data['Patient'] = [x for x in boxplot_patient_list]\n",
    "boxplot_data['Fraction'] = boxplot_fract_list\n",
    "boxplot_data['Class'] = boxplot_class_list\n",
    "\n",
    "boxplot_data.Fraction = [x if x > 0.0001 else 1e-4 for x in boxplot_data.Fraction]\n",
    "\n",
    "# Set plot\n",
    "fig = plt.figure(figsize = (7,4))\n",
    "ax = plt.gca()\n",
    "\n",
    "# Plot bar plot\n",
    "sns.barplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class', ci=95, capsize=.15,\n",
    "                palette={\"TCGA1\": \"#0000FF\",\n",
    "                         \"TCGA2\": \"#A4D3FC\",\n",
    "                         \"TCGA3\": \"#FF7DC2\", \n",
    "                         \"TCGA4\": \"#FF0000\",},\n",
    "                linewidth=1, ax = ax, \n",
    "                alpha=1,\n",
    "                errcolor='black',\n",
    "                errwidth=1)\n",
    "\n",
    "g = sns.stripplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class',\n",
    "                palette={\"TCGA1\": \"#0000FF\",\n",
    "                         \"TCGA2\": \"#A4D3FC\",\n",
    "                         \"TCGA3\": \"#FF7DC2\", \n",
    "                         \"TCGA4\": \"#FF0000\",},\n",
    "                linewidth = 0.5, edgecolor='white',\n",
    "                ax = ax, alpha=0.7,\n",
    "                dodge=True)\n",
    "\n",
    "fig.get_axes()[0].set_yscale('log')\n",
    "\n",
    "plt.ylim((1e-4,1.15))\n",
    "ax.set_ylabel(\"Cell Type Fraction\",fontsize=10)\n",
    "ax.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "ax.tick_params(labelsize=10)\n",
    "ax.set_xticklabels(boxplot_data['Patient'].unique(),rotation=90)\n",
    "sns.despine()\n",
    "\n",
    "ax.legend_.remove()\n",
    "plt.tight_layout()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'TCGA_Distribution_by_Patient_LOG_HARBOUR_PRIMARY'\n",
    "fn = '/workdir/uvmel_project/figures/Revision_2_Figures/' + figure_label \n",
    "plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X.index.get_level_values('Patient').value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### COLOR BY CELLTYPE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make Class Assignments\n",
    "X['cellType'] = X.index.get_level_values('cellType').values\n",
    "\n",
    "sns.set(font_scale=2.5)\n",
    "sns.set_style(\"white\")\n",
    "g = sns.pairplot(X[['Castle 1', 'Castle 2', 'Assignment', 'cellType']], \n",
    "                 hue = 'cellType',\n",
    "                 palette={\"Class 1A Primary Tumor Cells\": \"#2F75DF\",\n",
    "                          \"Class 1B PRAME+ Met Tumor Cells\": \"#21C108\",\n",
    "                          \"Class 2 PRAME- Primary Tumor Cells\": \"#B696B9\", \n",
    "                          \"Class 2 PRAME+ Primary Tumor Cells\": \"#D17070\",},\n",
    "                 diag_kind = 'kde', aspect = 1, size = 10); # SAVE FIGURE\n",
    "\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'HABROUR_PRIMARY_FULL_given_annotations'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "\n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "g.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "g.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### CONTINUE WITH INTRATUMOR HETEROGENEITY WITH FOR GEP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize = (5,5))\n",
    "sns.despine()\n",
    "sns.set(rc={'axes.facecolor':'white', 'figure.facecolor':'white'})\n",
    "ax = plt.gca()\n",
    "\n",
    "ax = plt.scatter(X[X['Assignment']=='Castle 1']['Castle 2'], X[X['Assignment']=='Castle 1']['Castle 1'],\n",
    "                 s = 1, c ='blue',alpha = 0.5)\n",
    "ax = sns.kdeplot(X[X['Assignment']=='Castle 1']['Castle 2'], X[X['Assignment']=='Castle 1']['Castle 1'],\n",
    "                 cmap=\"Blues\", shade=True, shade_lowest=False,n_levels = 20, alpha = 0.5)\n",
    "\n",
    "ax = plt.scatter(X[X['Assignment']=='Castle 2']['Castle 2'], X[X['Assignment']=='Castle 2']['Castle 1'],\n",
    "                 s = 1, c ='red',alpha = 0.5)\n",
    "ax = sns.kdeplot(X[X['Assignment']=='Castle 2']['Castle 2'], X[X['Assignment']=='Castle 2']['Castle 1'],\n",
    "                 cmap=\"Reds\", shade=True, shade_lowest=False, n_levels = 20, alpha = 0.5)\n",
    "\n",
    "\n",
    "# SAVE FIGURE\n",
    "# SAVE FIGURE\n",
    "figure_label = 'HARBOUR_PRIMARY_Bayesian_GMM'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "\n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update multi-index\n",
    "new_index = pd.MultiIndex.from_tuples(list(zip(X.index.get_level_values('Sample ID'), \n",
    "                                               X.index.get_level_values('Legend'),\n",
    "                                               X.index.get_level_values('Patient'),\n",
    "                                               X.index.get_level_values('Cell ID'),\n",
    "                                               X['Assignment'])), \n",
    "                                  names=['Sample ID','Legend', 'Patient','Cell ID','Assignment'])\n",
    "X = pd.DataFrame(data = X.values, columns = X.columns, index = new_index)\n",
    "\n",
    "# CLASS ASSIGNMENT BY PATIENT \n",
    "meta = 'Assignment'\n",
    "exec('tmp = X.groupby(level=[\\'Patient\\', \\'{}\\'], axis=0).size().unstack().fillna(0)'.format(meta))\n",
    "colors = [\"#0000FF\",\"#FF0000\"]\n",
    "\n",
    "plt.figure(figsize = (5,3))\n",
    "ax = plt.gca()\n",
    "tmp.div(tmp.sum(axis=1),axis=0).plot.barh(stacked=True, color=colors, ax = ax, width = 0.95)\n",
    "ax.legend_.remove()\n",
    "ax.set_frame_on(False)\n",
    "plt.xticks(fontsize=8)\n",
    "plt.yticks(fontsize=8)\n",
    "plt.xlabel('Cell Fraction', fontsize=12)\n",
    "plt.ylabel('Patient', fontsize=12)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'GEP_Distribution_by_Patient_HARBOUR_PRIMARY'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label\n",
    "plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exec('METADATA_HARBOUR = h5_data_harbour.load(\\'/METADATA_HARBOUR\\')')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exec('METADATA_HARBOUR_PRIMARY = h5_data_harbour.load(\\'/METADATA_HARBOUR_PRIMARY\\')')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set Harbour primary GEP assignments to results from combined model:\n",
    "METADATA_HARBOUR_PRIMARY['Assignment'] = [METADATA_HARBOUR.loc[\n",
    "                                          METADATA_HARBOUR.index.get_level_values('Legend') == 'primary'].loc[\n",
    "                                          METADATA_HARBOUR.loc[\n",
    "                                          METADATA_HARBOUR.index.get_level_values('Legend') == 'primary'].\\\n",
    "                                          index.get_level_values('Cell ID') == x].Assignment.values[0] \n",
    "                                          for x in METADATA_HARBOUR_PRIMARY.index.get_level_values('Cell ID').values]\n",
    "\n",
    "# Set Harbour primary TCGA assignments to results from combined model:\n",
    "METADATA_HARBOUR_PRIMARY['TCGA_Assignment'] = [METADATA_HARBOUR.loc[\n",
    "                                          METADATA_HARBOUR.index.get_level_values('Legend') == 'primary'].loc[\n",
    "                                          METADATA_HARBOUR.loc[\n",
    "                                          METADATA_HARBOUR.index.get_level_values('Legend') == 'primary'].\\\n",
    "                                          index.get_level_values('Cell ID') == x].TCGA_Assignment.values[0] \n",
    "                                          for x in METADATA_HARBOUR_PRIMARY.index.get_level_values('Cell ID').values]"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Down sampled assignment distribution (not used in paper anymore) \n",
    "\n",
    "# GENE-GENE COVARIANCE WITHIN NORMAL SUBSET\n",
    "n = 2000\n",
    "reps = np.arange(10)\n",
    "\n",
    "master_groupby = 'Patient'\n",
    "ix = [ind for ind,name in enumerate(X.index.names) if name==master_groupby][0]\n",
    "fraction_castle1 = np.zeros((len(reps),len(np.unique(X.index.get_level_values(master_groupby)))))\n",
    "fraction_castle2 = np.zeros((len(reps),len(np.unique(X.index.get_level_values(master_groupby)))))\n",
    "\n",
    "for rep in reps:\n",
    "    # RANDOMLY SAMPLE SAME NUMBER OF CELLS PER PATIENT\n",
    "    SUBSET = pd.DataFrame()\n",
    "    for class_selection in np.unique(X.index.get_level_values(master_groupby)):\n",
    "        CSUBSET = X.loc[X.index.map(lambda x: x[ix] in [class_selection])].sample(n=n,replace=True)\n",
    "        SUBSET = SUBSET.append(CSUBSET)\n",
    "    \n",
    "    bgmm = BayesianGaussianMixture(covariance_type='diag', n_components=2, random_state=1)\n",
    "    bgmm.fit(SUBSET[['Castle 1','Castle 2']])\n",
    "    clusters = bgmm.predict(SUBSET[['Castle 1','Castle 2']])\n",
    "    SUBSET['Assignment'] = clusters\n",
    "\n",
    "    # Make Class Assignments\n",
    "    lut ={}\n",
    "    for ii in np.unique(SUBSET['Assignment']):\n",
    "        if SUBSET[SUBSET['Assignment']==ii]['Castle 1'].mean()>SUBSET[SUBSET['Assignment']==ii]['Castle 2'].mean():\n",
    "            lut[ii] = 'Castle 1'\n",
    "        else:\n",
    "            lut[ii] = 'Castle 2'\n",
    "    \n",
    "    SUBSET['Assignment'] = SUBSET['Assignment'].map(lut)\n",
    "\n",
    "    # Update multi-index\n",
    "    new_index = pd.MultiIndex.from_tuples(list(zip(SUBSET.index.get_level_values('Sample ID'), \n",
    "                                                   SUBSET.index.get_level_values('Legend'),\n",
    "                                                   SUBSET.index.get_level_values('Patient'),\n",
    "                                                   SUBSET.index.get_level_values('Cell ID'),\n",
    "                                                   SUBSET['Assignment'])), \n",
    "                                          names=['Sample ID','Legend', 'Patient','Cell ID',\n",
    "                                                 'Assignment'])\n",
    "    SUBSET = pd.DataFrame(data = SUBSET.values, columns = SUBSET.columns, index = new_index)\n",
    "\n",
    "    # CASTLE ASSIGNMENT PER PATIENT\n",
    "    meta = 'Assignment'\n",
    "    exec('tmp = SUBSET.groupby(level=[\\'Patient\\', \\'{}\\'], axis=0).size().unstack().fillna(0)'.format(meta))\n",
    "\n",
    "    # SAVE PER DOWNSAMPLE\n",
    "    cell_type_fraction = tmp.div(tmp.sum(axis=1),axis=0)\n",
    "    fraction_castle1[rep,:] = cell_type_fraction['Castle 1'].values\n",
    "    fraction_castle2[rep,:] = cell_type_fraction['Castle 2'].values\n",
    "    \n",
    "# CONVER TO PANDAS DATAFRAME\n",
    "boxplot_data_fraction_castle1 = pd.DataFrame(data = fraction_castle1, \n",
    "                                             columns = ['{}_{}'.format(master_groupby,ind) \n",
    "                                                        for ind in \n",
    "                                                        np.unique(X.index.get_level_values(master_groupby))])\n",
    "boxplot_data_fraction_castle1 = boxplot_data_fraction_castle1.stack().rename_axis(('Rep', 'Patient')).\\\n",
    "                                                                      reset_index(name='Fraction')\n",
    "\n",
    "boxplot_data_fraction_castle2 = pd.DataFrame(data = fraction_castle2, \n",
    "                                             columns = ['{}_{}'.format(master_groupby,ind) \n",
    "                                                        for ind in \n",
    "                                                        np.unique(X.index.get_level_values(master_groupby))])\n",
    "boxplot_data_fraction_castle2 = boxplot_data_fraction_castle2.stack().rename_axis(('Rep', 'Patient')).\\\n",
    "                                                                      reset_index(name='Fraction')\n",
    "\n",
    "boxplot_data_fraction_castle1['Class'] = ['Castle 1']*len(boxplot_data_fraction_castle1)\n",
    "boxplot_data_fraction_castle2['Class'] = ['Castle 2']*len(boxplot_data_fraction_castle2)\n",
    "boxplot_data = boxplot_data_fraction_castle1.append(boxplot_data_fraction_castle2)\n",
    "boxplot_data['Patient'] = [x.replace('Patient_','') for x in boxplot_data.Patient]\n",
    "\n",
    "#boxplot_data[boxplot_data.Fraction < 0.0001] = 0\n",
    "boxplot_data.Fraction = [x if x > 0.001 else 1e-4 for x in boxplot_data.Fraction]\n",
    "\n",
    "# VIOLIN PLOT\n",
    "fig = plt.figure(figsize = (10,4))\n",
    "ax = plt.gca()\n",
    "\n",
    "# Plot violin plot\n",
    "sns.barplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class', ci=95, capsize=.2,\n",
    "                palette={\"Castle 1\": \"#0000FF\", \"Castle 2\": \"#FF0000\", 'Mixed': '#000000'},\n",
    "                linewidth=1, ax = ax, alpha=1)\n",
    "\n",
    "g = sns.stripplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class',\n",
    "                palette={\"Castle 1\": \"#0000FF\", \"Castle 2\": \"#FF0000\", 'Mixed': '#000000'},\n",
    "                linewidth = 0.5, edgecolor='white',\n",
    "                ax = ax, alpha=0.7,\n",
    "                dodge=True)\n",
    "\n",
    "fig.get_axes()[0].set_yscale('log')\n",
    "\n",
    "plt.ylim((1e-4,1.15))\n",
    "g.set_ylabel(\"Cell Type Fraction\",fontsize=10)\n",
    "g.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "g.tick_params(labelsize=10)\n",
    "g.set_xticklabels(ax.get_xticklabels(),rotation=90)\n",
    "sns.despine()\n",
    "\n",
    "ax.legend_.remove()\n",
    "#plt.tight_layout()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'GEP_Distribution_by_Patient_LOG_HARBOUR_PRIMARY'\n",
    "fn = '/workdir/uvmel_project/figures/Revision_2_Figures/' + figure_label \n",
    "plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GENE-GENE COVARIANCE WITHIN NORMAL SUBSET\n",
    "n = 1000\n",
    "reps = np.arange(10)\n",
    "\n",
    "boxplot_patient_list = []\n",
    "boxplot_fract_list = []\n",
    "boxplot_class_list = []\n",
    "GEP_CLASS_LIST = ['Castle 1', 'Castle 2']\n",
    "master_groupby = 'Patient'\n",
    "\n",
    "for rep in reps:\n",
    "    for patient in np.unique(METADATA_HARBOUR_PRIMARY.index.get_level_values('Patient')):\n",
    "        SUBSET = pd.DataFrame()\n",
    "        for class_selection in np.unique(METADATA_HARBOUR_PRIMARY.index.get_level_values(master_groupby)):\n",
    "            CSUBSET = METADATA_HARBOUR_PRIMARY.loc[METADATA_HARBOUR_PRIMARY.index.map(lambda x: x[ix] in \n",
    "                                                                  [class_selection])].sample(n=n,replace=True)\n",
    "            SUBSET = SUBSET.append(CSUBSET)\n",
    "        \n",
    "        tmp = [SUBSET.loc[SUBSET.index.get_level_values('Patient') == patient].\\\n",
    "               Assignment.value_counts().reindex(SUBSET.Assignment.unique(), fill_value=0)[x] \n",
    "               for x in GEP_CLASS_LIST]\n",
    "        tmp = list(tmp/sum(tmp))\n",
    "    \n",
    "        boxplot_patient_list = boxplot_patient_list + [patient for x in range(len(GEP_CLASS_LIST))]\n",
    "        boxplot_fract_list = boxplot_fract_list + tmp\n",
    "        boxplot_class_list = boxplot_class_list + GEP_CLASS_LIST\n",
    "\n",
    "boxplot_data = pd.DataFrame()\n",
    "boxplot_data['Patient'] = [x for x in boxplot_patient_list]\n",
    "boxplot_data['Fraction'] = boxplot_fract_list\n",
    "boxplot_data['Class'] = boxplot_class_list\n",
    "\n",
    "boxplot_data.Fraction = [x if x > 0.0001 else 1e-4 for x in boxplot_data.Fraction]\n",
    "\n",
    "# Set plot\n",
    "fig = plt.figure(figsize = (7,4))\n",
    "ax = plt.gca()\n",
    "\n",
    "# Plot bar plot\n",
    "sns.barplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class', ci=95, capsize=0.15,\n",
    "                palette={\"Castle 1\": \"#0000FF\", \"Castle 2\": \"#FF0000\", 'Mixed': '#000000'},\n",
    "                linewidth=1, ax = ax, alpha=1,\n",
    "                errcolor='black',\n",
    "                errwidth=1)\n",
    "\n",
    "g = sns.stripplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class',\n",
    "                palette={\"Castle 1\": \"#0000FF\", \"Castle 2\": \"#FF0000\", 'Mixed': '#000000'},\n",
    "                linewidth = 0.5, edgecolor='white',\n",
    "                ax = ax, alpha=0.7,\n",
    "                dodge=True)\n",
    "\n",
    "fig.get_axes()[0].set_yscale('log')\n",
    "\n",
    "plt.ylim((1e-4,1.15))\n",
    "ax.set_ylabel(\"Cell Type Fraction\",fontsize=10)\n",
    "ax.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "ax.tick_params(labelsize=10)\n",
    "ax.set_xticklabels(boxplot_data['Patient'].unique(),rotation=90)\n",
    "sns.despine()\n",
    "\n",
    "ax.legend_.remove()\n",
    "plt.tight_layout()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'GEP_Distribution_by_Patient_LOG_HARBOUR_PRIMARY'\n",
    "fn = '/workdir/uvmel_project/figures/Revision_2_Figures/' + figure_label \n",
    "plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# MSK Data\n",
    "n = 500\n",
    "reps = np.arange(20)\n",
    "\n",
    "boxplot_patient_list = []\n",
    "boxplot_fract_list = []\n",
    "boxplot_class_list = []\n",
    "GEP_CLASS_LIST = ['Castle 1', 'Castle 2']\n",
    "master_groupby = 'Patient'\n",
    "\n",
    "for rep in reps:\n",
    "    for patient in sort(METADATA_TUMOR.index.get_level_values('Patient').unique().tolist()):\n",
    "        SUBSET = pd.DataFrame()\n",
    "        for class_selection in np.unique(METADATA_TUMOR.index.get_level_values(master_groupby)):\n",
    "            CSUBSET = METADATA_TUMOR.loc[METADATA_TUMOR.index.map(lambda x: x[ix] in \n",
    "                                                                  [class_selection])].sample(n=n,replace=True)\n",
    "            SUBSET = SUBSET.append(CSUBSET)\n",
    "        \n",
    "        tmp = [SUBSET.loc[SUBSET.index.get_level_values('Patient') == patient].\\\n",
    "               Assignment.value_counts().reindex(SUBSET.Assignment.unique(), fill_value=0)[x] \n",
    "               for x in GEP_CLASS_LIST]\n",
    "        tmp = list(tmp/sum(tmp))\n",
    "    \n",
    "        boxplot_patient_list = boxplot_patient_list + [patient for x in range(len(GEP_CLASS_LIST))]\n",
    "        boxplot_fract_list = boxplot_fract_list + tmp\n",
    "        boxplot_class_list = boxplot_class_list + GEP_CLASS_LIST\n",
    "\n",
    "# Harbour Data\n",
    "for rep in reps:\n",
    "    for patient in sort(METADATA_HARBOUR_PRIMARY.index.get_level_values('Patient').unique().tolist()):\n",
    "        SUBSET = pd.DataFrame()\n",
    "        for class_selection in np.unique(METADATA_HARBOUR_PRIMARY.index.get_level_values(master_groupby)):\n",
    "            CSUBSET = METADATA_HARBOUR_PRIMARY.loc[METADATA_HARBOUR_PRIMARY.index.map(lambda x: x[ix] in \n",
    "                                                                  [class_selection])].sample(n=n,replace=True)\n",
    "            SUBSET = SUBSET.append(CSUBSET)\n",
    "        \n",
    "        tmp = [SUBSET.loc[SUBSET.index.get_level_values('Patient') == patient].\\\n",
    "               Assignment.value_counts().reindex(SUBSET.Assignment.unique(), fill_value=0)[x] \n",
    "               for x in GEP_CLASS_LIST]\n",
    "        tmp = list(tmp/sum(tmp))\n",
    "    \n",
    "        boxplot_patient_list = boxplot_patient_list + [patient for x in range(len(GEP_CLASS_LIST))]\n",
    "        boxplot_fract_list = boxplot_fract_list + tmp\n",
    "        boxplot_class_list = boxplot_class_list + GEP_CLASS_LIST\n",
    "\n",
    "        \n",
    "boxplot_data = pd.DataFrame()\n",
    "boxplot_data['Patient'] = ['MSK_'+x if 'MM' not in x else x for x in boxplot_patient_list]\n",
    "boxplot_data['Fraction'] = boxplot_fract_list\n",
    "boxplot_data['Class'] = boxplot_class_list\n",
    "\n",
    "boxplot_data.Fraction = [x if x > 0.0001 else 1e-4 for x in boxplot_data.Fraction]\n",
    "\n",
    "\n",
    "# PLOT\n",
    "fig = plt.figure(figsize = (16,6))\n",
    "ax = plt.gca()\n",
    "\n",
    "# Plot bar and strip plot\n",
    "sns.barplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class', ci=95, capsize=0.15,\n",
    "                palette={\"Castle 1\": \"#0000FF\", \"Castle 2\": \"#FF0000\", 'Mixed': '#000000'},\n",
    "                linewidth=1, ax = ax, alpha=1,\n",
    "                errcolor='black',\n",
    "                errwidth=1)\n",
    "\n",
    "g = sns.stripplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class',\n",
    "                palette={\"Castle 1\": \"#0000FF\", \"Castle 2\": \"#FF0000\", 'Mixed': '#000000'},\n",
    "                linewidth = 1, edgecolor='white',\n",
    "                s=10,\n",
    "                ax = ax, alpha=0.7,\n",
    "                dodge=True)\n",
    "\n",
    "fig.get_axes()[0].set_yscale('log')\n",
    "\n",
    "plt.ylim((1e-4,1.15))\n",
    "ax.set_ylabel(\"Cell Type Fraction\",fontsize=10)\n",
    "ax.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "ax.tick_params(labelsize=10)\n",
    "ax.set_xticklabels(boxplot_data['Patient'].unique(),rotation=90)\n",
    "sns.despine()\n",
    "\n",
    "ax.legend_.remove()\n",
    "plt.tight_layout()\n",
    "plt.minorticks_off()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'GEP_Distribution_by_Patient_LOG_COMBINED'\n",
    "fn = '/workdir/uvmel_project/figures/Revision_2_Figures/' + figure_label \n",
    "plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "METADATA_TUMOR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# MSK Data\n",
    "n = 500\n",
    "reps = np.arange(20)\n",
    "ix=2\n",
    "\n",
    "boxplot_patient_list = []\n",
    "boxplot_fract_list = []\n",
    "boxplot_class_list = []\n",
    "master_groupby = 'Patient'\n",
    "\n",
    "for rep in reps:\n",
    "    for patient in sort(METADATA_TUMOR.index.get_level_values('Patient').unique().tolist()):\n",
    "        SUBSET = pd.DataFrame()\n",
    "        for class_selection in np.unique(METADATA_TUMOR.index.get_level_values(master_groupby)):\n",
    "            CSUBSET = METADATA_TUMOR.loc[METADATA_TUMOR.index.map(lambda x: x[ix] in \n",
    "                                                                  [class_selection])].sample(n=n,replace=True)\n",
    "            SUBSET = SUBSET.append(CSUBSET)\n",
    "        \n",
    "        tmp = [SUBSET.loc[SUBSET.index.get_level_values('Patient') == patient].\\\n",
    "               TCGA_Assignment.value_counts().reindex(SUBSET.TCGA_Assignment.unique(), fill_value=0)[x] \n",
    "               for x in TCGA_CLASS_LIST]\n",
    "        tmp = list(tmp/sum(tmp))\n",
    "    \n",
    "        boxplot_patient_list = boxplot_patient_list + [patient for x in range(len(TCGA_CLASS_LIST))]\n",
    "        boxplot_fract_list = boxplot_fract_list + tmp\n",
    "        boxplot_class_list = boxplot_class_list + TCGA_CLASS_LIST\n",
    "\n",
    "# Harbour Data\n",
    "for rep in reps:\n",
    "    for patient in sort(METADATA_HARBOUR_PRIMARY.index.get_level_values('Patient').unique().tolist()):\n",
    "        SUBSET = pd.DataFrame()\n",
    "        for class_selection in np.unique(METADATA_HARBOUR_PRIMARY.index.get_level_values(master_groupby)):\n",
    "            CSUBSET = METADATA_HARBOUR_PRIMARY.loc[METADATA_HARBOUR_PRIMARY.index.map(lambda x: x[ix] in \n",
    "                                                                  [class_selection])].sample(n=n,replace=True)\n",
    "            SUBSET = SUBSET.append(CSUBSET)\n",
    "        \n",
    "        tmp = [SUBSET.loc[SUBSET.index.get_level_values('Patient') == patient].\\\n",
    "               TCGA_Assignment.value_counts().reindex(SUBSET.TCGA_Assignment.unique(), fill_value=0)[x] \n",
    "               for x in TCGA_CLASS_LIST]\n",
    "        tmp = list(tmp/sum(tmp))\n",
    "    \n",
    "        boxplot_patient_list = boxplot_patient_list + [patient for x in range(len(TCGA_CLASS_LIST))]\n",
    "        boxplot_fract_list = boxplot_fract_list + tmp\n",
    "        boxplot_class_list = boxplot_class_list + TCGA_CLASS_LIST\n",
    "\n",
    "        \n",
    "boxplot_data = pd.DataFrame()\n",
    "boxplot_data['Patient'] = ['MSK_'+x if 'MM' not in x else x for x in boxplot_patient_list]\n",
    "boxplot_data['Fraction'] = boxplot_fract_list\n",
    "boxplot_data['Class'] = boxplot_class_list\n",
    "\n",
    "boxplot_data.Fraction = [x if x > 0.0001 else 1e-4 for x in boxplot_data.Fraction]\n",
    "\n",
    "\n",
    "# PLOT\n",
    "fig = plt.figure(figsize = (16,6))\n",
    "ax = plt.gca()\n",
    "\n",
    "# Plot bar and strip plot\n",
    "sns.barplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class', ci=95, capsize=0.15,\n",
    "                palette={\"TCGA1\": \"#0000FF\",\n",
    "                         \"TCGA2\": \"#A4D3FC\",\n",
    "                         \"TCGA3\": \"#FF7DC2\", \n",
    "                         \"TCGA4\": \"#FF0000\"},\n",
    "                linewidth=1, ax = ax, alpha=1,\n",
    "                errcolor='black',\n",
    "                errwidth=1)\n",
    "\n",
    "g = sns.stripplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class',\n",
    "                palette={\"TCGA1\": \"#0000FF\",\n",
    "                         \"TCGA2\": \"#A4D3FC\",\n",
    "                         \"TCGA3\": \"#FF7DC2\", \n",
    "                         \"TCGA4\": \"#FF0000\"},\n",
    "                linewidth = 1, edgecolor='white',\n",
    "                s=10,\n",
    "                ax = ax, alpha=0.7,\n",
    "                dodge=True)\n",
    "\n",
    "fig.get_axes()[0].set_yscale('log')\n",
    "\n",
    "plt.ylim((1e-4,1.15))\n",
    "ax.set_ylabel(\"Cell Type Fraction\",fontsize=10)\n",
    "ax.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "ax.tick_params(labelsize=10)\n",
    "ax.set_xticklabels(boxplot_data['Patient'].unique(),rotation=90)\n",
    "sns.despine()\n",
    "\n",
    "ax.legend_.remove()\n",
    "plt.tight_layout()\n",
    "plt.minorticks_off()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'TCGA_Distribution_by_Patient_LOG_COMBINED'\n",
    "fn = '/workdir/uvmel_project/figures/Revision_2_Figures/' + figure_label \n",
    "#plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "#plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "boxplot_data.Patient.value_counts()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Set Harbour primary GEP assignments to results from combined model:\n",
    "METADATA_HARBOUR_PRIMARY['Assignment'] = [METADATA_HARBOUR.loc[\n",
    "                                          METADATA_HARBOUR.index.get_level_values('Legend') == 'primary'].loc[\n",
    "                                          METADATA_HARBOUR.loc[\n",
    "                                          METADATA_HARBOUR.index.get_level_values('Legend') == 'primary'].\\\n",
    "                                          index.get_level_values('Cell ID') == x].Assignment.values[0] \n",
    "                                          for x in METADATA_HARBOUR_PRIMARY.index.get_level_values('Cell ID').values]\n",
    "\n",
    "# Set Harbour primary TCGA assignments to results from combined model:\n",
    "METADATA_HARBOUR_PRIMARY['TCGA_Assignment'] = [METADATA_HARBOUR.loc[\n",
    "                                          METADATA_HARBOUR.index.get_level_values('Legend') == 'primary'].loc[\n",
    "                                          METADATA_HARBOUR.loc[\n",
    "                                          METADATA_HARBOUR.index.get_level_values('Legend') == 'primary'].\\\n",
    "                                          index.get_level_values('Cell ID') == x].TCGA_Assignment.values[0] \n",
    "                                          for x in METADATA_HARBOUR_PRIMARY.index.get_level_values('Cell ID').values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "boxplot_patient_list = []\n",
    "boxplot_fract_list = []\n",
    "boxplot_class_list = []\n",
    "\n",
    "for patient in sort(METADATA_HARBOUR_PRIMARY.index.get_level_values('Patient').unique().tolist()):\n",
    "    tmp = [METADATA_HARBOUR_PRIMARY.loc[METADATA_HARBOUR_PRIMARY.index.get_level_values('Patient') == patient].\\\n",
    "           Assignment.value_counts().reindex(METADATA_HARBOUR_PRIMARY.Assignment.unique(), fill_value=0)[x] \n",
    "           for x in GEP_CLASS_LIST]\n",
    "    tmp = list(tmp/sum(tmp))\n",
    "    \n",
    "    boxplot_patient_list = boxplot_patient_list + [patient for x in range(len(GEP_CLASS_LIST))]\n",
    "    boxplot_fract_list = boxplot_fract_list + tmp\n",
    "    boxplot_class_list = boxplot_class_list + GEP_CLASS_LIST\n",
    "\n",
    "boxplot_data = pd.DataFrame()\n",
    "boxplot_data['Patient'] = boxplot_patient_list\n",
    "boxplot_data['Fraction'] = boxplot_fract_list\n",
    "boxplot_data['Class'] = boxplot_class_list\n",
    "\n",
    "# VIOLIN PLOT\n",
    "fig = plt.figure(figsize = (12,5))\n",
    "ax = plt.gca()\n",
    "\n",
    "# Plot violin plot\n",
    "g = sns.barplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class', ci=95, capsize=.2,\n",
    "                palette={\"Castle 1\": \"#0000FF\", \"Castle 2\": \"#FF0000\", 'Mixed': '#000000'},\n",
    "                linewidth = 1, ax = ax);\n",
    "fig.get_axes()[0].set_yscale('log')\n",
    "\n",
    "plt.ylim((0,1))\n",
    "g.set_ylabel(\"\",fontsize=10)\n",
    "g.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "g.tick_params(labelsize=18)\n",
    "g.set_xticklabels(ax.get_xticklabels(),rotation=90)\n",
    "sns.despine()\n",
    "sns.set_style('white')\n",
    "plt.tick_params(size=0)\n",
    "\n",
    "ax.legend_.remove()\n",
    "plt.tight_layout()\n",
    "g.axis(True)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'GEP_Distribution_COMBINED'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "#plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "#plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "boxplot_patient_list = []\n",
    "boxplot_fract_list = []\n",
    "boxplot_class_list = []\n",
    "\n",
    "for patient in METADATA_TUMOR.index.get_level_values('Patient').unique():\n",
    "    tmp = [METADATA_TUMOR.loc[METADATA_TUMOR.index.get_level_values('Patient') == patient].\\\n",
    "           TCGA_Assignment.value_counts().reindex(METADATA_TUMOR.TCGA_Assignment.unique(), fill_value=0)[x] \n",
    "           for x in TCGA_CLASS_LIST]\n",
    "    tmp = list(tmp/sum(tmp))\n",
    "    \n",
    "    boxplot_patient_list = boxplot_patient_list + ['MSK-' + patient for x in range(len(TCGA_CLASS_LIST))]\n",
    "    boxplot_fract_list = boxplot_fract_list + tmp\n",
    "    boxplot_class_list = boxplot_class_list + TCGA_CLASS_LIST\n",
    "\n",
    "for patient in sort(METADATA_HARBOUR_PRIMARY.index.get_level_values('Patient').unique().tolist()):\n",
    "    tmp = [METADATA_HARBOUR_PRIMARY.loc[METADATA_HARBOUR_PRIMARY.index.get_level_values('Patient') == patient].\\\n",
    "           TCGA_Assignment.value_counts().reindex(METADATA_HARBOUR_PRIMARY.TCGA_Assignment.unique(), fill_value=0)[x] \n",
    "           for x in TCGA_CLASS_LIST]\n",
    "    tmp = list(tmp/sum(tmp))\n",
    "    \n",
    "    boxplot_patient_list = boxplot_patient_list + [patient for x in range(len(TCGA_CLASS_LIST))]\n",
    "    boxplot_fract_list = boxplot_fract_list + tmp\n",
    "    boxplot_class_list = boxplot_class_list + TCGA_CLASS_LIST\n",
    "\n",
    "boxplot_data = pd.DataFrame()\n",
    "boxplot_data['Patient'] = boxplot_patient_list\n",
    "boxplot_data['Fraction'] = boxplot_fract_list\n",
    "boxplot_data['Class'] = boxplot_class_list\n",
    "\n",
    "# VIOLIN PLOT\n",
    "fig = plt.figure(figsize = (12,5))\n",
    "ax = plt.gca()\n",
    "\n",
    "# Plot violin plot\n",
    "g = sns.barplot(data=boxplot_data,x = 'Patient', \n",
    "                y = 'Fraction', hue = 'Class', ci=95, capsize=.2,\n",
    "                palette={\"TCGA1\": \"#0000FF\",\n",
    "                         \"TCGA2\": \"#A4D3FC\",\n",
    "                         \"TCGA3\": \"#FF7DC2\", \n",
    "                         \"TCGA4\": \"#FF0000\"},\n",
    "                linewidth = 1, ax = ax);\n",
    "fig.get_axes()[0].set_yscale('log')\n",
    "\n",
    "plt.ylim((0,1))\n",
    "g.set_ylabel(\"\",fontsize=10)\n",
    "g.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "g.tick_params(labelsize=18)\n",
    "g.set_xticklabels(ax.get_xticklabels(),rotation=90)\n",
    "sns.despine()\n",
    "sns.set_style('ticks')\n",
    "plt.tick_params(size=0)\n",
    "\n",
    "ax.legend_.remove()\n",
    "plt.tight_layout()\n",
    "g.axis(True)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'TCGA_Distribution_COMBINED_BGMM_full_trained'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "TCGA_probs_x_HARBOUR = [X_norm.loc[X_norm.index.map(lambda x: x[2] == y)].TCGA2.values - \n",
    "                        X_norm.loc[X_norm.index.map(lambda x: x[2] == y)].TCGA4.values\n",
    "                        for y in X_norm.groupby(level='Patient').count().index]\n",
    "\n",
    "TCGA_probs_y_HARBOUR = [X_norm.loc[X_norm.index.map(lambda x: x[2] == y)].TCGA1.values - \n",
    "                        X_norm.loc[X_norm.index.map(lambda x: x[2] == y)].TCGA3.values\n",
    "                        for y in X_norm.groupby(level='Patient').count().index]\n",
    "\n",
    "TCGA_probs_per_patient_HARBOUR = [[(TCGA_probs_x_HARBOUR[x][y], TCGA_probs_y_HARBOUR[x][y]) \n",
    "                                   for y in range(len(TCGA_probs_x_HARBOUR[x]))]\n",
    "                                   for x in range(len(TCGA_probs_x_HARBOUR))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diamond_list = [plt.Rectangle((0,-1), np.sqrt(2), np.sqrt(2), 45, color='black', fill=False, linewidth=0.5) \n",
    "                for x in range(11)]\n",
    "\n",
    "fig1, axes = plt.subplots(2,3)\n",
    "\n",
    "for i, ax in enumerate(axes.flatten()):\n",
    "    if i in range(len(TCGA_probs_per_patient_HARBOUR)):\n",
    "        ax.scatter(*zip(*TCGA_probs_per_patient_HARBOUR[i]), s=0.2, color='black',\n",
    "                   alpha=0.3)\n",
    "        ax.set_frame_on(False)\n",
    "        ax.axes.get_xaxis().set_visible(False)\n",
    "        ax.axes.get_yaxis().set_visible(False)\n",
    "        ax.add_patch(diamond_list[i])\n",
    "        ax.set_title(X_norm.groupby(level='Patient').count().index[i], fontdict={'fontsize':6})\n",
    "    else:\n",
    "        ax.set_frame_on(False)\n",
    "        ax.axes.get_xaxis().set_visible(False)\n",
    "        ax.axes.get_yaxis().set_visible(False)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'TCGA_PROBS_DiamondPlot_HARBOUR_PRIMARY_1'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig1.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig1.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diamond_list = [plt.Rectangle((0,-1), np.sqrt(2), np.sqrt(2), 45, color='black', fill=False, linewidth=0.5) \n",
    "                for x in range(11)]\n",
    "\n",
    "fig1, axes = plt.subplots(2,3)\n",
    "\n",
    "for i, ax in enumerate(axes.flatten()):\n",
    "    i = i + 6\n",
    "    if i in range(len(TCGA_probs_per_patient_HARBOUR)):\n",
    "        ax.scatter(*zip(*TCGA_probs_per_patient_HARBOUR[i]), s=0.2, color='black',\n",
    "                   alpha=0.3)\n",
    "        ax.set_frame_on(False)\n",
    "        ax.axes.get_xaxis().set_visible(False)\n",
    "        ax.axes.get_yaxis().set_visible(False)\n",
    "        ax.add_patch(diamond_list[i])\n",
    "        ax.set_title(X_norm.groupby(level='Patient').count().index[i], fontdict={'fontsize':6})\n",
    "    else:\n",
    "        ax.set_frame_on(False)\n",
    "        ax.axes.get_xaxis().set_visible(False)\n",
    "        ax.axes.get_yaxis().set_visible(False)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'TCGA_PROBS_DiamondPlot_HARBOUR_PRIMARY_2'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig1.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig1.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PRC1 TARGET GENES HEATMAPS - HARBOUR DATASET"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GEP2 RANKED HEATMAP FOR PRC1 TARGET GENES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RANK BY GEP 2\n",
    "subset_type = 'TUMOR'\n",
    "QUERY = INDF_HARBOUR\n",
    "\n",
    "rank_by = 'Castle 2'#'Class 2 Onken v2'\n",
    "genes = np.unique(genesets[rank_by].values.ravel().tolist())\n",
    "detected_genes = [gene for gene in genes if gene in list(QUERY.columns)]\n",
    "\n",
    "vals = QUERY[detected_genes]\n",
    "SCORE = np.nanmedian(vals,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PRC1_target_genes = pd.read_csv(DATA_PATH+'PRC1_targets_Genesets_and_Chipseq.csv')\n",
    "plot_genes = PRC1_target_genes['PRC1_targets_Genesets_and_Chipseq'].dropna().values.tolist()\n",
    "plot_genes = [x for x in plot_genes if x in QUERY.columns.values.tolist()]\n",
    "\n",
    "# LOAD GENESETS\n",
    "path_to_genesets = DATA_PATH+'uveal_melanoma_MB_v3.csv'\n",
    "genesets = pd.read_csv(path_to_genesets,header='infer')\n",
    "\n",
    "# HEATMAP OF INDIVIDUAL CELLS RANKED BY LUNGE EP\n",
    "rank_by = 'Castle 2'#\n",
    "genesets_include = [rank_by]\n",
    "genes = np.unique(genesets[genesets_include].values.ravel().tolist())\n",
    "genes = [gene for gene in genes if gene in list(set(QUERY.columns))]\n",
    "vals = QUERY[genes]\n",
    "QUERY['GEP2_RANK'] =np.nanmean(vals,axis=1)\n",
    "\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "heatmap_data = pd.DataFrame( data = zscore(QUERY.sort_values(by=['GEP2_RANK'])[plot_genes],axis=0), \n",
    "                   columns = plot_genes, index = QUERY.sort_values(by=['GEP2_RANK']).index)\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns\n",
    "\n",
    "# LINKAGE \n",
    "method = 'centroid' # average, single centroid/euclidean\n",
    "metric = 'euclidean' # cosine\n",
    "linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "col_linkage = deepcopy(linkage)\n",
    "cl = hc.leaves_list(col_linkage)\n",
    "mat = heatmap_data.iloc[:,cl]\n",
    "\n",
    "window = 20\n",
    "mat = mat.rolling(window, win_type='triang',center = True).sum()\n",
    "half_window = int(window/2)\n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[ind] = mat.iloc[half_window]    \n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[-ind] = mat.iloc[-half_window]\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(4,10))\n",
    "\n",
    "# ADD MATRIX WITH LINEAGE NAMES\n",
    "axmatrix = fig.add_axes([0.12,0.1,0.4,0.6])\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=CM_DIVERGING,vmin=-10,vmax=10)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "axmatrix.set_xticklabels([]);\n",
    "axmatrix.grid(False)\n",
    "axmatrix.spines['top'].set_visible(False)\n",
    "axmatrix.spines['right'].set_visible(False)\n",
    "axmatrix.spines['bottom'].set_visible(False)\n",
    "axmatrix.spines['left'].set_visible(False)\n",
    "plt.tick_params(axis='both', size=0)\n",
    "\n",
    "\n",
    "# ADD ROW 1 COLOR BY SAMPLE TYPE\n",
    "sample_status = [QUERY.loc[QUERY.index.get_level_values('Cell ID') == x].index.get_level_values('Legend').\\\n",
    "                          values[0] for x in mat.index.get_level_values('Cell ID').values]\n",
    "\n",
    "lut_SAMPLE = dict({'primary': np.array([1, 1, 1]), 'metastatic': np.array([0, 0, 0])})\n",
    "\n",
    "row_colors1 = pd.Series(sample_status, index=sample_status).map(lut_SAMPLE)\n",
    "ax1 = fig.add_axes([0.07,0.1,0.045,0.6]) # [x0,y0,width,height]\n",
    "x = 0\n",
    "y = 0\n",
    "for c in row_colors1:\n",
    "    pos = (x, y / len(row_colors1))\n",
    "    ax1.add_patch(patches.Rectangle(pos, 1, 1 / len(row_colors1), color=c))\n",
    "    if y >= len(row_colors1)-1:\n",
    "        x += 1\n",
    "        y = 0\n",
    "    else:\n",
    "        y += 1\n",
    "plt.axis('off')\n",
    "\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'PRC1_Targets_Heatmap_UPDATED_HARBOUR'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/GEP2_Ranked_Heatmaps/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LOAD GENESETS\n",
    "plot_genes = ['IRF3', 'RELA', 'RELB', 'MB21D1', 'IKBKB', 'NFKB1', 'TMEM173', 'STAT2', 'CXCL10', 'NFKB2',\n",
    "              'CCL5', 'TBK1', 'JAK1', 'JAK2', 'IRF9', 'TYK2', 'IRF7', 'IRF1', 'STAT1'\n",
    "             ]\n",
    "\n",
    "path_to_genesets = DATA_PATH+'uveal_melanoma_MB_v3.csv'\n",
    "genesets = pd.read_csv(path_to_genesets,header='infer')\n",
    "\n",
    "datatype = 'INDF_HARBOUR'\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "\n",
    "# HEATMAP OF INDIVIDUAL CELLS RANKED BY LUNGE EP\n",
    "rank_by = 'Castle 2'#\n",
    "genesets_include = [rank_by]\n",
    "genes = np.unique(genesets[genesets_include].values.ravel().tolist())\n",
    "genes = [gene for gene in genes if gene in list(set(QUERY.columns))]\n",
    "vals = QUERY[genes]\n",
    "QUERY['GEP2_RANK'] =np.nanmean(vals,axis=1)\n",
    "heatmap_data = pd.DataFrame( data = QUERY.sort_values(by=['GEP2_RANK'])[plot_genes +['GEP2_RANK']], \n",
    "                   columns = plot_genes +['GEP2_RANK'], index = QUERY.sort_values(by=['GEP2_RANK']).index)\n",
    "\n",
    "y = heatmap_data['GEP2_RANK'].values\n",
    "x = np.arange(len(y))\n",
    "\n",
    "fig = plt.figure(figsize=(10,4))\n",
    "ax = plt.gca()\n",
    "\n",
    "ax.fill_between(x, 0, y,facecolor='#E0E0E0', interpolate=True)\n",
    "plt.xlim((0,x.max()))\n",
    "plt.ylim((0,y.max()))\n",
    "sns.despine()\n",
    "\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'PRC1_Targets_UPDATED_GEP2_RANK__HARBOUR_curve'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/GEP2_Ranked_Heatmaps/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_HARBOUR.index.get_level_values('cellType').value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_HARBOUR_PRIMARY.index.get_level_values('cellType').value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_HARBOUR_METASTASIS.index.get_level_values('cellType').value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### STING PATHWAY GENES HEATMAPS - HARBOUR DATASET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "plot_genes = ['IRF3', 'RELA', 'RELB', 'MB21D1', 'IKBKB', 'NFKB1', 'TMEM173', 'STAT2', 'CXCL10', 'NFKB2',\n",
    "              'CCL5', 'TBK1', 'JAK1', 'JAK2', 'IRF9', 'TYK2', 'IRF7', 'IRF1', 'STAT1'\n",
    "             ]\n",
    "\n",
    "print(len(plot_genes))\n",
    "\n",
    "# LOAD GENESETS\n",
    "path_to_genesets = DATA_PATH+'uveal_melanoma_MB_v3.csv'\n",
    "genesets = pd.read_csv(path_to_genesets,header='infer')\n",
    "\n",
    "subset_type = 'TUMOR'\n",
    "QUERY = INDF_HARBOUR\n",
    "\n",
    "# HEATMAP OF INDIVIDUAL CELLS RANKED BY LUNGE EP\n",
    "rank_by = 'Castle 2'#\n",
    "genesets_include = [rank_by]\n",
    "genes = np.unique(genesets[genesets_include].values.ravel().tolist())\n",
    "genes = [gene for gene in genes if gene in list(set(QUERY.columns))]\n",
    "vals = QUERY[genes]\n",
    "QUERY['GEP2_RANK'] =np.nanmean(vals,axis=1)\n",
    "\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "heatmap_data = pd.DataFrame( data = zscore(QUERY.sort_values(by=['GEP2_RANK'])[plot_genes],axis=0), \n",
    "                   columns = plot_genes, index = QUERY.sort_values(by=['GEP2_RANK']).index)\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns\n",
    "\n",
    "# LINKAGE \n",
    "#method = 'average' # average, single centroid/euclidean\n",
    "#metric = 'euclidean' # cosine\n",
    "#linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "#col_linkage = deepcopy(linkage)\n",
    "#cl = hc.leaves_list(col_linkage)\n",
    "#mat = heatmap_data.iloc[:,cl]\n",
    "mat = heatmap_data.copy()\n",
    "\n",
    "window = 20\n",
    "mat = mat.rolling(window, win_type='triang',center = True).sum()\n",
    "half_window = int(window/2)\n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[ind] = mat.iloc[half_window]    \n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[-ind] = mat.iloc[-half_window]\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(5,10))\n",
    "\n",
    "# ADD MATRIX WITH LINEAGE NAMES\n",
    "axmatrix = fig.add_axes([0.12,0.1,0.4,0.6])\n",
    "colors = [(1,1,1), np.divide(tuple(hex('FF0000').rgb),255)]\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=CM_DIVERGING,vmin=-10,vmax=10)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "#axmatrix.set_xticklabels([]);\n",
    "xtick = plt.xticks(range(len(labels)), labels, rotation = 90, fontsize = 6,fontname='Arial')\n",
    "ytick = plt.yticks([])\n",
    "axmatrix.grid(False)\n",
    "axmatrix.spines['top'].set_visible(False)\n",
    "axmatrix.spines['right'].set_visible(False)\n",
    "axmatrix.spines['bottom'].set_visible(False)\n",
    "axmatrix.spines['left'].set_visible(False)\n",
    "\n",
    "\n",
    "# ADD ROW 1 COLOR BY SAMPLE TYPE\n",
    "sample_status = [QUERY.loc[QUERY.index.get_level_values('Cell ID') == x].index.get_level_values('Legend').\\\n",
    "                          values[0] for x in mat.index.get_level_values('Cell ID').values]\n",
    "\n",
    "lut_SAMPLE = dict({'primary': np.array([1, 1, 1]), 'metastatic': np.array([0, 0, 0])})\n",
    "\n",
    "row_colors1 = pd.Series(sample_status, index=sample_status).map(lut_SAMPLE)\n",
    "ax1 = fig.add_axes([0.08,0.1,0.035,0.6]) # [x0,y0,width,height]\n",
    "x = 0\n",
    "y = 0\n",
    "for c in row_colors1:\n",
    "    pos = (x, y / len(row_colors1))\n",
    "    ax1.add_patch(patches.Rectangle(pos, 1, 1 / len(row_colors1), color=c))\n",
    "    if y >= len(row_colors1)-1:\n",
    "        x += 1\n",
    "        y = 0\n",
    "    else:\n",
    "        y += 1\n",
    "plt.axis('off')\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'STING_PATHWAY_Heatmap_HARBOUR'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/GEP2_Ranked_Heatmaps/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PRC1 TARGET GENES HEATMAPS - HARBOUR PRIMARY SAMPLES"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GEP2 RANKED HEATMAP FOR PRC1 TARGET GENES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RANK BY GEP 2\n",
    "subset_type = 'TUMOR'\n",
    "QUERY = INDF_HARBOUR_PRIMARY\n",
    "\n",
    "rank_by = 'Castle 2'#'Class 2 Onken v2'\n",
    "genes = np.unique(genesets[rank_by].values.ravel().tolist())\n",
    "detected_genes = [gene for gene in genes if gene in list(QUERY.columns)]\n",
    "\n",
    "vals = QUERY[detected_genes]\n",
    "SCORE = np.nanmedian(vals,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "PRC1_target_genes = pd.read_csv(DATA_PATH+'PRC1_targets_Genesets_and_Chipseq.csv')\n",
    "plot_genes = PRC1_target_genes['PRC1_targets_Genesets_and_Chipseq'].dropna().values.tolist()\n",
    "plot_genes = [x for x in plot_genes if x in QUERY.columns.values.tolist()]\n",
    "\n",
    "# LOAD GENESETS\n",
    "path_to_genesets = DATA_PATH+'uveal_melanoma_MB_v3.csv'\n",
    "genesets = pd.read_csv(path_to_genesets,header='infer')\n",
    "\n",
    "# HEATMAP OF INDIVIDUAL CELLS RANKED BY LUNGE EP\n",
    "rank_by = 'Castle 2'#\n",
    "genesets_include = [rank_by]\n",
    "genes = np.unique(genesets[genesets_include].values.ravel().tolist())\n",
    "genes = [gene for gene in genes if gene in list(set(QUERY.columns))]\n",
    "vals = QUERY[genes]\n",
    "QUERY['GEP2_RANK'] =np.nanmean(vals,axis=1)\n",
    "\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "heatmap_data = pd.DataFrame( data = zscore(QUERY.sort_values(by=['GEP2_RANK'])[plot_genes],axis=0), \n",
    "                   columns = plot_genes, index = QUERY.sort_values(by=['GEP2_RANK']).index)\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns\n",
    "\n",
    "# LINKAGE \n",
    "method = 'centroid' # average, single centroid/euclidean\n",
    "metric = 'euclidean' # cosine\n",
    "linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "col_linkage = deepcopy(linkage)\n",
    "cl = hc.leaves_list(col_linkage)\n",
    "mat = heatmap_data.iloc[:,cl]\n",
    "\n",
    "window = 20\n",
    "mat = mat.rolling(window, win_type='triang',center = True).sum()\n",
    "half_window = int(window/2)\n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[ind] = mat.iloc[half_window]    \n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[-ind] = mat.iloc[-half_window]\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(4,10))\n",
    "\n",
    "# ADD MATRIX WITH LINEAGE NAMES\n",
    "axmatrix = fig.add_axes([0.12,0.1,0.4,0.6])\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=CM_DIVERGING,vmin=-10,vmax=10)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "axmatrix.set_xticklabels([]);\n",
    "axmatrix.grid(False)\n",
    "plt.tick_params(axis='both', size=0)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'PRC1_Targets_Heatmap_UPDATED_HARBOUR_PRIMARY'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/GEP2_Ranked_Heatmaps/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### GEP2 RANKED HEATMAP NEW GENESET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RANK BY GEP 2\n",
    "subset_type = 'HARBOUR_PRIMARY'\n",
    "exec('QUERY = INDF_{}'.format(subset_type))\n",
    "rank_by = 'Castle 2'#'Class 2 Onken v2'\n",
    "genes = np.unique(genesets[rank_by].values.ravel().tolist())\n",
    "detected_genes = [gene for gene in genes if gene in list(QUERY.columns)]\n",
    "\n",
    "vals = QUERY[detected_genes]\n",
    "SCORE = np.nanmedian(vals,axis=1)\n",
    "\n",
    "target_genes = pd.read_csv(DATA_PATH+'829_PRE_92_1_geneset.txt', \n",
    "                                sep='\\t', header=None, names=['genes'])\n",
    "plot_genes = target_genes['genes'].dropna().values.tolist()\n",
    "plot_genes = [x for x in plot_genes if x in QUERY.columns.values.tolist()]\n",
    "\n",
    "# LOAD GENESETS\n",
    "path_to_genesets = DATA_PATH+'uveal_melanoma_MB_v3.csv'\n",
    "genesets = pd.read_csv(path_to_genesets,header='infer')\n",
    "\n",
    "# HEATMAP OF INDIVIDUAL CELLS RANKED BY LUNGE EP\n",
    "rank_by = 'Castle 2'#\n",
    "genesets_include = [rank_by]\n",
    "genes = np.unique(genesets[genesets_include].values.ravel().tolist())\n",
    "genes = [gene for gene in genes if gene in list(set(QUERY.columns))]\n",
    "vals = QUERY[genes]\n",
    "QUERY['GEP2_RANK'] =np.nanmean(vals,axis=1)\n",
    "\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "heatmap_data = pd.DataFrame( data = zscore(QUERY.sort_values(by=['GEP2_RANK'])[plot_genes],axis=0), \n",
    "                   columns = plot_genes, index = QUERY.sort_values(by=['GEP2_RANK']).index)\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns\n",
    "\n",
    "# LINKAGE \n",
    "method = 'average' # average, single centroid/euclidean\n",
    "metric = 'euclidean' # cosine\n",
    "linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "col_linkage = deepcopy(linkage)\n",
    "cl = hc.leaves_list(col_linkage)\n",
    "mat = heatmap_data.iloc[:,cl]\n",
    "\n",
    "window = 20\n",
    "mat = mat.rolling(window, win_type='triang',center = True).sum()\n",
    "half_window = int(window/2)\n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[ind] = mat.iloc[half_window]    \n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[-ind] = mat.iloc[-half_window]\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(4,10))\n",
    "\n",
    "# ADD MATRIX WITH LINEAGE NAMES\n",
    "axmatrix = fig.add_axes([0.12,0.1,0.4,0.6])\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=CM_DIVERGING,vmin=-10,vmax=10)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "axmatrix.set_xticklabels([]);\n",
    "axmatrix.grid(False)\n",
    "plt.tick_params(axis='both', size=0)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'average_euclidean_heatmap_NEW_GENESET_HARBOUR_PRIMARY'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/GEP2_Ranked_Heatmaps/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### GEP2 RANKED HEATMAP 5/5 GENESETS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RANK BY GEP 2\n",
    "subset_type = 'TUMOR'\n",
    "QUERY = INDF_HARBOUR_PRIMARY\n",
    "\n",
    "rank_by = 'Castle 2'#'Class 2 Onken v2'\n",
    "genes = np.unique(genesets[rank_by].values.ravel().tolist())\n",
    "detected_genes = [gene for gene in genes if gene in list(QUERY.columns)]\n",
    "\n",
    "vals = QUERY[detected_genes]\n",
    "QUERY['GEP2_RANK'] =np.nanmean(vals,axis=1)\n",
    "QUERY = QUERY.sort_values(by=['GEP2_RANK'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for col in target_genes:\n",
    "    print(col,':')\n",
    "    print(round(len([x for x in target_genes[col] if x in QUERY.columns])/len(target_genes[col].dropna())*100,3), '%')\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Harbour Primary Data')\n",
    "for col in target_genes.columns:\n",
    "    print(col,':')\n",
    "    plot_genes = target_genes[col].dropna().values.tolist()\n",
    "    print('Number of genes in set:',len(plot_genes))\n",
    "    plot_genes = [x for x in plot_genes if x in QUERY.columns.values.tolist()]\n",
    "    print('Number of genes in set in data:',len(plot_genes))\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LOAD GENESETS\n",
    "target_genes = pd.read_csv(DATA_PATH+'Genesets_5_5.csv', \n",
    "                                sep=',', header=0)\n",
    "\n",
    "# Iterate through each geneset:\n",
    "for col in target_genes.columns:\n",
    "    print(col)\n",
    "    plot_genes = target_genes[col].dropna().values.tolist()\n",
    "    plot_genes = [x for x in plot_genes if x in QUERY.columns.values.tolist()]\n",
    "    \n",
    "    if len(plot_genes) > 0:\n",
    "    \n",
    "        # CONSTRUCT HEATMAP DATA\n",
    "        heatmap_data = pd.DataFrame( data = zscore(QUERY[plot_genes],axis=0), \n",
    "                           columns = plot_genes, index = QUERY.index)\n",
    "        heatmap_data = heatmap_data.dropna(axis=1, how='any') \n",
    "        yticks = heatmap_data.index\n",
    "        xticks = heatmap_data.columns\n",
    "        \n",
    "        # LINKAGE \n",
    "        for m in [('average', 'euclidean'), ('average', 'cosine'), ('centroid', 'euclidean')]:\n",
    "            method = m[0] # average, single centroid/euclidean\n",
    "            metric = m[1] # cosine\n",
    "            linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "            col_linkage = deepcopy(linkage)\n",
    "            cl = hc.leaves_list(col_linkage)\n",
    "            mat = heatmap_data.iloc[:,cl]\n",
    "            \n",
    "            window = 20\n",
    "            mat = mat.rolling(window, win_type='triang',center = True).sum()\n",
    "            half_window = int(window/2)\n",
    "            for ind in np.arange(half_window):\n",
    "                mat.iloc[ind] = mat.iloc[half_window]    \n",
    "            for ind in np.arange(half_window):\n",
    "                mat.iloc[-ind] = mat.iloc[-half_window]\n",
    "            \n",
    "            # VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "            fig = plt.figure(figsize=(4,10))\n",
    "            \n",
    "            # ADD MATRIX WITH LINEAGE NAMES\n",
    "            axmatrix = fig.add_axes([0.12,0.1,0.4,0.6])\n",
    "            im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=CM_DIVERGING,vmin=-10,vmax=10)\n",
    "            labels = list(mat.columns)\n",
    "            axmatrix.xaxis.set_ticks_position('bottom')\n",
    "            axmatrix.set_yticklabels([]);\n",
    "            axmatrix.set_xticklabels([]);\n",
    "            #xtick = plt.xticks(range(len(labels)), labels, rotation = 90, fontsize = 6,fontname='Arial')\n",
    "            axmatrix.grid(False)\n",
    "            plt.tick_params(axis='both', size=0)\n",
    "            \n",
    "            # SAVE FIGURE\n",
    "            figure_label = col+'_'+method+'_'+metric+'_HARBOUR_PRIMARY'\n",
    "            fn = '/workdir/uvmel_project/figures/Revision_2_Figures/Genesets_5_5_Heatmaps/' + figure_label \n",
    "                \n",
    "            d = os.path.dirname(fn)\n",
    "            if not os.path.exists(d):\n",
    "                os.makedirs(d)\n",
    "                \n",
    "            fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "            #fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "            print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### GEP2 RANKED CURVE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "INDF_HARBOUR_PRIMARY.index.get_level_values('cellType').value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LOAD GENESETS\n",
    "plot_genes = ['IRF3', 'RELA', 'RELB', 'MB21D1', 'IKBKB', 'NFKB1', 'TMEM173', 'STAT2', 'CXCL10', 'NFKB2',\n",
    "              'CCL5', 'TBK1', 'JAK1', 'JAK2', 'IRF9', 'TYK2', 'IRF7', 'IRF1', 'STAT1'\n",
    "             ]\n",
    "\n",
    "path_to_genesets = DATA_PATH+'uveal_melanoma_MB_v3.csv'\n",
    "genesets = pd.read_csv(path_to_genesets,header='infer')\n",
    "\n",
    "datatype = 'INDF_HARBOUR_PRIMARY'\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "\n",
    "# HEATMAP OF INDIVIDUAL CELLS RANKED BY LUNGE EP\n",
    "rank_by = 'Castle 2'#\n",
    "genesets_include = [rank_by]\n",
    "genes = np.unique(genesets[genesets_include].values.ravel().tolist())\n",
    "genes = [gene for gene in genes if gene in list(set(QUERY.columns))]\n",
    "vals = QUERY[genes]\n",
    "QUERY['GEP2_RANK'] =np.nanmean(vals,axis=1)\n",
    "heatmap_data = pd.DataFrame( data = QUERY.sort_values(by=['GEP2_RANK'])[plot_genes +['GEP2_RANK']], \n",
    "                   columns = plot_genes +['GEP2_RANK'], index = QUERY.sort_values(by=['GEP2_RANK']).index)\n",
    "\n",
    "y = heatmap_data['GEP2_RANK'].values\n",
    "x = np.arange(len(y))\n",
    "\n",
    "fig = plt.figure(figsize=(10,4))\n",
    "ax = plt.gca()\n",
    "\n",
    "ax.fill_between(x, 0, y,facecolor='#E0E0E0', interpolate=True)\n",
    "plt.xlim((0,x.max()))\n",
    "plt.ylim((0,y.max()))\n",
    "sns.despine()\n",
    "\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'PRC1_Targets_UPDATED_GEP2_RANK__HARBOUR_PRIMARY_curve'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/GEP2_Ranked_Heatmaps/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### STING PATHWAY GENES HEATMAPS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "plot_genes = ['MB21D1','TMEM173','IRF3','RELA','RELB','IKBKB','NFKB1',\n",
    "              'STAT2','CXCL10','NFKB2','CCL5',\n",
    "              'TBK1','JAK1','JAK2','IRF9',\n",
    "              'TYK2','IRF7','IRF1','STAT1',\n",
    "              'PPARG','DDIT3','NUPR1','RAB3B','IGFBP4','LRRC8C',\n",
    "              'TCP11L2','MAFK','NRG1','F2R','KRT19','CTGF','ZFC3H1']\n",
    "\n",
    "print(len(plot_genes))\n",
    "\n",
    "# LOAD GENESETS\n",
    "path_to_genesets = DATA_PATH+'uveal_melanoma_MB_v3.csv'\n",
    "genesets = pd.read_csv(path_to_genesets,header='infer')\n",
    "\n",
    "subset_type = 'TUMOR'\n",
    "QUERY = INDF_HARBOUR_PRIMARY\n",
    "\n",
    "# HEATMAP OF INDIVIDUAL CELLS RANKED BY LUNGE EP\n",
    "rank_by = 'Castle 2'#\n",
    "genesets_include = [rank_by]\n",
    "genes = np.unique(genesets[genesets_include].values.ravel().tolist())\n",
    "genes = [gene for gene in genes if gene in list(set(QUERY.columns))]\n",
    "vals = QUERY[genes]\n",
    "QUERY['GEP2_RANK'] =np.nanmean(vals,axis=1)\n",
    "\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "heatmap_data = pd.DataFrame( data = zscore(QUERY.sort_values(by=['GEP2_RANK'])[plot_genes],axis=0), \n",
    "                   columns = plot_genes, index = QUERY.sort_values(by=['GEP2_RANK']).index)\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns\n",
    "\n",
    "# LINKAGE \n",
    "method = 'centroid' # average, single centroid/euclidean\n",
    "metric = 'euclidean' # cosine\n",
    "linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "col_linkage = deepcopy(linkage)\n",
    "cl = hc.leaves_list(col_linkage)\n",
    "mat = heatmap_data.iloc[:,cl]\n",
    "#mat = heatmap_data.copy()\n",
    "\n",
    "window = 20\n",
    "mat = mat.rolling(window, win_type='triang',center = True).sum()\n",
    "half_window = int(window/2)\n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[ind] = mat.iloc[half_window]    \n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[-ind] = mat.iloc[-half_window]\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(5,10))\n",
    "\n",
    "# ADD MATRIX WITH LINEAGE NAMES\n",
    "axmatrix = fig.add_axes([0.12,0.1,0.5,0.7])\n",
    "colors = [(1,1,1), np.divide(tuple(hex('FF0000').rgb),255)]\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=CM_DIVERGING,vmin=-10,vmax=10)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "#axmatrix.set_xticklabels([]);\n",
    "xtick = plt.xticks(range(len(labels)), labels, rotation = 90, fontsize = 5,fontname='Arial')\n",
    "ytick = plt.yticks([])\n",
    "axmatrix.grid(False)\n",
    "plt.tick_params(size=0)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '/STING_PATHWAY_Heatmap_HARBOUR_PRIMARY_EXTENDED_CLUSTERED_centroid_euclidean'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/GEP2_Ranked_Heatmaps/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GEP PROGNOSTIFICATION OF INDIVIDUAL TUMOR CELLS - HARBOUR"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GEP SIGNATURE CLUSTERED HEATMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class_genes = GEP_markerfile_df.index.values.tolist()\n",
    "\n",
    "QUERY = INDF_HARBOUR\n",
    "\n",
    "# COLUMN INDEX AND COLORS\n",
    "genes = [gene for gene in class_genes if gene in QUERY.columns]#gene_array['Gene'].values\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "heatmap_data = pd.DataFrame(data = zscore(QUERY[genes].values,axis=0),columns = genes, index = QUERY.index)\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns\n",
    "\n",
    "# LINKAGE \n",
    "method = 'average' # average, single\n",
    "metric = 'euclidean' # cosine\n",
    "linkage = hc.linkage(heatmap_data, method=method, metric = metric)\n",
    "row_linkage = deepcopy(linkage)\n",
    "linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "col_linkage = deepcopy(linkage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# REORDER HEATMAP ACCORDING TO LINKAGE (OPTIONAL, STILL SLOW)\n",
    "r1 = hc.leaves_list(row_linkage)\n",
    "mat = heatmap_data.iloc[r1,c1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GEP_assignments_sorted = [METADATA_HARBOUR.loc\n",
    "                          [METADATA_HARBOUR.index.get_level_values('Cell ID') == x].\\\n",
    "                          Assignment.\\\n",
    "                          values[0] for x in mat.index.get_level_values('Cell ID').values]\n",
    "\n",
    "row_colors = pd.Series(GEP_assignments_sorted, index=GEP_assignments_sorted).map(lut_CASTLE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(4,10))\n",
    "plt.rcParams[\"axes.grid\"] = False\n",
    "\n",
    "# ADD ROW COLOR INDEX 1 (GEP CLASS)\n",
    "ax1 = fig.add_axes([0,0.1,0.05,0.6]) # [x0,y0,width,height]\n",
    "x = 0\n",
    "y = 0\n",
    "for c in row_colors:\n",
    "    pos = (x, y / len(row_colors))\n",
    "    ax1.add_patch(patches.Rectangle(pos, 1, 1 / len(row_colors), color=c))\n",
    "    if y >= len(row_colors)-1:\n",
    "        x += 1\n",
    "        y = 0\n",
    "    else:\n",
    "        y += 1\n",
    "plt.axis('off')\n",
    "\n",
    "# ADD ROW 2 COLOR BY SAMPLE TYPE\n",
    "sample_status = [QUERY.loc[QUERY.index.get_level_values('Cell ID') == x].index.get_level_values('Legend').\\\n",
    "                          values[0] for x in mat.index.get_level_values('Cell ID').values]\n",
    "\n",
    "lut_SAMPLE = dict({'primary': np.array([1, 1, 1]), 'metastatic': np.array([0, 0, 0])})\n",
    "\n",
    "row_colors2 = pd.Series(sample_status, index=sample_status).map(lut_SAMPLE)\n",
    "ax2 = fig.add_axes([0.95,0.1,0.05,0.6]) # [x0,y0,width,height]\n",
    "x = 0\n",
    "y = 0\n",
    "for c in row_colors2:\n",
    "    pos = (x, y / len(row_colors2))\n",
    "    ax2.add_patch(patches.Rectangle(pos, 1, 1 / len(row_colors2), color=c))\n",
    "    if y >= len(row_colors2)-1:\n",
    "        x += 1\n",
    "        y = 0\n",
    "    else:\n",
    "        y += 1\n",
    "plt.axis('off')\n",
    "\n",
    "# ADD MATRIX WITH GENE NAMES\n",
    "axmatrix = fig.add_axes([0.05,0.1,0.9,0.6])\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=plt.cm.RdBu_r,vmin=-1,vmax=1)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "xtick = plt.xticks(range(len(labels)), labels, rotation = 90, fontsize = 14)\n",
    "\n",
    "# ADD COLORBAR\n",
    "axcolor = fig.add_axes([1.1,0.1,0.1,0.1])\n",
    "cbar = plt.colorbar(im, cax=axcolor)\n",
    "#cbar.ax.get_yaxis().set_ticks([])\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'GEP_HARBOUR_BayesianGMM'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### PAIRWISE CORRELATION BETWEEN MONOSOMY 3 SIGNATURE AND GEP 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# VIOLIN PLOTS\n",
    "subset_type = 'HARBOUR'\n",
    "datatype = 'INDF_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "\n",
    "meta = 'Assignment'\n",
    "\n",
    "gene1 = 'Castle 2'\n",
    "signature_genes = genesets[gene1].values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "G1 = (np.nansum(QUERY[detected_genes],axis=1))#*complexity\n",
    "\n",
    "gene2 = 'Monosomy 3 Up'\n",
    "signature_genes = genesets[gene2].values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "G2 = (np.nansum(QUERY[detected_genes],axis=1))#*complexity\n",
    "\n",
    "# ROW INDEX\n",
    "meta = 'Assignment'\n",
    "groupby_type = meta # 'Class', 'Meta-Source', 'Legend'\n",
    "FLATUI = ['0000FF','FF0000']#FLATUI_CLASS, 'FLATUI_SOURCE, FLATUI_SAMPLES\n",
    "\n",
    "violin_data = pd.DataFrame({\n",
    "                            meta:list(META[meta]),\n",
    "                            gene1: G1,\n",
    "                            gene2: G2,\n",
    "                           })\n",
    "\n",
    "ind = violin_data[groupby_type]\n",
    "\n",
    "# CONVERT HEX TO RGB (FLATUI_CLASS)\n",
    "colors = np.zeros((len(FLATUI),3))\n",
    "for ii,hexcolor in enumerate(FLATUI):\n",
    "    colors[ii,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "# Palatte for Class METADATA\n",
    "cix = (np.linspace(0,shape(colors)[0],len(np.unique(ind)))).astype(int)\n",
    "if cix[len(cix)-1]==shape(colors)[0]:\n",
    "    cix[len(cix)-1]=shape(colors)[0]-1\n",
    "lut = dict(zip(np.unique(ind), colors[cix,:]))\n",
    "dot_colors = pd.Series(ind).map(lut)\n",
    "metacell_colors = [rgb2hex(int(color[0]*255), int(color[1]*255), int(color[2]*255)) for color in dot_colors]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# DISTRIBUTION OF SOX2/SOX9\n",
    "x = gene1\n",
    "y = gene2\n",
    "\n",
    "plt.figure(figsize = (20,20))\n",
    "sns.despine()\n",
    "sns.set(rc={'axes.facecolor':'white', 'figure.facecolor':'white'})\n",
    "\n",
    "# JOINTPLOT\n",
    "g = sns.jointplot(x,y, data=violin_data.fillna(0), kind=\"reg\", stat_func=None,ratio=2, color='k', size=10)\n",
    "g.ax_joint.cla() # or g.ax_joint.collections[0].set_visible(False), as per mwaskom's comment\n",
    "\n",
    "# REPLOT SCATTER WITH LINEAGE COLORED\n",
    "plt.sca(g.ax_joint)\n",
    "plt.scatter(violin_data[x], violin_data[y], c=dot_colors)\n",
    "\n",
    "# ADD REGRESSION\n",
    "xx = violin_data[x]\n",
    "yy = violin_data[y]\n",
    "f = lambda x, *p: polyval(p, x)\n",
    "p, cov = curve_fit(f, xx, yy, [1, 1])\n",
    "\n",
    "# simulated draws from the probability density function of the regression\n",
    "xi = linspace(np.min(xx), np.max(xx), 100)\n",
    "ps = np.random.multivariate_normal(p, cov, 10000)\n",
    "ysample = np.asarray([f(xi, *pi) for pi in ps])\n",
    "lower = percentile(ysample, 10, axis=0)\n",
    "upper = percentile(ysample, 90, axis=0)\n",
    "\n",
    "# regression estimate line\n",
    "y_fit = poly1d(p)(xi)\n",
    "\n",
    "# PLOT REGRESSION LINE\n",
    "plt.fill_between(xi, lower, upper, facecolor='k', alpha=0.5)\n",
    "plt.plot(xi, y_fit, 'k-')\n",
    "#plt.xlim((0,5))\n",
    "#plt.ylim((0,5))\n",
    "\n",
    "# ADD AXIS LABELS\n",
    "g.set_axis_labels(x, y, fontsize=10,fontname = 'Arial')\n",
    "\n",
    "# COMPUTE STATISTICS\n",
    "(r, p) = stats.pearsonr(violin_data[x].fillna(0), violin_data[y].fillna(0))\n",
    "print('{} vs {}, R: {}, p: {}'.format(x,y,r,p))\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'M3_GEP2_Correlation_BGMM_HARBOUR'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "plt.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### EVALUATE GENE EXPRESSION PER GEP CLASS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_type = 'box'\n",
    "title = 'BAP1'\n",
    "meta = 'Assignment'\n",
    "datatype = 'INDF_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "\n",
    "fig = plt.figure(figsize = (1,5))\n",
    "ax = plt.gca()\n",
    "\n",
    "scale_type = 'count'\n",
    "palette = dict(zip(['Castle 1','Castle 2'],['#0000FF','#FF0000'])) \n",
    "\n",
    "\n",
    "genes = [title]\n",
    "genes = np.unique([x for x in genes if str(x) != 'NAN'])\n",
    "detected_genes = list(set(genes).intersection(set(QUERY.columns)))\n",
    "vals = QUERY[detected_genes]\n",
    "SCORE = np.nansum(vals,axis=1)\n",
    "\n",
    "# Format data structure for violin plot\n",
    "violin_data = []\n",
    "for ind,v in enumerate(SCORE):\n",
    "    violin_data.append({'gene': title, 'Z-normalized Expression': v,\n",
    "                        meta:META[meta].values[ind]}) \n",
    "violin_data = pd.DataFrame(violin_data)  \n",
    "\n",
    "# BOXPLOT GENE EXPRESSION\n",
    "if plot_type == 'box':\n",
    "    g = sns.boxplot(x=\"gene\", y=\"Z-normalized Expression\", \n",
    "                    hue=meta,data=violin_data, palette=palette,notch = True, \n",
    "                    hue_order = ['Castle 1','Castle 2'], \n",
    "                    fliersize = 4, showmeans=False,linewidth = 1, ax = ax) #order = labels, \n",
    "    g.set_ylabel(\"{}\".format(datatype),fontsize=10)\n",
    "    g.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "    g.tick_params(labelsize=10)\n",
    "    sns.despine()\n",
    "    ax.set(ylim=(0, SCORE.max()*0.95))\n",
    "    g.legend(loc='upper right',prop={'size':6},bbox_to_anchor=(2.0, 0.95),fancybox=True) \n",
    "\n",
    "\n",
    "elif plot_type == 'violin':\n",
    "    # VIOLIN GENE EXPRESSION\n",
    "    g = sns.violinplot(x=\"gene\", y=\"Z-normalized Expression\", \n",
    "                       hue=meta,data=violin_data, palette=palette,notch = True, \n",
    "                    hue_order = ['Castle 1','Castle 2'],fliersize = 4, showmeans=True,linewidth = 1) #order = labels, \n",
    "\n",
    "    g.set_ylabel(\"{}\".format(datatype),fontsize=10)\n",
    "    g.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "    ax.set(ylim=(0, SCORE.max()*0.95))\n",
    "    g.tick_params(labelsize=10)\n",
    "    sns.despine()\n",
    "\n",
    "# COMPARE DISTRIBUTIONS\n",
    "CLASS1 = violin_data.loc[violin_data[meta].isin(['Castle 1'])]['Z-normalized Expression'].values\n",
    "CLASS2 = violin_data.loc[violin_data[meta].isin(['Castle 2'])]['Z-normalized Expression'].values\n",
    "print(title + ' CLASS1 vs. CLASS2')\n",
    "print(stats.mannwhitneyu(CLASS1,CLASS2))\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'BAP1_Expression_by_GEP_BGMM_HARBOUR'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "plt.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GEP PROGNOSTIFICATION OF INDIVIDUAL TUMOR CELLS - HARBOUR PRIMARY"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GEP SIGNATURE CLUSTERED HEATMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class_genes = GEP_markerfile_df.index.values.tolist()\n",
    "\n",
    "QUERY = INDF_HARBOUR_PRIMARY\n",
    "\n",
    "# COLUMN INDEX AND COLORS\n",
    "genes = [gene for gene in class_genes if gene in QUERY.columns]#gene_array['Gene'].values\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "heatmap_data = pd.DataFrame(data = zscore(QUERY[genes].values,axis=0),columns = genes, index = QUERY.index)\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns\n",
    "\n",
    "# LINKAGE \n",
    "method = 'centroid' # average, single\n",
    "metric = 'euclidean' # cosine\n",
    "linkage = hc.linkage(heatmap_data, method=method, metric = metric)\n",
    "row_linkage = deepcopy(linkage)\n",
    "linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "col_linkage = deepcopy(linkage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# REORDER HEATMAP ACCORDING TO LINKAGE (OPTIONAL, STILL SLOW)\n",
    "r1 = hc.leaves_list(row_linkage)\n",
    "mat = heatmap_data.iloc[r1,c1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GEP_assignments_sorted = [METADATA_HARBOUR.loc[METADATA_HARBOUR.index.get_level_values('Legend') == 'primary'].loc\n",
    "                          [METADATA_HARBOUR.loc[METADATA_HARBOUR.index.get_level_values('Legend') == 'primary'].\\\n",
    "                           index.get_level_values('Cell ID') == x].Assignment.\\\n",
    "                          values[0] for x in mat.index.get_level_values('Cell ID').values]\n",
    "\n",
    "row_colors = pd.Series(GEP_assignments_sorted, index=GEP_assignments_sorted).map(lut_CASTLE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(4,10))\n",
    "plt.rcParams[\"axes.grid\"] = False\n",
    "\n",
    "# ADD ROW COLOR INDEX 1 (PHENOGRAPH CLASS)\n",
    "ax1 = fig.add_axes([0,0.1,0.05,0.6]) # [x0,y0,width,height]\n",
    "x = 0\n",
    "y = 0\n",
    "for c in row_colors:\n",
    "    pos = (x, y / len(row_colors))\n",
    "    ax1.add_patch(patches.Rectangle(pos, 1, 1 / len(row_colors), color=c))\n",
    "    if y >= len(row_colors)-1:\n",
    "        x += 1\n",
    "        y = 0\n",
    "    else:\n",
    "        y += 1\n",
    "plt.axis('off')\n",
    "\n",
    "# ADD MATRIX WITH GENE NAMES\n",
    "axmatrix = fig.add_axes([0.05,0.1,0.9,0.6], frameon=False)\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=plt.cm.RdBu_r,vmin=-1,vmax=1)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "xtick = plt.xticks(range(len(labels)), labels, rotation = 90, fontsize = 14)\n",
    "plt.tick_params(size=0)\n",
    "\n",
    "# ADD COLORBAR\n",
    "axcolor = fig.add_axes([1.1,0.1,0.1,0.1])\n",
    "cbar = plt.colorbar(im, cax=axcolor)\n",
    "#cbar.ax.get_yaxis().set_ticks([])\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'GEP_HARBOUR_PRIMARY_BayesianGMM'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### PAIRWISE CORRELATION BETWEEN MONOSOMY 3 AND GEP 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "METADATA_HARBOUR_PRIMARY['Assignment'] = [METADATA_HARBOUR.loc[\n",
    "                                          METADATA_HARBOUR.index.get_level_values('Legend') == 'primary'].loc[\n",
    "                                          METADATA_HARBOUR.loc[\n",
    "                                          METADATA_HARBOUR.index.get_level_values('Legend') == 'primary'].\\\n",
    "                                          index.get_level_values('Cell ID') == x].Assignment.values[0] \n",
    "                                          for x in METADATA_HARBOUR_PRIMARY.index.get_level_values('Cell ID').values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# VIOLIN PLOTS\n",
    "subset_type = 'HARBOUR_PRIMARY'\n",
    "datatype = 'INDF_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "\n",
    "meta = 'Assignment'\n",
    "\n",
    "gene1 = 'Castle 2'\n",
    "signature_genes = genesets[gene1].values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "G1 = (np.nansum(QUERY[detected_genes],axis=1))#*complexity\n",
    "\n",
    "gene2 = 'Monosomy 3 Up'\n",
    "signature_genes = genesets[gene2].values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "G2 = (np.nansum(QUERY[detected_genes],axis=1))#*complexity\n",
    "\n",
    "# ROW INDEX\n",
    "meta = 'Assignment'\n",
    "groupby_type = meta # 'Class', 'Meta-Source', 'Legend'\n",
    "FLATUI = ['0000FF','FF0000']#FLATUI_CLASS, 'FLATUI_SOURCE, FLATUI_SAMPLES\n",
    "\n",
    "violin_data = pd.DataFrame({\n",
    "                            meta:list(META[meta]),\n",
    "                            gene1: G1,\n",
    "                            gene2: G2,\n",
    "                           })\n",
    "\n",
    "ind = violin_data[groupby_type]\n",
    "\n",
    "# CONVERT HEX TO RGB (FLATUI_CLASS)\n",
    "colors = np.zeros((len(FLATUI),3))\n",
    "for ii,hexcolor in enumerate(FLATUI):\n",
    "    colors[ii,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "# Palatte for Class METADATA\n",
    "cix = (np.linspace(0,shape(colors)[0],len(np.unique(ind)))).astype(int)\n",
    "if cix[len(cix)-1]==shape(colors)[0]:\n",
    "    cix[len(cix)-1]=shape(colors)[0]-1\n",
    "lut = dict(zip(np.unique(ind), colors[cix,:]))\n",
    "dot_colors = pd.Series(ind).map(lut)\n",
    "metacell_colors = [rgb2hex(int(color[0]*255), int(color[1]*255), int(color[2]*255)) for color in dot_colors]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# DISTRIBUTION OF SOX2/SOX9\n",
    "x = gene1\n",
    "y = gene2\n",
    "\n",
    "plt.figure(figsize = (20,20))\n",
    "#sns.set(rc={'axes.facecolor':'white', 'figure.facecolor':'white'})\n",
    "sns.set_style('ticks')\n",
    "\n",
    "# JOINTPLOT\n",
    "g = sns.jointplot(x,y, data=violin_data.fillna(0), kind=\"reg\", stat_func=None,ratio=2, color='k', size=10)\n",
    "g.ax_joint.cla() # or g.ax_joint.collections[0].set_visible(False), as per mwaskom's comment\n",
    "\n",
    "# REPLOT SCATTER WITH LINEAGE COLORED\n",
    "plt.sca(g.ax_joint)\n",
    "plt.scatter(violin_data[x], violin_data[y], c=dot_colors)\n",
    "\n",
    "# ADD REGRESSION\n",
    "xx = violin_data[x]\n",
    "yy = violin_data[y]\n",
    "f = lambda x, *p: polyval(p, x)\n",
    "p, cov = curve_fit(f, xx, yy, [1, 1])\n",
    "\n",
    "# simulated draws from the probability density function of the regression\n",
    "xi = linspace(np.min(xx), np.max(xx), 100)\n",
    "ps = np.random.multivariate_normal(p, cov, 10000)\n",
    "ysample = np.asarray([f(xi, *pi) for pi in ps])\n",
    "lower = percentile(ysample, 10, axis=0)\n",
    "upper = percentile(ysample, 90, axis=0)\n",
    "\n",
    "# regression estimate line\n",
    "y_fit = poly1d(p)(xi)\n",
    "\n",
    "# PLOT REGRESSION LINE\n",
    "plt.fill_between(xi, lower, upper, facecolor='k', alpha=0.5)\n",
    "plt.plot(xi, y_fit, 'k-')\n",
    "#plt.xlim((0,5))\n",
    "#plt.ylim((0,5))\n",
    "plt.tick_params(size=0)\n",
    "plt.yticks(ticks=[0, 10, 20, 30, 40, 50, 60], \n",
    "           labels=['0', '10', '20', '30', '40', '50', '60'], size=20)\n",
    "plt.xticks(ticks=[0, 2, 4, 6, 8], labels=['0', '2', '4', '6', '8'], size=20)\n",
    "\n",
    "# ADD AXIS LABELS\n",
    "g.set_axis_labels(x, y, fontsize=10,fontname = 'Arial')\n",
    "\n",
    "# COMPUTE STATISTICS\n",
    "(r, p) = stats.pearsonr(violin_data[x].fillna(0), violin_data[y].fillna(0))\n",
    "print('{} vs {}, R: {}, p: {}'.format(x,y,r,p))\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'M3_GEP2_Correlation_BGMM_HARBOUR_PRIMARY'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "plt.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### EVALUATE GENE EXPRESSION PER GEP CLASS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_type = 'box'\n",
    "title = 'BAP1'\n",
    "meta = 'Assignment'\n",
    "datatype = 'INDF_{}'.format(subset_type)\n",
    "exec('QUERY = {}'.format(datatype))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "\n",
    "fig = plt.figure(figsize = (1,5))\n",
    "ax = plt.gca()\n",
    "\n",
    "scale_type = 'count'\n",
    "palette = dict(zip(['Castle 1','Castle 2'],['#0000FF','#FF0000'])) \n",
    "\n",
    "\n",
    "genes = [title]\n",
    "genes = np.unique([x for x in genes if str(x) != 'NAN'])\n",
    "detected_genes = list(set(genes).intersection(set(QUERY.columns)))\n",
    "vals = QUERY[detected_genes]\n",
    "SCORE = np.nansum(vals,axis=1)\n",
    "\n",
    "# Format data structure for violin plot\n",
    "violin_data = []\n",
    "for ind,v in enumerate(SCORE):\n",
    "    violin_data.append({'gene': title, 'Z-normalized Expression': v,\n",
    "                        meta:META[meta].values[ind]}) \n",
    "violin_data = pd.DataFrame(violin_data)  \n",
    "\n",
    "# BOXPLOT GENE EXPRESSION\n",
    "if plot_type == 'box':\n",
    "    g = sns.boxplot(x=\"gene\", y=\"Z-normalized Expression\", \n",
    "                    hue=meta,data=violin_data, palette=palette,notch = True, \n",
    "                    hue_order = ['Castle 1','Castle 2'], \n",
    "                    fliersize = 4, showmeans=False,linewidth = 1, ax = ax) #order = labels, \n",
    "    g.set_ylabel(\"BAP1 Imputed Expression\", fontsize=14)\n",
    "    g.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "    g.tick_params(labelsize=11)\n",
    "    sns.despine()\n",
    "    ax.set(ylim=(0, SCORE.max()*0.95))\n",
    "    g.legend(loc='upper right',prop={'size':6},bbox_to_anchor=(2.0, 0.95),fancybox=True) \n",
    "\n",
    "\n",
    "elif plot_type == 'violin':\n",
    "    # VIOLIN GENE EXPRESSION\n",
    "    g = sns.violinplot(x=\"gene\", y=\"Z-normalized Expression\", \n",
    "                       hue=meta,data=violin_data, palette=palette,notch = True, \n",
    "                    hue_order = ['Castle 1','Castle 2'],fliersize = 4, showmeans=True,linewidth = 1) #order = labels, \n",
    "\n",
    "    g.set_ylabel(\"BAP1 Imputed Expression\", fontsize=14)\n",
    "    g.set_xlabel(\" \",fontsize=10,rotation = 90)\n",
    "    ax.set(ylim=(0, SCORE.max()*0.95))\n",
    "    g.tick_params(labelsize=11)\n",
    "    sns.despine()\n",
    "    \n",
    "plt.tick_params(size=0)\n",
    "\n",
    "# COMPARE DISTRIBUTIONS\n",
    "CLASS1 = violin_data.loc[violin_data[meta].isin(['Castle 1'])]['Z-normalized Expression'].values\n",
    "CLASS2 = violin_data.loc[violin_data[meta].isin(['Castle 2'])]['Z-normalized Expression'].values\n",
    "print(title + ' CLASS1 vs. CLASS2')\n",
    "print(stats.mannwhitneyu(CLASS1,CLASS2))\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'BAP1_Expression_by_GEP_BGMM_HARBOUR_PRIMARY'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "plt.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### PHENOTYPIC VOLUME BY PATIENT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_type = 'HARBOUR_PRIMARY'\n",
    "meta ='Assignment'\n",
    "\n",
    "exec('QUERY = h5_data_harbour.load(\\'/NDF_TUMOR_PRIMARY\\')')\n",
    "exec('METADATA_HARBOUR = h5_data_harbour.load(\\'/METADATA_HARBOUR\\')')\n",
    "META = METADATA_HARBOUR.loc[METADATA_HARBOUR.index.map(lambda x: x[1] == 'primary')]\n",
    "new_index = pd.MultiIndex.from_tuples(list(zip(META.index.get_level_values('Sample ID'), \n",
    "                                               META.index.get_level_values('Legend'),\n",
    "                                               META.index.get_level_values('Patient'),\n",
    "                                               META.index.get_level_values('Cell ID'),\n",
    "                                               META[meta])), \n",
    "                                  names=['Sample ID','Legend', 'Patient','Cell ID',meta])\n",
    "QUERY = pd.DataFrame(data = QUERY.values, index = new_index, columns = QUERY.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SPECIFY SUBSET\n",
    "fig_dpi = 400\n",
    "\n",
    "# FIT BINOMIAL DISTRIBUTION AND FILTER BASED ON MEAN/STD OF SECOND\n",
    "plt.figure(figsize = (10,3))\n",
    "gs1 = gridspec.GridSpec(1, 3)\n",
    "gs1.update(wspace=0.7, hspace=0.7) # set the spacing between axes. \n",
    "\n",
    "# (4) PLOT LOG NUMBER OF CELLS CONTRIBUTING TO EACH GENE\n",
    "num_cells_per_gene = np.log(np.sum(QUERY.values > 0,axis=0))\n",
    "num_cells_per_gene[(np.isinf(num_cells_per_gene)) | (np.isnan(num_cells_per_gene))] = 0\n",
    "rmv_genes1 = np.where(num_cells_per_gene<=0)[0] # GENES MUST BE DETECTED IN AT LEAST 10 CELLS\n",
    "\n",
    "ax = plt.subplot(gs1[0])\n",
    "bins = np.linspace(num_cells_per_gene.min(), num_cells_per_gene.max()*0.95, 20)\n",
    "plt.hist(num_cells_per_gene, bins, alpha=0.5, label='keep')\n",
    "\n",
    "if rmv_genes1.any():\n",
    "    plt.hist(num_cells_per_gene[rmv_genes1], bins, alpha=1, label='remove')\n",
    "    \n",
    "#ax.set_axis_bgcolor('white')\n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Frequency')\n",
    "plt.xlabel('Gene Filter 1: Remove Genes Singletons \\n(Log. # Expressing Cells)')\n",
    "plt.grid(True)\n",
    "sns.despine()\n",
    "\n",
    "# (2) PLOT LOG COUNTS PER GENE - REMOVE LOW ABUNDANCE GENES\n",
    "log_counts_per_gene = np.log(np.log(np.sum(QUERY.values,axis=0)))\n",
    "log_counts_per_gene[(np.isinf(log_counts_per_gene)) | (np.isnan(log_counts_per_gene))] = 0\n",
    "data = log_counts_per_gene\n",
    "\n",
    "ax = plt.subplot(gs1[1])\n",
    "bins = np.linspace(data.min(), data.max()*0.95, 100)\n",
    "y,x,_=hist(data,bins,alpha=.3,label='data')\n",
    "\n",
    "x=(x[1:]+x[:-1])/2 \n",
    "expected=(0,.1,3500,1.75,.2,500)\n",
    "params,cov=curve_fit(bimodal,x,y,expected)\n",
    "sigma=sqrt(diag(cov))\n",
    "plot(x,bimodal(x,*params),color='red',lw=3,label='model')\n",
    "\n",
    "mu1 = params[0]\n",
    "std1 = params[1]\n",
    "mu2 = params[3]\n",
    "std2 = params[4]\n",
    "rmv_genes_neg = np.where(data<mu2-3*std2)[0]\n",
    "rmv_genes2 = np.sort(list(set(list(rmv_genes_neg))))\n",
    "\n",
    "if rmv_genes2.any():\n",
    "    plt.hist(data[rmv_genes2], bins, alpha=1, label='remove')\n",
    "    \n",
    "#ax.set_axis_bgcolor('white')\n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Frequency')\n",
    "plt.xlabel('Gene Filter 2: \\nLog-Log Counts/Gene')\n",
    "plt.grid(True)\n",
    "sns.despine()\n",
    "\n",
    "# (3) PLOT STDEV PER GENE\n",
    "pseudocount = 0.001\n",
    "std_per_gene = np.log10(QUERY.std(axis=0, skipna=False)+pseudocount)\n",
    "data = std_per_gene\n",
    "\n",
    "ax = plt.subplot(gs1[2])\n",
    "bins = np.linspace(data.min(), data.max()*0.95, 100)\n",
    "y,x,_=hist(data,bins,alpha=.3,label='data')\n",
    "\n",
    "x=(x[1:]+x[:-1])/2 \n",
    "expected=(0,-3,3000,-1,.2,500)\n",
    "params,cov=curve_fit(bimodal,x,y,expected,maxfev=3000)\n",
    "sigma=sqrt(diag(cov))\n",
    "plot(x,bimodal(x,*params),color='red',lw=3,label='model')\n",
    "\n",
    "mu1 = params[0]\n",
    "std1 = params[1]\n",
    "mu2 = params[3]\n",
    "std2 = params[4]\n",
    "rmv_genes_neg = np.where(data<mu2-3*abs(std2))[0]\n",
    "rmv_genes3 = np.sort(list(set(list(rmv_genes_neg))))\n",
    "\n",
    "if rmv_genes3.any():\n",
    "    plt.hist(data[rmv_genes3], bins, alpha=1, label='remove')\n",
    "    \n",
    "#ax.set_axis_bgcolor('white')\n",
    "plt.xticks(rotation=70)\n",
    "plt.ylabel('Frequency')\n",
    "plt.xlabel('Gene Filter 3: \\nSDEV/Gene')\n",
    "plt.grid(True)\n",
    "sns.despine()\n",
    "\n",
    "# Add (abbreviated) legend bottom left\n",
    "L = plt.legend(loc='upper right',prop={'size':12},bbox_to_anchor=(1.6, 1.05),fancybox=True) \n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '_filter_genes'\n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label + '.png'\n",
    "#plt.savefig(fn, dpi=fig_dpi)\n",
    "print(fn)\n",
    "\n",
    "# EVALUATE NUMBER OF CELLS/GENES REMOVED\n",
    "print('Count Gene Filter 1: {}'.format(len(rmv_genes1)))\n",
    "print('Count Gene Filter 2: {}'.format(len(rmv_genes2)))\n",
    "print('Count Gene Filter 3: {}'.format(len(rmv_genes3)))\n",
    "\n",
    "# REMOVE SELECTED OUTLIER GENES(HIGHLIGHTED IN GREEN)\n",
    "CUT_DF = deepcopy(QUERY)\n",
    "rmv_genes = np.sort(list(set(list(rmv_genes1) + list(rmv_genes2)+list(rmv_genes3))))\n",
    "if rmv_genes.any():\n",
    "    CUT_DF = CUT_DF.drop(CUT_DF.columns[rmv_genes],axis=1)\n",
    "    print(CUT_DF.shape)\n",
    "    \n",
    "# Remove empty genes if they exist\n",
    "drop_genes = np.where(CUT_DF.sum(axis=0)==0)[0]\n",
    "CUT_DF = CUT_DF.drop(CUT_DF.columns[drop_genes],axis=1)\n",
    "print(CUT_DF.shape)\n",
    "\n",
    "# GENES EXCLUDING LOW ABUNDANCE GENES\n",
    "variable_goi = list(CUT_DF.columns)\n",
    "print(len(variable_goi))"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "n = 900\n",
    "reps = np.arange(20)\n",
    "gene_subset = variable_goi"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# GENE-GENE COVARIANCE WITHIN PATIENT 1\n",
    "master_groupby = 'Patient'\n",
    "class_selection = ['UMM059']\n",
    "ix = [ind for ind,name in enumerate(QUERY.index.names) if name==master_groupby][0]\n",
    "logvolume_castle1 = np.zeros(len(reps))\n",
    "\n",
    "for rep in reps:\n",
    "    \n",
    "    SUBSET = QUERY.loc[QUERY.index.map(lambda x: x[ix] in class_selection)][gene_subset].sample(n=n,replace=True)\n",
    "    GENE_COV = SUBSET.cov(min_periods=None)\n",
    "    eigenvalues = np.linalg.eigvals(GENE_COV)\n",
    "    logvolume = np.float(np.sum(0.5*np.log10(eigenvalues[eigenvalues>0]**2))/len(eigenvalues))\n",
    "    logvolume_castle1[rep] = logvolume\n",
    "    print(rep)\n",
    "    \n",
    "# GENE-GENE COVARIANCE WITHIN PATIENT 2\n",
    "master_groupby = 'Patient'\n",
    "class_selection = ['UMM061']\n",
    "ix = [ind for ind,name in enumerate(QUERY.index.names) if name==master_groupby][0]\n",
    "logvolume_castle2 = np.zeros(len(reps))\n",
    "\n",
    "for rep in reps:\n",
    "    \n",
    "    SUBSET = QUERY.loc[QUERY.index.map(lambda x: x[ix] in class_selection)][gene_subset].sample(n=n,replace=True)\n",
    "    GENE_COV = SUBSET.cov(min_periods=None)\n",
    "    eigenvalues = np.linalg.eigvals(GENE_COV)\n",
    "    logvolume = np.float(np.sum(0.5*np.log10(eigenvalues[eigenvalues>0]**2))/len(eigenvalues))\n",
    "    logvolume_castle2[rep] = logvolume\n",
    "    print(rep)\n",
    "    \n",
    "# GENE-GENE COVARIANCE WITHIN PATIENT 3\n",
    "master_groupby = 'Patient'\n",
    "class_selection = ['UMM062']\n",
    "ix = [ind for ind,name in enumerate(QUERY.index.names) if name==master_groupby][0]\n",
    "logvolume_castle3 = np.zeros(len(reps))\n",
    "\n",
    "for rep in reps:\n",
    "    \n",
    "    SUBSET = QUERY.loc[QUERY.index.map(lambda x: x[ix] in class_selection)][gene_subset].sample(n=n,replace=True)\n",
    "    GENE_COV = SUBSET.cov(min_periods=None)\n",
    "    eigenvalues = np.linalg.eigvals(GENE_COV)\n",
    "    logvolume = np.float(np.sum(0.5*np.log10(eigenvalues[eigenvalues>0]**2))/len(eigenvalues))\n",
    "    logvolume_castle3[rep] = logvolume\n",
    "    print(rep)\n",
    "    \n",
    "# GENE-GENE COVARIANCE WITHIN PATIENT 4\n",
    "master_groupby = 'Patient'\n",
    "class_selection = ['UMM063']\n",
    "ix = [ind for ind,name in enumerate(QUERY.index.names) if name==master_groupby][0]\n",
    "logvolume_castle4 = np.zeros(len(reps))\n",
    "\n",
    "for rep in reps:\n",
    "    \n",
    "    SUBSET = QUERY.loc[QUERY.index.map(lambda x: x[ix] in class_selection)][gene_subset].sample(n=n,replace=True)\n",
    "    GENE_COV = SUBSET.cov(min_periods=None)\n",
    "    eigenvalues = np.linalg.eigvals(GENE_COV)\n",
    "    logvolume = np.float(np.sum(0.5*np.log10(eigenvalues[eigenvalues>0]**2))/len(eigenvalues))\n",
    "    logvolume_castle4[rep] = logvolume\n",
    "    print(rep)\n",
    "    \n",
    "# GENE-GENE COVARIANCE WITHIN PATIENT 5\n",
    "master_groupby = 'Patient'\n",
    "class_selection = ['UMM064']\n",
    "ix = [ind for ind,name in enumerate(QUERY.index.names) if name==master_groupby][0]\n",
    "logvolume_castle5 = np.zeros(len(reps))\n",
    "\n",
    "for rep in reps:\n",
    "    \n",
    "    SUBSET = QUERY.loc[QUERY.index.map(lambda x: x[ix] in class_selection)][gene_subset].sample(n=n,replace=True)\n",
    "    GENE_COV = SUBSET.cov(min_periods=None)\n",
    "    eigenvalues = np.linalg.eigvals(GENE_COV)\n",
    "    logvolume = np.float(np.sum(0.5*np.log10(eigenvalues[eigenvalues>0]**2))/len(eigenvalues))\n",
    "    logvolume_castle5[rep] = logvolume\n",
    "    print(rep)\n",
    "    \n",
    "# GENE-GENE COVARIANCE WITHIN PATIENT 6\n",
    "master_groupby = 'Patient'\n",
    "class_selection = ['UMM065']\n",
    "ix = [ind for ind,name in enumerate(QUERY.index.names) if name==master_groupby][0]\n",
    "logvolume_castle6 = np.zeros(len(reps))\n",
    "\n",
    "for rep in reps:\n",
    "    \n",
    "    SUBSET = QUERY.loc[QUERY.index.map(lambda x: x[ix] in class_selection)][gene_subset].sample(n=n,replace=True)\n",
    "    GENE_COV = SUBSET.cov(min_periods=None)\n",
    "    eigenvalues = np.linalg.eigvals(GENE_COV)\n",
    "    logvolume = np.float(np.sum(0.5*np.log10(eigenvalues[eigenvalues>0]**2))/len(eigenvalues))\n",
    "    logvolume_castle6[rep] = logvolume\n",
    "    print(rep)\n",
    "    \n",
    "# GENE-GENE COVARIANCE WITHIN PATIENT 6\n",
    "master_groupby = 'Patient'\n",
    "class_selection = ['UMM066']\n",
    "ix = [ind for ind,name in enumerate(QUERY.index.names) if name==master_groupby][0]\n",
    "logvolume_castle7 = np.zeros(len(reps))\n",
    "\n",
    "for rep in reps:\n",
    "    \n",
    "    SUBSET = QUERY.loc[QUERY.index.map(lambda x: x[ix] in class_selection)][gene_subset].sample(n=n,replace=True)\n",
    "    GENE_COV = SUBSET.cov(min_periods=None)\n",
    "    eigenvalues = np.linalg.eigvals(GENE_COV)\n",
    "    logvolume = np.float(np.sum(0.5*np.log10(eigenvalues[eigenvalues>0]**2))/len(eigenvalues))\n",
    "    logvolume_castle7[rep] = logvolume\n",
    "    print(rep)\n",
    "    \n",
    "# GENE-GENE COVARIANCE WITHIN PATIENT 6\n",
    "master_groupby = 'Patient'\n",
    "class_selection = ['UMM069']\n",
    "ix = [ind for ind,name in enumerate(QUERY.index.names) if name==master_groupby][0]\n",
    "logvolume_castle8 = np.zeros(len(reps))\n",
    "\n",
    "for rep in reps:\n",
    "    \n",
    "    SUBSET = QUERY.loc[QUERY.index.map(lambda x: x[ix] in class_selection)][gene_subset].sample(n=n,replace=True)\n",
    "    GENE_COV = SUBSET.cov(min_periods=None)\n",
    "    eigenvalues = np.linalg.eigvals(GENE_COV)\n",
    "    logvolume = np.float(np.sum(0.5*np.log10(eigenvalues[eigenvalues>0]**2))/len(eigenvalues))\n",
    "    logvolume_castle8[rep] = logvolume\n",
    "    print(rep)\n",
    "\n",
    "# WRITE RESULTS TO DATAFRAME\n",
    "boxplot_data_goi = pd.DataFrame()\n",
    "boxplot_data_goi['Patient'] = ['UMM059']*len(logvolume_castle1)+['UMM061']*len(logvolume_castle2)+\\\n",
    "                                 ['UMM062']*len(logvolume_castle3)+['UMM063']*len(logvolume_castle4)+\\\n",
    "                                 ['UMM064']*len(logvolume_castle5)+['UMM065']*len(logvolume_castle6)+\\\n",
    "                                 ['UMM066']*len(logvolume_castle7)+['UMM069']*len(logvolume_castle8)\n",
    "boxplot_data_goi['Log Phenotypic Volume'] = np.concatenate([logvolume_castle1,logvolume_castle2,\n",
    "                                                            logvolume_castle3,logvolume_castle4,\n",
    "                                                            logvolume_castle5,logvolume_castle6,\n",
    "                                                            logvolume_castle7,logvolume_castle8],axis = 0)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# STATISTIC\n",
    "C1_harbour = boxplot_data_goi.loc[boxplot_data_goi['Patient'].isin(['UMM059'])]['Log Phenotypic Volume'].values\n",
    "C2_harbour = boxplot_data_goi.loc[boxplot_data_goi['Patient'].isin(['UMM061'])]['Log Phenotypic Volume'].values\n",
    "C3_harbour = boxplot_data_goi.loc[boxplot_data_goi['Patient'].isin(['UMM062'])]['Log Phenotypic Volume'].values\n",
    "C4_harbour = boxplot_data_goi.loc[boxplot_data_goi['Patient'].isin(['UMM063'])]['Log Phenotypic Volume'].values\n",
    "C5_harbour = boxplot_data_goi.loc[boxplot_data_goi['Patient'].isin(['UMM064'])]['Log Phenotypic Volume'].values\n",
    "C6_harbour = boxplot_data_goi.loc[boxplot_data_goi['Patient'].isin(['UMM065'])]['Log Phenotypic Volume'].values\n",
    "C7_harbour = boxplot_data_goi.loc[boxplot_data_goi['Patient'].isin(['UMM066'])]['Log Phenotypic Volume'].values\n",
    "C8_harbour = boxplot_data_goi.loc[boxplot_data_goi['Patient'].isin(['UMM069'])]['Log Phenotypic Volume'].values\n",
    "\n",
    "# Pickle results incase needed later:\n",
    "joblib.dump(C1_harbour,'/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM059')\n",
    "joblib.dump(C2_harbour,'/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM061')\n",
    "joblib.dump(C3_harbour,'/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM062')\n",
    "joblib.dump(C4_harbour,'/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM063')\n",
    "joblib.dump(C5_harbour,'/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM064')\n",
    "joblib.dump(C6_harbour,'/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM065')\n",
    "joblib.dump(C7_harbour,'/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM066')\n",
    "joblib.dump(C8_harbour,'/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM069')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C1_harbour = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM059')\n",
    "C2_harbour = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM061')\n",
    "C3_harbour = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM062')\n",
    "C4_harbour = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM063')\n",
    "C5_harbour = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM064')\n",
    "C6_harbour = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM065')\n",
    "C7_harbour = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM066')\n",
    "C8_harbour = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM069')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"white\")\n",
    "fig = plt.figure(figsize = (5,5))\n",
    "ax = plt.gca()\n",
    "\n",
    "#sns.kdeplot(IM, shade=True, color=\"#B0C4DE\")\n",
    "sns.kdeplot(C1_harbour, shade=True, color=\"#\"+FLATUI_PATIENT[0])\n",
    "sns.kdeplot(C2_harbour, shade=True, color=\"#\"+FLATUI_PATIENT[1])\n",
    "sns.kdeplot(C3_harbour, shade=True, color=\"#\"+FLATUI_PATIENT[2])\n",
    "sns.kdeplot(C4_harbour, shade=True, color=\"#\"+FLATUI_PATIENT[3])\n",
    "sns.kdeplot(C5_harbour, shade=True, color=\"#\"+FLATUI_PATIENT[4])\n",
    "sns.kdeplot(C6_harbour, shade=True, color=\"#\"+FLATUI_PATIENT[5])\n",
    "sns.kdeplot(C7_harbour, shade=True, color=\"#\"+FLATUI_PATIENT[5])\n",
    "sns.kdeplot(C8_harbour, shade=True, color=\"#\"+FLATUI_PATIENT[5])\n",
    "\n",
    "ax.set_ylabel(\"Frequency\",fontsize=10)\n",
    "ax.set_xlabel(\"Log Phenotypic Volume\",fontsize=10)\n",
    "ax.tick_params(labelsize=10)\n",
    "#g.set_xticklabels(ax.get_xticklabels(),rotation=90)\n",
    "sns.despine()\n",
    "plt.ylim((0,30))\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'Log_Phenotypic_Volume_BY_PATIENT_HARBOUR_PRIMARY'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label\n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PHENOTYPIC VOLUME INTRAPATIENT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Patient UMM062:\n",
    "QUERY_FULL = QUERY.copy()\n",
    "QUERY = QUERY_FULL.loc[QUERY_FULL.index.map(lambda x: x[2] == 'UMM062')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "QUERY.index.get_level_values('Assignment').value_counts()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "n = 80\n",
    "reps = np.arange(50)\n",
    "gene_subset = variable_goi"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# GENE-GENE COVARIANCE WITHIN PATIENT 1\n",
    "master_groupby = 'Assignment'\n",
    "class_selection = ['Castle 1']\n",
    "ix = [ind for ind,name in enumerate(QUERY.index.names) if name==master_groupby][0]\n",
    "logvolume_castle1 = np.zeros(len(reps))\n",
    "\n",
    "for rep in reps:\n",
    "    \n",
    "    SUBSET = QUERY.loc[QUERY.index.map(lambda x: x[ix] in class_selection)][gene_subset].sample(n=n,replace=True)\n",
    "    GENE_COV = SUBSET.cov(min_periods=None)\n",
    "    eigenvalues = np.linalg.eigvals(GENE_COV)\n",
    "    logvolume = np.float(np.sum(0.5*np.log10(eigenvalues[eigenvalues>0]**2))/len(eigenvalues))\n",
    "    logvolume_castle1[rep] = logvolume\n",
    "    print(rep)\n",
    "    \n",
    "# GENE-GENE COVARIANCE WITHIN PATIENT 2\n",
    "master_groupby = 'Assignment'\n",
    "class_selection = ['Castle 2']\n",
    "ix = [ind for ind,name in enumerate(QUERY.index.names) if name==master_groupby][0]\n",
    "logvolume_castle2 = np.zeros(len(reps))\n",
    "\n",
    "for rep in reps:\n",
    "    \n",
    "    SUBSET = QUERY.loc[QUERY.index.map(lambda x: x[ix] in class_selection)][gene_subset].sample(n=n,replace=True)\n",
    "    GENE_COV = SUBSET.cov(min_periods=None)\n",
    "    eigenvalues = np.linalg.eigvals(GENE_COV)\n",
    "    logvolume = np.float(np.sum(0.5*np.log10(eigenvalues[eigenvalues>0]**2))/len(eigenvalues))\n",
    "    logvolume_castle2[rep] = logvolume\n",
    "    print(rep)\n",
    "    \n",
    "# WRITE RESULTS TO DATAFRAME\n",
    "boxplot_data_goi_UMM062 = pd.DataFrame()\n",
    "boxplot_data_goi_UMM062['Assignment'] = ['Castle 1']*len(logvolume_castle1)+['Castle 2']*len(logvolume_castle2)\n",
    "boxplot_data_goi_UMM062['Log Phenotypic Volume'] = np.concatenate([logvolume_castle1,logvolume_castle2],axis = 0)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# STATISTIC\n",
    "C1_UMM062 = boxplot_data_goi_UMM062.loc[boxplot_data_goi_UMM062['Assignment'].\\\n",
    "                                    isin(['Castle 1'])]['Log Phenotypic Volume'].values\n",
    "C2_UMM062 = boxplot_data_goi_UMM062.loc[boxplot_data_goi_UMM062['Assignment'].\\\n",
    "                                    isin(['Castle 2'])]['Log Phenotypic Volume'].values\n",
    "\n",
    "joblib.dump(C1_UMM062,'/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM062_GEP1')\n",
    "joblib.dump(C2_UMM062,'/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM062_GEP2')\n",
    "\n",
    "stats.mannwhitneyu(C1_UMM062, C2_UMM062)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C1_UMM062 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM062_GEP1')\n",
    "C2_UMM062 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM062_GEP2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"white\")\n",
    "fig = plt.figure(figsize = (5,5))\n",
    "ax = plt.gca()\n",
    "\n",
    "#sns.kdeplot(IM, shade=True, color=\"#B0C4DE\")\n",
    "sns.kdeplot(C1_UMM062, shade=True, color=\"#0000FF\")\n",
    "sns.kdeplot(C2_UMM062, shade=True, color=\"#FF0000\")\n",
    "\n",
    "ax.set_ylabel(\"Frequency\",fontsize=10)\n",
    "ax.set_xlabel(\"Log Phenotypic Volume\",fontsize=10)\n",
    "ax.tick_params(labelsize=10)\n",
    "#g.set_xticklabels(ax.get_xticklabels(),rotation=90)\n",
    "sns.despine()\n",
    "plt.ylim((0,10))\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'Log_Phenotypic_Volume_UMM062'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label\n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Patient UMM065:\n",
    "QUERY = QUERY_FULL.loc[QUERY_FULL.index.map(lambda x: x[2] == 'UMM065')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "QUERY.index.get_level_values('Assignment').value_counts()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "n = 70\n",
    "reps = np.arange(100)\n",
    "gene_subset = variable_goi"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# GENE-GENE COVARIANCE WITHIN PATIENT 1\n",
    "master_groupby = 'Assignment'\n",
    "class_selection = ['Castle 1']\n",
    "ix = [ind for ind,name in enumerate(QUERY.index.names) if name==master_groupby][0]\n",
    "#logvolume_castle1 = np.zeros(len(reps))\n",
    "\n",
    "for rep in reps[]:\n",
    "    \n",
    "    SUBSET = QUERY.loc[QUERY.index.map(lambda x: x[ix] in class_selection)][gene_subset].sample(n=n,replace=True)\n",
    "    GENE_COV = SUBSET.cov(min_periods=None)\n",
    "    eigenvalues = np.linalg.eigvals(GENE_COV)\n",
    "    logvolume = np.float(np.sum(0.5*np.log10(eigenvalues[eigenvalues>0]**2))/len(eigenvalues))\n",
    "    logvolume_castle1[rep] = logvolume\n",
    "    print(rep)\n",
    "    \n",
    "# GENE-GENE COVARIANCE WITHIN PATIENT 2\n",
    "master_groupby = 'Assignment'\n",
    "class_selection = ['Castle 2']\n",
    "ix = [ind for ind,name in enumerate(QUERY.index.names) if name==master_groupby][0]\n",
    "logvolume_castle2 = np.zeros(len(reps))\n",
    "\n",
    "for rep in reps[]:\n",
    "    \n",
    "    SUBSET = QUERY.loc[QUERY.index.map(lambda x: x[ix] in class_selection)][gene_subset].sample(n=n,replace=True)\n",
    "    GENE_COV = SUBSET.cov(min_periods=None)\n",
    "    eigenvalues = np.linalg.eigvals(GENE_COV)\n",
    "    logvolume = np.float(np.sum(0.5*np.log10(eigenvalues[eigenvalues>0]**2))/len(eigenvalues))\n",
    "    logvolume_castle2[rep] = logvolume\n",
    "    print(rep)\n",
    "    \n",
    "# WRITE RESULTS TO DATAFRAME\n",
    "boxplot_data_goi_UMM065 = pd.DataFrame()\n",
    "boxplot_data_goi_UMM065['Assignment'] = ['Castle 1']*len(logvolume_castle1)+['Castle 2']*len(logvolume_castle2)\n",
    "boxplot_data_goi_UMM065['Log Phenotypic Volume'] = np.concatenate([logvolume_castle1,logvolume_castle2],axis = 0)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# STATISTIC\n",
    "C1_UMM065 = boxplot_data_goi_UMM065.loc[boxplot_data_goi_UMM065['Assignment'].\\\n",
    "                                    isin(['Castle 1'])]['Log Phenotypic Volume'].values\n",
    "C2_UMM065 = boxplot_data_goi_UMM065.loc[boxplot_data_goi_UMM065['Assignment'].\\\n",
    "                                    isin(['Castle 2'])]['Log Phenotypic Volume'].values\n",
    "\n",
    "joblib.dump(C1_UMM065, '/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM065_GEP1')\n",
    "joblib.dump(C2_UMM065, '/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM065_GEP2')\n",
    "\n",
    "stats.mannwhitneyu(C1_UMM065, C2_UMM065)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C1_UMM065 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM065_GEP1')\n",
    "C2_UMM065 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM065_GEP2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"white\")\n",
    "fig = plt.figure(figsize = (5,5))\n",
    "ax = plt.gca()\n",
    "\n",
    "#sns.kdeplot(IM, shade=True, color=\"#B0C4DE\")\n",
    "sns.kdeplot(C1_UMM065, shade=True, color=\"#0000FF\")\n",
    "sns.kdeplot(C2_UMM065, shade=True, color=\"#FF0000\")\n",
    "\n",
    "ax.set_ylabel(\"Frequency\",fontsize=10)\n",
    "ax.set_xlabel(\"Log Phenotypic Volume\",fontsize=10)\n",
    "ax.tick_params(labelsize=10)\n",
    "#g.set_xticklabels(ax.get_xticklabels(),rotation=90)\n",
    "sns.despine()\n",
    "plt.ylim((0,10))\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'Log_Phenotypic_Volume_UMM065'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label\n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TESTING STING AND IFN IN CLASS 1 SAMPLES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "INDF_HARBOUR_CLASS1 = INDF_HARBOUR.loc[INDF_HARBOUR.index.map(lambda x: x[2] in ['BSSR0022', 'UMM062', 'UMM065'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(INDF_HARBOUR_CLASS1.shape)\n",
    "INDF_HARBOUR_CLASS1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_genes = ['IRF3', 'RELA', 'RELB', 'MB21D1', 'IKBKB', 'NFKB1', 'TMEM173', 'STAT2', 'CXCL10', 'NFKB2',\n",
    "              'CCL5', 'TBK1', 'JAK1', 'JAK2', 'IRF9', 'TYK2', 'IRF7', 'IRF1', 'STAT1',\n",
    "              'HLA-A', 'HLA-B', 'HLA-C', 'HLA-E', 'HLA-F', 'HLA-G'\n",
    "             ]\n",
    "\n",
    "print(len(plot_genes))\n",
    "\n",
    "subset_type = 'TUMOR'\n",
    "QUERY = INDF_HARBOUR_CLASS1\n",
    "\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "heatmap_data = pd.DataFrame(data = zscore(QUERY[plot_genes],axis=0), \n",
    "                            columns = plot_genes, index = QUERY.index)\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns\n",
    "\n",
    "# LINKAGE \n",
    "method = 'centroid' # average, single centroid/euclidean\n",
    "metric = 'euclidean' # cosine\n",
    "#linkage = hc.linkage(heatmap_data, method=method, metric = metric)\n",
    "#row_linkage = deepcopy(linkage)\n",
    "#rl = hc.leaves_list(row_linkage)\n",
    "#mat = heatmap_data.iloc[rl,:]\n",
    "mat = heatmap_data.copy()\n",
    "\n",
    "window = 20\n",
    "mat = mat.rolling(window, win_type='triang',center = True).sum()\n",
    "half_window = int(window/2)\n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[ind] = mat.iloc[half_window]    \n",
    "for ind in np.arange(half_window):\n",
    "    mat.iloc[-ind] = mat.iloc[-half_window]\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(5,10))\n",
    "\n",
    "# ADD MATRIX WITH LINEAGE NAMES\n",
    "axmatrix = fig.add_axes([0.12,0.1,0.4,0.6])\n",
    "colors = [(1,1,1), np.divide(tuple(hex('FF0000').rgb),255)]\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=CM_DIVERGING,vmin=-10,vmax=10)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "#axmatrix.set_xticklabels([]);\n",
    "xtick = plt.xticks(range(len(labels)), labels, rotation = 90, fontsize = 6,fontname='Arial')\n",
    "ytick = plt.yticks([])\n",
    "axmatrix.grid(False)\n",
    "axmatrix.spines['top'].set_visible(False)\n",
    "axmatrix.spines['right'].set_visible(False)\n",
    "axmatrix.spines['bottom'].set_visible(False)\n",
    "axmatrix.spines['left'].set_visible(False)\n",
    "\n",
    "# ADD ROW 1 COLOR BY SAMPLE TYPE\n",
    "patient_status = [QUERY.loc[QUERY.index.get_level_values('Cell ID') == x].index.get_level_values('Patient').\\\n",
    "                          values[0] for x in mat.index.get_level_values('Cell ID').values]\n",
    "\n",
    "lut_PATIENT_STATUS_TEMP = dict({'BSSR0022': np.array([51, 0, 255])/255, \n",
    "                                'UMM062': np.array([147, 124, 242])/255,\n",
    "                                'UMM065': np.array([214, 207, 243])/255})\n",
    "\n",
    "row_colors1 = pd.Series(patient_status, index=patient_status).map(lut_PATIENT_STATUS_TEMP)\n",
    "ax1 = fig.add_axes([0.08,0.1,0.035,0.6]) # [x0,y0,width,height]\n",
    "x = 0\n",
    "y = 0\n",
    "for c in row_colors1:\n",
    "    pos = (x, y / len(row_colors1))\n",
    "    ax1.add_patch(patches.Rectangle(pos, 1, 1 / len(row_colors1), color=c))\n",
    "    if y >= len(row_colors1)-1:\n",
    "        x += 1\n",
    "        y = 0\n",
    "    else:\n",
    "        y += 1\n",
    "plt.axis('off')\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '/STING_IFN_Heatmap_HARBOUR_Class1'.format(meta,subset_type,method,metric)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/TEST_Heatmaps/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## KARYOTYPIC ANALYSIS PER ARCHETYPE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TEST FOR CANONICAL VARIANTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "CNV_DF = pd.DataFrame()\n",
    "\n",
    "for x in range(1,7):\n",
    "    tmp = pd.read_csv('/workdir/uvmel_project/data/CNV_data/uvmel'+ str(x) +'_cnv_data_075.csv', index_col=0)\n",
    "    tmp.index.name = 'Cell'\n",
    "    tmp['Patient'] = ['UM0'+str(x) for y in range(len(tmp))]\n",
    "    tmp.set_index(['Patient'], append=True, inplace=True)\n",
    "    tmp = tmp.reorder_levels(['Patient','Cell']).sort_index()\n",
    "\n",
    "    CNV_DF = pd.concat([CNV_DF, tmp])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "CNV_FREQ_DF = pd.DataFrame()\n",
    "patient_list = []\n",
    "freq_list = [[] for x in range(3*len(CNV_DF.columns))]\n",
    "variant_list = []\n",
    "\n",
    "for x in CNV_DF.index.get_level_values('Patient').unique():\n",
    "    patient_list = patient_list + [x]\n",
    "    \n",
    "    for y in CNV_DF.columns.tolist():\n",
    "        variant_list = variant_list + [y+'-', y+'+']\n",
    "        vals_tmp = CNV_DF[CNV_DF.index.get_level_values('Patient') == \n",
    "                                                           x][y].value_counts(normalize=True).reindex([-1,0,1], \n",
    "                                                           fill_value=0).sort_index().values.tolist()\n",
    "        \n",
    "        print(CNV_DF[CNV_DF.index.get_level_values('Patient') == \n",
    "                                                           x][y].value_counts(normalize=True).reindex([-1,0,1], \n",
    "                                                           fill_value=0).sort_index())\n",
    "        \n",
    "        freq_list[variant_list.index(y+'-')].append(vals_tmp[0])\n",
    "        freq_list[variant_list.index(y+'+')].append(vals_tmp[2])\n",
    "        \n",
    "CNV_FREQ_DF['Patient'] = patient_list\n",
    "\n",
    "for x in variant_list:\n",
    "    CNV_FREQ_DF[x] = freq_list[variant_list.index(x)]\n",
    "\n",
    "CNV_FREQ_DF.set_index(['Patient'], inplace=True)\n",
    "\n",
    "merged_3_loss = []\n",
    "merged_8_gain = []\n",
    "\n",
    "for x in CNV_DF.index.get_level_values('Patient').unique():\n",
    "    \n",
    "    merged_3_loss.append(sum([1 if ((CNV_DF[CNV_DF.index.get_level_values('Patient') == x]['3p'][y] == -1) |\n",
    "                             (CNV_DF[CNV_DF.index.get_level_values('Patient') == x]['3q'][y] == -1))\n",
    "                               else 0 \n",
    "                               for y in range(len(CNV_DF[CNV_DF.index.get_level_values('Patient') == x]))])/\n",
    "                        len(CNV_DF[CNV_DF.index.get_level_values('Patient') == x]))\n",
    "    \n",
    "    merged_8_gain.append(sum([1 if ((CNV_DF[CNV_DF.index.get_level_values('Patient') == x]['8p'][y] == 1) |\n",
    "                             (CNV_DF[CNV_DF.index.get_level_values('Patient') == x]['8q'][y] == -1))\n",
    "                               else 0 \n",
    "                               for y in range(len(CNV_DF[CNV_DF.index.get_level_values('Patient') == x]))])/\n",
    "                        len(CNV_DF[CNV_DF.index.get_level_values('Patient') == x]))\n",
    "    \n",
    "CNV_FREQ_DF['3-'] = merged_3_loss\n",
    "CNV_FREQ_DF['8+'] = merged_8_gain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "CNV_FREQ_DF[['6q-']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CONSTRUCT HEATMAP DATA\n",
    "canonical_list = ['1p-', '3-', '6p+', '6q-', '8p-', '8+',\n",
    "                  '11q-', '16q-', '17p-']\n",
    "\n",
    "\n",
    "mat = CNV_FREQ_DF[canonical_list].T.iloc[::-1]                     # Canonical Only\n",
    "#mat = CNV_FREQ_DF.T.iloc[::-1]                                      # All\n",
    "mat = mat[['UM01', 'UM02', 'UM04', 'UM05', 'UM06']]\n",
    "#mat_full = mat[['UM01', 'UM02', 'UM04', 'UM05', 'UM06']]\n",
    "#mat = mat[['UM03']]\n",
    "\n",
    "mat = mat.loc[~(mat==0).all(axis=1)]\n",
    "#mat_full = mat_full.loc[~(mat_full==0).all(axis=1)]\n",
    "\n",
    "#mat = mat[mat.index.map(lambda x: x in mat_full.index)]\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(5,4))\n",
    "\n",
    "# ADD MATRIX WITH GENE NAMES\n",
    "axmatrix = fig.add_axes([0.08,0.1,0.7,0.6])\n",
    "#cnv_colormap = mcolors.LinearSegmentedColormap('cnv_colormap',['white',\n",
    "#                                       plt.cm.RdBu_r(0),\n",
    "#                                      plt.cm.RdBu(0)])\n",
    "\n",
    "color_cutoff = 50\n",
    "reds = cm.get_cmap('Reds')\n",
    "cnv_colormap = reds(np.linspace(0, 1, 100))\n",
    "\n",
    "#cnv_colormap[color_cutoff:, :] = plt.cm.RdBu(0)\n",
    "#cnv_colormap = reds(np.linspace(0,1,100))\n",
    "cnv_colormap = mcolors.ListedColormap(cnv_colormap)\n",
    "\n",
    "#norm = mcolors.TwoSlopeNorm(vmin=0, vmax=color_cutoff/100)\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=cnv_colormap, vmax=0.5)\n",
    "labels = list(mat.columns)\n",
    "labels_y = list(mat.index)\n",
    "axmatrix.xaxis.set_ticks_position('top')\n",
    "axmatrix.set_yticklabels(mat.index)\n",
    "xtick = plt.xticks(range(len(labels)), labels, rotation = 0, fontsize = 8, fontname='Arial')\n",
    "ytick = plt.yticks(range(len(labels_y)), labels_y, fontsize=6, fontname='Arial')\n",
    "#axmatrix.grid(False)\n",
    "#axmatrix.tick_params(axis=u'y', which=u'both',length=0)\n",
    "\n",
    "# Add colorbar\n",
    "axcolor = fig.add_axes([0.8,0.1,0.05,0.3])\n",
    "cbar = plt.colorbar(im, cax=axcolor, ticks=[0,0.5,1])\n",
    "\n",
    "#plt.axis('off')\n",
    "fig.axes.facecolor:'white'\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'CNV_Frequencies_Canonical_All_Others'\n",
    "fn = '/workdir/uvmel_project/figures/Revision_2_Figures/CNV_FIGURES/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "canonical_list = ['1p-', '3-', '6p+', '6q-', '8p-', '8+',\n",
    "                  '11q-', '16q-', '17p-']\n",
    "\n",
    "\n",
    "mat = CNV_FREQ_DF[canonical_list].T.iloc[::-1]                     # Canonical Only\n",
    "#mat = CNV_FREQ_DF.T.iloc[::-1]                                      # All\n",
    "#mat = mat[['UM01', 'UM02', 'UM04', 'UM05', 'UM06']]\n",
    "#mat_full = mat[['UM01', 'UM02', 'UM04', 'UM05', 'UM06']]\n",
    "#mat = mat[['UM03']]\n",
    "\n",
    "mat = mat.loc[~(mat==0).all(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mat.index = ['1p-', '3-', '6p-', '6p+', '8+', '11q-', '16q-', '17p-']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(np.arange(0,1,1.0/100), np.arange(0,1,1.0/100),\n",
    "            cmap=cnv_colormap, c=np.arange(0,1,1.0/100))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### LOAD CNV RESULTS FROM inferCNV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "inferCNV_df = pd.read_csv('/workdir/uvmel_project/inferCNV/results_no_cluster_by_grouping_0.1/'+\n",
    "                          'expr.infercnv.dat',\n",
    "                           sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "inferCNV_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_TUMOR.index.get_level_values('Patient').value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### LOAD CNV RESULTS FROM CaSpER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_input = pd.read_csv('/workdir/uvmel_project/data/CNV_data/uvmel2_cnv_data_075.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_input.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### VISUALIZE RESULTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CONSTRUCT HEATMAP DATA\n",
    "heatmap_data = mat_input\n",
    "\n",
    "# LINKAGE \n",
    "method = 'average' # average, single centroid/euclidean\n",
    "metric = 'euclidean' # cosine\n",
    "linkage = hc.linkage(heatmap_data, method=method, metric = metric)\n",
    "row_linkage = deepcopy(linkage)\n",
    "\n",
    "# REORDER HEATMAP ACCORDING TO LINKAGE (OPTIONAL, STILL SLOW)\n",
    "r1 = hc.leaves_list(row_linkage)\n",
    "#mat = heatmap_data                                                # Not Clustered\n",
    "mat = heatmap_data.iloc[r1,:]                                      # Clustered\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(14,6))\n",
    "\n",
    "\n",
    "# ADD MATRIX WITH GENE NAMES\n",
    "axmatrix = fig.add_axes([0.08,0.1,0.7,0.6])\n",
    "cnv_colormap = mcolors.ListedColormap([plt.cm.RdBu_r(0),\n",
    "                                       'white',\n",
    "                                       plt.cm.RdBu(0)])\n",
    "\n",
    "norm = mcolors.TwoSlopeNorm(vmin=-1, vmax = 1, vcenter=0)\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=cnv_colormap)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "xtick = plt.xticks(range(len(labels)), labels, rotation = 0, fontsize = 8, fontname='Arial')\n",
    "axmatrix.grid(False)\n",
    "axmatrix.tick_params(axis=u'y', which=u'both',length=0)\n",
    "\n",
    "# ADD DENDROGRAM\n",
    "ax1 = fig.add_axes([0.00,0.1,0.08,0.6]) # [x0,y0,width,height]\n",
    "Z1 = sch.dendrogram(row_linkage, orientation='left', color_threshold = 0, above_threshold_color='#808080',\n",
    "                    p=6, truncate_mode='level')\n",
    "ax1.set_xticks([])\n",
    "ax1.set_yticks([])\n",
    "plt.axis('off')\n",
    "\n",
    "fig.axes.facecolor:'white'\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'UM02_CLUSTERED_075'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/CNV_FIGURES/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### VISUALIZE ADJACENCY GRAPH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "patientNum = 2\n",
    "CNVadjMatx = pd.read_csv('/workdir/uvmel_project/data/CNV_data/uvmel'+str(patientNum)+'_adjMatx_data_075.csv', \n",
    "                         index_col=0)\n",
    "CNVvarFreq = pd.read_csv('/workdir/uvmel_project/data/CNV_data/uvmel'+str(patientNum)+'_varFreq_data_075.csv', \n",
    "                         index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_p = 0.01\n",
    "min_freq = 0.10\n",
    "\n",
    "CNVadjMatxFiltered = CNVadjMatx.copy()\n",
    "\n",
    "CNVadjMatxFiltered[abs(CNVadjMatxFiltered) > max_p] = 0\n",
    "\n",
    "filterVars = [x for x in CNVvarFreq[CNVvarFreq.varFreq < min_freq].V1.values if x in CNVadjMatxFiltered.columns]\n",
    "CNVadjMatxFiltered.loc[:,filterVars] = 0\n",
    "CNVadjMatxFiltered.loc[filterVars,:] = 0\n",
    "\n",
    "CNVadjMatxFiltered = CNVadjMatxFiltered.loc[:, (CNVadjMatxFiltered != 0).any(axis=0)]\n",
    "CNVadjMatxFiltered = CNVadjMatxFiltered.loc[(CNVadjMatxFiltered.T != 0).any(axis=0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "G = nx.from_pandas_adjacency(CNVadjMatxFiltered)\n",
    "\n",
    "weights = [G[u][v]['weight'] for u,v in G.edges()]\n",
    "label_dict = dict(zip(list(G.nodes),[x[:2] if len(x) == 5 else x[:3] for x in list(G.nodes)]))\n",
    "\n",
    "fig = plt.figure(figsize=(8, 6))\n",
    "plt.tight_layout()\n",
    "\n",
    "try:\n",
    "    nx.draw_planar(G, \n",
    "            with_labels=True,\n",
    "            labels=label_dict,\n",
    "            font_color='white',\n",
    "            font_size=17,\n",
    "            font_weight='bold',\n",
    "            node_color=[plt.cm.RdBu(0) if 'amp' in node else plt.cm.RdBu_r(0) for node in G],\n",
    "            node_size=[8000.0*(CNVvarFreq.loc[CNVvarFreq.V1 == label].varFreq.values[0])\n",
    "                       for label in G],\n",
    "            width=[-0.1*np.log2(abs(x)) for x in weights],\n",
    "            style=['solid' if x < 0 else 'dashed' for x in weights])\n",
    "    \n",
    "    \n",
    "except:\n",
    "    nx.draw_shell(G, \n",
    "            with_labels=True,\n",
    "            labels=label_dict,\n",
    "            font_color='white',\n",
    "            font_size=17,\n",
    "            font_weight='bold',\n",
    "            node_color=[plt.cm.RdBu(0) if 'amp' in node else plt.cm.RdBu_r(0) for node in G],\n",
    "            node_size=[8000.0*(CNVvarFreq.loc[CNVvarFreq.V1 == label].varFreq.values[0])\n",
    "                       for label in G],\n",
    "            width=[-0.1*np.log2(abs(x)) for x in weights],\n",
    "            style=['solid' if x < 0 else 'dashed' for x in weights])\n",
    "\n",
    "    \n",
    "plt.axis('off')\n",
    "axis = plt.gca()\n",
    "axis.set_xlim([1.2*x for x in axis.get_xlim()])\n",
    "axis.set_ylim([1.2*y for y in axis.get_ylim()])\n",
    "plt.tight_layout()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'UM0'+str(patientNum)+'_Network_10_075'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/CNV_FIGURES/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### KARYOTYPIC VOLUME"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "patientNum = 6\n",
    "chr_smoothed_map = pd.DataFrame()\n",
    "#current_list = [patientNum]\n",
    "current_list = [1,2,3,4,5,6]\n",
    "\n",
    "for i in current_list:\n",
    "    chr_smoothed_map = chr_smoothed_map.append(pd.read_csv('/workdir/uvmel_project/data/CNV_data/uvmel'+\n",
    "                                           str(i)+'_chrSmoothedExp_data.csv', \n",
    "                                           index_col=0))\n",
    "\n",
    "chr_smoothed_map.index = DF_TUMOR.loc[DF_TUMOR.index.map(lambda x: x[2] in ['UM0'+str(x) for x in current_list\n",
    "                                                                    ])].index\n",
    "chr_smoothed_map['Assignment'] = METADATA_TUMOR.loc[METADATA_TUMOR.index.map(lambda x: x[2] in \n",
    "                                                                     ['UM0'+str(x) for x in current_list\n",
    "                                                                     ])].Assignment\n",
    "chr_smoothed_map.set_index('Assignment', append=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chr_smoothed_map.index.get_level_values('Assignment').value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chr_smoothed_map.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(chr_smoothed_map[chr_smoothed_map.index.map(lambda x: x[2] == 'UM05')]['3'].values, bins=100)\n",
    "plt.axvline(-0.25, ymin=0, ymax=100, color='red')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "thr = 0.25\n",
    "\n",
    "CNV_EXP_DF = chr_smoothed_map\n",
    "CNV_EXP_DF[CNV_EXP_DF < -1*thr] = -1\n",
    "CNV_EXP_DF[CNV_EXP_DF > thr] = 1\n",
    "CNV_EXP_DF[abs(CNV_EXP_DF) < thr] = 0\n",
    "\n",
    "CNV_EXP_FREQ_DF = pd.DataFrame()\n",
    "patient_list = []\n",
    "freq_list = [[] for x in range(3*len(CNV_EXP_DF.columns))]\n",
    "variant_list = []\n",
    "\n",
    "for x in CNV_EXP_DF.index.get_level_values('Patient').unique():\n",
    "    patient_list = patient_list + [x]\n",
    "    \n",
    "    for y in CNV_EXP_DF.columns.tolist():\n",
    "        variant_list = variant_list + [y+'-', y+'+']\n",
    "        vals_tmp = CNV_EXP_DF[CNV_EXP_DF.index.get_level_values('Patient') == \n",
    "                                                           x][y].value_counts(normalize=True).reindex([-1,0,1], \n",
    "                                                           fill_value=0).sort_index().values.tolist()\n",
    "        \n",
    "        print(CNV_EXP_DF[CNV_EXP_DF.index.get_level_values('Patient') == \n",
    "                                                           x][y].value_counts(normalize=True).reindex([-1,0,1], \n",
    "                                                           fill_value=0).sort_index())\n",
    "        \n",
    "        freq_list[variant_list.index(y+'-')].append(vals_tmp[0])\n",
    "        freq_list[variant_list.index(y+'+')].append(vals_tmp[2])\n",
    "        \n",
    "CNV_EXP_FREQ_DF['Patient'] = patient_list\n",
    "\n",
    "for x in variant_list:\n",
    "    CNV_EXP_FREQ_DF[x] = freq_list[variant_list.index(x)]\n",
    "\n",
    "CNV_EXP_FREQ_DF.set_index(['Patient'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "CNV_EXP_FREQ_DF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CONSTRUCT HEATMAP DATA\n",
    "\n",
    "canonical_list = ['1-', '3-', '6+', '6-', '8-', '8+',\n",
    "                  '11-', '16-', '17-']\n",
    "\n",
    "\n",
    "mat = CNV_EXP_FREQ_DF\n",
    "\n",
    "mat = mat[canonical_list].T.iloc[::-1]\n",
    "mat = mat[['UM03']]\n",
    "#mat = mat[['UM01', 'UM02', 'UM04', 'UM05', 'UM06']]\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(1,4))\n",
    "\n",
    "# ADD MATRIX WITH GENE NAMES\n",
    "axmatrix = fig.add_axes([0.08,0.1,0.7,0.6])\n",
    "#cnv_colormap = mcolors.LinearSegmentedColormap('cnv_colormap',['white',\n",
    "#                                       plt.cm.RdBu_r(0),\n",
    "#                                      plt.cm.RdBu(0)])\n",
    "\n",
    "color_cutoff = 50\n",
    "reds = cm.get_cmap('Reds')\n",
    "cnv_colormap = reds(np.linspace(0, 1, 100))\n",
    "\n",
    "#cnv_colormap[color_cutoff:, :] = plt.cm.RdBu(0)\n",
    "#cnv_colormap = reds(np.linspace(0,1,100))\n",
    "cnv_colormap = mcolors.ListedColormap(cnv_colormap)\n",
    "\n",
    "#norm = mcolors.TwoSlopeNorm(vmin=0, vmax=color_cutoff/100)\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=cnv_colormap, vmax=1)\n",
    "labels = list(mat.columns)\n",
    "labels_y = list(mat.index)\n",
    "axmatrix.xaxis.set_ticks_position('top')\n",
    "axmatrix.set_yticklabels(mat.index)\n",
    "xtick = plt.xticks(range(len(labels)), labels, rotation = 0, fontsize = 10, fontname='Arial')\n",
    "ytick = plt.yticks(range(len(labels_y)), labels_y, fontsize=10, fontname='Arial')\n",
    "#axmatrix.grid(False)\n",
    "#axmatrix.tick_params(axis=u'y', which=u'both',length=0)\n",
    "\n",
    "# Add colorbar\n",
    "axcolor = fig.add_axes([0.8,0.1,0.05,0.3])\n",
    "cbar = plt.colorbar(im, cax=axcolor, ticks=[0,0.5,1])\n",
    "\n",
    "#plt.axis('off')\n",
    "fig.axes.facecolor:'white'\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'CNV_Heatmap_from_Expression_UM03_unsaturated'\n",
    "fn = '/workdir/uvmel_project/figures/Revision_2_Figures/CNV_FIGURES/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CONSTRUCT HEATMAP DATA\n",
    "mat = chr_smoothed_map\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(12,8))\n",
    "\n",
    "# ADD MATRIX WITH GENE NAMES\n",
    "axmatrix = fig.add_axes([0.08,0.1,0.7,0.6])\n",
    "#cnv_colormap = mcolors.LinearSegmentedColormap('cnv_colormap',['white',\n",
    "#                                       plt.cm.RdBu_r(0),\n",
    "#                                      plt.cm.RdBu(0)])\n",
    "\n",
    "color_cutoff = 50\n",
    "reds = cm.get_cmap('Reds')\n",
    "cnv_colormap = reds(np.linspace(0, 1, 100))\n",
    "\n",
    "#cnv_colormap[color_cutoff:, :] = plt.cm.RdBu(0)\n",
    "#cnv_colormap = reds(np.linspace(0,1,100))\n",
    "cnv_colormap = mcolors.ListedColormap(cnv_colormap)\n",
    "\n",
    "#norm = mcolors.TwoSlopeNorm(vmin=0, vmax=color_cutoff/100)\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap='RdBu_r', vmin=-1, vmax=1)\n",
    "labels = list(mat.columns)\n",
    "labels_y = list(mat.index)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels(mat.index)\n",
    "xtick = plt.xticks(range(len(labels)), labels, rotation = 0, fontsize = 8, fontname='Arial')\n",
    "ytick = plt.yticks(range(len(labels_y)), [], fontsize=6, fontname='Arial')\n",
    "#axmatrix.grid(False)\n",
    "#axmatrix.tick_params(axis=u'y', which=u'both',length=0)\n",
    "\n",
    "# Add colorbar\n",
    "axcolor = fig.add_axes([0.8,0.1,0.05,0.3])\n",
    "cbar = plt.colorbar(im, cax=axcolor, ticks=[0,0.5,1])\n",
    "\n",
    "#plt.axis('off')\n",
    "fig.axes.facecolor:'white'\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'UM0'+str(patientNum)+'_Heatmap'\n",
    "fn = '/workdir/uvmel_project/figures/Revision_2_Figures/CNV_FIGURES/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n = 50\n",
    "reps = np.arange(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mat = chr_smoothed_map.copy()\n",
    "\n",
    "# CLASS I TUMORS\n",
    "logvolume_tumor_class1 = np.zeros(len(reps))\n",
    "class_selection = 'Castle 1'\n",
    "SUBSET = mat.loc[mat.index.map(lambda x: x[10] == class_selection)]\n",
    "\n",
    "for rep in reps:\n",
    "    CSUBSET = SUBSET.sample(n=n,replace=True)\n",
    "    GENE_COV = CSUBSET.cov(min_periods=None)\n",
    "    eigenvalues = np.linalg.eigvals(GENE_COV)\n",
    "    logvolume = np.float(np.sum(0.5*np.log10(eigenvalues[eigenvalues>0]**2))/len(eigenvalues))\n",
    "    logvolume_tumor_class1[rep] = logvolume\n",
    "    \n",
    "# CLASS II TUMORS\n",
    "logvolume_tumor_class2 = np.zeros(len(reps))\n",
    "class_selection = 'Castle 2'\n",
    "SUBSET = mat.loc[mat.index.map(lambda x: x[10] == class_selection)]\n",
    "\n",
    "for rep in reps:\n",
    "    CSUBSET = SUBSET.sample(n=n,replace=True)\n",
    "    GENE_COV = CSUBSET.cov(min_periods=None)\n",
    "    eigenvalues = np.linalg.eigvals(GENE_COV)\n",
    "    logvolume = np.float(np.sum(0.5*np.log10(eigenvalues[eigenvalues>0]**2))/len(eigenvalues))\n",
    "    logvolume_tumor_class2[rep] = logvolume\n",
    "\n",
    "\n",
    "# WRITE RESULTS TO DATAFRAME\n",
    "boxplot_data_goi = pd.DataFrame()\n",
    "boxplot_data_goi['Meta-Source'] = ['Group 1']*len(logvolume_tumor_class1)+['Group 2']*len(logvolume_tumor_class2)\n",
    "boxplot_data_goi['Log Karyotypic Volume'] = np.concatenate([logvolume_tumor_class1,logvolume_tumor_class2],axis = 0)\n",
    "\n",
    "# STATISTIC\n",
    "T1 = boxplot_data_goi.loc[boxplot_data_goi['Meta-Source'].isin(['Group 1'])]['Log Karyotypic Volume'].values\n",
    "T2 = boxplot_data_goi.loc[boxplot_data_goi['Meta-Source'].isin(['Group 2'])]['Log Karyotypic Volume'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(T1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CLASS I vs. CLASS 2\n",
    "print('Castle 1 vs. Castle 2')\n",
    "print(stats.mannwhitneyu(T1, T2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"white\")\n",
    "fig = plt.figure(figsize = (5,5))\n",
    "ax = plt.gca()\n",
    "\n",
    "sns.kdeplot(T1, shade=True, color=\"#0000FF\")\n",
    "sns.kdeplot(T2, shade=True, color=\"#FF0000\")\n",
    "#sns.kdeplot(T3, shade=True, color=\"#FF0000\")\n",
    "#sns.kdeplot(T4, shade=True, color=\"#FF0000\")\n",
    "\n",
    "ax.set_ylabel(\"Frequency\",fontsize=10)\n",
    "ax.set_xlabel(\"Log Karyotypic Volume\",fontsize=10)\n",
    "ax.tick_params(labelsize=10)\n",
    "#ax.set_xticklabels(ax.get_xticklabels(),rotation=90)\n",
    "sns.despine()\n",
    "plt.ylim((0,10))\n",
    "#plt.xlim((-0.95,-0.4))\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'Log_Karyotypic_Volume_KDE_PATIENT_'+str(patientNum)\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label\n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"white\")\n",
    "fig = plt.figure(figsize = (5,5))\n",
    "ax = plt.gca()\n",
    "\n",
    "sns.pointplot('Meta-Source', 'Log Karyotypic Volume',data=boxplot_data_goi, color='black', \n",
    "              dodge=True, join=False,\n",
    "              scale=0.8, errwidth=1, capsize=0.08, ci='sd')\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'Karyotypic_Volume_Scatter_UM03'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label\n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(7,7))\n",
    "\n",
    "ax = sns.violinplot(x='Meta-Source', y='Log Karyotypic Volume', data=boxplot_data_goi,\n",
    "                    hue='Meta-Source',\n",
    "                    palette={'Group 1': 'white', 'Group 2': 'gray'},\n",
    "                    dodge=False,\n",
    "                    scale='area',)\n",
    "\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.get_legend().remove()\n",
    "\n",
    "plt.xticks(rotation=90)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'Karyotypic_Volume_Violin_UM03'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label\n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SHANNON DIVERSITY INDEX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cnv_mat = pd.DataFrame()\n",
    "current_list = [1,2,3,4,5,6]\n",
    "\n",
    "for i in current_list:\n",
    "    cnv_mat = cnv_mat.append(pd.read_csv('/workdir/uvmel_project/data/CNV_data/uvmel'+\n",
    "                                           str(i)+'_cnv_data_075.csv', \n",
    "                                           index_col=0))\n",
    "\n",
    "cnv_mat.index = DF_TUMOR.loc[DF_TUMOR.index.map(lambda x: x[2] in ['UM0'+str(x) for x in current_list\n",
    "                                                                    ])].index.get_level_values('Patient')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#unique_kar_df = pd.DataFrame(cnv_mat.groupby(['Patient']+list(cnv_mat.columns)).size(), columns=['Counts'])\n",
    "kar_df = cnv_mat.copy()\n",
    "\n",
    "kar_df.reset_index(inplace=True)\n",
    "kar_df['State'] = kar_df[kar_df.columns[1:-1]].apply(\n",
    "                                       lambda x: ','.join(x.dropna().astype(str)), axis=1)\n",
    "\n",
    "#unique_kar_df = unique_kar_df[['Patient', 'State', 'Counts']]\n",
    "\n",
    "kar_df = kar_df[['Patient', 'State']]\n",
    "kar_df.set_index('Patient', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kar_df.State[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_size = 700\n",
    "sample_rounds= 20\n",
    "\n",
    "sdi_df = pd.DataFrame()\n",
    "\n",
    "for i in range(sample_rounds):\n",
    "    sdi_list = []\n",
    "    for patient in kar_df.index.unique():\n",
    "        tmp_df = kar_df.loc[kar_df.index.get_level_values('Patient') == patient]\n",
    "        tmp_df = tmp_df.sample(sample_size, replace=True)\n",
    "        p_vect = tmp_df['State'].value_counts(normalize=True)\n",
    "        sdi_list.append(-1*sum(list(map(lambda x: np.log(x)*x, p_vect))))\n",
    "    \n",
    "    sdi_df[str(i)] = sdi_list\n",
    "\n",
    "sdi_df.index = kar_df.index.unique()\n",
    "#sdi_df['Ave'] = sdi_df.mean(axis=1).values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sdi_boxplot_data = pd.DataFrame()\n",
    "sdi_boxplot_data['Patient'] = [item for sublist in [[x]*sample_rounds for x in sdi_df.index] for item in sublist]\n",
    "sdi_boxplot_data['SDI'] = [item for sublist in [sdi_df.iloc[x].values for x in range(len(sdi_df.index))]\n",
    "                           for item in sublist]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"white\")\n",
    "fig = plt.figure(figsize = (5,5))\n",
    "ax = plt.gca()\n",
    "\n",
    "sns.pointplot('Patient', 'SDI',data=sdi_boxplot_data, color='black', \n",
    "              dodge=True, join=False,\n",
    "              scale=0.7, errwidth=1, capsize=0.1, ci='sd')\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'Karyotypic_SDI_Scatter'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label\n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kar_df['Cell ID'] = DF_TUMOR.index.get_level_values('Cell ID')\n",
    "kar_df.set_index('Cell ID', append=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kar_unique_df = pd.DataFrame(kar_df.groupby(['Patient','State']).size())\n",
    "kar_unique_df.rename(columns={0:'Count'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kar_ds_df = kar_df[kar_df.reset_index().duplicated(['Patient','State'], keep=False).values == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kar_ds_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_size = 300\n",
    "sample_rounds= 50\n",
    "\n",
    "sdi_ds_df = pd.DataFrame()\n",
    "\n",
    "for i in range(sample_rounds):\n",
    "    sdi_list = []\n",
    "    for patient in kar_ds_df.index.get_level_values('Patient').unique():\n",
    "        tmp_df = kar_ds_df.loc[kar_ds_df.index.get_level_values('Patient') == patient]\n",
    "        tmp_df = tmp_df.sample(sample_size, replace=True)\n",
    "        p_vect = tmp_df['State'].value_counts(normalize=True)\n",
    "        sdi_list.append(-1*sum(list(map(lambda x: np.log(x)*x, p_vect))))\n",
    "    \n",
    "    sdi_ds_df[str(i)] = sdi_list\n",
    "\n",
    "sdi_ds_df.index = kar_ds_df.index.get_level_values('Patient').unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sdi_boxplot_ds_data = pd.DataFrame()\n",
    "sdi_boxplot_ds_data['Patient'] = [item for sublist in [[x]*sample_rounds \n",
    "                                                       for x in sdi_ds_df.index] for item in sublist]\n",
    "sdi_boxplot_ds_data['SDI'] = [item for sublist in [sdi_ds_df.iloc[x].values \n",
    "                                                   for x in range(len(sdi_ds_df.index))]\n",
    "                              for item in sublist]\n",
    "\n",
    "sns.set_style(\"white\")\n",
    "fig = plt.figure(figsize = (5,5))\n",
    "ax = plt.gca()\n",
    "\n",
    "sns.pointplot('Patient', 'SDI',data=sdi_boxplot_ds_data, color='black', \n",
    "              dodge=True, join=False,\n",
    "              scale=0.7, errwidth=1, capsize=0.1, ci='sd')\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'Karyotypic_SDI_Scatter_DROP_SINGLETONS'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label\n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SHANNON DIVERSITY WITHIN UM03 FOR GEP ASSIGNMENTS\n",
    "\n",
    "kar_df_UM03 = kar_df.loc[kar_df.index.get_level_values('Patient') == 'UM03']\n",
    "kar_df_UM03['Assignment'] = METADATA_TUMOR.loc[METADATA_TUMOR.index.get_level_values('Patient') == \n",
    "                                               'UM03'].Assignment.values\n",
    "\n",
    "kar_df_UM03.set_index('Assignment', append=False, inplace=True)\n",
    "\n",
    "sample_size = 50\n",
    "sample_rounds= 100\n",
    "\n",
    "sdi_df = pd.DataFrame()\n",
    "\n",
    "for i in range(sample_rounds):\n",
    "    sdi_list = []\n",
    "    for gep_class in kar_df_UM03.index.unique():\n",
    "        tmp_df = kar_df_UM03.loc[kar_df_UM03.index.get_level_values('Assignment') == gep_class]\n",
    "        tmp_df = tmp_df.sample(sample_size, replace=True)\n",
    "        p_vect = tmp_df['State'].value_counts(normalize=True)\n",
    "        sdi_list.append(-1*sum(list(map(lambda x: np.log(x)*x, p_vect))))\n",
    "    \n",
    "    sdi_df[str(i)] = sdi_list\n",
    "\n",
    "sdi_df.index = kar_df_UM03.index.unique()\n",
    "#sdi_df['Ave'] = sdi_df.mean(axis=1).values\n",
    "\n",
    "sdi_boxplot_data = pd.DataFrame()\n",
    "sdi_boxplot_data['GEP'] = [item for sublist in [[x]*sample_rounds for x in sdi_df.index] for item in sublist]\n",
    "sdi_boxplot_data['SDI'] = [item for sublist in [sdi_df.iloc[x].values for x in range(len(sdi_df.index))]\n",
    "                           for item in sublist]\n",
    "\n",
    "sns.set_style(\"white\")\n",
    "fig = plt.figure(figsize = (5,5))\n",
    "ax = plt.gca()\n",
    "\n",
    "sns.kdeplot(sdi_boxplot_data.loc[sdi_boxplot_data.GEP == 'Castle 1']['SDI'].values, shade=True, color=\"#0000FF\")\n",
    "sns.kdeplot(sdi_boxplot_data.loc[sdi_boxplot_data.GEP == 'Castle 2']['SDI'].values, shade=True, color=\"#FF0000\")\n",
    "\n",
    "ax.set_ylabel(\"Frequency\",fontsize=10)\n",
    "ax.set_xlabel(\"SDI\",fontsize=10)\n",
    "ax.tick_params(labelsize=10)\n",
    "#ax.set_xticklabels(ax.get_xticklabels(),rotation=90)\n",
    "sns.despine()\n",
    "plt.ylim((0,10))\n",
    "#plt.xlim((-0.95,-0.4))\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'Karyotypic_SDI_KDE_UM03'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label\n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CLASS I vs. CLASS 2\n",
    "print('Castle 1 vs. Castle 2')\n",
    "print(stats.mannwhitneyu(sdi_boxplot_data.loc[sdi_boxplot_data.GEP == 'Castle 1']['SDI'].values,\n",
    "                         sdi_boxplot_data.loc[sdi_boxplot_data.GEP == 'Castle 2']['SDI'].values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(2,7))\n",
    "\n",
    "ax = sns.barplot(x='GEP', y='SDI', data=sdi_boxplot_data,\n",
    "                 hue='GEP',\n",
    "                 edgecolor='black',\n",
    "                 palette={'Castle 1': 'white', 'Castle 2': 'gray'},\n",
    "                 dodge=False,\n",
    "                 ci=95,\n",
    "                 errwidth=1,\n",
    "                 capsize=0.1,\n",
    "                 estimator=median)\n",
    " \n",
    "ax.set_yscale('log')\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.get_legend().remove()\n",
    "\n",
    "plt.xticks(rotation=90)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'Karyotypic_SDI_Barplot_UM03_median'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label\n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PHEN_VOL_UM01 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UM01')\n",
    "PHEN_VOL_UM02 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UM02')\n",
    "PHEN_VOL_UM03 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UM03')\n",
    "PHEN_VOL_UM04 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UM04')\n",
    "PHEN_VOL_UM05 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UM05')\n",
    "PHEN_VOL_UM06 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UM06')\n",
    "\n",
    "PHEN_VOL_LIST = [PHEN_VOL_UM01, PHEN_VOL_UM02, PHEN_VOL_UM03, PHEN_VOL_UM04, PHEN_VOL_UM05, PHEN_VOL_UM06]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_TUMOR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "METADATA_TUMOR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phen_vol_data = pd.DataFrame()\n",
    "\n",
    "phen_vol_data['Patient'] = ['UM01']*len(PHEN_VOL_LIST[0])+['UM02']*len(PHEN_VOL_LIST[1])+\\\n",
    "                           ['UM03']*len(PHEN_VOL_LIST[2])+['UM04']*len(PHEN_VOL_LIST[3])+\\\n",
    "                           ['UM05']*len(PHEN_VOL_LIST[4])+['UM06']*len(PHEN_VOL_LIST[5])\n",
    "\n",
    "phen_vol_data['Log Phenotypic Volume'] = [x for l in PHEN_VOL_LIST for x in l]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"white\")\n",
    "fig = plt.figure(figsize = (5,5))\n",
    "ax = plt.gca()\n",
    "\n",
    "sns.pointplot('Patient', 'Log Phenotypic Volume',data=phen_vol_data, color='black', \n",
    "              dodge=True, join=False,\n",
    "              scale=0.7, errwidth=1, capsize=0.1, ci='sd')\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'Phenotypic_Volume_Scatter'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label\n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PHENOTYPIC VOLUME HARBOUR"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "PHEN_VOL_UMM059 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM059')\n",
    "PHEN_VOL_UMM061 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM061')\n",
    "PHEN_VOL_UMM062 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM062')\n",
    "PHEN_VOL_UMM063 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM063')\n",
    "PHEN_VOL_UMM064 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM064')\n",
    "PHEN_VOL_UMM065 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM065')\n",
    "PHEN_VOL_UMM066 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM066')\n",
    "PHEN_VOL_UMM069 = joblib.load('/workdir/uvmel_project/data/MSK_Phenotypic_Volume_UMM069')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "PHEN_VOL_LIST_COMBINED = PHEN_VOL_LIST + [PHEN_VOL_UMM059, PHEN_VOL_UMM061, PHEN_VOL_UMM062, PHEN_VOL_UMM063,\n",
    "                                          PHEN_VOL_UMM064, PHEN_VOL_UMM065, PHEN_VOL_UMM066, PHEN_VOL_UMM069]"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "phen_vol_data_combined = pd.DataFrame()\n",
    "\n",
    "phen_vol_data_combined['Patient'] = ['MSK-UM01']*len(PHEN_VOL_LIST[0])+['MSK-UM02']*len(PHEN_VOL_LIST[1])+\\\n",
    "                                    ['MSK-UM03']*len(PHEN_VOL_LIST[2])+['MSK-UM04']*len(PHEN_VOL_LIST[3])+\\\n",
    "                                    ['MSK-UM05']*len(PHEN_VOL_LIST[4])+['MSK-UM06']*len(PHEN_VOL_LIST[5])+\\\n",
    "                                    ['UMM059']*len(PHEN_VOL_LIST_COMBINED[6])+\\\n",
    "                                    ['UMM061']*len(PHEN_VOL_LIST_COMBINED[7])+\\\n",
    "                                    ['UMM062']*len(PHEN_VOL_LIST_COMBINED[8])+\\\n",
    "                                    ['UMM063']*len(PHEN_VOL_LIST_COMBINED[9])+\\\n",
    "                                    ['UMM064']*len(PHEN_VOL_LIST_COMBINED[10])+\\\n",
    "                                    ['UMM065']*len(PHEN_VOL_LIST_COMBINED[11])+\\\n",
    "                                    ['UMM066']*len(PHEN_VOL_LIST_COMBINED[12])+\\\n",
    "                                    ['UMM069']*len(PHEN_VOL_LIST_COMBINED[13])\n",
    "\n",
    "phen_vol_data_combined['Log Phenotypic Volume'] = [x for l in PHEN_VOL_LIST_COMBINED for x in l]\n",
    "\n",
    "sns.set_style(\"white\")\n",
    "fig = plt.figure(figsize = (12,5))\n",
    "#ax = plt.gca()\n",
    "\n",
    "ax = sns.pointplot('Patient', 'Log Phenotypic Volume',data=phen_vol_data_combined, color='black', \n",
    "              dodge=True, join=False,\n",
    "              scale=0.7, errwidth=1, capsize=0.1, ci='sd')\n",
    "\n",
    "ax.set_xticklabels(ax.get_xticklabels(),rotation=90)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'Phenotypic_Volume_Scatter_COMBINED'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label\n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "phen_vol_data_combined_MSK = phen_vol_data_combined.loc[phen_vol_data_combined.Patient.map(lambda x: 'MSK' in x)]"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "phen_vol_data_combined_MSK['Status'] = ['Alive' if x in ['MSK-UM01', 'MSK-UM05', 'MSK-UM03', 'MSK-UM04']\n",
    "                                        else 'Deceased' for x in phen_vol_data_combined_MSK.Patient.values]"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "phen_boxplot_data = phen_vol_data_combined_MSK.groupby('Patient').mean().sort_values(by='Log Phenotypic Volume')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "phen_boxplot_data['SD'] = phen_vol_data_combined_MSK.groupby('Patient').std().sort_values(\n",
    "                                                     by='Log Phenotypic Volume').values"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "fig, ax = plt.subplots(figsize=(6,4))\n",
    "\n",
    "plt.bar(x=phen_boxplot_data.Patient, \n",
    "             height=phen_boxplot_data['Log Phenotypic Volume'],\n",
    "             yerr=phen_boxplot_data['SD'],\n",
    "             capsize=2,\n",
    "             error_kw={'elinewidth': 1},\n",
    "             color=['white', 'white', 'white', 'white', 'gray', 'gray'],\n",
    "             edgecolor='black')\n",
    "\n",
    "ax.spines['bottom'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'Log_Phenotypic_Volume_Barplot'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label \n",
    "plt.savefig(fn + '.png', dpi=400, transparent=True)\n",
    "plt.savefig(fn + '.pdf', dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "phen_boxplot_data['Status'] = ['Alive' if x in ['MSK-UM01', 'MSK-UM05', 'MSK-UM03', 'MSK-UM04']\n",
    "                                        else 'Deceased' for x in phen_boxplot_data.index.values]"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "phen_boxplot_data.reset_index(inplace=True)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "fig, ax = plt.subplots(figsize = (12,5))\n",
    "\n",
    "ax.errorbar(x = phen_vol_data_combined.Patient.unique(),\n",
    "                  y = [mean(x) for x in PHEN_VOL_LIST_COMBINED],\n",
    "                  yerr=[np.std(x)/len(x) for x in PHEN_VOL_LIST_COMBINED],\n",
    "                  color='black',\n",
    "                  fmt='o')\n",
    "\n",
    "plt.setp(ax.get_xticklabels(), rotation='vertical', fontsize=18)\n",
    "plt.ylabel('Log Phenotypic Volume', fontsize=18)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'Phenotypic_Volume_Scatter_COMBINED_SME'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label\n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "fig, ax = plt.subplots(figsize=(7,7))\n",
    "\n",
    "ax = sns.violinplot(x='Patient', y='Log Phenotypic Volume', data=phen_vol_data_combined_MSK,\n",
    "                    order=phen_vol_data_combined_MSK.groupby('Patient').mean().\\\n",
    "                          sort_values(by='Log Phenotypic Volume').index.values,\n",
    "                    hue='Status',\n",
    "                    palette={'Alive': 'white', 'Deceased': 'gray'},\n",
    "                    dodge=False,\n",
    "                    scale='count',\n",
    "                    cut=0)\n",
    "\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "plt.xticks(rotation=90)\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'Phenotypic_Volume_Violin_MSK_count_scale_cut'\n",
    "fn = '/workdir/uvmel_project/figures/final_figures_10_27_20/' + figure_label\n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exec('METADATA_TUMOR = h5_data.load(\\'/METADATA_TUMOR\\')')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "state_map_HMMi3 = {1: 'loss',\n",
    "                   2: 'neutral',\n",
    "                   3: 'gain'}\n",
    "\n",
    "for patient in ['UM01', 'UM02', 'UM03', 'UM04', 'UM05', 'UM06']:\n",
    "\n",
    "    cnv_mat = pd.DataFrame()\n",
    "    \n",
    "    QUERY = pd.read_csv('/workdir/uvmel_project/single_sample_results/results_HMMi3_merged_ref/'+patient+'/'+\n",
    "                                'HMM_CNV_predictions.HMMi3.rand_trees.hmm_mode-subclusters.'+\n",
    "                                'Pnorm_0.5.pred_cnv_regions.dat',\n",
    "                                 sep='\\t')\n",
    "    \n",
    "    cytoband = pd.read_csv('/workdir/uvmel_project/casper/data/cytoband.txt', sep='\\t', header=None)\n",
    "    cytoband.columns = ['chr', 'start', 'end', 'arm', 'col5']\n",
    "    \n",
    "    QUERY_cell_groupings = pd.read_csv('/workdir/uvmel_project/single_sample_results/results_HMMi3_merged_ref/'+\n",
    "                                             patient+'/'+\n",
    "                                            '17_HMM_predHMMi3.rand_trees.hmm_mode-subclusters.cell_groupings',\n",
    "                                             sep='\\t')\n",
    "    \n",
    "    prop_df = pd.DataFrame(QUERY_cell_groupings.cell_group_name.value_counts(normalize=True)).reset_index()\n",
    "    prop_df.columns = ['cell_group_name', 'proportion']\n",
    "    \n",
    "    QUERY['arm'] = ['p' if 'p' in cytoband[(cytoband.chr == QUERY.chr[x]) & \n",
    "                               (cytoband.start <= QUERY.start[x]) &\n",
    "                               (cytoband.end >= QUERY.start[x])].arm.values[0]\n",
    "                            else\n",
    "                            'q'\n",
    "                            for x in range(len(UM03_HMMi3_df))]\n",
    "    \n",
    "    QUERY['CNV'] = [QUERY.chr[x].replace('chr', '')+QUERY.arm[x]+ ' ' + \n",
    "                              state_map_HMMi3[QUERY.state[x]] \n",
    "                              for x in range(len(QUERY))]\n",
    "    \n",
    "    cell_state_map = dict(zip(QUERY.cell_group_name.unique(),\n",
    "                              [','.join(QUERY[QUERY.cell_group_name==x].CNV.values)\n",
    "                               for x in QUERY.cell_group_name.unique()]))\n",
    "    \n",
    "    cell_groupings_cell_names = [int(x[5:]) for x in QUERY_cell_groupings.cell.values]\n",
    "    \n",
    "    cnv_mat['Patient'] = [patient for x in range(len(QUERY_cell_groupings))]\n",
    "    cnv_mat['State'] = [cell_state_map[x] for x in QUERY_cell_groupings.cell_group_name.values]\n",
    "    \n",
    "    QUERY_META = METADATA_TUMOR[METADATA_TUMOR.index.map(lambda x: x[2] == patient)]\n",
    "    \n",
    "    cnv_mat.index = [QUERY_META[QUERY_META.index.map(lambda x: x[3] == y)].index.\\\n",
    "                     get_level_values('Patient').values[0]\n",
    "                     for y in cell_groupings_cell_names]\n",
    "    \n",
    "    #unique_kar_df = pd.DataFrame(cnv_mat.groupby(['Patient']+list(cnv_mat.columns)).size(), columns=['Counts'])\n",
    "    kar_df = cnv_mat.copy()\n",
    "    \n",
    "    kar_df.reset_index(inplace=True)\n",
    "    #kar_df['State'] = kar_df[kar_df.columns[1:-1]].apply(\n",
    "    #                                       lambda x: ','.join(x.dropna().astype(str)), axis=1)\n",
    "    \n",
    "    #unique_kar_df = unique_kar_df[['Patient', 'State', 'Counts']]\n",
    "    \n",
    "    kar_df = kar_df[['Patient', 'State']]\n",
    "    kar_df.set_index('Patient', inplace=True)\n",
    "    \n",
    "    \n",
    "    \n",
    "    kar_df['Assignment'] = [QUERY_META[QUERY_META.index.map(lambda x: x[3] == y)].Assignment.values[0]\n",
    "                            for y in cell_groupings_cell_names]\n",
    "    \n",
    "    kar_df.set_index('Assignment', append=False, inplace=True)\n",
    "    \n",
    "    \n",
    "    # Compute karyotypic SDI:\n",
    "    \n",
    "    sample_size = 50\n",
    "    sample_rounds= 100\n",
    "    \n",
    "    sdi_df = pd.DataFrame()\n",
    "    \n",
    "    for i in range(sample_rounds):\n",
    "        sdi_list = []\n",
    "        for gep_class in kar_df.index.unique():\n",
    "            tmp_df = kar_df.loc[kar_df.index.get_level_values('Assignment') == gep_class]\n",
    "            tmp_df = tmp_df.sample(sample_size, replace=True)\n",
    "            p_vect = tmp_df['State'].value_counts(normalize=True)\n",
    "            sdi_list.append(-1*sum(list(map(lambda x: np.log(x)*x, p_vect))))\n",
    "        \n",
    "        sdi_df[str(i)] = sdi_list\n",
    "    \n",
    "    sdi_df.index = kar_df.index.unique()\n",
    "    #sdi_df['Ave'] = sdi_df.mean(axis=1).values\n",
    "    \n",
    "    sdi_boxplot_data = pd.DataFrame()\n",
    "    sdi_boxplot_data['GEP'] = [item for sublist in [[x]*sample_rounds for x in sdi_df.index] for item in sublist]\n",
    "    sdi_boxplot_data['SDI'] = [item for sublist in [sdi_df.iloc[x].values for x in range(len(sdi_df.index))]\n",
    "                               for item in sublist]\n",
    "    \n",
    "    # Save data:\n",
    "    sdi_boxplot_data.to_csv('/workdir/uvmel_project/data/UM03_Karyotypic_SDI_inferCNV.csv')\n",
    "    \n",
    "    \n",
    "    # Plot KDE:\n",
    "    sns.set_style(\"white\")\n",
    "    fig = plt.figure(figsize = (5,5))\n",
    "    ax = plt.gca()\n",
    "    \n",
    "    sns.kdeplot(sdi_boxplot_data.loc[sdi_boxplot_data.GEP == 'Castle 1']['SDI'].values, shade=True, color=\"#0000FF\")\n",
    "    sns.kdeplot(sdi_boxplot_data.loc[sdi_boxplot_data.GEP == 'Castle 2']['SDI'].values, shade=True, color=\"#FF0000\")\n",
    "    \n",
    "    ax.set_ylabel(\"Frequency\",fontsize=10)\n",
    "    ax.set_xlabel(\"SDI\",fontsize=10)\n",
    "    ax.tick_params(labelsize=10)\n",
    "    #ax.set_xticklabels(ax.get_xticklabels(),rotation=90)\n",
    "    sns.despine()\n",
    "    plt.ylim((0,10))\n",
    "    #plt.xlim((-0.95,-0.4))\n",
    "    \n",
    "    # SAVE FIGURE\n",
    "    figure_label = 'Karyotypic_SDI_KDE_UM03'\n",
    "    fn = '/workdir/uvmel_project/figures/Revision_2_Figures/' + figure_label\n",
    "        \n",
    "    d = os.path.dirname(fn)\n",
    "    if not os.path.exists(d):\n",
    "        os.makedirs(d)\n",
    "        \n",
    "    fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "    fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "    print(fn)\n",
    "    \n",
    "    \n",
    "    # CLASS I vs. CLASS 2\n",
    "    print('Castle 1 vs. Castle 2')\n",
    "    \n",
    "    print(sdi_boxplot_data.loc[sdi_boxplot_data.GEP == 'Castle 1']['SDI'].values)\n",
    "    print(sdi_boxplot_data.loc[sdi_boxplot_data.GEP == 'Castle 2']['SDI'].values)\n",
    "    print(stats.mannwhitneyu(sdi_boxplot_data.loc[sdi_boxplot_data.GEP == 'Castle 1']['SDI'].values,\n",
    "                             sdi_boxplot_data.loc[sdi_boxplot_data.GEP == 'Castle 2']['SDI'].values))\n",
    "    \n",
    "    \n",
    "    # Plot barplot:\n",
    "    fig, ax = plt.subplots(figsize=(2,7))\n",
    "    \n",
    "    ax = sns.barplot(x='GEP', y='SDI', data=sdi_boxplot_data,\n",
    "                     hue='GEP',\n",
    "                     edgecolor='black',\n",
    "                     palette={'Castle 1': 'white', 'Castle 2': 'gray'},\n",
    "                     dodge=False,\n",
    "                     ci=95,\n",
    "                     errwidth=1,\n",
    "                     capsize=0.1,\n",
    "                     estimator=median)\n",
    "     \n",
    "    ax.set_yscale('log')\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.get_legend().remove()\n",
    "    \n",
    "    plt.xticks(rotation=90)\n",
    "    \n",
    "    # SAVE FIGURE\n",
    "    figure_label = 'Karyotypic_SDI_Barplot_UM03_median'\n",
    "    fn = '/workdir/uvmel_project/figures/Revision_2_Figures/' + figure_label\n",
    "        \n",
    "    d = os.path.dirname(fn)\n",
    "    if not os.path.exists(d):\n",
    "        os.makedirs(d)\n",
    "        \n",
    "    fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "    fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "    print(fn)\n",
    "    \n",
    "    print(\"Script Finished\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "state_map_HMMi3 = {1: 'loss',\n",
    "                   2: 'neutral',\n",
    "                   3: 'gain'}\n",
    "\n",
    "cnv_mat = pd.DataFrame()\n",
    "\n",
    "UM03_HMMi3_df = pd.read_csv('/workdir/uvmel_project/single_sample_results/results_HMMi3_merged_ref/UM03/'+\n",
    "                            'HMM_CNV_predictions.HMMi3.rand_trees.hmm_mode-subclusters.'+\n",
    "                            'Pnorm_0.5.pred_cnv_regions.dat',\n",
    "                             sep='\\t')\n",
    "\n",
    "cytoband = pd.read_csv('/workdir/uvmel_project/casper/data/cytoband.txt', sep='\\t', header=None)\n",
    "cytoband.columns = ['chr', 'start', 'end', 'arm', 'col5']\n",
    "\n",
    "UM03_HMMi3_cell_groupings = pd.read_csv('/workdir/uvmel_project/single_sample_results/results_HMMi3_merged_ref/'+\n",
    "                                        'UM03/'+\n",
    "                                        '17_HMM_predHMMi3.rand_trees.hmm_mode-subclusters.cell_groupings',\n",
    "                                         sep='\\t')\n",
    "\n",
    "prop_df_3 = pd.DataFrame(UM03_HMMi3_cell_groupings.cell_group_name.value_counts(normalize=True)).reset_index()\n",
    "prop_df_3.columns = ['cell_group_name', 'proportion']\n",
    "\n",
    "UM03_HMMi3_df['arm'] = ['p' if 'p' in cytoband[(cytoband.chr == UM03_HMMi3_df.chr[x]) & \n",
    "                           (cytoband.start <= UM03_HMMi3_df.start[x]) &\n",
    "                           (cytoband.end >= UM03_HMMi3_df.start[x])].arm.values[0]\n",
    "                        else\n",
    "                        'q'\n",
    "                        for x in range(len(UM03_HMMi3_df))]\n",
    "\n",
    "UM03_HMMi3_df['CNV'] = [UM03_HMMi3_df.chr[x].replace('chr', '')+UM03_HMMi3_df.arm[x]+ ' ' + \n",
    "                          state_map_HMMi3[UM03_HMMi3_df.state[x]] \n",
    "                          for x in range(len(UM03_HMMi3_df))]\n",
    "\n",
    "cell_state_map = dict(zip(UM03_HMMi3_df.cell_group_name.unique(),\n",
    "                          [','.join(UM03_HMMi3_df[UM03_HMMi3_df.cell_group_name==x].CNV.values)\n",
    "                           for x in UM03_HMMi3_df.cell_group_name.unique()]))\n",
    "\n",
    "cell_groupings_cell_names = [int(x[5:]) for x in UM03_HMMi3_cell_groupings.cell.values]\n",
    "\n",
    "cnv_mat['Patient'] = ['UM03' for x in range(len(UM03_HMMi3_cell_groupings))]\n",
    "cnv_mat['State'] = [cell_state_map[x] for x in UM03_HMMi3_cell_groupings.cell_group_name.values]\n",
    "\n",
    "METADATA_TUMOR = METADATA_TUMOR[METADATA_TUMOR.index.map(lambda x: x[2] == 'UM03')]\n",
    "\n",
    "cnv_mat.index = [METADATA_TUMOR[METADATA_TUMOR.index.map(lambda x: x[3] == y)].index.get_level_values('Patient').values[0]\n",
    "                 for y in cell_groupings_cell_names]\n",
    "\n",
    "#unique_kar_df = pd.DataFrame(cnv_mat.groupby(['Patient']+list(cnv_mat.columns)).size(), columns=['Counts'])\n",
    "kar_df = cnv_mat.copy()\n",
    "\n",
    "kar_df.reset_index(inplace=True)\n",
    "#kar_df['State'] = kar_df[kar_df.columns[1:-1]].apply(\n",
    "#                                       lambda x: ','.join(x.dropna().astype(str)), axis=1)\n",
    "\n",
    "#unique_kar_df = unique_kar_df[['Patient', 'State', 'Counts']]\n",
    "\n",
    "kar_df = kar_df[['Patient', 'State']]\n",
    "kar_df.set_index('Patient', inplace=True)\n",
    "\n",
    "kar_df['Assignment'] = [METADATA_TUMOR[METADATA_TUMOR.index.map(lambda x: x[3] == y)].Assignment.values[0]\n",
    "                        for y in cell_groupings_cell_names]\n",
    "\n",
    "kar_df.set_index('Assignment', append=False, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kar_df[kar_df.index == 'Castle 2'].State.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kar_df[kar_df.index == 'Castle 1'].State.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cnv_mat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "METADATA_TUMOR.Assignment.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "METADATA_TUMOR.index.get_level_values('Cell ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "UM03_expr = pd.read_csv('/workdir/uvmel_project/single_sample_results/results_HMMi6_merged_ref/UM03/'+\n",
    "                                'expr.infercnv.dat',\n",
    "                        sep='\\t').T\n",
    "\n",
    "UM03_expr.index = [int(x[5:]) for x in UM03_expr.index]\n",
    "UM03_expr = UM03_expr.sort_index()\n",
    "UM03_expr = UM03_expr[UM03_expr.index.map(lambda x: x in METADATA_TUMOR.index.get_level_values('Cell ID').tolist())]\n",
    "UM03_expr.index = METADATA_TUMOR.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "UM03_expr_prelim = pd.read_csv('/workdir/uvmel_project/single_sample_results/results_HMMi6_merged_ref/UM03/'+\n",
    "                                'expr.infercnv.preliminary.dat',\n",
    "                        sep='\\t').T\n",
    "\n",
    "UM03_expr_prelim.index = [int(x[5:]) for x in UM03_expr_prelim.index]\n",
    "UM03_expr_prelim = UM03_expr_prelim.sort_index()\n",
    "UM03_expr_prelim = UM03_expr_prelim[UM03_expr_prelim.index.map(lambda x: x in \n",
    "                                    METADATA_TUMOR.index.get_level_values('Cell ID').tolist())]\n",
    "UM03_expr_prelim.index = METADATA_TUMOR.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "QUERY = UM03_expr\n",
    "\n",
    "# COLUMN INDEX AND COLORS\n",
    "genes = QUERY.columns.tolist()\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "heatmap_data = QUERY\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns\n",
    "\n",
    "# LINKAGE \n",
    "method = 'centroid' # average, single\n",
    "metric = 'euclidean' # cosine\n",
    "linkage = hc.linkage(heatmap_data, method=method, metric = metric)\n",
    "row_linkage = deepcopy(linkage)\n",
    "#linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "#col_linkage = deepcopy(linkage)\n",
    "\n",
    "# REORDER HEATMAP ACCORDING TO LINKAGE (OPTIONAL, STILL SLOW)\n",
    "r1 = hc.leaves_list(row_linkage)\n",
    "mat = heatmap_data.iloc[r1,:]\n",
    "\n",
    "GEP_assignments_sorted = [METADATA_TUMOR[METADATA_TUMOR.index.get_level_values('Cell ID') == x].Assignment.values[0] \n",
    "                          for x in mat.index.get_level_values('Cell ID').values]\n",
    "\n",
    "row_colors = pd.Series(GEP_assignments_sorted, index=GEP_assignments_sorted).map(lut_CASTLE)\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(4,10))\n",
    "plt.rcParams[\"axes.grid\"] = False\n",
    "\n",
    "# ADD ROW COLOR INDEX 1 (PHENOGRAPH CLASS)\n",
    "ax1 = fig.add_axes([0,0.1,0.05,0.6]) # [x0,y0,width,height]\n",
    "x = 0\n",
    "y = 0\n",
    "for c in row_colors:\n",
    "    pos = (x, y / len(row_colors))\n",
    "    ax1.add_patch(patches.Rectangle(pos, 1, 1 / len(row_colors), color=c))\n",
    "    if y >= len(row_colors)-1:\n",
    "        x += 1\n",
    "        y = 0\n",
    "    else:\n",
    "        y += 1\n",
    "plt.axis('off')\n",
    "\n",
    "# ADD MATRIX WITH GENE NAMES\n",
    "axmatrix = fig.add_axes([0.05,0.1,0.9,0.6], frameon=False)\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=plt.cm.RdBu_r,\n",
    "                      vmin=0.890275611964209,\n",
    "                      vmax=1.11449552904841)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "axmatrix.set_xticklabels([]);\n",
    "#xtick = plt.xticks(range(len(labels)), labels, rotation = 90, fontsize = 14)\n",
    "plt.tick_params(size=0)\n",
    "\n",
    "# ADD COLORBAR\n",
    "axcolor = fig.add_axes([1.1,0.1,0.1,0.1])\n",
    "cbar = plt.colorbar(im, cax=axcolor)\n",
    "#cbar.ax.get_yaxis().set_ticks([])\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'UM03_inferCNV_expr_centroid_euclidean'\n",
    "fn = '/workdir/uvmel_project/figures/Revision_2_Figures/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "QUERY = UM03_expr_prelim\n",
    "\n",
    "# COLUMN INDEX AND COLORS\n",
    "genes = QUERY.columns.tolist()\n",
    "# CONSTRUCT HEATMAP DATA\n",
    "heatmap_data = QUERY\n",
    "yticks = heatmap_data.index\n",
    "xticks = heatmap_data.columns\n",
    "\n",
    "# LINKAGE \n",
    "method = 'centroid' # average, single\n",
    "metric = 'euclidean' # cosine\n",
    "linkage = hc.linkage(heatmap_data, method=method, metric = metric)\n",
    "row_linkage = deepcopy(linkage)\n",
    "#linkage = hc.linkage(heatmap_data.T, method=method, metric = metric)\n",
    "#col_linkage = deepcopy(linkage)\n",
    "\n",
    "# REORDER HEATMAP ACCORDING TO LINKAGE (OPTIONAL, STILL SLOW)\n",
    "r1 = hc.leaves_list(row_linkage)\n",
    "mat = heatmap_data.iloc[r1,:]\n",
    "\n",
    "GEP_assignments_sorted = [METADATA_TUMOR[METADATA_TUMOR.index.get_level_values('Cell ID') == x].Assignment.values[0] \n",
    "                          for x in mat.index.get_level_values('Cell ID').values]\n",
    "\n",
    "row_colors = pd.Series(GEP_assignments_sorted, index=GEP_assignments_sorted).map(lut_CASTLE)\n",
    "\n",
    "# VIEW CLUSTERED LABELED HEATMAP AND DENDROGRAM \n",
    "fig = plt.figure(figsize=(4,10))\n",
    "plt.rcParams[\"axes.grid\"] = False\n",
    "\n",
    "# ADD ROW COLOR INDEX 1 (PHENOGRAPH CLASS)\n",
    "ax1 = fig.add_axes([0,0.1,0.05,0.6]) # [x0,y0,width,height]\n",
    "x = 0\n",
    "y = 0\n",
    "for c in row_colors:\n",
    "    pos = (x, y / len(row_colors))\n",
    "    ax1.add_patch(patches.Rectangle(pos, 1, 1 / len(row_colors), color=c))\n",
    "    if y >= len(row_colors)-1:\n",
    "        x += 1\n",
    "        y = 0\n",
    "    else:\n",
    "        y += 1\n",
    "plt.axis('off')\n",
    "\n",
    "# ADD MATRIX WITH GENE NAMES\n",
    "axmatrix = fig.add_axes([0.05,0.1,0.9,0.6], frameon=False)\n",
    "im = axmatrix.matshow(mat, aspect='auto', origin='lower', cmap=plt.cm.RdBu_r,\n",
    "                      vmin=0.885504470951586,\n",
    "                      vmax=1.11449552904841)\n",
    "labels = list(mat.columns)\n",
    "axmatrix.xaxis.set_ticks_position('bottom')\n",
    "axmatrix.set_yticklabels([]);\n",
    "axmatrix.set_xticklabels([]);\n",
    "#xtick = plt.xticks(range(len(labels)), labels, rotation = 90, fontsize = 14)\n",
    "plt.tick_params(size=0)\n",
    "\n",
    "# ADD COLORBAR\n",
    "axcolor = fig.add_axes([1.1,0.1,0.1,0.1])\n",
    "cbar = plt.colorbar(im, cax=axcolor)\n",
    "#cbar.ax.get_yaxis().set_ticks([])\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = 'UM03_inferCNV_expr_PRELIM_centroid_euclidean'\n",
    "fn = '/workdir/uvmel_project/figures/Revision_2_Figures/' + figure_label \n",
    "    \n",
    "d = os.path.dirname(fn)\n",
    "if not os.path.exists(d):\n",
    "    os.makedirs(d)\n",
    "    \n",
    "fig.savefig(fn + '.png', bbox_inches='tight',dpi=400)\n",
    "fig.savefig(fn + '.pdf', bbox_inches='tight',dpi=400)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RANK DEG PER ARCHETYPE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### VISUALIZE ARCHETYPES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FLATUI_ARCHETYPES = [\n",
    "                 'E6E6FA', # not assgined\n",
    "                 'FF66FF', # pink\n",
    "                 'D1C756', # purple\n",
    "                 '9ACD32', # Green\n",
    "                 '00FFFF', # skye blue\n",
    "                 'FF8800', # orange\n",
    "                 '0000FF', # dark blue\n",
    "                 'FFD700', # yellow\n",
    "                 '20B2AA', # teal\n",
    "    ]\n",
    "\n",
    "# CONVERT HEX TO RGB (FLATUI_SAMPLE)\n",
    "colors = np.zeros((len(FLATUI_ARCHETYPES),3))\n",
    "for ind,hexcolor in enumerate(FLATUI_ARCHETYPES):\n",
    "    colors[ind,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "CM_ARCHETYPES = LinearSegmentedColormap.from_list('FLATUI_ARCHETYPES', colors, N=len(colors))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FLATUI_NEAREST_ARCHETYPES = [\n",
    "                 'FF66FF', # pink\n",
    "                 '6600CC', # purple\n",
    "                 '9ACD32', # Green\n",
    "                 '00FFFF', # skye blue\n",
    "                 'FF8800', # orange\n",
    "                 '0000FF', # dark blue\n",
    "                 'FFD700', # yellow\n",
    "                 '20B2AA', # teal\n",
    "    ]\n",
    "\n",
    "# CONVERT HEX TO RGB (FLATUI_SAMPLE)\n",
    "colors = np.zeros((len(FLATUI_NEAREST_ARCHETYPES),3))\n",
    "for ind,hexcolor in enumerate(FLATUI_NEAREST_ARCHETYPES):\n",
    "    colors[ind,:] = tuple(hex(hexcolor).rgb)\n",
    "colors = np.divide(colors,255)\n",
    "CM_NEAREST_ARCHETYPES = LinearSegmentedColormap.from_list('FLATUI_NEAREST_ARCHETYPES', colors, N=len(colors))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visuaulize distance to each archetype\n",
    "# PLOT CATEGORICAL CLUSTER ASSIGNMENTS\n",
    "# Note: Any rows assigned -1 were identified as outliers and should not be considered as a member of any community.\n",
    "plt.figure(figsize = (20,10))\n",
    "nrow = 1\n",
    "ncol = 2\n",
    "dot_size = 2\n",
    "gs1 = gridspec.GridSpec(nrow, ncol)\n",
    "gs1.update(wspace=0.05, hspace=0.1) # set the spacing between axes. \n",
    "\n",
    "cm = plt.cm.bwr \n",
    "dimtype = 'ForceDirected' \n",
    "meta = 'Phenograph_Class'\n",
    "exec('DIM = DIMENSIONS_{}'.format(subset_type))\n",
    "exec('QUERY= INDF_{}'.format(subset_type))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "x = DIM['{}0'.format(dimtype)]\n",
    "y = DIM['{}1'.format(dimtype)]\n",
    "axis_min = -ceil(ceil(max(np.abs(x).max(),np.abs(y).max()))/10)*10\n",
    "axis_max = ceil(ceil(max(np.abs(x).max(),np.abs(y).max()))/10)*10\n",
    "\n",
    "# MONOSOMY 3 SIGNATURE\n",
    "ax = plt.subplot(gs1[0])\n",
    "gene_sig = 'Monosomy 3 Up'\n",
    "signature_genes = genesets[gene_sig].values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "phase = zscore(np.nanmedian(QUERY[detected_genes],axis=1))\n",
    "# Plot scatter map with phase score\n",
    "mask = np.isnan(phase)\n",
    "    \n",
    "# Randomize/Sort Color Values before Plotting\n",
    "color_values = np.array(phase)\n",
    "i = np.random.permutation(len(color_values))\n",
    "xx = np.array(x)[i]\n",
    "yy = np.array(y)[i]\n",
    "color_values = color_values[i]\n",
    "\n",
    "plt.scatter(xx, yy, c = '#E6E6FA', cmap = CM_DIVERGING, s=dot_size, alpha =0.4)\n",
    "plt.clim(-1.5,1.5)\n",
    "plt.xlim(axis_min,axis_max)\n",
    "plt.ylim(axis_min,axis_max)\n",
    "plt.axis('off')\n",
    "plt.title('Monosomy 3', fontname='Helvetica', size=12, weight='normal')\n",
    "\n",
    "# Cell Nearest to Each Archetype\n",
    "for ii in np.unique(META['Archetype_Nearest_Cell'])[1:]:\n",
    "    c=META['Archetype_Nearest_Cell']==ii\n",
    "    cc=META['Archetype_Soft_Cluster']==ii\n",
    "    plt.scatter(x[c.values],y[c.values],s= cc.sum()*0.5, color = '#'+FLATUI_ARCHETYPES[ii])\n",
    "    #offset = 1000\n",
    "    #plt.text(x[c]+offset,y[c]+offset,s = '{}'.format(ii), fontsize = 20)\n",
    "    sns.despine()\n",
    "    plt.xlabel('tSNE-1', fontname='Helvetica', size=10, weight='normal')\n",
    "    plt.ylabel('tSNE-2', fontname='Helvetica', size=10, weight='normal')\n",
    "    plt.title('Archetypes', fontname='Helvetica', size=12, weight='normal')\n",
    "    plt.xlim(axis_min,axis_max)\n",
    "    plt.ylim(axis_min,axis_max)\n",
    "    plt.axis('off')\n",
    "\n",
    "# Soft Cluster Around Each Archetype Defined by Multi-Scale Diffusion Distance\n",
    "ax = plt.subplot(gs1[1])\n",
    "seqc.plot.scatter.categorical(x, y, c=META['Archetype_Soft_Cluster'], \n",
    "                              cmap=CM_ARCHETYPES,legend_kwargs={'ncol': 1}, s=dot_size, ax=ax);\n",
    "sns.despine()\n",
    "plt.xlabel('tSNE-1', fontname='Helvetica', size=10, weight='normal')\n",
    "plt.ylabel('tSNE-2', fontname='Helvetica', size=10, weight='normal')\n",
    "plt.title('Sample ID', fontname='Helvetica', size=12, weight='normal')\n",
    "plt.xlim(axis_min,axis_max)\n",
    "plt.ylim(axis_min,axis_max)\n",
    "plt.axis('off')\n",
    "#ax.legend_.remove()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '_NEARESTArchetypes_Grey_INDF_{}_{}_{}'.format(subset_type,meta,dimtype) \n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label \n",
    "plt.savefig(fn + '.png', bbox_inches='tight',dpi=fig_dpi)\n",
    "plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=fig_dpi)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visuaulize distance to each archetype\n",
    "# PLOT CATEGORICAL CLUSTER ASSIGNMENTS\n",
    "# Note: Any rows assigned -1 were identified as outliers and should not be considered as a member of any community.\n",
    "plt.figure(figsize = (20,10))\n",
    "nrow = 1\n",
    "ncol = 2\n",
    "dot_size = 2\n",
    "gs1 = gridspec.GridSpec(nrow, ncol)\n",
    "gs1.update(wspace=0.05, hspace=0.1) # set the spacing between axes. \n",
    "\n",
    "cm = plt.cm.bwr \n",
    "dimtype = 'ForceDirected' \n",
    "meta = 'Phenograph_Class'\n",
    "exec('DIM = DIMENSIONS_{}'.format(subset_type))\n",
    "exec('QUERY= INDF_{}'.format(subset_type))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "x = DIM['{}0'.format(dimtype)]\n",
    "y = DIM['{}1'.format(dimtype)]\n",
    "axis_min = -ceil(ceil(max(np.abs(x).max(),np.abs(y).max()))/10)*10\n",
    "axis_max = ceil(ceil(max(np.abs(x).max(),np.abs(y).max()))/10)*10\n",
    "\n",
    "# MONOSOMY 3 SIGNATURE\n",
    "ax = plt.subplot(gs1[0])\n",
    "gene_sig = 'Monosomy 3 Up'\n",
    "signature_genes = genesets[gene_sig].values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "phase = zscore(np.nanmedian(QUERY[detected_genes],axis=1))\n",
    "# Plot scatter map with phase score\n",
    "mask = np.isnan(phase)\n",
    "    \n",
    "# Randomize/Sort Color Values before Plotting\n",
    "color_values = np.array(phase)\n",
    "i = np.random.permutation(len(color_values))\n",
    "xx = np.array(x)[i]\n",
    "yy = np.array(y)[i]\n",
    "color_values = color_values[i]\n",
    "\n",
    "plt.scatter(xx, yy, c = '#E6E6FA', cmap = CM_DIVERGING, s=dot_size, alpha =0.4)\n",
    "plt.clim(-1.5,1.5)\n",
    "plt.xlim(axis_min,axis_max)\n",
    "plt.ylim(axis_min,axis_max)\n",
    "plt.axis('off')\n",
    "plt.title('Monosomy 3', fontname='Helvetica', size=12, weight='normal')\n",
    "\n",
    "# Cell Nearest to Each Archetype\n",
    "for ii in np.unique(META['Archetype_Nearest_Cell'])[1:]:\n",
    "    c=META['Archetype_Nearest_Cell']==ii\n",
    "    cc=META['Archetype_Soft_Cluster']==ii\n",
    "    plt.scatter(x[c.values],y[c.values],s= cc.sum()*0.5, color = '#'+FLATUI_ARCHETYPES[ii])\n",
    "    #offset = 1000\n",
    "    #plt.text(x[c]+offset,y[c]+offset,s = '{}'.format(ii), fontsize = 20)\n",
    "    sns.despine()\n",
    "    plt.xlabel('tSNE-1', fontname='Helvetica', size=10, weight='normal')\n",
    "    plt.ylabel('tSNE-2', fontname='Helvetica', size=10, weight='normal')\n",
    "    plt.title('Archetypes', fontname='Helvetica', size=12, weight='normal')\n",
    "    plt.xlim(axis_min,axis_max)\n",
    "    plt.ylim(axis_min,axis_max)\n",
    "    plt.axis('off')\n",
    "\n",
    "# Nearest Archetype Defined by Multi-Scale Diffusion Distance\n",
    "ax = plt.subplot(gs1[1])\n",
    "seqc.plot.scatter.categorical(x, y, c=META['Nearest_Archetype'], \n",
    "                              cmap=CM_NEAREST_ARCHETYPES,legend_kwargs={'ncol': 1}, s=dot_size, ax=ax);\n",
    "sns.despine()\n",
    "plt.xlabel('tSNE-1', fontname='Helvetica', size=10, weight='normal')\n",
    "plt.ylabel('tSNE-2', fontname='Helvetica', size=10, weight='normal')\n",
    "plt.title('Sample ID', fontname='Helvetica', size=12, weight='normal')\n",
    "plt.xlim(axis_min,axis_max)\n",
    "plt.ylim(axis_min,axis_max)\n",
    "plt.axis('off')\n",
    "#ax.legend_.remove()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '_NEARESTArchetypes_Grey_INDF_{}_{}_{}'.format(subset_type,meta,dimtype) \n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label \n",
    "plt.savefig(fn + '.png', bbox_inches='tight',dpi=fig_dpi)\n",
    "plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=fig_dpi)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visuaulize distance to each archetype colored by GEP1 Probability\n",
    "# PLOT CATEGORICAL CLUSTER ASSIGNMENTS\n",
    "# Note: Any rows assigned -1 were identified as outliers and should not be considered as a member of any community.\n",
    "plt.figure(figsize = (20,10))\n",
    "nrow = 1\n",
    "ncol = 2\n",
    "dot_size = 2\n",
    "gs1 = gridspec.GridSpec(nrow, ncol)\n",
    "gs1.update(wspace=0.05, hspace=0.1) # set the spacing between axes. \n",
    "\n",
    "cm = plt.cm.bwr \n",
    "dimtype = 'ForceDirected' \n",
    "meta = 'Phenograph_Class'\n",
    "exec('DIM = DIMENSIONS_{}'.format(subset_type))\n",
    "exec('QUERY= INDF_{}'.format(subset_type))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "x = DIM['{}0'.format(dimtype)]\n",
    "y = DIM['{}1'.format(dimtype)]\n",
    "axis_min = -ceil(ceil(max(np.abs(x).max(),np.abs(y).max()))/10)*10\n",
    "axis_max = ceil(ceil(max(np.abs(x).max(),np.abs(y).max()))/10)*10\n",
    "\n",
    "# MONOSOMY 3 SIGNATURE\n",
    "ax = plt.subplot(gs1[0])\n",
    "gene_sig = 'Monosomy 3 Up'\n",
    "signature_genes = genesets[gene_sig].values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "phase = zscore(np.nanmedian(QUERY[detected_genes],axis=1))\n",
    "# Plot scatter map with phase score\n",
    "mask = np.isnan(phase)\n",
    "    \n",
    "# Randomize/Sort Color Values before Plotting\n",
    "color_values = np.array(phase)\n",
    "i = np.random.permutation(len(color_values))\n",
    "xx = np.array(x)[i]\n",
    "yy = np.array(y)[i]\n",
    "color_values = color_values[i]\n",
    "\n",
    "plt.scatter(xx, yy, c = '#E6E6FA', cmap = CM_DIVERGING, s=dot_size, alpha =0.4)\n",
    "plt.clim(-1.5,1.5)\n",
    "plt.xlim(axis_min,axis_max)\n",
    "plt.ylim(axis_min,axis_max)\n",
    "plt.axis('off')\n",
    "plt.title('Monosomy 3', fontname='Helvetica', size=12, weight='normal')\n",
    "\n",
    "# Cell Nearest to Each Archetype\n",
    "for ii in np.unique(META['Archetype_Nearest_Cell'])[1:]:\n",
    "    c=META['Archetype_Nearest_Cell']==ii\n",
    "    cc=META['Archetype_Soft_Cluster']==ii\n",
    "    plt.scatter(x[c.values],y[c.values],s= cc.sum()*0.5, color = '#'+FLATUI_ARCHETYPES[ii])\n",
    "    #offset = 1000\n",
    "    #plt.text(x[c]+offset,y[c]+offset,s = '{}'.format(ii), fontsize = 20)\n",
    "    sns.despine()\n",
    "    plt.xlabel('tSNE-1', fontname='Helvetica', size=10, weight='normal')\n",
    "    plt.ylabel('tSNE-2', fontname='Helvetica', size=10, weight='normal')\n",
    "    plt.title('Archetypes', fontname='Helvetica', size=12, weight='normal')\n",
    "    plt.xlim(axis_min,axis_max)\n",
    "    plt.ylim(axis_min,axis_max)\n",
    "    plt.axis('off')\n",
    "\n",
    "# Nearest Archetype Defined by Multi-Scale Diffusion Distance\n",
    "ax = plt.subplot(gs1[1])\n",
    "seqc.plot.scatter.categorical(x, y, c=METADATA_TUMOR.Assignment, \n",
    "                              cmap=CM_CASTLE, legend=False, s=dot_size, ax=ax,);\n",
    "sns.despine()\n",
    "plt.xlabel('tSNE-1', fontname='Helvetica', size=10, weight='normal')\n",
    "plt.ylabel('tSNE-2', fontname='Helvetica', size=10, weight='normal')\n",
    "plt.title('Sample ID', fontname='Helvetica', size=12, weight='normal')\n",
    "plt.xlim(axis_min,axis_max)\n",
    "plt.ylim(axis_min,axis_max)\n",
    "plt.axis('off')\n",
    "#ax.legend_.remove()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '_NEARESTArchetypes_Grey_INDF_{}_{}_{}'.format(subset_type,meta,dimtype) \n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label \n",
    "plt.savefig(fn + '.png', bbox_inches='tight',dpi=fig_dpi)\n",
    "plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=fig_dpi)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visuaulize distance to each archetype colored by GEP1 Probability\n",
    "# PLOT CATEGORICAL CLUSTER ASSIGNMENTS\n",
    "# Note: Any rows assigned -1 were identified as outliers and should not be considered as a member of any community.\n",
    "plt.figure(figsize = (20,10))\n",
    "nrow = 1\n",
    "ncol = 2\n",
    "dot_size = 2\n",
    "gs1 = gridspec.GridSpec(nrow, ncol)\n",
    "gs1.update(wspace=0.05, hspace=0.1) # set the spacing between axes. \n",
    "\n",
    "cm = plt.cm.bwr \n",
    "dimtype = 'ForceDirected' \n",
    "meta = 'Phenograph_Class'\n",
    "exec('DIM = DIMENSIONS_{}'.format(subset_type))\n",
    "exec('QUERY= INDF_{}'.format(subset_type))\n",
    "exec('META = METADATA_{}'.format(subset_type))\n",
    "x = DIM['{}0'.format(dimtype)]\n",
    "y = DIM['{}1'.format(dimtype)]\n",
    "axis_min = -ceil(ceil(max(np.abs(x).max(),np.abs(y).max()))/10)*10\n",
    "axis_max = ceil(ceil(max(np.abs(x).max(),np.abs(y).max()))/10)*10\n",
    "\n",
    "# MONOSOMY 3 SIGNATURE\n",
    "ax = plt.subplot(gs1[0])\n",
    "gene_sig = 'Monosomy 3 Up'\n",
    "signature_genes = genesets[gene_sig].values\n",
    "signature_genes = [x for x in signature_genes if str(x) != 'nan']\n",
    "detected_genes = list(set(signature_genes).intersection(set(QUERY.columns)))\n",
    "phase = zscore(np.nanmedian(QUERY[detected_genes],axis=1))\n",
    "# Plot scatter map with phase score\n",
    "mask = np.isnan(phase)\n",
    "    \n",
    "# Randomize/Sort Color Values before Plotting\n",
    "color_values = np.array(phase)\n",
    "i = np.random.permutation(len(color_values))\n",
    "xx = np.array(x)[i]\n",
    "yy = np.array(y)[i]\n",
    "color_values = color_values[i]\n",
    "\n",
    "plt.scatter(xx, yy, c = '#E6E6FA', cmap = CM_DIVERGING, s=dot_size, alpha =0.4)\n",
    "plt.clim(-1.5,1.5)\n",
    "plt.xlim(axis_min,axis_max)\n",
    "plt.ylim(axis_min,axis_max)\n",
    "plt.axis('off')\n",
    "plt.title('Monosomy 3', fontname='Helvetica', size=12, weight='normal')\n",
    "\n",
    "# Cell Nearest to Each Archetype\n",
    "for ii in np.unique(META['Archetype_Nearest_Cell'])[1:]:\n",
    "    c=META['Archetype_Nearest_Cell']==ii\n",
    "    cc=META['Archetype_Soft_Cluster']==ii\n",
    "    plt.scatter(x[c.values],y[c.values],s= cc.sum()*0.5, color = '#'+FLATUI_ARCHETYPES[ii])\n",
    "    #offset = 1000\n",
    "    #plt.text(x[c]+offset,y[c]+offset,s = '{}'.format(ii), fontsize = 20)\n",
    "    sns.despine()\n",
    "    plt.xlabel('tSNE-1', fontname='Helvetica', size=10, weight='normal')\n",
    "    plt.ylabel('tSNE-2', fontname='Helvetica', size=10, weight='normal')\n",
    "    plt.title('Archetypes', fontname='Helvetica', size=12, weight='normal')\n",
    "    plt.xlim(axis_min,axis_max)\n",
    "    plt.ylim(axis_min,axis_max)\n",
    "    plt.axis('off')\n",
    "\n",
    "# Nearest Archetype Defined by Multi-Scale Diffusion Distance\n",
    "ax = plt.subplot(gs1[1])\n",
    "seqc.plot.scatter.categorical(x, y, c=METADATA_TUMOR.TCGA_Assignment, \n",
    "                              cmap=CM_TCGA, legend=False, s=dot_size, ax=ax,);\n",
    "sns.despine()\n",
    "plt.xlabel('tSNE-1', fontname='Helvetica', size=10, weight='normal')\n",
    "plt.ylabel('tSNE-2', fontname='Helvetica', size=10, weight='normal')\n",
    "plt.title('Sample ID', fontname='Helvetica', size=12, weight='normal')\n",
    "plt.xlim(axis_min,axis_max)\n",
    "plt.ylim(axis_min,axis_max)\n",
    "plt.axis('off')\n",
    "#ax.legend_.remove()\n",
    "\n",
    "# SAVE FIGURE\n",
    "figure_label = '_NEARESTArchetypes_Grey_INDF_{}_{}_{}'.format(subset_type,meta,dimtype) \n",
    "fn = FIG_output_stem + FN.replace(\".h5\", \"\") + figure_label \n",
    "plt.savefig(fn + '.png', bbox_inches='tight',dpi=fig_dpi)\n",
    "plt.savefig(fn + '.pdf', bbox_inches='tight',dpi=fig_dpi)\n",
    "print(fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Table of Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "665.333px",
    "left": "59px",
    "top": "110px",
    "width": "235.996px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
